<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Introduction to data analysis with R and Bioconductor: R and RStudio</title><meta name="viewport" content="width=device-width, initial-scale=1"><link rel="stylesheet" type="text/css" href="assets/styles.css"><script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png"><link rel="manifest" href="site.webmanifest"><link rel="mask-icon" href="safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" content="#ffffff"></head><body>
    <header id="top" class="navbar navbar-expand-md navbar-light bg-white top-nav incubator"><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-6">
      <div class="large-logo">
        <img alt="Carpentries Incubator" src="assets/images/incubator-logo.svg"></div>
    </div>
    <div class="selector-container">
      
      
      <div class="dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/20-r-rstudio.html';">Instructor View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl navbar-light bg-white bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Introduction to data analysis with R and Bioconductor
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Introduction to data analysis with R and Bioconductor
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><li><a class="dropdown-item" href="discuss.html">Discussion</a></li><li><a class="dropdown-item" href="reference.html"></a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a class="btn btn-primary" href="aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Introduction to data analysis with R and Bioconductor
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 10%" class="percentage">
    10%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 10%" aria-valuenow="10" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/20-r-rstudio.html">Instructor View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->
      
            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="10-data-organisation.html">1. Data organisation with spreadsheets</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        2. R and RStudio
        </span>
      </button>
    </div><!--/div.accordion-header-->
        
    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#what-is-r-what-is-rstudio">What is R? What is RStudio?</a></li>
<li><a href="#why-learn-r">Why learn R?</a></li>
<li><a href="#knowing-your-way-around-rstudio">Knowing your way around RStudio</a></li>
<li><a href="#getting-set-up">Getting set up</a></li>
<li><a href="#interacting-with-r">Interacting with R</a></li>
<li><a href="#how-to-learn-more-during-and-after-the-course">How to learn more during and after the course?</a></li>
<li><a href="#seeking-help">Seeking help</a></li>
<li><a href="#r-packages">R packages</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="23-starting-with-r.html">3. Introduction to R</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="25-starting-with-data.html">4. Starting with data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="30-dplyr.html">5. Manipulating and analysing data with dplyr</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="40-visualization.html">6. Data visualization</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="60-next-steps.html">7. Next steps</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="discuss.html">Discussion</a></li><li><a href="reference.html"></a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width resources"><a href="aio.html">See all in one page</a>
            

            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">
            
            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="10-data-organisation.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="23-starting-with-r.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="10-data-organisation.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Data organisation
        </a>
        <a class="chapter-link float-end" href="23-starting-with-r.html" rel="next">
          Next: Introduction to R... 
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>R and RStudio</h1>
        <p>Last updated on 2024-04-02 |
        
        <a href="https://github.com/carpentries-incubator/bioc-intro/edit/main/episodes/20-r-rstudio.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
        
        
        
        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>What are R and RStudio?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Describe the purpose of the RStudio Script, Console, Environment,
and Plots panes.</li>
<li>Organise files and directories for a set of analyses as an R
project, and understand the purpose of the working directory.</li>
<li>Use the built-in RStudio help interface to search for more
information on R functions.</li>
<li>Demonstrate how to provide sufficient information for
troubleshooting with the R user community.</li>
</ul></div>
</div>
</div>
</div>
</div>
<blockquote>
<p>This episode is based on the Data Carpentries’s <em>Data Analysis and
Visualisation in R for Ecologists</em> lesson.</p>
</blockquote>
<section id="what-is-r-what-is-rstudio"><h2 class="section-heading">What is R? What is RStudio?<a class="anchor" aria-label="anchor" href="#what-is-r-what-is-rstudio"></a>
</h2>
<hr class="half-width"><p>The term <a href="https://www.r-project.org/" class="external-link">R</a> is used to refer
to the <em>programming language</em>, the <em>environment for
statistical computing</em> and <em>the software</em> that interprets the
scripts written using it.</p>
<p><a href="https://rstudio.com" class="external-link">RStudio</a> is currently a very popular
way to not only write your R scripts but also to interact with the R
software<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>. To function correctly, RStudio needs R and
therefore both need to be installed on your computer.</p>
<p>The <a href="https://raw.githubusercontent.com/rstudio/cheatsheets/main/rstudio-ide.pdf" class="external-link">RStudio
IDE Cheat Sheet</a> provides much more information than will be covered
here, but can be useful to learn keyboard shortcuts and discover new
features.</p>
</section><section id="why-learn-r"><h2 class="section-heading">Why learn R?<a class="anchor" aria-label="anchor" href="#why-learn-r"></a>
</h2>
<hr class="half-width"><div class="section level3">
<h3 id="r-does-not-involve-lots-of-pointing-and-clicking-and-thats-a-good-thing">R does not involve lots of pointing and clicking, and that’s a good
thing<a class="anchor" aria-label="anchor" href="#r-does-not-involve-lots-of-pointing-and-clicking-and-thats-a-good-thing"></a></h3>
<p>The learning curve might be steeper than with other software, but
with R, the results of your analysis do not rely on remembering a
succession of pointing and clicking, but instead on a series of written
commands, and that’s a good thing! So, if you want to redo your analysis
because you collected more data, you don’t have to remember which button
you clicked in which order to obtain your results; you just have to run
your script again.</p>
<p>Working with scripts makes the steps you used in your analysis clear,
and the code you write can be inspected by someone else who can give you
feedback and spot mistakes.</p>
<p>Working with scripts forces you to have a deeper understanding of
what you are doing, and facilitates your learning and comprehension of
the methods you use.</p>
</div>
<div class="section level3">
<h3 id="r-code-is-great-for-reproducibility">R code is great for reproducibility<a class="anchor" aria-label="anchor" href="#r-code-is-great-for-reproducibility"></a></h3>
<p>Reproducibility means that someone else (including your future self)
can obtain the same results from the same dataset when using the same
analysis code.</p>
<p>R integrates with other tools to generate manuscripts or reports from
your code. If you collect more data, or fix a mistake in your dataset,
the figures and the statistical tests in your manuscript or report are
updated automatically.</p>
<p>An increasing number of journals and funding agencies expect analyses
to be reproducible, so knowing R will give you an edge with these
requirements.</p>
</div>
<div class="section level3">
<h3 id="r-is-interdisciplinary-and-extensible">R is interdisciplinary and extensible<a class="anchor" aria-label="anchor" href="#r-is-interdisciplinary-and-extensible"></a></h3>
<p>With 10000+ packages<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a> that can be installed to extend its
capabilities, R provides a framework that allows you to combine
statistical approaches from many scientific disciplines to best suit the
analytical framework you need to analyse your data. For instance, R has
packages for image analysis, GIS, time series, population genetics, and
a lot more.</p>
<figure style="text-align: center"><img src="fig/cran.png" alt="Exponential increase of the number of packages available on [CRAN](https://cran.r-project.org/), the Comprehensive R Archive Network. From the R Journal, Volume 10/2, December 2018." class="figure mx-auto d-block"><figcaption>
Exponential increase of the number of packages available on <a href="https://cran.r-project.org/" class="external-link">CRAN</a>, the Comprehensive R Archive
Network. From the R Journal, Volume 10/2, December 2018.
</figcaption></figure></div>
<div class="section level3">
<h3 id="r-works-on-data-of-all-shapes-and-sizes">R works on data of all shapes and sizes<a class="anchor" aria-label="anchor" href="#r-works-on-data-of-all-shapes-and-sizes"></a></h3>
<p>The skills you learn with R scale easily with the size of your
dataset. Whether your dataset has hundreds or millions of lines, it
won’t make much difference to you.</p>
<p>R is designed for data analysis. It comes with special data
structures and data types that make handling of missing data and
statistical factors convenient.</p>
<p>R can connect to spreadsheets, databases, and many other data
formats, on your computer or on the web.</p>
</div>
<div class="section level3">
<h3 id="r-produces-high-quality-graphics">R produces high-quality graphics<a class="anchor" aria-label="anchor" href="#r-produces-high-quality-graphics"></a></h3>
<p>The plotting functionalities in R are extensive, and allow you to
adjust any aspect of your graph to convey most effectively the message
from your data.</p>
</div>
<div class="section level3">
<h3 id="r-has-a-large-and-welcoming-community">R has a large and welcoming community<a class="anchor" aria-label="anchor" href="#r-has-a-large-and-welcoming-community"></a></h3>
<p>Thousands of people use R daily. Many of them are willing to help you
through mailing lists and websites such as <a href="https://stackoverflow.com/" class="external-link">Stack Overflow</a>, or on the <a href="https://community.rstudio.com/" class="external-link">RStudio community</a>. These broad
user communities extend to specialised areas such as bioinformatics. One
such subset of the R community is <a href="https://bioconductor.org/" class="external-link">Bioconductor</a>, a scientific project
for analysis and comprehension “of data from current and emerging
biological assays.” This workshop was developed by members of the
Bioconductor community; for more information on Bioconductor, please see
the companion workshop <a href="https://carpentries-incubator.github.io/bioc-project/" class="external-link">“The
Bioconductor Project”</a>.</p>
</div>
<div class="section level3">
<h3 id="not-only-is-r-free-but-it-is-also-open-source-and-cross-platform">Not only is R free, but it is also open-source and
cross-platform<a class="anchor" aria-label="anchor" href="#not-only-is-r-free-but-it-is-also-open-source-and-cross-platform"></a></h3>
<p>Anyone can inspect the source code to see how R works. Because of
this transparency, there is less chance for mistakes, and if you (or
someone else) find some, you can report and fix bugs.</p>
</div>
</section><section id="knowing-your-way-around-rstudio"><h2 class="section-heading">Knowing your way around RStudio<a class="anchor" aria-label="anchor" href="#knowing-your-way-around-rstudio"></a>
</h2>
<hr class="half-width"><p>Let’s start by learning about <a href="https://www.rstudio.com/" class="external-link">RStudio</a>, which is an Integrated
Development Environment (IDE) for working with R.</p>
<p>The RStudio IDE open-source product is free under the <a href="https://www.gnu.org/licenses/agpl-3.0.en.html" class="external-link">Affero General
Public License (AGPL) v3</a>. The RStudio IDE is also available with a
commercial license and priority email support from Posit, Inc.</p>
<p>We will use the RStudio IDE to write code, navigate the files on our
computer, inspect the variables we are going to create, and visualise
the plots we will generate. RStudio can also be used for other things
(e.g., version control, developing packages, writing Shiny apps) that we
will not cover during the workshop.</p>
<figure style="text-align: center"><img src="fig/rstudio-screenshot.png" alt="RStudio interface screenshot. Clockwise from top left: Source, Environment/History, Files/Plots/Packages/Help/Viewer, Console." width="100%" class="figure mx-auto d-block"><figcaption>
RStudio interface screenshot. Clockwise from top left: Source,
Environment/History, Files/Plots/Packages/Help/Viewer, Console.
</figcaption></figure><p>The RStudio window is divided into 4 “Panes”:</p>
<ul><li>the <strong>Source</strong> for your scripts and documents
(top-left, in the default layout)</li>
<li>your <strong>Environment/History</strong> (top-right),</li>
<li>your <strong>Files/Plots/Packages/Help/Viewer</strong>
(bottom-right), and</li>
<li>the R <strong>Console</strong> (bottom-left).</li>
</ul><p>The placement of these panes and their content can be customised (see
menu, <code>Tools -&gt; Global Options -&gt; Pane Layout</code>).</p>
<p>One of the advantages of using RStudio is that all the information
you need to write code is available in a single window. Additionally,
with many shortcuts, <strong>autocompletion</strong>, and
<strong>highlighting</strong> for the major file types you use while
developing in R, RStudio will make typing easier and less
error-prone.</p>
</section><section id="getting-set-up"><h2 class="section-heading">Getting set up<a class="anchor" aria-label="anchor" href="#getting-set-up"></a>
</h2>
<hr class="half-width"><p>It is good practice to keep a set of related data, analyses, and text
self-contained in a single folder, called the <strong>working
directory</strong>. All of the scripts within this folder can then use
<strong>relative paths</strong> to files that indicate where inside the
project a file is located (as opposed to absolute paths, which point to
where a file is on a specific computer). Working this way makes it a lot
easier to move your project around on your computer and share it with
others without worrying about whether or not the underlying scripts will
still work.</p>
<p>RStudio provides a helpful set of tools to do this through its
“Projects” interface, which not only creates a working directory for
you, but also remembers its location (allowing you to quickly navigate
to it) and optionally preserves custom settings and open files to make
it easier to resume work after a break. Go through the steps for
creating an “R Project” for this tutorial below.</p>
<ol style="list-style-type: decimal"><li>Start RStudio.</li>
<li>Under the <code>File</code> menu, click on <code>New project</code>.
Choose <code>New directory</code>, then <code>New project</code>.</li>
<li>Enter a name for this new folder (or “directory”), and choose a
convenient location for it. This will be your <strong>working
directory</strong> for this session (or whole course) (e.g.,
<code>bioc-intro</code>).</li>
<li>Click on <code>Create project</code>.</li>
<li>(Optional) Set Preferences to ‘Never’ save workspace in
RStudio.</li>
</ol><p>RStudio’s default preferences generally work well, but saving a
workspace to .RData can be cumbersome, especially if you are working
with larger datasets. To turn that off, go to Tools –&gt; ‘Global
Options’ and select the ‘Never’ option for ‘Save workspace to .RData’ on
exit.</p>
<figure style="text-align: center"><img src="fig/rstudio-preferences.png" alt="Set 'Save workspace to .RData on exit' to 'Never'" width="70%" class="figure mx-auto d-block"><figcaption>
Set ‘Save workspace to .RData on exit’ to ‘Never’
</figcaption></figure><p>To avoid <a href="https://yihui.name/en/2018/11/biggest-regret-knitr/" class="external-link">character
encoding issues between Windows and other operating systems</a>, we are
going to set UTF-8 by default:</p>
<figure style="text-align: center"><img src="fig/utf8.png" alt="Set the default text encoding to UTF-8 to save us headache in the coming future. (Figure from the link above)." width="70%" class="figure mx-auto d-block"><figcaption>
Set the default text encoding to UTF-8 to save us headache in the coming
future. (Figure from the link above).
</figcaption></figure><div class="section level3">
<h3 id="organizing-your-working-directory">Organizing your working directory<a class="anchor" aria-label="anchor" href="#organizing-your-working-directory"></a></h3>
<p>Using a consistent folder structure across your projects will help
keep things organised, and will also make it easy to find/file things in
the future. This can be especially helpful when you have multiple
projects. In general, you may create directories (folders) for
<strong>scripts</strong>, <strong>data</strong>, and
<strong>documents</strong>.</p>
<ul><li>
<strong><code>data/</code></strong> Use this folder to store your
raw data and intermediate datasets you may create for the need of a
particular analysis. For the sake of transparency and <a href="https://en.wikipedia.org/wiki/Provenance" class="external-link">provenance</a>, you
should <em>always</em> keep a copy of your raw data accessible and do as
much of your data cleanup and preprocessing programmatically (i.e., with
scripts, rather than manually) as possible. Separating raw data from
processed data is also a good idea. For example, you could have files
<code>data/raw/tree_survey.plot1.txt</code> and
<code>...plot2.txt</code> kept separate from a
<code>data/processed/tree.survey.csv</code> file generated by the
<code>scripts/01.preprocess.tree_survey.R</code> script.</li>
<li>
<strong><code>documents/</code></strong> This would be a place to
keep outlines, drafts, and other text.</li>
<li>
<strong><code>scripts/</code></strong> (or <code>src</code>) This
would be the location to keep your R scripts for different analyses or
plotting, and potentially a separate folder for your functions (more on
that later).</li>
</ul><p>You may want additional directories or subdirectories depending on
your project needs, but these should form the backbone of your working
directory.</p>
<figure style="text-align: center"><img src="fig/working-directory-structure.png" alt="Example of a working directory structure." width="100%" class="figure mx-auto d-block"><figcaption>
Example of a working directory structure.
</figcaption></figure><p>For this course, we will need a <code>data/</code> folder to store
our raw data, and we will use <code>data_output/</code> for when we
learn how to export data as CSV files, and <code>fig_output/</code>
folder for the figures that we will save.</p>
<div id="challenge-create-your-project-directory-structure" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="challenge-create-your-project-directory-structure" class="callout-inner">
<h3 class="callout-title">Challenge: create your project directory
structure<a class="anchor" aria-label="anchor" href="#challenge-create-your-project-directory-structure"></a>
</h3>
<div class="callout-content">
<p>Under the <code>Files</code> tab on the right of the screen, click on
<code>New Folder</code> and create a folder named <code>data</code>
within your newly created working directory (e.g.,
<code>~/bioc-intro/data</code>). (Alternatively, type
<code>dir.create("data")</code> at your R console.) Repeat these
operations to create a <code>data_output/</code> and a
<code>fig_output</code> folders.</p>
</div>
</div>
</div>
<p>We are going to keep the script in the root of our working directory
because we are only going to use one file and it will make things
easier.</p>
<p>Your working directory should now look like this:</p>
<figure style="text-align: center"><img src="fig/r-starting-how-it-should-look-like.png" alt="How it should look like at the beginning of this lesson" width="100%" class="figure mx-auto d-block"><figcaption>
How it should look like at the beginning of this lesson
</figcaption></figure><p><strong>Project management</strong> is also applicable to
bioinformatics projects, of course<a href="#fn3" class="footnote-ref" id="fnref3"><sup>3</sup></a>. William Noble (<span class="citation">@Noble:2009</span>) proposes the following directory
structure:</p>
<blockquote>
<p>Directory names are in large typeface, and filenames are in smaller
typeface. Only a subset of the files are shown here. Note that the dates
are formatted <code>&lt;year&gt;-&lt;month&gt;-&lt;day&gt;</code> so
that they can be sorted in chronological order. The source code
<code>src/ms-analysis.c</code> is compiled to create
<code>bin/ms-analysis</code> and is documented in
<code>doc/ms-analysis.html</code>. The <code>README</code> files in the
data directories specify who downloaded the data files from what URL on
what date. The driver script <code>results/2009-01-15/runall</code>
automatically generates the three subdirectories split1, split2, and
split3, corresponding to three cross-validation splits. The
<code>bin/parse-sqt.py</code> script is called by both of the
<code>runall</code> driver scripts.</p>
</blockquote>
<figure style="text-align: center"><img src="fig/noble-bioinfo-project.png" alt="Directory structure for a sample bioinformatics project." width="100%" class="figure mx-auto d-block"><figcaption>
Directory structure for a sample bioinformatics project.
</figcaption></figure><p>The most important aspect of a well defined and well documented
project directory is to enable someone unfamiliar with the project<a href="#fn4" class="footnote-ref" id="fnref4"><sup>4</sup></a> to</p>
<ol style="list-style-type: decimal"><li><p>understand what the project is about, what data are available,
what analyses were run, and what results were produced and, most
importantly to</p></li>
<li><p>repeat the analysis over again - with new data, or changing some
analysis parameters.</p></li>
</ol></div>
<div class="section level3">
<h3 id="the-working-directory">The working directory<a class="anchor" aria-label="anchor" href="#the-working-directory"></a></h3>
<p>The working directory is an important concept to understand. It is
the place from where R will be looking for and saving the files. When
you write code for your project, it should refer to files in relation to
the root of your working directory and only need files within this
structure.</p>
<p>Using RStudio projects makes this easy and ensures that your working
directory is set properly. If you need to check it, you can use
<code>getwd()</code>. If for some reason your working directory is not
what it should be, you can change it in the RStudio interface by
navigating in the file browser where your working directory should be,
and clicking on the blue gear icon <code>More</code>, and select
<code>Set As Working Directory</code>. Alternatively you can use
<code>setwd("/path/to/working/directory")</code> to reset your working
directory. However, your scripts should not include this line because it
will fail on someone else’s computer.</p>
<p><strong>Example</strong></p>
<p>The schema below represents the working directory
<code>bioc-intro</code> with the <code>data</code> and
<code>fig_output</code> sub-directories, and 2 files in the latter:</p>
<pre><code>bioc-intro/data/
          /fig_output/fig1.pdf
          /fig_output/fig2.png</code></pre>
<p>If we were in the working directory, we could refer to the
<code>fig1.pdf</code> file using the relative path
<code>bioc-intro/fig_output/fig1.pdf</code> or the absolute path
<code>/home/user/bioc-intro/fig_output/fig1.pdf</code>.</p>
<p>If we were in the <code>data</code> directory, we would use the
relative path <code>../fig_output/fig1.pdf</code> or the same absolute
path <code>/home/user/bioc-intro/fig_output/fig1.pdf</code>.</p>
</div>
</section><section id="interacting-with-r"><h2 class="section-heading">Interacting with R<a class="anchor" aria-label="anchor" href="#interacting-with-r"></a>
</h2>
<hr class="half-width"><p>The basis of programming is that we write down instructions for the
computer to follow, and then we tell the computer to follow those
instructions. We write, or <em>code</em>, instructions in R because it
is a common language that both the computer and we can understand. We
call the instructions <em>commands</em> and we tell the computer to
follow the instructions by <em>executing</em> (also called
<em>running</em>) those commands.</p>
<p>There are two main ways of interacting with R: by using the
<strong>console</strong> or by using <strong>scripts</strong> (plain
text files that contain your code). The console pane (in RStudio, the
bottom left panel) is the place where commands written in the R language
can be typed and executed immediately by the computer. It is also where
the results will be shown for commands that have been executed. You can
type commands directly into the console and press <code>Enter</code> to
execute those commands, but they will be forgotten when you close the
session.</p>
<p>Because we want our code and workflow to be reproducible, it is
better to type the commands we want in the script editor, and save the
script. This way, there is a complete record of what we did, and anyone
(including our future selves!) can easily replicate the results on their
computer. Note, however, that merely typing the commands in the script
does not automatically <em>run</em> them - they still need to be sent to
the console for execution.</p>
<p>RStudio allows you to execute commands directly from the script
editor by using the <code>Ctrl</code> + <code>Enter</code> shortcut (on
Macs, <code>Cmd</code> + <code>Return</code> will work, too). The
command on the current line in the script (indicated by the cursor) or
all of the commands in the currently selected text will be sent to the
console and executed when you press <code>Ctrl</code> +
<code>Enter</code>. You can find other keyboard shortcuts in this <a href="https://raw.githubusercontent.com/rstudio/cheatsheets/main/rstudio-ide.pdf" class="external-link">RStudio
cheatsheet about the RStudio IDE</a>.</p>
<p>At some point in your analysis you may want to check the content of a
variable or the structure of an object, without necessarily keeping a
record of it in your script. You can type these commands and execute
them directly in the console. RStudio provides the <code>Ctrl</code> +
<code>1</code> and <code>Ctrl</code> + <code>2</code> shortcuts allow
you to jump between the script and the console panes.</p>
<p>If R is ready to accept commands, the R console shows a
<code>&gt;</code> prompt. If it receives a command (by typing,
copy-pasting or sending from the script editor using <code>Ctrl</code> +
<code>Enter</code>), R will try to execute it, and when ready, will show
the results and come back with a new <code>&gt;</code> prompt to wait
for new commands.</p>
<p>If R is still waiting for you to enter more data because it isn’t
complete yet, the console will show a <code>+</code> prompt. It means
that you haven’t finished entering a complete command. This is because
you have not ‘closed’ a parenthesis or quotation, i.e. you don’t have
the same number of left-parentheses as right-parentheses, or the same
number of opening and closing quotation marks. When this happens, and
you thought you finished typing your command, click inside the console
window and press <code>Esc</code>; this will cancel the incomplete
command and return you to the <code>&gt;</code> prompt.</p>
</section><section id="how-to-learn-more-during-and-after-the-course"><h2 class="section-heading">How to learn more during and after the course?<a class="anchor" aria-label="anchor" href="#how-to-learn-more-during-and-after-the-course"></a>
</h2>
<hr class="half-width"><p>The material we cover during this course will give you an initial
taste of how you can use R to analyse data for your own research.
However, you will need to learn more to do advanced operations such as
cleaning your dataset, using statistical methods, or creating beautiful
graphics<a href="#fn5" class="footnote-ref" id="fnref5"><sup>5</sup></a>. The best way to become proficient and
efficient at R, as with any other tool, is to use it to address your
actual research questions. As a beginner, it can feel daunting to have
to write a script from scratch, and given that many people make their
code available online, modifying existing code to suit your purpose
might make it easier for you to get started.</p>
<figure><img src="fig/kitten-try-things.jpg" width="400px" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></section><section id="seeking-help"><h2 class="section-heading">Seeking help<a class="anchor" aria-label="anchor" href="#seeking-help"></a>
</h2>
<hr class="half-width"><div class="section level3">
<h3 id="use-the-built-in-rstudio-help-interface-to-search-for-more-information-on-r-functions">Use the built-in RStudio help interface to search for more
information on R functions<a class="anchor" aria-label="anchor" href="#use-the-built-in-rstudio-help-interface-to-search-for-more-information-on-r-functions"></a></h3>
<figure style="text-align: center"><img src="fig/rstudiohelp.png" alt="RStudio help interface." width="70%" class="figure mx-auto d-block"><figcaption>
RStudio help interface.
</figcaption></figure><p>One of the fastest ways to get help, is to use the RStudio help
interface. This panel by default can be found at the lower right hand
panel of RStudio. As seen in the screenshot, by typing the word “Mean”,
RStudio tries to also give a number of suggestions that you might be
interested in. The description is then shown in the display window.</p>
</div>
<div class="section level3">
<h3 id="i-know-the-name-of-the-function-i-want-to-use-but-im-not-sure-how-to-use-it">I know the name of the function I want to use, but I’m not sure how
to use it<a class="anchor" aria-label="anchor" href="#i-know-the-name-of-the-function-i-want-to-use-but-im-not-sure-how-to-use-it"></a></h3>
<p>If you need help with a specific function, let’s say
<code>barplot()</code>, you can type:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">?</span><span class="va">barplot</span></span></code></pre>
</div>
<p>If you just need to remind yourself of the names of the arguments,
you can use:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">args</span><span class="op">(</span><span class="va">lm</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="i-want-to-use-a-function-that-does-x-there-must-be-a-function-for-it-but-i-dont-know-which-one">I want to use a function that does X, there must be a function for
it but I don’t know which one…<a class="anchor" aria-label="anchor" href="#i-want-to-use-a-function-that-does-x-there-must-be-a-function-for-it-but-i-dont-know-which-one"></a></h3>
<p>If you are looking for a function to do a particular task, you can
use the <code>help.search()</code> function, which is called by the
double question mark <code>??</code>. However, this only looks through
the installed packages for help pages with a match to your search
request</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">?</span><span class="op">?</span><span class="va">kruskal</span></span></code></pre>
</div>
<p>If you can’t find what you are looking for, you can use the <a href="https://www.rdocumentation.org" class="external-link">rdocumentation.org</a> website
that searches through the help files across all packages available.</p>
<p>Finally, a generic Google or internet search “R &lt;task&gt;” will
often either send you to the appropriate package documentation or a
helpful forum where someone else has already asked your question.</p>
</div>
<div class="section level3">
<h3 id="i-am-stuck-i-get-an-error-message-that-i-dont-understand">I am stuck… I get an error message that I don’t understand<a class="anchor" aria-label="anchor" href="#i-am-stuck-i-get-an-error-message-that-i-dont-understand"></a></h3>
<p>Start by googling the error message. However, this doesn’t always
work very well because often, package developers rely on the error
catching provided by R. You end up with general error messages that
might not be very helpful to diagnose a problem (e.g. “subscript out of
bounds”). If the message is very generic, you might also include the
name of the function or package you’re using in your query.</p>
<p>However, you should check Stack Overflow. Search using the
<code>[r]</code> tag. Most questions have already been answered, but the
challenge is to use the right words in the search to find the
answers:</p>
<p><a href="https://stackoverflow.com/questions/tagged/r" class="external-link">http://stackoverflow.com/questions/tagged/r</a></p>
<p>The <a href="https://cran.r-project.org/doc/manuals/R-intro.pdf" class="external-link">Introduction
to R</a> can also be dense for people with little programming experience
but it is a good place to understand the underpinnings of the R
language.</p>
<p>The <a href="https://cran.r-project.org/doc/FAQ/R-FAQ.html" class="external-link">R FAQ</a>
is dense and technical but it is full of useful information.</p>
</div>
<div class="section level3">
<h3 id="asking-for-help">Asking for help<a class="anchor" aria-label="anchor" href="#asking-for-help"></a></h3>
<p>The key to receiving help from someone is for them to rapidly grasp
your problem. You should make it as easy as possible to pinpoint where
the issue might be.</p>
<p>Try to use the correct words to describe your problem. For instance,
a package is not the same thing as a library. Most people will
understand what you meant, but others have really strong feelings about
the difference in meaning. The key point is that it can make things
confusing for people trying to help you. Be as precise as possible when
describing your problem.</p>
<p>If possible, try to reduce what doesn’t work to a simple
<em>reproducible example</em>. If you can reproduce the problem using a
very small data frame instead of your 50000 rows and 10000 columns one,
provide the small one with the description of your problem. When
appropriate, try to generalise what you are doing so even people who are
not in your field can understand the question. For instance instead of
using a subset of your real dataset, create a small (3 columns, 5 rows)
generic one. For more information on how to write a reproducible example
see <a href="https://adv-r.had.co.nz/Reproducibility.html" class="external-link">this article
by Hadley Wickham</a>.</p>
<p>To share an object with someone else, if it’s relatively small, you
can use the function <code>dput()</code>. It will output R code that can
be used to recreate the exact same object as the one in memory:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## iris is an example data frame that comes with R and head() is a</span></span>
<span><span class="co">## function that returns the first part of the data frame</span></span>
<span><span class="fu">dput</span><span class="op">(</span><span class="fu">head</span><span class="op">(</span><span class="va">iris</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="fu">structure</span><span class="op">(</span><span class="fu">list</span><span class="op">(</span>Sepal.Length <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">5.1</span>, <span class="fl">4.9</span>, <span class="fl">4.7</span>, <span class="fl">4.6</span>, <span class="fl">5</span>, <span class="fl">5.4</span><span class="op">)</span>, </span>
<span>    Sepal.Width <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">3.5</span>, <span class="fl">3</span>, <span class="fl">3.2</span>, <span class="fl">3.1</span>, <span class="fl">3.6</span>, <span class="fl">3.9</span><span class="op">)</span>, Petal.Length <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1.4</span>, </span>
<span>    <span class="fl">1.4</span>, <span class="fl">1.3</span>, <span class="fl">1.5</span>, <span class="fl">1.4</span>, <span class="fl">1.7</span><span class="op">)</span>, Petal.Width <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">0.2</span>, <span class="fl">0.2</span>, <span class="fl">0.2</span>, </span>
<span>    <span class="fl">0.2</span>, <span class="fl">0.2</span>, <span class="fl">0.4</span><span class="op">)</span>, Species <span class="op">=</span> <span class="fu">structure</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="fl">1L</span>, <span class="fl">1L</span>, <span class="fl">1L</span>, <span class="fl">1L</span>, <span class="fl">1L</span>, </span>
<span>    <span class="fl">1L</span><span class="op">)</span>, levels <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"setosa"</span>, <span class="st">"versicolor"</span>, <span class="st">"virginica"</span><span class="op">)</span>, class <span class="op">=</span> <span class="st">"factor"</span><span class="op">)</span><span class="op">)</span>, row.names <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="cn">NA</span>, </span>
<span><span class="fl">6L</span><span class="op">)</span>, class <span class="op">=</span> <span class="st">"data.frame"</span><span class="op">)</span></span></code></pre>
</div>
<p>If the object is larger, provide either the raw file (i.e., your CSV
file) with your script up to the point of the error (and after removing
everything that is not relevant to your issue). Alternatively, in
particular if your question is not related to a data frame, you can save
any R object to a file[^export]:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">saveRDS</span><span class="op">(</span><span class="va">iris</span>, file<span class="op">=</span><span class="st">"/tmp/iris.rds"</span><span class="op">)</span></span></code></pre>
</div>
<p>The content of this file is however not human readable and cannot be
posted directly on Stack Overflow. Instead, it can be sent to someone by
email who can read it with the <code>readRDS()</code> command (here it
is assumed that the downloaded file is in a <code>Downloads</code>
folder in the user’s home directory):</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">some_data</span> <span class="op">&lt;-</span> <span class="fu">readRDS</span><span class="op">(</span>file<span class="op">=</span><span class="st">"~/Downloads/iris.rds"</span><span class="op">)</span></span></code></pre>
</div>
<p>Last, but certainly not least, <strong>always include the output of
<code>sessionInfo()</code></strong> as it provides critical information
about your platform, the versions of R and the packages that you are
using, and other information that can be very helpful to understand your
problem.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sessionInfo</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>R version 4.3.3 (2024-02-29)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.4 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0 
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   

time zone: UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
[1] BiocManager_1.30.22 compiler_4.3.3      tools_4.3.3        
[4] yaml_2.3.8          highr_0.10          knitr_1.45         
[7] xfun_0.43           renv_1.0.5          evaluate_0.23      </code></pre>
</div>
</div>
<div class="section level3">
<h3 id="where-to-ask-for-help">Where to ask for help?<a class="anchor" aria-label="anchor" href="#where-to-ask-for-help"></a></h3>
<ul><li>The person sitting next to you during the course. Don’t hesitate to
talk to your neighbour during the workshop, compare your answers, and
ask for help.</li>
<li>Your friendly colleagues: if you know someone with more experience
than you, they might be able and willing to help you.</li>
<li>
<a href="https://stackoverflow.com/questions/tagged/r" class="external-link">Stack
Overflow</a>: if your question hasn’t been answered before and is well
crafted, chances are you will get an answer in less than 5 min. Remember
to follow their guidelines on <a href="https://stackoverflow.com/help/how-to-ask" class="external-link">how to ask a good
question</a>.</li>
<li>The <a href="https://stat.ethz.ch/mailman/listinfo/r-help" class="external-link">R-help
mailing list</a>: it is read by a lot of people (including most of the R
core team), a lot of people post to it, but the tone can be pretty dry,
and it is not always very welcoming to new users. If your question is
valid, you are likely to get an answer very fast but don’t expect that
it will come with smiley faces. Also, here more than anywhere else, be
sure to use correct vocabulary (otherwise you might get an answer
pointing to the misuse of your words rather than answering your
question). You will also have more success if your question is about a
base function rather than a specific package.</li>
<li>If your question is about a specific package, see if there is a
mailing list for it. Usually it’s included in the DESCRIPTION file of
the package that can be accessed using
<code>packageDescription("name-of-package")</code>. You may also want to
try to email the author of the package directly, or open an issue on the
code repository (e.g., GitHub).</li>
<li>There are also some topic-specific mailing lists (GIS,
phylogenetics, etc…), the complete list is <a href="https://www.r-project.org/mail.html" class="external-link">here</a>.</li>
</ul></div>
<div class="section level3">
<h3 id="more-resources">More resources<a class="anchor" aria-label="anchor" href="#more-resources"></a></h3>
<ul><li><p>The <a href="https://www.r-project.org/posting-guide.html" class="external-link">Posting Guide</a>
for the R mailing lists.</p></li>
<li><p><a href="https://blog.revolutionanalytics.com/2014/01/how-to-ask-for-r-help.html" class="external-link">How
to ask for R help</a> useful guidelines.</p></li>
<li><p><a href="https://codeblog.jonskeet.uk/2010/08/29/writing-the-perfect-question/" class="external-link">This
blog post by Jon Skeet</a> has quite comprehensive advice on how to ask
programming questions.</p></li>
<li><p>The <a href="https://cran.rstudio.com/web/packages/reprex/" class="external-link">reprex</a> package
is very helpful to create reproducible examples when asking for help.
The rOpenSci community call “How to ask questions so they get answered”
(<a href="https://github.com/ropensci/commcalls/issues/14" class="external-link">Github
link</a> and <a href="https://vimeo.com/208749032" class="external-link">video recording</a>)
includes a presentation of the reprex package and of its
philosophy.</p></li>
</ul></div>
</section><section id="r-packages"><h2 class="section-heading">R packages<a class="anchor" aria-label="anchor" href="#r-packages"></a>
</h2>
<hr class="half-width"><div class="section level3">
<h3 id="loading-packages">Loading packages<a class="anchor" aria-label="anchor" href="#loading-packages"></a></h3>
<p>As we have seen above, R packages play a fundamental role in R. The
make use of a package’s functionality, assuming it is installed, we
first need to load it to be able to use it. This is done with the
<code>library()</code> function. Below, we load
<code>ggplot2</code>.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="st">"ggplot2"</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="installing-packages">Installing packages<a class="anchor" aria-label="anchor" href="#installing-packages"></a></h3>
<p>The default package repository is The <em>Comprehensive R Archive
Network</em> (CRAN), and any package that is available on CRAN can be
installed with the <code>install.packages()</code> function. Below, for
example, we install the <code>dplyr</code> package that we will learn
about later.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"dplyr"</span><span class="op">)</span></span></code></pre>
</div>
<p>This command will install the <code>dplyr</code> package as well as
all its dependencies, i.e. all the packages that it relies on to
function.</p>
<p>Another major R package repository is maintained by Bioconductor. <a href="https://bioconductor.org/packages/release/BiocViews.html#___Software" class="external-link">Bioconductor
packages</a> are managed and installed using a dedicated package, namely
<code>BiocManager</code>, that can be installed from CRAN with</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span></code></pre>
</div>
<p>Individual packages such as <code>SummarizedExperiment</code> (we
will use it later), <code>DESeq2</code> (for RNA-Seq analysis), and any
others from either Bioconductor or CRAN can then be installed with
<code>BiocManager::install</code>.</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"SummarizedExperiment"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"DESeq2"</span><span class="op">)</span></span></code></pre>
</div>
<p>By default, <code>BiocManager::install()</code> will also check all
your installed packages and see if there are newer versions available.
If there are, it will show them to you and ask you if you want to
<code>Update all/some/none? [a/s/n]:</code> and then wait for your
answer. While you should strive to have the most up-to-date package
versions, in practice we recommend only updating packages in a fresh R
session before any packages are loaded.</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul><li>Start using R and RStudio</li>
</ul></div>
</div>
</div>
</div>
</section><div class="footnotes footnotes-end-of-document">
<hr><ol><li id="fn1"><p>As opposed to using R directly from the command line
console. There exist other software that interface and integrate with R,
but RStudio is particularly well suited for beginners while providing
numerous very advanced features.<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>i.e. add-ons that confer R with new functionality, such
as bioinformatics data analysis.<a href="#fnref2" class="footnote-back">↩︎</a></p></li>
<li id="fn3"><p>In this course, we consider bioinformatics as data
science applied to biological or bio-medical data.<a href="#fnref3" class="footnote-back">↩︎</a></p></li>
<li id="fn4"><p>That someone could be, and very likely will be your
future self, a couple of months or years after the analyses were run.<a href="#fnref4" class="footnote-back">↩︎</a></p></li>
<li id="fn5"><p>We will introduce most of these (except statistics)
here, but will only manage to scratch the surface of the wealth of what
is possible to do with R.<a href="#fnref5" class="footnote-back">↩︎</a></p></li>
</ol></div>



      </div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="10-data-organisation.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="23-starting-with-r.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="10-data-organisation.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Data organisation
        </a>
        <a class="chapter-link float-end" href="23-starting-with-r.html" rel="next">
          Next: Introduction to R... 
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>
        
        <a href="https://github.com/carpentries-incubator/bioc-intro/edit/main/episodes/20-r-rstudio.Rmd" class="external-link">Edit on GitHub</a>
        
	
        | <a href="https://github.com/carpentries-incubator/bioc-intro/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/carpentries-incubator/bioc-intro/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries-incubator/bioc-intro/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:laurent.gatto@uclouvain.be">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">
        
        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>
        
        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.4" class="external-link">sandpaper (0.16.4)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.5" class="external-link">pegboard (0.7.5)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.2" class="external-link">varnish (1.0.2)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://carpentries-incubator.github.io/bioc-intro/20-r-rstudio.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "R and RStudio",
  "creativeWorkStatus": "active",
  "url": "https://carpentries-incubator.github.io/bioc-intro/20-r-rstudio.html",
  "identifier": "https://carpentries-incubator.github.io/bioc-intro/20-r-rstudio.html",
  "dateCreated": "2020-09-14",
  "dateModified": "2024-04-02",
  "datePublished": "2024-04-16"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo
    2022-11-07: we have gotten a notification that we have an overage for our
    tracking and I'm pretty sure this has to do with Workbench usage.
    Considering that I am not _currently_ using this tracking because I do not
    yet know how to access the data, I am turning this off for now.
  <script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
    _paq.push(["setDomains", ["*.preview.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
    _paq.push(["setDoNotTrack", true]);
    _paq.push(["disableCookies"]);
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
          var u="https://carpentries.matomo.cloud/";
          _paq.push(['setTrackerUrl', u+'matomo.php']);
          _paq.push(['setSiteId', '1']);
          var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
          g.async=true; g.src='https://cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
        })();
  </script>
  End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

