<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<title>Introduction to data analysis with R and Bioconductor: All in One View</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<script src="assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="assets/styles.css">
<script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="favicons/incubator/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicons/incubator/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="favicons/incubator/favicon-16x16.png">
<link rel="manifest" href="favicons/incubator/site.webmanifest">
<link rel="mask-icon" href="favicons/incubator/safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" media="(prefers-color-scheme: light)" content="white">
<meta name="theme-color" media="(prefers-color-scheme: dark)" content="black">
</head>
<body>
    <header id="top" class="navbar navbar-expand-md top-nav incubator"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="incubator-logo" alt="Carpentries Incubator" src="assets/images/incubator-logo.svg">
</div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text">
<li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul>
</li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1">
<li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/aio.html';">Instructor View</button></li>
        </ul>
</div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="assets/images/incubator-logo-sm.svg">
</div>
    <div class="lesson-title-md">
      Introduction to data analysis with R and Bioconductor
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0">
<li class="nav-item">
          <span class="lesson-title">
            Introduction to data analysis with R and Bioconductor
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown">
<li><a class="dropdown-item" href="discuss.html">Discussion</a></li>
<li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul>
</li>
      </ul>
</div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div>
<!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Introduction to data analysis with R and Bioconductor
</div>

<aside class="col-md-12 lesson-progress"><div style="width: %" class="percentage">
    %
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: %" aria-valuenow="" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text">
<li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul>
</li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/aio.html">Instructor View</a>
                      </div>
                    </div>
                  </div>
<!--/div.accordion-item-->
                </div>
<!--/div.accordion-flush-->
              </div>
<!--div.sidenav-view-selector -->
            </div>
<!--/div.col -->

            <hr>
</div>
<!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="10-data-organisation.html">1. Data organisation with spreadsheets</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="20-r-rstudio.html">2. R and RStudio</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="23-starting-with-r.html">3. Introduction to R</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="25-starting-with-data.html">4. Starting with data</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="30-dplyr.html">5. Manipulating and analysing data with dplyr</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="40-visualization.html">6. Data visualization</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="60-next-steps.html">7. Next steps</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width">
<div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul>
<li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="discuss.html">Discussion</a></li>
<li><a href="reference.html">Reference</a></li>
                    </ul>
</div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources">
<a href="aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none">
<div class="d-grid gap-1">

            </div>
          </div>
<!-- /div.accordion -->
        </div>
<!-- /div.sidebar-inner -->
      </nav>
</div>
<!-- /div.sidebar -->
  </div>
<!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-extra.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <main id="main-content" class="main-content"><div class="container lesson-content">
        
        
<section id="aio-10-data-organisation"><p>Content from <a href="10-data-organisation.html">Data organisation with spreadsheets</a></p>
<hr>
<p>Last updated on 2025-06-15 |

        <a href="https://github.com/carpentries-incubator/bioc-intro/edit/main/episodes/10-data-organisation.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How to organise tabular data?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Learn about spreadsheets, their strengths and weaknesses.</li>
<li>How do we format data in spreadsheets for effective data use?</li>
<li>Learn about common spreadsheet errors and how to correct them.</li>
<li>Organise your data according to tidy data principles.</li>
<li>Learn about text-based spreadsheet formats such as the
comma-separated (CSV) or tab-separated (TSV) formats.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<blockquote>
<p>This episode is based on the Data Carpentries’s <em>Data Analysis and
Visualisation in R for Ecologists</em> lesson.</p>
</blockquote>
<section><h2 class="section-heading" id="spreadsheet-programs">Spreadsheet programs<a class="anchor" aria-label="anchor" href="#spreadsheet-programs"></a>
</h2>
<hr class="half-width">
<p><strong>Question</strong></p>
<ul>
<li>What are basic principles for using spreadsheets for good data
organization?</li>
</ul>
<p><strong>Objective</strong></p>
<ul>
<li>Describe best practices for organizing data so computers can make
the best use of datasets.</li>
</ul>
<p><strong>Keypoint</strong></p>
<ul>
<li>Good data organization is the foundation of any research
project.</li>
</ul>
<p>Good data organization is the foundation of your research project.
Most researchers have data or do data entry in spreadsheets. Spreadsheet
programs are very useful graphical interfaces for designing data tables
and handling very basic data quality control functions. See also <span class="citation">@Broman:2018</span>.</p>
<div class="section level3">
<h3 id="spreadsheet-outline">Spreadsheet outline<a class="anchor" aria-label="anchor" href="#spreadsheet-outline"></a>
</h3>
<p>Spreadsheets are good for data entry. Therefore we have a lot of data
in spreadsheets. Much of your time as a researcher will be spent in this
‘data wrangling’ stage. It’s not the most fun, but it’s necessary. We’ll
teach you how to think about data organization and some practices for
more effective data wrangling.</p>
</div>
<div class="section level3">
<h3 id="what-this-lesson-will-not-teach-you">What this lesson will not teach you<a class="anchor" aria-label="anchor" href="#what-this-lesson-will-not-teach-you"></a>
</h3>
<ul>
<li>How to do <em>statistics</em> in a spreadsheet</li>
<li>How to do <em>plotting</em> in a spreadsheet</li>
<li>How to <em>write code</em> in spreadsheet programs</li>
</ul>
<p>If you’re looking to do this, a good reference is <a href="https://www.amazon.com/Head-First-Excel-learners-spreadsheets/dp/0596807694/" class="external-link">Head
First Excel</a>, published by O’Reilly.</p>
</div>
<div class="section level3">
<h3 id="why-arent-we-teaching-data-analysis-in-spreadsheets">Why aren’t we teaching data analysis in spreadsheets<a class="anchor" aria-label="anchor" href="#why-arent-we-teaching-data-analysis-in-spreadsheets"></a>
</h3>
<ul>
<li><p>Data analysis in spreadsheets usually requires a lot of manual
work. If you want to change a parameter or run an analysis with a new
dataset, you usually have to redo everything by hand. (We do know that
you can create macros, but see the next point.)</p></li>
<li><p>It is also difficult to track or reproduce statistical or
plotting analyses done in spreadsheet programs when you want to go back
to your work or someone asks for details of your analysis.</p></li>
</ul>
<p>Many spreadsheet programs are available. Since most participants
utilise Excel as their primary spreadsheet program, this lesson will
make use of Excel examples. A free spreadsheet program that can also be
used is LibreOffice. Commands may differ a bit between programs, but the
general idea is the same.</p>
<p>Spreadsheet programs encompass a lot of the things we need to be able
to do as researchers. We can use them for:</p>
<ul>
<li>Data entry</li>
<li>Organizing data</li>
<li>Subsetting and sorting data</li>
<li>Statistics</li>
<li>Plotting</li>
</ul>
<p>Spreadsheet programs use tables to represent and display data. Data
formatted as tables is also the main theme of this chapter, and we will
see how to organise data into tables in a standardised way to ensure
efficient downstream analysis.</p>
<div id="challenge-discuss-the-following-points-with-your-neighbour" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="challenge-discuss-the-following-points-with-your-neighbour" class="callout-inner">
<h3 class="callout-title">Challenge: Discuss the following points with your neighbour</h3>
<div class="callout-content">
<ul>
<li>Have you used spreadsheets, in your research, courses, or at
home?</li>
<li>What kind of operations do you do in spreadsheets?</li>
<li>Which ones do you think spreadsheets are good for?</li>
<li>Have you accidentally done something in a spreadsheet program that
made you frustrated or sad?</li>
</ul>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="problems-with-spreadsheets">Problems with spreadsheets<a class="anchor" aria-label="anchor" href="#problems-with-spreadsheets"></a>
</h3>
<p>Spreadsheets are good for data entry, but in reality we tend to use
spreadsheet programs for much more than data entry. We use them to
create data tables for publications, to generate summary statistics, and
make figures.</p>
<p>Generating tables for publications in a spreadsheet is not optimal -
often, when formatting a data table for publication, we’re reporting key
summary statistics in a way that is not really meant to be read as data,
and often involves special formatting (merging cells, creating borders,
making it pretty). We advise you to do this sort of operation within
your document editing software.</p>
<p>The latter two applications, generating statistics and figures,
should be used with caution: because of the graphical, drag and drop
nature of spreadsheet programs, it can be very difficult, if not
impossible, to replicate your steps (much less retrace anyone else’s),
particularly if your stats or figures require you to do more complex
calculations. Furthermore, in doing calculations in a spreadsheet, it’s
easy to accidentally apply a slightly different formula to multiple
adjacent cells. When using a command-line based statistics program like
R or SAS, it’s practically impossible to apply a calculation to one
observation in your dataset but not another unless you’re doing it on
purpose.</p>
</div>
<div class="section level3">
<h3 id="using-spreadsheets-for-data-entry-and-cleaning">Using spreadsheets for data entry and cleaning<a class="anchor" aria-label="anchor" href="#using-spreadsheets-for-data-entry-and-cleaning"></a>
</h3>
<p>In this lesson, we will assume that you are most likely using Excel
as your primary spreadsheet program - there are others (gnumeric, Calc
from OpenOffice), and their functionality is similar, but Excel seems to
be the program most used by biologists and biomedical researchers.</p>
<p>In this lesson we’re going to talk about:</p>
<ol style="list-style-type: decimal">
<li>Formatting data tables in spreadsheets</li>
<li>Formatting problems</li>
<li>Exporting data</li>
</ol>
</div>
</section><section><h2 class="section-heading" id="formatting-data-tables-in-spreadsheets">Formatting data tables in spreadsheets<a class="anchor" aria-label="anchor" href="#formatting-data-tables-in-spreadsheets"></a>
</h2>
<hr class="half-width">
<p><strong>Questions</strong></p>
<ul>
<li>How do we format data in spreadsheets for effective data use?</li>
</ul>
<p><strong>Objectives</strong></p>
<ul>
<li><p>Describe best practices for data entry and formatting in
spreadsheets.</p></li>
<li><p>Apply best practices to arrange variables and observations in a
spreadsheet.</p></li>
</ul>
<p><strong>Keypoints</strong></p>
<ul>
<li><p>Never modify your raw data. Always make a copy before making any
changes.</p></li>
<li><p>Keep track of all of the steps you take to clean your data in a
plain text file.</p></li>
<li><p>Organise your data according to tidy data principles.</p></li>
</ul>
<p>The most common mistake made is treating spreadsheet programs like
lab notebooks, that is, relying on context, notes in the margin, spatial
layout of data and fields to convey information. As humans, we can
(usually) interpret these things, but computers don’t view information
the same way, and unless we explain to the computer what every single
thing means (and that can be hard!), it will not be able to see how our
data fits together.</p>
<p>Using the power of computers, we can manage and analyse data in much
more effective and faster ways, but to use that power, we have to set up
our data for the computer to be able to understand it (and computers are
very literal).</p>
<p>This is why it’s extremely important to set up well-formatted tables
from the outset - before you even start entering data from your very
first preliminary experiment. Data organization is the foundation of
your research project. It can make it easier or harder to work with your
data throughout your analysis, so it’s worth thinking about when you’re
doing your data entry or setting up your experiment. You can set things
up in different ways in spreadsheets, but some of these choices can
limit your ability to work with the data in other programs or have the
you-of-6-months-from-now or your collaborator work with the data.</p>
<p><strong>Note:</strong> the best layouts/formats (as well as software
and interfaces) for data entry and data analysis might be different. It
is important to take this into account, and ideally automate the
conversion from one to another.</p>
<div class="section level3">
<h3 id="keeping-track-of-your-analyses">Keeping track of your analyses<a class="anchor" aria-label="anchor" href="#keeping-track-of-your-analyses"></a>
</h3>
<p>When you’re working with spreadsheets, during data clean up or
analyses, it’s very easy to end up with a spreadsheet that looks very
different from the one you started with. In order to be able to
reproduce your analyses or figure out what you did when a reviewer or
instructor asks for a different analysis, you should</p>
<ul>
<li><p>create a new file with your cleaned or analysed data. Don’t
modify the original dataset, or you will never know where you
started!</p></li>
<li><p>keep track of the steps you took in your clean up or analysis.
You should track these steps as you would any step in an experiment. We
recommend that you do this in a plain text file stored in the same
folder as the data file.</p></li>
</ul>
<p>This might be an example of a spreadsheet setup:</p>
<figure style="text-align: center"><img src="fig/spreadsheet-setup-updated.png" alt="SCreenshot of a spreadsheet and a README.txt file." width="70%" class="figure mx-auto d-block"><figcaption>
Spreadsheet example setip
</figcaption></figure><p>Put these principles into practice today during your exercises.</p>
<p>While versioning is out of scope for this course, you can look at the
Carpentries lesson on <a href="https://swcarpentry.github.io/git-novice/" class="external-link">‘Git’</a> to learn how
to maintain <strong>version control</strong> over your data. See also
this <a href="https://lgatto.github.io/github-intro/" class="external-link">blog post</a> for
a quick tutorial or <span class="citation">@Perez-Riverol:2016</span>
for a more research-oriented use-case.</p>
</div>
<div class="section level3">
<h3 id="structuring-data-in-spreadsheets">Structuring data in spreadsheets<a class="anchor" aria-label="anchor" href="#structuring-data-in-spreadsheets"></a>
</h3>
<p>The cardinal rules of using spreadsheet programs for data:</p>
<ol style="list-style-type: decimal">
<li>Put all your variables in columns - the thing you’re measuring, like
‘weight’ or ‘temperature’.</li>
<li>Put each observation in its own row.</li>
<li>Don’t combine multiple pieces of information in one cell. Sometimes
it just seems like one thing, but think if that’s the only way you’ll
want to be able to use or sort that data.</li>
<li>Leave the raw data raw - don’t change it!</li>
<li>Export the cleaned data to a text-based format like CSV
(comma-separated values) format. This ensures that anyone can use the
data, and is required by most data repositories.</li>
</ol>
<p>For instance, we have data from patients that visited several
hospitals in Brussels, Belgium. They recorded the date of the visit, the
hospital, the patients’ gender, weight and blood group.</p>
<p>If we were to keep track of the data like this:</p>
<figure style="text-align: center"><img src="fig/multiple-info.png" alt="A table showing the blood and rhesus types encoded together in one column." width="70%" class="figure mx-auto d-block"><figcaption>
Multiple variable in a single column.
</figcaption></figure><p>the problem is that the ABO and Rhesus groups are in the same
<code>Blood</code> type column. So, if they wanted to look at all
observations of the A group or look at weight distributions by ABO
group, it would be tricky to do this using this data setup. If instead
we put the ABO and Rhesus groups in different columns, you can see that
it would be much easier.</p>
<figure style="text-align: center"><img src="fig/single-info.png" alt="A table showing the blood and rhesus types encoded together in different columns." width="70%" class="figure mx-auto d-block"><figcaption>
Variables encoded in a single column each.
</figcaption></figure><p>An important rule when setting up a datasheet, is that
<strong>columns are used for variables</strong> and <strong>rows are
used for observations</strong>:</p>
<ul>
<li>columns are variables</li>
<li>rows are observations</li>
<li>cells are individual values</li>
</ul>
<div id="challenge-were-going-to-take-a-messy-dataset-and-describe-how-we-would-clean-it-up." class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="challenge-were-going-to-take-a-messy-dataset-and-describe-how-we-would-clean-it-up." class="callout-inner">
<h3 class="callout-title">Challenge: We’re going to take a messy dataset and describe how we would clean it up.</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li><p>Download a messy dataset by <a href="https://github.com/UCLouvain-CBIO/WSBIM1207/raw/master/data/messy_covid.xlsx" class="external-link">clicking
this link</a>.</p></li>
<li><p>Open up the data in a spreadsheet program.</p></li>
<li><p>You can see that there are two tabs. The data contains various
clinical variables recorded in various hospitals in Brussels during the
first and second COVID-19 waves in 2020. As you can see, the data have
been recorded differently during the March and November waves. Now
you’re the person in charge of this project and you want to be able to
start analyzing the data.</p></li>
<li><p>With the person next to you, identify what is wrong with this
spreadsheet. Also discuss the steps you would need to take to clean up
first and second wave tabs, and to put them all together in one
spreadsheet.</p></li>
</ol>
<p><strong>Important:</strong> Do not forget our first piece of advice:
to create a new file (or tab) for the cleaned data, never modify your
original (raw) data.</p>
</div>
</div>
</div>
<p>After you go through this exercise, we’ll discuss as a group what was
wrong with this data and how you would fix it.</p>
<!-- - Take about 10 minutes to work on this exercise. -->
<!-- - All the mistakes in the *common mistakes* section below are present -->
<!--   in the messy dataset. If the exercise is done during a workshop, ask -->
<!--   people what they saw as wrong with the data. As they bring up -->
<!--   different points, you can refer to the common mistakes or expand a -->
<!--   bit on the point they brought up. -->
<!-- - If you get a response where they've fixed the date, you can pause -->
<!--   and go to the dates lesson. Or you can say you'll come back to dates -->
<!--   at the end. -->
<div id="challenge-once-you-have-tidied-up-the-data-answer-the-following-questions" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="challenge-once-you-have-tidied-up-the-data-answer-the-following-questions" class="callout-inner">
<h3 class="callout-title">Challenge: Once you have tidied up the data, answer the following questions:</h3>
<div class="callout-content">
<ul>
<li>How many men and women took part in the study?</li>
<li>How many A, AB, and B types have been tested?</li>
<li>As above, but disregarding the contaminated samples?</li>
<li>How many Rhesus + and - have been tested?</li>
<li>How many universal donors (O-) have been tested?</li>
<li>What is the average weight of AB men?</li>
<li>How many samples have been tested in the different hospitals?</li>
</ul>
</div>
</div>
</div>
<p>An <strong>excellent reference</strong>, in particular with regard to
R scripting is the <em>Tidy Data</em> paper <span class="citation">@Wickham:2014</span>.</p>
</div>
</section><section><h2 class="section-heading" id="common-spreadsheet-errors">Common spreadsheet errors<a class="anchor" aria-label="anchor" href="#common-spreadsheet-errors"></a>
</h2>
<hr class="half-width">
<p><strong>Questions</strong></p>
<ul>
<li>What are some common challenges with formatting data in spreadsheets
and how can we avoid them?</li>
</ul>
<p><strong>Objectives</strong></p>
<ul>
<li>Recognise and resolve common spreadsheet formatting problems.</li>
</ul>
<p><strong>Keypoints</strong></p>
<ul>
<li>Avoid using multiple tables within one spreadsheet.</li>
<li>Avoid spreading data across multiple tabs.</li>
<li>Record zeros as zeros.</li>
<li>Use an appropriate null value to record missing data.</li>
<li>Don’t use formatting to convey information or to make your
spreadsheet look pretty.</li>
<li>Place comments in a separate column.</li>
<li>Record units in column headers.</li>
<li>Include only one piece of information in a cell.</li>
<li>Avoid spaces, numbers and special characters in column headers.</li>
<li>Avoid special characters in your data.</li>
<li>Record metadata in a separate plain text file.</li>
</ul>
<!-- This lesson is meant to be used as a reference for discussion as --><!-- learners identify issues with the messy dataset discussed in the --><!-- previous lesson. Instructors: don't go through this lesson except to --><!-- refer to responses to the exercise in the previous lesson. --><p>There are a few potential errors to be on the lookout for in your own
data as well as data from collaborators or the Internet. If you are
aware of the errors and the possible negative effect on downstream data
analysis and result interpretation, it might motivate yourself and your
project members to try and avoid them. Making small changes to the way
you format your data in spreadsheets, can have a great impact on
efficiency and reliability when it comes to data cleaning and
analysis.</p>
<ul>
<li><a href="#tables">Using multiple tables</a></li>
<li><a href="#tabs">Using multiple tabs</a></li>
<li><a href="#zeros">Not filling in zeros</a></li>
<li><a href="#null">Using problematic null values</a></li>
<li><a href="#formatting">Using formatting to convey
information</a></li>
<li><a href="#formattingpretty">Using formatting to make the data sheet
look pretty</a></li>
<li><a href="#units">Placing comments or units in cells</a></li>
<li><a href="#info">Entering more than one piece of information in a
cell</a></li>
<li><a href="#fieldname">Using problematic field names</a></li>
<li><a href="#special">Using special characters in data</a></li>
<li><a href="#metadata">Inclusion of metadata in data table</a></li>
</ul>
<div class="section level3">
<h3 id="tables">Using multiple tables<a class="anchor" aria-label="anchor" href="#tables"></a>
</h3>
<p>A common strategy is creating multiple data tables within one
spreadsheet. This confuses the computer, so don’t do this! When you
create multiple tables within one spreadsheet, you’re drawing false
associations between things for the computer, which sees each row as an
observation. You’re also potentially using the same field name in
multiple places, which will make it harder to clean your data up into a
usable form. The example below depicts the problem:</p>
<figure style="text-align: center"><img src="fig/2_datasheet_example.jpg" alt="A sheet with many subtable." width="70%" class="figure mx-auto d-block"><figcaption>
A sheet with many subtable.
</figcaption></figure><p>In the example above, the computer will see (for example) row 4 and
assume that all columns A-AF refer to the same sample. This row actually
represents four distinct samples (sample 1 for each of four different
collection dates - May 29th, June 12th, June 19th, and June 26th), as
well as some calculated summary statistics (an average (avr) and
standard error of measurement (SEM)) for two of those samples. Other
rows are similarly problematic.</p>
</div>
<div class="section level3">
<h3 id="tabs">Using multiple tabs<a class="anchor" aria-label="anchor" href="#tabs"></a>
</h3>
<p>But what about workbook tabs? That seems like an easy way to organise
data, right? Well, yes and no. When you create extra tabs, you fail to
allow the computer to see connections in the data that are there (you
have to introduce spreadsheet application-specific functions or
scripting to ensure this connection). Say, for instance, you make a
separate tab for each day you take a measurement.</p>
<p>This isn’t good practice for two reasons:</p>
<ol style="list-style-type: decimal">
<li><p>you are more likely to accidentally add inconsistencies to your
data if each time you take a measurement, you start recording data in a
new tab, and</p></li>
<li><p>even if you manage to prevent all inconsistencies from creeping
in, you will add an extra step for yourself before you analyse the data
because you will have to combine these data into a single datatable. You
will have to explicitly tell the computer how to combine tabs - and if
the tabs are inconsistently formatted, you might even have to do it
manually.</p></li>
</ol>
<p>The next time you’re entering data, and you go to create another tab
or table, ask yourself if you could avoid adding this tab by adding
another column to your original spreadsheet. We used multiple tabs in
our example of a messy data file, but now you’ve seen how you can
reorganise your data to consolidate across tabs.</p>
<p>Your data sheet might get very long over the course of the
experiment. This makes it harder to enter data if you can’t see your
headers at the top of the spreadsheet. But don’t repeat your header row.
These can easily get mixed into the data, leading to problems down the
road. Instead you can <a href="https://support.office.com/en-ca/article/Freeze-column-headings-for-easy-scrolling-57ccce0c-cf85-4725-9579-c5d13106ca6a" class="external-link">freeze
the column headers</a> so that they remain visible even when you have a
spreadsheet with many rows.</p>
</div>
<div class="section level3">
<h3 id="zeros">Not filling in zeros<a class="anchor" aria-label="anchor" href="#zeros"></a>
</h3>
<p>It might be that when you’re measuring something, it’s usually a
zero, say the number of times a rabbit is observed in the survey. Why
bother writing in the number zero in that column, when it’s mostly
zeros?</p>
<p>However, there’s a difference between a zero and a blank cell in a
spreadsheet. To the computer, a zero is actually data. You measured or
counted it. A blank cell means that it wasn’t measured and the computer
will interpret it as an unknown value (also known as a null or missing
value).</p>
<p>The spreadsheets or statistical programs will likely misinterpret
blank cells that you intend to be zeros. By not entering the value of
your observation, you are telling your computer to represent that data
as unknown or missing (null). This can cause problems with subsequent
calculations or analyses. For example, the average of a set of numbers
which includes a single null value is always null (because the computer
can’t guess the value of the missing observations). Because of this,
it’s very important to record zeros as zeros and truly missing data as
nulls.</p>
</div>
<div class="section level3">
<h3 id="null">Using problematic null values<a class="anchor" aria-label="anchor" href="#null"></a>
</h3>
<p><strong>Example</strong>: using -999 or other numerical values (or
zero) to represent missing data.</p>
<p><strong>Solutions</strong>:</p>
<p>There are a few reasons why null values get represented differently
within a dataset. Sometimes confusing null values are automatically
recorded from the measuring device. If that’s the case, there’s not much
you can do, but it can be addressed in data cleaning with a tool like <a href="https://www.datacarpentry.org/OpenRefine-ecology-lesson/" class="external-link">OpenRefine</a>
before analysis. Other times different null values are used to convey
different reasons why the data isn’t there. This is important
information to capture, but is in effect using one column to capture two
pieces of information. Like for <a href="#formatting">using formatting
to convey information</a> it would be good here to create a new column
like ‘data_missing’ and use that column to capture the different
reasons.</p>
<p>Whatever the reason, it’s a problem if unknown or missing data is
recorded as -999, 999, or 0.</p>
<p>Many statistical programs will not recognise that these are intended
to represent missing (null) values. How these values are interpreted
will depend on the software you use to analyse your data. It is
essential to use a clearly defined and consistent null indicator.</p>
<p>Blanks (most applications) and NA (for R) are good choices. <span class="citation">@White:2013</span> explain good choices for indicating
null values for different software applications in their article:</p>
<figure style="text-align: center"><img src="fig/3_white_table_1.jpg" alt="A table extracted from a paper showing commonly used null values." width="70%" class="figure mx-auto d-block"><figcaption>
Commonly used null values.
</figcaption></figure>
</div>
<div class="section level3">
<h3 id="formatting">Using formatting to convey information<a class="anchor" aria-label="anchor" href="#formatting"></a>
</h3>
<p><strong>Example</strong>: highlighting cells, rows or columns that
should be excluded from an analysis, leaving blank rows to indicate
separations in data.</p>
<figure style="text-align: center"><img src="fig/formatting.png" alt="A table showing several cells highlighted in yellow used to encode sample contamination." width="70%" class="figure mx-auto d-block"><figcaption>
Colours used to ecode information.
</figcaption></figure><p><strong>Solution</strong>: create a new field to encode which data
should be excluded.</p>
<figure style="text-align: center"><img src="fig/good_formatting.png" alt="Same table as above with a new variable encoding contambination." width="70%" class="figure mx-auto d-block"><figcaption>
A new variable encoding sample contamination
</figcaption></figure>
</div>
<div class="section level3">
<h3 id="formattingpretty">Using formatting to make the data sheet look pretty<a class="anchor" aria-label="anchor" href="#formattingpretty"></a>
</h3>
<p><strong>Example</strong>: merging cells.</p>
<p><strong>Solution</strong>: If you’re not careful, formatting a
worksheet to be more aesthetically pleasing can compromise your
computer’s ability to see associations in the data. Merged cells will
make your data unreadable by statistics software. Consider restructuring
your data in such a way that you will not need to merge cells to
organise your data.</p>
</div>
<div class="section level3">
<h3 id="units">Placing comments or units in cells<a class="anchor" aria-label="anchor" href="#units"></a>
</h3>
<p>Most analysis software can’t see Excel or LibreOffice comments, and
would be confused by comments placed within your data cells. As
described above for formatting, create another field if you need to add
notes to cells. Similarly, don’t include units in cells: ideally, all
the measurements you place in one column should be in the same unit, but
if for some reason they aren’t, create another field and specify the
units the cell is in.</p>
</div>
<div class="section level3">
<h3 id="info">Entering more than one piece of information in a cell<a class="anchor" aria-label="anchor" href="#info"></a>
</h3>
<p><strong>Example</strong>: Recording ABO and Rhesus groups in one
cell, such as A+, B+, A-, …</p>
<p><strong>Solution</strong>: Don’t include more than one piece of
information in a cell. This will limit the ways in which you can analyse
your data. If you need both these measurements, design your data sheet
to include this information. For example, include one column for the ABO
group and one for the Rhesus group.</p>
</div>
<div class="section level3">
<h3 id="fieldname">Using problematic field names<a class="anchor" aria-label="anchor" href="#fieldname"></a>
</h3>
<p>Choose descriptive field names, but be careful not to include spaces,
numbers, or special characters of any kind. Spaces can be misinterpreted
by parsers that use whitespace as delimiters and some programs don’t
like field names that are text strings that start with numbers.</p>
<p>Underscores (<code>_</code>) are a good alternative to spaces.
Consider writing names in camel case (like this: ExampleFileName) to
improve readability. Remember that abbreviations that make sense at the
moment may not be so obvious in 6 months, but don’t overdo it with names
that are excessively long. Including the units in the field names avoids
confusion and enables others to readily interpret your fields.</p>
<p><strong>Examples</strong></p>
<table class="table">
<thead><tr class="header">
<th>Good Name</th>
<th>Good Alternative</th>
<th>Avoid</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Max_temp_C</td>
<td>MaxTemp</td>
<td>Maximum Temp (°C)</td>
</tr>
<tr class="even">
<td>Precipitation_mm</td>
<td>Precipitation</td>
<td>precmm</td>
</tr>
<tr class="odd">
<td>Mean_year_growth</td>
<td>MeanYearGrowth</td>
<td>Mean growth/year</td>
</tr>
<tr class="even">
<td>sex</td>
<td>sex</td>
<td>M/F</td>
</tr>
<tr class="odd">
<td>weight</td>
<td>weight</td>
<td>w.</td>
</tr>
<tr class="even">
<td>cell_type</td>
<td>CellType</td>
<td>Cell Type</td>
</tr>
<tr class="odd">
<td>Observation_01</td>
<td>first_observation</td>
<td>1st Obs</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="special">Using special characters in data<a class="anchor" aria-label="anchor" href="#special"></a>
</h3>
<p><strong>Example</strong>: You treat your spreadsheet program as a
word processor when writing notes, for example copying data directly
from Word or other applications.</p>
<p><strong>Solution</strong>: This is a common strategy. For example,
when writing longer text in a cell, people often include line breaks,
em-dashes, etc. in their spreadsheet. Also, when copying data in from
applications such as Word, formatting and fancy non-standard characters
(such as left- and right-aligned quotation marks) are included. When
exporting this data into a coding/statistical environment or into a
relational database, dangerous things may occur, such as lines being cut
in half and encoding errors being thrown.</p>
<p>General best practice is to avoid adding characters such as newlines,
tabs, and vertical tabs. In other words, treat a text cell as if it were
a simple web form that can only contain text and spaces.</p>
</div>
<div class="section level3">
<h3 id="metadata">Inclusion of metadata in data table<a class="anchor" aria-label="anchor" href="#metadata"></a>
</h3>
<p><strong>Example</strong>: You add a legend at the top or bottom of
your data table explaining column meaning, units, exceptions, etc.</p>
<p><strong>Solution</strong>: Recording data about your data
(“metadata”) is essential. You may be on intimate terms with your
dataset while you are collecting and analysing it, but the chances that
you will still remember that the variable “sglmemgp” means single member
of group, for example, or the exact algorithm you used to transform a
variable or create a derived one, after a few months, a year, or more
are slim.</p>
<p>As well, there are many reasons other people may want to examine or
use your data - to understand your findings, to verify your findings, to
review your submitted publication, to replicate your results, to design
a similar study, or even to archive your data for access and re-use by
others. While digital data by definition are machine-readable,
understanding their meaning is a job for human beings. The importance of
documenting your data during the collection and analysis phase of your
research cannot be overestimated, especially if your research is going
to be part of the scholarly record.</p>
<p>However, metadata should not be contained in the data file itself.
Unlike a table in a paper or a supplemental file, metadata (in the form
of legends) should not be included in a data file since this information
is not data, and including it can disrupt how computer programs
interpret your data file. Rather, metadata should be stored as a
separate file in the same directory as your data file, preferably in
plain text format with a name that clearly associates it with your data
file. Because metadata files are free text format, they also allow you
to encode comments, units, information about how null values are
encoded, etc. that are important to document but can disrupt the
formatting of your data file.</p>
<p>Additionally, file or database level metadata describes how files
that make up the dataset relate to each other; what format they are in;
and whether they supercede or are superceded by previous files. A
folder-level readme.txt file is the classic way of accounting for all
the files and folders in a project.</p>
<p>(Text on metadata adapted from the online course Research Data <a href="https://datalib.edina.ac.uk/mantra" class="external-link">MANTRA</a> by EDINA and Data
Library, University of Edinburgh. MANTRA is licensed under a <a href="https://creativecommons.org/licenses/by/4.0/" class="external-link">Creative Commons
Attribution 4.0 International License</a>.)</p>
</div>
</section><section><h2 class="section-heading" id="exporting-data">Exporting data<a class="anchor" aria-label="anchor" href="#exporting-data"></a>
</h2>
<hr class="half-width">
<p><strong>Question</strong></p>
<ul>
<li>How can we export data from spreadsheets in a way that is useful for
downstream applications?</li>
</ul>
<p><strong>Objectives</strong></p>
<ul>
<li>Store spreadsheet data in universal file formats.</li>
<li>Export data from a spreadsheet to a CSV file.</li>
</ul>
<p><strong>Keypoints</strong></p>
<ul>
<li><p>Data stored in common spreadsheet formats will often not be read
correctly into data analysis software, introducing errors into your
data.</p></li>
<li><p>Exporting data from spreadsheets to formats like CSV or TSV puts
it in a format that can be used consistently by most programs.</p></li>
</ul>
<p>Storing the data you’re going to work with for your analyses in Excel
default file format (<code>*.xls</code> or <code>*.xlsx</code> -
depending on the Excel version) isn’t a good idea. Why?</p>
<ul>
<li><p>Because it is a proprietary format, and it is possible that in
the future, technology won’t exist (or will become sufficiently rare) to
make it inconvenient, if not impossible, to open the file.</p></li>
<li><p>Other spreadsheet software may not be able to open files saved in
a proprietary Excel format.</p></li>
<li><p>Different versions of Excel may handle data differently, leading
to inconsistencies. <a href="https://datacarpentry.org/spreadsheet-ecology-lesson/03-dates-as-data/index.html" class="external-link">Dates</a>
is a well-documented example of inconsistencies in data
storage.</p></li>
<li><p>Finally, more journals and grant agencies are requiring you to
deposit your data in a data repository, and most of them don’t accept
Excel format. It needs to be in one of the formats discussed
below.</p></li>
<li><p>The above points also apply to other formats such as open data
formats used by LibreOffice / Open Office. These formats are not static
and do not get parsed the same way by different software
packages.</p></li>
</ul>
<p>Storing data in a universal, open, and static format will help deal
with this problem. Try tab-delimited (tab separated values or TSV) or
comma-delimited (comma separated values or CSV). CSV files are plain
text files where the columns are separated by commas, hence ‘comma
separated values’ or CSV. The advantage of a CSV file over an
Excel/SPSS/etc. file is that we can open and read a CSV file using just
about any software, including plain text editors like TextEdit or
NotePad. Data in a CSV file can also be easily imported into other
formats and environments, such as SQLite and R. We’re not tied to a
certain version of a certain expensive program when we work with CSV
files, so it’s a good format to work with for maximum portability and
endurance. Most spreadsheet programs can save to delimited text formats
like CSV easily, although they may give you a warning during the file
export.</p>
<p>To save a file you have opened in Excel in CSV format:</p>
<ol style="list-style-type: decimal">
<li>From the top menu select ‘File’ and ‘Save as’.</li>
<li>In the ‘Format’ field, from the list, select ‘Comma Separated
Values’ (<code>*.csv</code>).</li>
<li>Double check the file name and the location where you want to save
it and hit ‘Save’.</li>
</ol>
<p>An important note for backwards compatibility: you can open CSV files
in Excel!</p>
<figure style="text-align: center"><img src="fig/excel-to-csv.png" alt="Excel menu to export data in a csv format on MacOS " width="70%" class="figure mx-auto d-block"><figcaption>
Saving an Excel file to CSV.
</figcaption></figure><p><strong>A note on R and <code>xls</code></strong>: There are R
packages that can read <code>xls</code> files (as well as Google
spreadsheets). It is even possible to access different worksheets in the
<code>xls</code> documents.</p>
<p><strong>But</strong></p>
<ul>
<li>some of these only work on Windows.</li>
<li>this equates to replacing a (simple but manual) export to
<code>csv</code> with additional complexity/dependencies in the data
analysis R code.</li>
<li>data formatting best practice still apply.</li>
<li>Is there really a good reason why <code>csv</code> (or similar) is
not adequate?</li>
</ul>
<div class="section level3">
<h3 id="caveats-on-commas">Caveats on commas<a class="anchor" aria-label="anchor" href="#caveats-on-commas"></a>
</h3>
<p>In some datasets, the data values themselves may include commas (,).
In that case, the software which you use (including Excel) will most
likely incorrectly display the data in columns. This is because the
commas which are a part of the data values will be interpreted as
delimiters.</p>
<p>For example, our data might look like this:</p>
<pre><code>species_id,genus,species,taxa
AB,Amphispiza,bilineata,Bird
AH,Ammospermophilus,harrisi,Rodent, not censused
AS,Ammodramus,savannarum,Bird
BA,Baiomys,taylori,Rodent</code></pre>
<p>In the record
<code>AH,Ammospermophilus,harrisi,Rodent, not censused</code> the value
for <code>taxa</code> includes a comma
(<code>Rodent, not censused</code>). If we try to read the above into
Excel (or other spreadsheet program), we will get something like
this:</p>
<figure style="text-align: center"><img src="fig/csv-mistake.png" alt="A table illustrating ill-imported data that contained a comma in a cell. That cell has been split in two cells." width="80%" class="figure mx-auto d-block"><figcaption>
The risks of having commas inside comma-separated data.
</figcaption></figure><p>The value for <code>taxa</code> was split into two columns (instead
of being put in one column <code>D</code>). This can propagate to a
number of further errors. For example, the extra column will be
interpreted as a column with many missing values (and without a proper
header). In addition to that, the value in column <code>D</code> for the
record in row 3 (so the one where the value for ‘taxa’ contained the
comma) is now incorrect.</p>
<p>If you want to store your data in <code>csv</code> format and expect
that your data values may contain commas, you can avoid the problem
discussed above by putting the values in quotes (““). Applying this
rule, our data might look like this:</p>
<pre><code>species_id,genus,species,taxa
"AB","Amphispiza","bilineata","Bird"
"AH","Ammospermophilus","harrisi","Rodent, not censused"
"AS","Ammodramus","savannarum","Bird"
"BA","Baiomys","taylori","Rodent"</code></pre>
<p>Now opening this file as a <code>csv</code> in Excel will not lead to
an extra column, because Excel will only use commas that fall outside of
quotation marks as delimiting characters.</p>
<p>Alternatively, if you are working with data that contains commas, you
likely will need to use another delimiter when working in a
spreadsheet<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>. In this case, consider using tabs as your
delimiter and working with TSV files. TSV files can be exported from
spreadsheet programs in the same way as CSV files.</p>
<p>If you are working with an already existing dataset in which the data
values are not included in “” but which have commas as both delimiters
and parts of data values, you are potentially facing a major problem
with data cleaning. If the dataset you’re dealing with contains hundreds
or thousands of records, cleaning them up manually (by either removing
commas from the data values or putting the values into quotes - ““) is
not only going to take hours and hours but may potentially end up with
you accidentally introducing many errors.</p>
<p>Cleaning up datasets is one of the major problems in many scientific
disciplines. The approach almost always depends on the particular
context. However, it is a good practice to clean the data in an
automated fashion, for example by writing and running a script. The
Python and R lessons will give you the basis for developing skills to
build relevant scripts.</p>
</div>
</section><section><h2 class="section-heading" id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h2>
<hr class="half-width">
<figure style="text-align: center"><img src="fig/analysis.png" alt="The tidy data analysis circle, iterating over data transformation, visualisation and modelling." width="70%" class="figure mx-auto d-block"><figcaption>
A typical data analysis workflow.
</figcaption></figure><p>A typical data analysis workflow is illustrated in the figure above,
where data is repeatedly transformed, visualised, and modelled. This
iteration is repeated multiple times until the data is understood. In
many real-life cases, however, most time is spent cleaning up and
preparing the data, rather than actually analysing and understanding
it.</p>
<p>An agile data analysis workflow, with several fast iterations of the
transform/visualise/model cycle is only feasible if the data is
formatted in a predictable way and one can reason about the data without
having to look at it and/or fix it.</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>Good data organization is the foundation of any research
project.</li>
</ul>
</div>
</div>
</div>
</section><div class="footnotes footnotes-end-of-document">
<hr>
<ol>
<li id="fn1"><p>This is particularly relevant in European countries
where the comma is used as a decimal separator. In such cases, the
default value separator in a csv file will be the semi-colon (;), or
values will be systematically quoted.<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
</ol>
</div></section><section id="aio-20-r-rstudio"><p>Content from <a href="20-r-rstudio.html">R and RStudio</a></p>
<hr>
<p>Last updated on 2025-06-15 |

        <a href="https://github.com/carpentries-incubator/bioc-intro/edit/main/episodes/20-r-rstudio.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What are R and RStudio?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Describe the purpose of the RStudio Script, Console, Environment,
and Plots panes.</li>
<li>Organise files and directories for a set of analyses as an R
project, and understand the purpose of the working directory.</li>
<li>Use the built-in RStudio help interface to search for more
information on R functions.</li>
<li>Demonstrate how to provide sufficient information for
troubleshooting with the R user community.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<blockquote>
<p>This episode is based on the Data Carpentries’s <em>Data Analysis and
Visualisation in R for Ecologists</em> lesson.</p>
</blockquote>
<section><h2 class="section-heading" id="what-is-r-what-is-rstudio">What is R? What is RStudio?<a class="anchor" aria-label="anchor" href="#what-is-r-what-is-rstudio"></a>
</h2>
<hr class="half-width">
<p>The term <a href="https://www.r-project.org/" class="external-link">R</a> is used to refer
to the <em>programming language</em>, the <em>environment for
statistical computing</em> and <em>the software</em> that interprets the
scripts written using it.</p>
<p><a href="https://rstudio.com" class="external-link">RStudio</a> is currently a very popular
way to not only write your R scripts but also to interact with the R
software<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>. To function correctly, RStudio needs R and
therefore both need to be installed on your computer.</p>
<p>The <a href="https://raw.githubusercontent.com/rstudio/cheatsheets/main/rstudio-ide.pdf" class="external-link">RStudio
IDE Cheat Sheet</a> provides much more information than will be covered
here, but can be useful to learn keyboard shortcuts and discover new
features.</p>
</section><section><h2 class="section-heading" id="why-learn-r">Why learn R?<a class="anchor" aria-label="anchor" href="#why-learn-r"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="r-does-not-involve-lots-of-pointing-and-clicking-and-thats-a-good-thing">R does not involve lots of pointing and clicking, and that’s a good
thing<a class="anchor" aria-label="anchor" href="#r-does-not-involve-lots-of-pointing-and-clicking-and-thats-a-good-thing"></a>
</h3>
<p>The learning curve might be steeper than with other software, but
with R, the results of your analysis do not rely on remembering a
succession of pointing and clicking, but instead on a series of written
commands, and that’s a good thing! So, if you want to redo your analysis
because you collected more data, you don’t have to remember which button
you clicked in which order to obtain your results; you just have to run
your script again.</p>
<p>Working with scripts makes the steps you used in your analysis clear,
and the code you write can be inspected by someone else who can give you
feedback and spot mistakes.</p>
<p>Working with scripts forces you to have a deeper understanding of
what you are doing, and facilitates your learning and comprehension of
the methods you use.</p>
</div>
<div class="section level3">
<h3 id="r-code-is-great-for-reproducibility">R code is great for reproducibility<a class="anchor" aria-label="anchor" href="#r-code-is-great-for-reproducibility"></a>
</h3>
<p>Reproducibility means that someone else (including your future self)
can obtain the same results from the same dataset when using the same
analysis code.</p>
<p>R integrates with other tools to generate manuscripts or reports from
your code. If you collect more data, or fix a mistake in your dataset,
the figures and the statistical tests in your manuscript or report are
updated automatically.</p>
<p>An increasing number of journals and funding agencies expect analyses
to be reproducible, so knowing R will give you an edge with these
requirements.</p>
</div>
<div class="section level3">
<h3 id="r-is-interdisciplinary-and-extensible">R is interdisciplinary and extensible<a class="anchor" aria-label="anchor" href="#r-is-interdisciplinary-and-extensible"></a>
</h3>
<p>With 10000+ packages<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a> that can be installed to extend its
capabilities, R provides a framework that allows you to combine
statistical approaches from many scientific disciplines to best suit the
analytical framework you need to analyse your data. For instance, R has
packages for image analysis, GIS, time series, population genetics, and
a lot more.</p>
<figure style="text-align: center"><img src="fig/cran.png" alt="Two graphs showing the exponential increase of the number of CRAN packages from 2000 to 2018. The figure on the right shows linear trend for the log of number of packages." class="figure mx-auto d-block"><figcaption>
Exponential increase of the number of packages available on <a href="https://cran.r-project.org/" class="external-link">CRAN</a>, the Comprehensive R Archive
Network. From the R Journal, Volume 10/2, December 2018.
</figcaption></figure>
</div>
<div class="section level3">
<h3 id="r-works-on-data-of-all-shapes-and-sizes">R works on data of all shapes and sizes<a class="anchor" aria-label="anchor" href="#r-works-on-data-of-all-shapes-and-sizes"></a>
</h3>
<p>The skills you learn with R scale easily with the size of your
dataset. Whether your dataset has hundreds or millions of lines, it
won’t make much difference to you.</p>
<p>R is designed for data analysis. It comes with special data
structures and data types that make handling of missing data and
statistical factors convenient.</p>
<p>R can connect to spreadsheets, databases, and many other data
formats, on your computer or on the web.</p>
</div>
<div class="section level3">
<h3 id="r-produces-high-quality-graphics">R produces high-quality graphics<a class="anchor" aria-label="anchor" href="#r-produces-high-quality-graphics"></a>
</h3>
<p>The plotting functionalities in R are extensive, and allow you to
adjust any aspect of your graph to convey most effectively the message
from your data.</p>
</div>
<div class="section level3">
<h3 id="r-has-a-large-and-welcoming-community">R has a large and welcoming community<a class="anchor" aria-label="anchor" href="#r-has-a-large-and-welcoming-community"></a>
</h3>
<p>Thousands of people use R daily. Many of them are willing to help you
through mailing lists and websites such as <a href="https://stackoverflow.com/" class="external-link">Stack Overflow</a>, or on the <a href="https://community.rstudio.com/" class="external-link">RStudio community</a>. These broad
user communities extend to specialised areas such as bioinformatics. One
such subset of the R community is <a href="https://bioconductor.org/" class="external-link">Bioconductor</a>, a scientific project
for analysis and comprehension “of data from current and emerging
biological assays.” This workshop was developed by members of the
Bioconductor community; for more information on Bioconductor, please see
the companion workshop <a href="https://carpentries-incubator.github.io/bioc-project/" class="external-link">“The
Bioconductor Project”</a>.</p>
</div>
<div class="section level3">
<h3 id="not-only-is-r-free-but-it-is-also-open-source-and-cross-platform">Not only is R free, but it is also open-source and
cross-platform<a class="anchor" aria-label="anchor" href="#not-only-is-r-free-but-it-is-also-open-source-and-cross-platform"></a>
</h3>
<p>Anyone can inspect the source code to see how R works. Because of
this transparency, there is less chance for mistakes, and if you (or
someone else) find some, you can report and fix bugs.</p>
</div>
</section><section><h2 class="section-heading" id="knowing-your-way-around-rstudio">Knowing your way around RStudio<a class="anchor" aria-label="anchor" href="#knowing-your-way-around-rstudio"></a>
</h2>
<hr class="half-width">
<p>Let’s start by learning about <a href="https://www.rstudio.com/" class="external-link">RStudio</a>, which is an Integrated
Development Environment (IDE) for working with R.</p>
<p>The RStudio IDE open-source product is free under the <a href="https://www.gnu.org/licenses/agpl-3.0.en.html" class="external-link">Affero General
Public License (AGPL) v3</a>. The RStudio IDE is also available with a
commercial license and priority email support from Posit, Inc.</p>
<p>We will use the RStudio IDE to write code, navigate the files on our
computer, inspect the variables we are going to create, and visualise
the plots we will generate. RStudio can also be used for other things
(e.g., version control, developing packages, writing Shiny apps) that we
will not cover during the workshop.</p>
<figure style="text-align: center"><img src="fig/rstudio-screenshot.png" alt="Screenshot of the RStudio interface showing the typical four panel setup." width="100%" class="figure mx-auto d-block"><figcaption>
RStudio interface screenshot. Clockwise from top left: Source,
Environment/History, Files/Plots/Packages/Help/Viewer, Console.
</figcaption></figure><p>The RStudio window is divided into 4 “Panes”:</p>
<ul>
<li>the <strong>Source</strong> for your scripts and documents
(top-left, in the default layout)</li>
<li>your <strong>Environment/History</strong> (top-right),</li>
<li>your <strong>Files/Plots/Packages/Help/Viewer</strong>
(bottom-right), and</li>
<li>the R <strong>Console</strong> (bottom-left).</li>
</ul>
<p>The placement of these panes and their content can be customised (see
menu, <code>Tools -&gt; Global Options -&gt; Pane Layout</code>).</p>
<p>One of the advantages of using RStudio is that all the information
you need to write code is available in a single window. Additionally,
with many shortcuts, <strong>autocompletion</strong>, and
<strong>highlighting</strong> for the major file types you use while
developing in R, RStudio will make typing easier and less
error-prone.</p>
</section><section><h2 class="section-heading" id="getting-set-up">Getting set up<a class="anchor" aria-label="anchor" href="#getting-set-up"></a>
</h2>
<hr class="half-width">
<p>It is good practice to keep a set of related data, analyses, and text
self-contained in a single folder, called the <strong>working
directory</strong>. All of the scripts within this folder can then use
<strong>relative paths</strong> to files that indicate where inside the
project a file is located (as opposed to absolute paths, which point to
where a file is on a specific computer). Working this way makes it a lot
easier to move your project around on your computer and share it with
others without worrying about whether or not the underlying scripts will
still work.</p>
<p>RStudio provides a helpful set of tools to do this through its
“Projects” interface, which not only creates a working directory for
you, but also remembers its location (allowing you to quickly navigate
to it) and optionally preserves custom settings and open files to make
it easier to resume work after a break. Go through the steps for
creating an “R Project” for this tutorial below.</p>
<ol style="list-style-type: decimal">
<li>Start RStudio.</li>
<li>Under the <code>File</code> menu, click on <code>New project</code>.
Choose <code>New directory</code>, then <code>New project</code>.</li>
<li>Enter a name for this new folder (or “directory”), and choose a
convenient location for it. This will be your <strong>working
directory</strong> for this session (or whole course) (e.g.,
<code>bioc-intro</code>).</li>
<li>Click on <code>Create project</code>.</li>
<li>(Optional) Set Preferences to ‘Never’ save workspace in
RStudio.</li>
</ol>
<p>RStudio’s default preferences generally work well, but saving a
workspace to .RData can be cumbersome, especially if you are working
with larger datasets. To turn that off, go to Tools –&gt; ‘Global
Options’ and select the ‘Never’ option for ‘Save workspace to .RData’ on
exit.</p>
<figure style="text-align: center"><img src="fig/rstudio-preferences.png" alt="Screenshot of the RStudio General Options panel." width="70%" class="figure mx-auto d-block"><figcaption>
Set ‘Save workspace to .RData on exit’ to ‘Never’
</figcaption></figure><p>To avoid <a href="https://yihui.name/en/2018/11/biggest-regret-knitr/" class="external-link">character
encoding issues between Windows and other operating systems</a>, we are
going to set UTF-8 by default:</p>
<figure style="text-align: center"><img src="fig/utf8.png" alt="Screenshot of the RStudio Code Options panel." width="70%" class="figure mx-auto d-block"><figcaption>
Set the default text encoding to UTF-8 to save us headache in the coming
future. (Figure from the link above).
</figcaption></figure><div class="section level3">
<h3 id="organizing-your-working-directory">Organizing your working directory<a class="anchor" aria-label="anchor" href="#organizing-your-working-directory"></a>
</h3>
<p>Using a consistent folder structure across your projects will help
keep things organised, and will also make it easy to find/file things in
the future. This can be especially helpful when you have multiple
projects. In general, you may create directories (folders) for
<strong>scripts</strong>, <strong>data</strong>, and
<strong>documents</strong>.</p>
<ul>
<li>
<strong><code>data/</code></strong> Use this folder to store your
raw data and intermediate datasets you may create for the need of a
particular analysis. For the sake of transparency and <a href="https://en.wikipedia.org/wiki/Provenance" class="external-link">provenance</a>, you
should <em>always</em> keep a copy of your raw data accessible and do as
much of your data cleanup and preprocessing programmatically (i.e., with
scripts, rather than manually) as possible. Separating raw data from
processed data is also a good idea. For example, you could have files
<code>data/raw/tree_survey.plot1.txt</code> and
<code>...plot2.txt</code> kept separate from a
<code>data/processed/tree.survey.csv</code> file generated by the
<code>scripts/01.preprocess.tree_survey.R</code> script.</li>
<li>
<strong><code>documents/</code></strong> This would be a place to
keep outlines, drafts, and other text.</li>
<li>
<strong><code>scripts/</code></strong> (or <code>src</code>) This
would be the location to keep your R scripts for different analyses or
plotting, and potentially a separate folder for your functions (more on
that later).</li>
</ul>
<p>You may want additional directories or subdirectories depending on
your project needs, but these should form the backbone of your working
directory.</p>
<figure style="text-align: center"><img src="fig/working-directory-structure.png" alt="Flow diagram showing the structure of the working directory and various directory of a typical RStuio project." width="100%" class="figure mx-auto d-block"><figcaption>
Example of a working directory structure.
</figcaption></figure><p>For this course, we will need a <code>data/</code> folder to store
our raw data, and we will use <code>data_output/</code> for when we
learn how to export data as CSV files, and <code>fig_output/</code>
folder for the figures that we will save.</p>
<div id="challenge-create-your-project-directory-structure" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="challenge-create-your-project-directory-structure" class="callout-inner">
<h3 class="callout-title">Challenge: create your project directory structure</h3>
<div class="callout-content">
<p>Under the <code>Files</code> tab on the right of the screen, click on
<code>New Folder</code> and create a folder named <code>data</code>
within your newly created working directory (e.g.,
<code>~/bioc-intro/data</code>). (Alternatively, type
<code>dir.create("data")</code> at your R console.) Repeat these
operations to create a <code>data_output/</code> and a
<code>fig_output</code> folders.</p>
</div>
</div>
</div>
<p>We are going to keep the script in the root of our working directory
because we are only going to use one file and it will make things
easier.</p>
<p>Your working directory should now look like this:</p>
<figure style="text-align: center"><img src="fig/r-starting-how-it-should-look-like.png" alt="Screenshot of RStudio's Files panel." width="100%" class="figure mx-auto d-block"><figcaption>
How it should look like at the beginning of this lesson
</figcaption></figure><p><strong>Project management</strong> is also applicable to
bioinformatics projects, of course<a href="#fn3" class="footnote-ref" id="fnref3"><sup>3</sup></a>. William Noble (<span class="citation">@Noble:2009</span>) proposes the following directory
structure:</p>
<blockquote>
<p>Directory names are in large typeface, and filenames are in smaller
typeface. Only a subset of the files are shown here. Note that the dates
are formatted <code>&lt;year&gt;-&lt;month&gt;-&lt;day&gt;</code> so
that they can be sorted in chronological order. The source code
<code>src/ms-analysis.c</code> is compiled to create
<code>bin/ms-analysis</code> and is documented in
<code>doc/ms-analysis.html</code>. The <code>README</code> files in the
data directories specify who downloaded the data files from what URL on
what date. The driver script <code>results/2009-01-15/runall</code>
automatically generates the three subdirectories split1, split2, and
split3, corresponding to three cross-validation splits. The
<code>bin/parse-sqt.py</code> script is called by both of the
<code>runall</code> driver scripts.</p>
</blockquote>
<figure style="text-align: center"><img src="fig/noble-bioinfo-project.png" alt="Figure form Nobel et al. (2009) showing a complex bioinformatics project structure." width="100%" class="figure mx-auto d-block"><figcaption>
Directory structure for a sample bioinformatics project.
</figcaption></figure><p>The most important aspect of a well defined and well documented
project directory is to enable someone unfamiliar with the project<a href="#fn4" class="footnote-ref" id="fnref4"><sup>4</sup></a> to</p>
<ol style="list-style-type: decimal">
<li><p>understand what the project is about, what data are available,
what analyses were run, and what results were produced and, most
importantly to</p></li>
<li><p>repeat the analysis over again - with new data, or changing some
analysis parameters.</p></li>
</ol>
</div>
<div class="section level3">
<h3 id="the-working-directory">The working directory<a class="anchor" aria-label="anchor" href="#the-working-directory"></a>
</h3>
<p>The working directory is an important concept to understand. It is
the place from where R will be looking for and saving the files. When
you write code for your project, it should refer to files in relation to
the root of your working directory and only need files within this
structure.</p>
<p>Using RStudio projects makes this easy and ensures that your working
directory is set properly. If you need to check it, you can use
<code>getwd()</code>. If for some reason your working directory is not
what it should be, you can change it in the RStudio interface by
navigating in the file browser where your working directory should be,
and clicking on the blue gear icon <code>More</code>, and select
<code>Set As Working Directory</code>. Alternatively you can use
<code>setwd("/path/to/working/directory")</code> to reset your working
directory. However, your scripts should not include this line because it
will fail on someone else’s computer.</p>
<p><strong>Example</strong></p>
<p>The schema below represents the working directory
<code>bioc-intro</code> with the <code>data</code> and
<code>fig_output</code> sub-directories, and 2 files in the latter:</p>
<pre><code>bioc-intro/data/
          /fig_output/fig1.pdf
          /fig_output/fig2.png</code></pre>
<p>If we were in the working directory, we could refer to the
<code>fig1.pdf</code> file using the relative path
<code>bioc-intro/fig_output/fig1.pdf</code> or the absolute path
<code>/home/user/bioc-intro/fig_output/fig1.pdf</code>.</p>
<p>If we were in the <code>data</code> directory, we would use the
relative path <code>../fig_output/fig1.pdf</code> or the same absolute
path <code>/home/user/bioc-intro/fig_output/fig1.pdf</code>.</p>
</div>
</section><section><h2 class="section-heading" id="interacting-with-r">Interacting with R<a class="anchor" aria-label="anchor" href="#interacting-with-r"></a>
</h2>
<hr class="half-width">
<p>The basis of programming is that we write down instructions for the
computer to follow, and then we tell the computer to follow those
instructions. We write, or <em>code</em>, instructions in R because it
is a common language that both the computer and we can understand. We
call the instructions <em>commands</em> and we tell the computer to
follow the instructions by <em>executing</em> (also called
<em>running</em>) those commands.</p>
<p>There are two main ways of interacting with R: by using the
<strong>console</strong> or by using <strong>scripts</strong> (plain
text files that contain your code). The console pane (in RStudio, the
bottom left panel) is the place where commands written in the R language
can be typed and executed immediately by the computer. It is also where
the results will be shown for commands that have been executed. You can
type commands directly into the console and press <code>Enter</code> to
execute those commands, but they will be forgotten when you close the
session.</p>
<p>Because we want our code and workflow to be reproducible, it is
better to type the commands we want in the script editor, and save the
script. This way, there is a complete record of what we did, and anyone
(including our future selves!) can easily replicate the results on their
computer. Note, however, that merely typing the commands in the script
does not automatically <em>run</em> them - they still need to be sent to
the console for execution.</p>
<p>RStudio allows you to execute commands directly from the script
editor by using the <code>Ctrl</code> + <code>Enter</code> shortcut (on
Macs, <code>Cmd</code> + <code>Return</code> will work, too). The
command on the current line in the script (indicated by the cursor) or
all of the commands in the currently selected text will be sent to the
console and executed when you press <code>Ctrl</code> +
<code>Enter</code>. You can find other keyboard shortcuts in this <a href="https://raw.githubusercontent.com/rstudio/cheatsheets/main/rstudio-ide.pdf" class="external-link">RStudio
cheatsheet about the RStudio IDE</a>.</p>
<p>At some point in your analysis you may want to check the content of a
variable or the structure of an object, without necessarily keeping a
record of it in your script. You can type these commands and execute
them directly in the console. RStudio provides the <code>Ctrl</code> +
<code>1</code> and <code>Ctrl</code> + <code>2</code> shortcuts allow
you to jump between the script and the console panes.</p>
<p>If R is ready to accept commands, the R console shows a
<code>&gt;</code> prompt. If it receives a command (by typing,
copy-pasting or sending from the script editor using <code>Ctrl</code> +
<code>Enter</code>), R will try to execute it, and when ready, will show
the results and come back with a new <code>&gt;</code> prompt to wait
for new commands.</p>
<p>If R is still waiting for you to enter more data because it isn’t
complete yet, the console will show a <code>+</code> prompt. It means
that you haven’t finished entering a complete command. This is because
you have not ‘closed’ a parenthesis or quotation, i.e. you don’t have
the same number of left-parentheses as right-parentheses, or the same
number of opening and closing quotation marks. When this happens, and
you thought you finished typing your command, click inside the console
window and press <code>Esc</code>; this will cancel the incomplete
command and return you to the <code>&gt;</code> prompt.</p>
</section><section><h2 class="section-heading" id="how-to-learn-more-during-and-after-the-course">How to learn more during and after the course?<a class="anchor" aria-label="anchor" href="#how-to-learn-more-during-and-after-the-course"></a>
</h2>
<hr class="half-width">
<p>The material we cover during this course will give you an initial
taste of how you can use R to analyse data for your own research.
However, you will need to learn more to do advanced operations such as
cleaning your dataset, using statistical methods, or creating beautiful
graphics<a href="#fn5" class="footnote-ref" id="fnref5"><sup>5</sup></a>. The best way to become proficient and
efficient at R, as with any other tool, is to use it to address your
actual research questions. As a beginner, it can feel daunting to have
to write a script from scratch, and given that many people make their
code available online, modifying existing code to suit your purpose
might make it easier for you to get started.</p>
<figure><img src="fig/kitten-try-things.jpg" alt="Fake book cover showing a kitten with title 'Changing Stuff and Seeing What Happens.'" width="400px" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></section><section><h2 class="section-heading" id="seeking-help">Seeking help<a class="anchor" aria-label="anchor" href="#seeking-help"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="use-the-built-in-rstudio-help-interface-to-search-for-more-information-on-r-functions">Use the built-in RStudio help interface to search for more
information on R functions<a class="anchor" aria-label="anchor" href="#use-the-built-in-rstudio-help-interface-to-search-for-more-information-on-r-functions"></a>
</h3>
<figure style="text-align: center"><img src="fig/rstudiohelp.png" alt="Screenshot of RStudio's Help panel." width="70%" class="figure mx-auto d-block"><figcaption>
RStudio help interface.
</figcaption></figure><p>One of the fastest ways to get help, is to use the RStudio help
interface. This panel by default can be found at the lower right hand
panel of RStudio. As seen in the screenshot, by typing the word “Mean”,
RStudio tries to also give a number of suggestions that you might be
interested in. The description is then shown in the display window.</p>
</div>
<div class="section level3">
<h3 id="i-know-the-name-of-the-function-i-want-to-use-but-im-not-sure-how-to-use-it">I know the name of the function I want to use, but I’m not sure how
to use it<a class="anchor" aria-label="anchor" href="#i-know-the-name-of-the-function-i-want-to-use-but-im-not-sure-how-to-use-it"></a>
</h3>
<p>If you need help with a specific function, let’s say
<code>barplot()</code>, you can type:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">?</span><span class="va">barplot</span></span></code></pre>
</div>
<p>If you just need to remind yourself of the names of the arguments,
you can use:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">args</span><span class="op">(</span><span class="va">lm</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="i-want-to-use-a-function-that-does-x-there-must-be-a-function-for-it-but-i-dont-know-which-one">I want to use a function that does X, there must be a function for
it but I don’t know which one…<a class="anchor" aria-label="anchor" href="#i-want-to-use-a-function-that-does-x-there-must-be-a-function-for-it-but-i-dont-know-which-one"></a>
</h3>
<p>If you are looking for a function to do a particular task, you can
use the <code>help.search()</code> function, which is called by the
double question mark <code>??</code>. However, this only looks through
the installed packages for help pages with a match to your search
request</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">?</span><span class="op">?</span><span class="va">kruskal</span></span></code></pre>
</div>
<p>If you can’t find what you are looking for, you can use the <a href="https://www.rdocumentation.org" class="external-link">rdocumentation.org</a> website
that searches through the help files across all packages available.</p>
<p>Finally, a generic Google or internet search “R &lt;task&gt;” will
often either send you to the appropriate package documentation or a
helpful forum where someone else has already asked your question.</p>
</div>
<div class="section level3">
<h3 id="i-am-stuck-i-get-an-error-message-that-i-dont-understand">I am stuck… I get an error message that I don’t understand<a class="anchor" aria-label="anchor" href="#i-am-stuck-i-get-an-error-message-that-i-dont-understand"></a>
</h3>
<p>Start by googling the error message. However, this doesn’t always
work very well because often, package developers rely on the error
catching provided by R. You end up with general error messages that
might not be very helpful to diagnose a problem (e.g. “subscript out of
bounds”). If the message is very generic, you might also include the
name of the function or package you’re using in your query.</p>
<p>However, you should check Stack Overflow. Search using the
<code>[r]</code> tag. Most questions have already been answered, but the
challenge is to use the right words in the search to find the
answers:</p>
<p><a href="https://stackoverflow.com/questions/tagged/r" class="external-link">https://stackoverflow.com/questions/tagged/r</a></p>
<p>The <a href="https://cran.r-project.org/doc/manuals/R-intro.pdf" class="external-link">Introduction
to R</a> can also be dense for people with little programming experience
but it is a good place to understand the underpinnings of the R
language.</p>
<p>The <a href="https://cran.r-project.org/doc/FAQ/R-FAQ.html" class="external-link">R FAQ</a>
is dense and technical but it is full of useful information.</p>
</div>
<div class="section level3">
<h3 id="asking-for-help">Asking for help<a class="anchor" aria-label="anchor" href="#asking-for-help"></a>
</h3>
<p>The key to receiving help from someone is for them to rapidly grasp
your problem. You should make it as easy as possible to pinpoint where
the issue might be.</p>
<p>Try to use the correct words to describe your problem. For instance,
a package is not the same thing as a library. Most people will
understand what you meant, but others have really strong feelings about
the difference in meaning. The key point is that it can make things
confusing for people trying to help you. Be as precise as possible when
describing your problem.</p>
<p>If possible, try to reduce what doesn’t work to a simple
<em>reproducible example</em>. If you can reproduce the problem using a
very small data frame instead of your 50000 rows and 10000 columns one,
provide the small one with the description of your problem. When
appropriate, try to generalise what you are doing so even people who are
not in your field can understand the question. For instance instead of
using a subset of your real dataset, create a small (3 columns, 5 rows)
generic one. For more information on how to write a reproducible example
see <a href="https://adv-r.had.co.nz/Reproducibility.html" class="external-link">this article
by Hadley Wickham</a>.</p>
<p>To share an object with someone else, if it’s relatively small, you
can use the function <code>dput()</code>. It will output R code that can
be used to recreate the exact same object as the one in memory:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## iris is an example data frame that comes with R and head() is a</span></span>
<span><span class="co">## function that returns the first part of the data frame</span></span>
<span><span class="fu">dput</span><span class="op">(</span><span class="fu">head</span><span class="op">(</span><span class="va">iris</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="fu">structure</span><span class="op">(</span><span class="fu">list</span><span class="op">(</span>Sepal.Length <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">5.1</span>, <span class="fl">4.9</span>, <span class="fl">4.7</span>, <span class="fl">4.6</span>, <span class="fl">5</span>, <span class="fl">5.4</span><span class="op">)</span>,</span>
<span>    Sepal.Width <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">3.5</span>, <span class="fl">3</span>, <span class="fl">3.2</span>, <span class="fl">3.1</span>, <span class="fl">3.6</span>, <span class="fl">3.9</span><span class="op">)</span>, Petal.Length <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1.4</span>,</span>
<span>    <span class="fl">1.4</span>, <span class="fl">1.3</span>, <span class="fl">1.5</span>, <span class="fl">1.4</span>, <span class="fl">1.7</span><span class="op">)</span>, Petal.Width <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">0.2</span>, <span class="fl">0.2</span>, <span class="fl">0.2</span>,</span>
<span>    <span class="fl">0.2</span>, <span class="fl">0.2</span>, <span class="fl">0.4</span><span class="op">)</span>, Species <span class="op">=</span> <span class="fu">structure</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="fl">1L</span>, <span class="fl">1L</span>, <span class="fl">1L</span>, <span class="fl">1L</span>, <span class="fl">1L</span>,</span>
<span>    <span class="fl">1L</span><span class="op">)</span>, levels <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"setosa"</span>, <span class="st">"versicolor"</span>, <span class="st">"virginica"</span><span class="op">)</span>, class <span class="op">=</span> <span class="st">"factor"</span><span class="op">)</span><span class="op">)</span>, row.names <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="cn">NA</span>,</span>
<span><span class="fl">6L</span><span class="op">)</span>, class <span class="op">=</span> <span class="st">"data.frame"</span><span class="op">)</span></span></code></pre>
</div>
<p>If the object is larger, provide either the raw file (i.e., your CSV
file) with your script up to the point of the error (and after removing
everything that is not relevant to your issue). Alternatively, in
particular if your question is not related to a data frame, you can save
any R object to a file[^export]:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">saveRDS</span><span class="op">(</span><span class="va">iris</span>, file<span class="op">=</span><span class="st">"/tmp/iris.rds"</span><span class="op">)</span></span></code></pre>
</div>
<p>The content of this file is however not human readable and cannot be
posted directly on Stack Overflow. Instead, it can be sent to someone by
email who can read it with the <code>readRDS()</code> command (here it
is assumed that the downloaded file is in a <code>Downloads</code>
folder in the user’s home directory):</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">some_data</span> <span class="op">&lt;-</span> <span class="fu">readRDS</span><span class="op">(</span>file<span class="op">=</span><span class="st">"~/Downloads/iris.rds"</span><span class="op">)</span></span></code></pre>
</div>
<p>Last, but certainly not least, <strong>always include the output of
<code>sessionInfo()</code></strong> as it provides critical information
about your platform, the versions of R and the packages that you are
using, and other information that can be very helpful to understand your
problem.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sessionInfo</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>R version 4.5.0 (2025-04-11)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C

time zone: UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

loaded via a namespace (and not attached):
[1] BiocManager_1.30.25 compiler_4.5.0      tools_4.5.0
[4] yaml_2.3.10         knitr_1.50          xfun_0.52
[7] renv_1.1.4          evaluate_1.0.3     </code></pre>
</div>
</div>
<div class="section level3">
<h3 id="where-to-ask-for-help">Where to ask for help?<a class="anchor" aria-label="anchor" href="#where-to-ask-for-help"></a>
</h3>
<ul>
<li>The person sitting next to you during the course. Don’t hesitate to
talk to your neighbour during the workshop, compare your answers, and
ask for help.</li>
<li>Your friendly colleagues: if you know someone with more experience
than you, they might be able and willing to help you.</li>
<li>
<a href="https://stackoverflow.com/questions/tagged/r" class="external-link">Stack
Overflow</a>: if your question hasn’t been answered before and is well
crafted, chances are you will get an answer in less than 5 min. Remember
to follow their guidelines on <a href="https://stackoverflow.com/help/how-to-ask" class="external-link">how to ask a good
question</a>.</li>
<li>The <a href="https://stat.ethz.ch/mailman/listinfo/r-help" class="external-link">R-help
mailing list</a>: it is read by a lot of people (including most of the R
core team), a lot of people post to it, but the tone can be pretty dry,
and it is not always very welcoming to new users. If your question is
valid, you are likely to get an answer very fast but don’t expect that
it will come with smiley faces. Also, here more than anywhere else, be
sure to use correct vocabulary (otherwise you might get an answer
pointing to the misuse of your words rather than answering your
question). You will also have more success if your question is about a
base function rather than a specific package.</li>
<li>If your question is about a specific package, see if there is a
mailing list for it. Usually it’s included in the DESCRIPTION file of
the package that can be accessed using
<code>packageDescription("name-of-package")</code>. You may also want to
try to email the author of the package directly, or open an issue on the
code repository (e.g., GitHub).</li>
<li>There are also some topic-specific mailing lists (GIS,
phylogenetics, etc…), the complete list is available on the <a href="https://www.r-project.org/mail.html" class="external-link">R-project mailing list
page</a>.</li>
</ul>
</div>
<div class="section level3">
<h3 id="more-resources">More resources<a class="anchor" aria-label="anchor" href="#more-resources"></a>
</h3>
<ul>
<li><p>The <a href="https://www.r-project.org/posting-guide.html" class="external-link">Posting Guide</a>
for the R mailing lists.</p></li>
<li><p><a href="https://blog.revolutionanalytics.com/2014/01/how-to-ask-for-r-help.html" class="external-link">How
to ask for R help</a> useful guidelines.</p></li>
<li><p><a href="https://codeblog.jonskeet.uk/2010/08/29/writing-the-perfect-question/" class="external-link">This
blog post by Jon Skeet</a> has quite comprehensive advice on how to ask
programming questions.</p></li>
<li><p>The <a href="https://cran.rstudio.com/web/packages/reprex/" class="external-link">reprex</a> package
is very helpful to create reproducible examples when asking for help.
The rOpenSci community call “How to ask questions so they get answered”
(<a href="https://github.com/ropensci/commcalls/issues/14" class="external-link">Github
link</a> and <a href="https://vimeo.com/208749032" class="external-link">video recording</a>)
includes a presentation of the reprex package and of its
philosophy.</p></li>
</ul>
</div>
</section><section><h2 class="section-heading" id="r-packages">R packages<a class="anchor" aria-label="anchor" href="#r-packages"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="loading-packages">Loading packages<a class="anchor" aria-label="anchor" href="#loading-packages"></a>
</h3>
<p>As we have seen above, R packages play a fundamental role in R. The
make use of a package’s functionality, assuming it is installed, we
first need to load it to be able to use it. This is done with the
<code>library()</code> function. Below, we load
<code>ggplot2</code>.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="st">"ggplot2"</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="installing-packages">Installing packages<a class="anchor" aria-label="anchor" href="#installing-packages"></a>
</h3>
<p>The default package repository is The <em>Comprehensive R Archive
Network</em> (CRAN), and any package that is available on CRAN can be
installed with the <code>install.packages()</code> function. Below, for
example, we install the <code>dplyr</code> package that we will learn
about later.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"dplyr"</span><span class="op">)</span></span></code></pre>
</div>
<p>This command will install the <code>dplyr</code> package as well as
all its dependencies, i.e. all the packages that it relies on to
function.</p>
<p>Another major R package repository is maintained by Bioconductor. <a href="https://bioconductor.org/packages/release/BiocViews.html#___Software" class="external-link">Bioconductor
packages</a> are managed and installed using a dedicated package, namely
<code>BiocManager</code>, that can be installed from CRAN with</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span></code></pre>
</div>
<p>Individual packages such as <code>SummarizedExperiment</code> (we
will use it later), <code>DESeq2</code> (for RNA-Seq analysis), and any
others from either Bioconductor or CRAN can then be installed with
<code>BiocManager::install</code>.</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"SummarizedExperiment"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"DESeq2"</span><span class="op">)</span></span></code></pre>
</div>
<p>By default, <code>BiocManager::install()</code> will also check all
your installed packages and see if there are newer versions available.
If there are, it will show them to you and ask you if you want to
<code>Update all/some/none? [a/s/n]:</code> and then wait for your
answer. While you should strive to have the most up-to-date package
versions, in practice we recommend only updating packages in a fresh R
session before any packages are loaded.</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>Start using R and RStudio</li>
</ul>
</div>
</div>
</div>
</div>
</section><div class="footnotes footnotes-end-of-document">
<hr>
<ol>
<li id="fn1"><p>As opposed to using R directly from the command line
console. There exist other software that interface and integrate with R,
but RStudio is particularly well suited for beginners while providing
numerous very advanced features.<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>i.e. add-ons that confer R with new functionality, such
as bioinformatics data analysis.<a href="#fnref2" class="footnote-back">↩︎</a></p></li>
<li id="fn3"><p>In this course, we consider bioinformatics as data
science applied to biological or bio-medical data.<a href="#fnref3" class="footnote-back">↩︎</a></p></li>
<li id="fn4"><p>That someone could be, and very likely will be your
future self, a couple of months or years after the analyses were run.<a href="#fnref4" class="footnote-back">↩︎</a></p></li>
<li id="fn5"><p>We will introduce most of these (except statistics)
here, but will only manage to scratch the surface of the wealth of what
is possible to do with R.<a href="#fnref5" class="footnote-back">↩︎</a></p></li>
</ol>
</div></section><section id="aio-23-starting-with-r"><p>Content from <a href="23-starting-with-r.html">Introduction to R</a></p>
<hr>
<p>Last updated on 2025-06-15 |

        <a href="https://github.com/carpentries-incubator/bioc-intro/edit/main/episodes/23-starting-with-r.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>First commands in R</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Define the following terms as they relate to R: object, assign,
call, function, arguments, options.</li>
<li>Assign values to objects in R.</li>
<li>Learn how to <em>name</em> objects</li>
<li>Use comments to inform script.</li>
<li>Solve simple arithmetic operations in R.</li>
<li>Call functions and use arguments to change their default
options.</li>
<li>Inspect the content of vectors and manipulate their content.</li>
<li>Subset and extract values from vectors.</li>
<li>Analyze vectors with missing data.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<blockquote>
<p>This episode is based on the Data Carpentries’s <em>Data Analysis and
Visualisation in R for Ecologists</em> lesson.</p>
</blockquote>
<section><h2 class="section-heading" id="creating-objects-in-r">Creating objects in R<a class="anchor" aria-label="anchor" href="#creating-objects-in-r"></a>
</h2>
<hr class="half-width">
<p>You can get output from R simply by typing math in the console:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fl">3</span> <span class="op">+</span> <span class="fl">5</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 8</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fl">12</span> <span class="op">/</span> <span class="fl">7</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1.714286</code></pre>
</div>
<p>However, to do useful and interesting things, we need to assign
<em>values</em> to <em>objects</em>. To create an object, we need to
give it a name followed by the assignment operator <code>&lt;-</code>,
and the value we want to give it:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">weight_kg</span> <span class="op">&lt;-</span> <span class="fl">55</span></span></code></pre>
</div>
<p><code>&lt;-</code> is the assignment operator. It assigns values on
the right to objects on the left. So, after executing
<code>x &lt;- 3</code>, the value of <code>x</code> is <code>3</code>.
The arrow can be read as 3 <strong>goes into</strong> <code>x</code>.
For historical reasons, you can also use <code>=</code> for assignments,
but not in every context. Because of the <a href="https://blog.revolutionanalytics.com/2008/12/use-equals-or-arrow-for-assignment.html" class="external-link">slight
differences</a> in syntax, it is good practice to always use
<code>&lt;-</code> for assignments.</p>
<p>In RStudio, typing <kbd>Alt</kbd> + <kbd>-</kbd> (push <kbd>Alt</kbd>
at the same time as the <kbd>-</kbd> key) will write <code>&lt;-</code>
in a single keystroke in a PC, while typing <kbd>Option</kbd> +
<kbd>-</kbd> (push <kbd>Option</kbd> at the same time as the
<kbd>-</kbd> key) does the same in a Mac.</p>
<div class="section level3">
<h3 id="naming-variables">Naming variables<a class="anchor" aria-label="anchor" href="#naming-variables"></a>
</h3>
<p>Objects can be given any name such as <code>x</code>,
<code>current_temperature</code>, or <code>subject_id</code>. You want
your object names to be explicit and not too long. They cannot start
with a number (<code>2x</code> is not valid, but <code>x2</code> is). R
is case sensitive (e.g., <code>weight_kg</code> is different from
<code>Weight_kg</code>). There are some names that cannot be used
because they are the names of fundamental functions in R (e.g.,
<code>if</code>, <code>else</code>, <code>for</code>, see the <a href="https://stat.ethz.ch/R-manual/R-devel/library/base/html/Reserved.html" class="external-link">Reserved
Words in R</a> manual page, for a complete list). In general, even if
it’s allowed, it’s best to not use other function names (e.g.,
<code>c</code>, <code>T</code>, <code>mean</code>, <code>data</code>,
<code>df</code>, <code>weights</code>). If in doubt, check the help to
see if the name is already in use. It’s also best to avoid dots
(<code>.</code>) within an object name as in <code>my.dataset</code>.
There are many functions in R with dots in their names for historical
reasons, but because dots have a special meaning in R (for methods) and
other programming languages, it’s best to avoid them. It is also
recommended to use nouns for object names, and verbs for function names.
It’s important to be consistent in the styling of your code (where you
put spaces, how you name objects, etc.). Using a consistent coding style
makes your code clearer to read for your future self and your
collaborators. In R, some popular style guides are <a href="https://google.github.io/styleguide/Rguide.xml" class="external-link">Google’s</a>, the
<a href="https://style.tidyverse.org/" class="external-link">tidyverse’s</a> style and the <a href="https://bioconductor.org/developers/how-to/coding-style/" class="external-link">Bioconductor
style guide</a>. The tidyverse’s is very comprehensive and may seem
overwhelming at first. You can install the <a href="https://github.com/jimhester/lintr" class="external-link"><strong><code>lintr</code></strong></a>
package to automatically check for issues in the styling of your
code.</p>
<blockquote>
<p><strong>Objects vs. variables</strong>: What are known as
<code>objects</code> in <code>R</code> are known as
<code>variables</code> in many other programming languages. Depending on
the context, <code>object</code> and <code>variable</code> can have
drastically different meanings. However, in this lesson, the two words
are used synonymously. For more information <a href="https://cran.r-project.org/doc/manuals/r-release/R-lang.html#Objects" class="external-link">see
here.</a></p>
</blockquote>
<p>When assigning a value to an object, R does not print anything. You
can force R to print the value by using parentheses or by typing the
object name:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">weight_kg</span> <span class="op">&lt;-</span> <span class="fl">55</span>    <span class="co"># doesn't print anything</span></span>
<span><span class="op">(</span><span class="va">weight_kg</span> <span class="op">&lt;-</span> <span class="fl">55</span><span class="op">)</span>  <span class="co"># but putting parenthesis around the call prints the value of `weight_kg`</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 55</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">weight_kg</span>          <span class="co"># and so does typing the name of the object</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 55</code></pre>
</div>
<p>Now that R has <code>weight_kg</code> in memory, we can do arithmetic
with it. For instance, we may want to convert this weight into pounds
(weight in pounds is 2.2 times the weight in kg):</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fl">2.2</span> <span class="op">*</span> <span class="va">weight_kg</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 121</code></pre>
</div>
<p>We can also change an object’s value by assigning it a new one:</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">weight_kg</span> <span class="op">&lt;-</span> <span class="fl">57.5</span></span>
<span><span class="fl">2.2</span> <span class="op">*</span> <span class="va">weight_kg</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 126.5</code></pre>
</div>
<p>This means that assigning a value to one object does not change the
values of other objects For example, let’s store the animal’s weight in
pounds in a new object, <code>weight_lb</code>:</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">weight_lb</span> <span class="op">&lt;-</span> <span class="fl">2.2</span> <span class="op">*</span> <span class="va">weight_kg</span></span></code></pre>
</div>
<p>and then change <code>weight_kg</code> to 100.</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">weight_kg</span> <span class="op">&lt;-</span> <span class="fl">100</span></span></code></pre>
</div>
<div id="challenge" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="challenge" class="callout-inner">
<h3 class="callout-title">Challenge:</h3>
<div class="callout-content">
<p>What do you think is the current content of the object
<code>weight_lb</code>? 126.5 or 220?</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="comments">Comments<a class="anchor" aria-label="anchor" href="#comments"></a>
</h2>
<hr class="half-width">
<p>The comment character in R is <code>#</code>, anything to the right
of a <code>#</code> in a script will be ignored by R. It is useful to
leave notes, and explanations in your scripts.</p>
<p>RStudio makes it easy to comment or uncomment a paragraph: after
selecting the lines you want to comment, press at the same time on your
keyboard <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>C</kbd>. If you only
want to comment out one line, you can put the cursor at any location of
that line (i.e. no need to select the whole line), then press
<kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>C</kbd>.</p>
<div id="challenge-1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="challenge-1" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>What are the values after each statement in the following?</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mass</span> <span class="op">&lt;-</span> <span class="fl">47.5</span>            <span class="co"># mass?</span></span>
<span><span class="va">age</span>  <span class="op">&lt;-</span> <span class="fl">122</span>             <span class="co"># age?</span></span>
<span><span class="va">mass</span> <span class="op">&lt;-</span> <span class="va">mass</span> <span class="op">*</span> <span class="fl">2.0</span>      <span class="co"># mass?</span></span>
<span><span class="va">age</span>  <span class="op">&lt;-</span> <span class="va">age</span> <span class="op">-</span> <span class="fl">20</span>        <span class="co"># age?</span></span>
<span><span class="va">mass_index</span> <span class="op">&lt;-</span> <span class="va">mass</span><span class="op">/</span><span class="va">age</span>  <span class="co"># mass_index?</span></span></code></pre>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="functions-and-their-arguments">Functions and their arguments<a class="anchor" aria-label="anchor" href="#functions-and-their-arguments"></a>
</h2>
<hr class="half-width">
<p>Functions are “canned scripts” that automate more complicated sets of
commands including operations assignments, etc. Many functions are
predefined, or can be made available by importing R <em>packages</em>
(more on that later). A function usually gets one or more inputs called
<em>arguments</em>. Functions often (but not always) return a
<em>value</em>. A typical example would be the function
<code>sqrt()</code>. The input (the argument) must be a number, and the
return value (in fact, the output) is the square root of that number.
Executing a function (‘running it’) is called <em>calling</em> the
function. An example of a function call is:</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">b</span> <span class="op">&lt;-</span> <span class="fu">sqrt</span><span class="op">(</span><span class="va">a</span><span class="op">)</span></span></code></pre>
</div>
<p>Here, the value of <code>a</code> is given to the <code>sqrt()</code>
function, the <code>sqrt()</code> function calculates the square root,
and returns the value which is then assigned to the object
<code>b</code>. This function is very simple, because it takes just one
argument.</p>
<p>The return ‘value’ of a function need not be numerical (like that of
<code>sqrt()</code>), and it also does not need to be a single item: it
can be a set of things, or even a dataset. We’ll see that when we read
data files into R.</p>
<p>Arguments can be anything, not only numbers or filenames, but also
other objects. Exactly what each argument means differs per function,
and must be looked up in the documentation (see below). Some functions
take arguments which may either be specified by the user, or, if left
out, take on a <em>default</em> value: these are called
<em>options</em>. Options are typically used to alter the way the
function operates, such as whether it ignores ‘bad values’, or what
symbol to use in a plot. However, if you want something specific, you
can specify a value of your choice which will be used instead of the
default.</p>
<p>Let’s try a function that can take multiple arguments:
<code>round()</code>.</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">round</span><span class="op">(</span><span class="fl">3.14159</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 3</code></pre>
</div>
<p>Here, we’ve called <code>round()</code> with just one argument,
<code>3.14159</code>, and it has returned the value <code>3</code>.
That’s because the default is to round to the nearest whole number. If
we want more digits we can see how to do that by getting information
about the <code>round</code> function. We can use
<code>args(round)</code> or look at the help for this function using
<code>?round</code>.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">args</span><span class="op">(</span><span class="va">round</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="kw">function</span> <span class="op">(</span><span class="va">x</span>, <span class="va">digits</span> <span class="op">=</span> <span class="fl">0</span>, <span class="va">...</span><span class="op">)</span></span>
<span><span class="cn">NULL</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">?</span><span class="va">round</span></span></code></pre>
</div>
<p>We see that if we want a different number of digits, we can type
<code>digits=2</code> or however many we want.</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">round</span><span class="op">(</span><span class="fl">3.14159</span>, digits <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 3.14</code></pre>
</div>
<p>If you provide the arguments in the exact same order as they are
defined you don’t have to name them:</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">round</span><span class="op">(</span><span class="fl">3.14159</span>, <span class="fl">2</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 3.14</code></pre>
</div>
<p>And if you do name the arguments, you can switch their order:</p>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">round</span><span class="op">(</span>digits <span class="op">=</span> <span class="fl">2</span>, x <span class="op">=</span> <span class="fl">3.14159</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 3.14</code></pre>
</div>
<p>It’s good practice to put the non-optional arguments (like the number
you’re rounding) first in your function call, and to specify the names
of all optional arguments. If you don’t, someone reading your code might
have to look up the definition of a function with unfamiliar arguments
to understand what you’re doing. By specifying the name of the arguments
you are also safeguarding against possible future changes in the
function interface, which may potentially add new arguments in between
the existing ones.</p>
</section><section><h2 class="section-heading" id="vectors-and-data-types">Vectors and data types<a class="anchor" aria-label="anchor" href="#vectors-and-data-types"></a>
</h2>
<hr class="half-width">
<p>A vector is the most common and basic data type in R, and is pretty
much the workhorse of R. A vector is composed by a series of values,
such as numbers or characters. We can assign a series of values to a
vector using the <code>c()</code> function. For example we can create a
vector of animal weights and assign it to a new object
<code>weight_g</code>:</p>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">weight_g</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fl">50</span>, <span class="fl">60</span>, <span class="fl">65</span>, <span class="fl">82</span><span class="op">)</span></span>
<span><span class="va">weight_g</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 50 60 65 82</code></pre>
</div>
<p>A vector can also contain characters:</p>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">molecules</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">"dna"</span>, <span class="st">"rna"</span>, <span class="st">"protein"</span><span class="op">)</span></span>
<span><span class="va">molecules</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "dna"     "rna"     "protein"</code></pre>
</div>
<p>The quotes around “dna”, “rna”, etc. are essential here. Without the
quotes R will assume there are objects called <code>dna</code>,
<code>rna</code> and <code>protein</code>. As these objects don’t exist
in R’s memory, there will be an error message.</p>
<p>There are many functions that allow you to inspect the content of a
vector. <code>length()</code> tells you how many elements are in a
particular vector:</p>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">length</span><span class="op">(</span><span class="va">weight_g</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 4</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">length</span><span class="op">(</span><span class="va">molecules</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 3</code></pre>
</div>
<p>An important feature of a vector, is that all of the elements are the
same type of data. The function <code>class()</code> indicates the class
(the type of element) of an object:</p>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">class</span><span class="op">(</span><span class="va">weight_g</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "numeric"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb39">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">class</span><span class="op">(</span><span class="va">molecules</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "character"</code></pre>
</div>
<p>The function <code>str()</code> provides an overview of the structure
of an object and its elements. It is a useful function when working with
large and complex objects:</p>
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">str</span><span class="op">(</span><span class="va">weight_g</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> num [1:4] 50 60 65 82</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb43">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">str</span><span class="op">(</span><span class="va">molecules</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> chr [1:3] "dna" "rna" "protein"</code></pre>
</div>
<p>You can use the <code>c()</code> function to add other elements to
your vector:</p>
<div class="codewrapper sourceCode" id="cb45">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">weight_g</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="va">weight_g</span>, <span class="fl">90</span><span class="op">)</span> <span class="co"># add to the end of the vector</span></span>
<span><span class="va">weight_g</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fl">30</span>, <span class="va">weight_g</span><span class="op">)</span> <span class="co"># add to the beginning of the vector</span></span>
<span><span class="va">weight_g</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 30 50 60 65 82 90</code></pre>
</div>
<p>In the first line, we take the original vector <code>weight_g</code>,
add the value <code>90</code> to the end of it, and save the result back
into <code>weight_g</code>. Then we add the value <code>30</code> to the
beginning, again saving the result back into <code>weight_g</code>.</p>
<p>We can do this over and over again to grow a vector, or assemble a
dataset. As we program, this may be useful to add results that we are
collecting or calculating.</p>
<p>An <strong>atomic vector</strong> is the simplest R <strong>data
type</strong> and is a linear vector of a single type. Above, we saw 2
of the 6 main <strong>atomic vector</strong> types that R uses:
<code>"character"</code> and <code>"numeric"</code> (or
<code>"double"</code>). These are the basic building blocks that all R
objects are built from. The other 4 <strong>atomic vector</strong> types
are:</p>
<ul>
<li>
<code>"logical"</code> for <code>TRUE</code> and <code>FALSE</code>
(the boolean data type)</li>
<li>
<code>"integer"</code> for integer numbers (e.g., <code>2L</code>,
the <code>L</code> indicates to R that it’s an integer)</li>
<li>
<code>"complex"</code> to represent complex numbers with real and
imaginary parts (e.g., <code>1 + 4i</code>) and that’s all we’re going
to say about them</li>
<li>
<code>"raw"</code> for bitstreams that we won’t discuss further</li>
</ul>
<p>You can check the type of your vector using the <code>typeof()</code>
function and inputting your vector as the argument.</p>
<p>Vectors are one of the many <strong>data structures</strong> that R
uses. Other important ones are lists (<code>list</code>), matrices
(<code>matrix</code>), data frames (<code>data.frame</code>), factors
(<code>factor</code>) and arrays (<code>array</code>).</p>
<div id="challenge-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2" class="callout-inner">
<h3 class="callout-title">Challenge:</h3>
<div class="callout-content">
<p>We’ve seen that atomic vectors can be of type character, numeric (or
double), integer, and logical. But what happens if we try to mix these
types in a single vector?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>R implicitly converts them to all be the same type</p>
</div>
</div>
</div>
</div>
<div id="challenge-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-3" class="callout-inner">
<h3 class="callout-title">Challenge:</h3>
<div class="callout-content">
<p>What will happen in each of these examples? (hint: use
<code>class()</code> to check the data type of your objects and type in
their names to see what happens):</p>
<div class="codewrapper sourceCode" id="cb47">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">num_char</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="st">"a"</span><span class="op">)</span></span>
<span><span class="va">num_logical</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="cn">TRUE</span>, <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">char_logical</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">"a"</span>, <span class="st">"b"</span>, <span class="st">"c"</span>, <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">tricky</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="st">"4"</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb48">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">class</span><span class="op">(</span><span class="va">num_char</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "character"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb50">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">num_char</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "1" "2" "3" "a"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb52">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">class</span><span class="op">(</span><span class="va">num_logical</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "numeric"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb54">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">num_logical</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1 2 3 1 0</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb56">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">class</span><span class="op">(</span><span class="va">char_logical</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "character"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb58">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">char_logical</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "a"    "b"    "c"    "TRUE"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb60">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">class</span><span class="op">(</span><span class="va">tricky</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "character"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb62">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tricky</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "1" "2" "3" "4"</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="challenge-4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-4" class="callout-inner">
<h3 class="callout-title">Challenge:</h3>
<div class="callout-content">
<p>Why do you think it happens?</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<p>Vectors can be of only one data type. R tries to convert (coerce) the
content of this vector to find a <em>common denominator</em> that
doesn’t lose any information.</p>
</div>
</div>
</div>
</div>
<div id="challenge-5" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-5" class="callout-inner">
<h3 class="callout-title">Challenge:</h3>
<div class="callout-content">
<p>How many values in <code>combined_logical</code> are
<code>"TRUE"</code> (as a character) in the following example:</p>
<div class="codewrapper sourceCode" id="cb64">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">num_logical</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">char_logical</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">"a"</span>, <span class="st">"b"</span>, <span class="st">"c"</span>, <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">combined_logical</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="va">num_logical</span>, <span class="va">char_logical</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<p>Only one. There is no memory of past data types, and the coercion
happens the first time the vector is evaluated. Therefore, the
<code>TRUE</code> in <code>num_logical</code> gets converted into a
<code>1</code> before it gets converted into <code>"1"</code> in
<code>combined_logical</code>.</p>
<div class="codewrapper sourceCode" id="cb65">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">combined_logical</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "1"    "2"    "3"    "1"    "a"    "b"    "c"    "TRUE"</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="challenge-6" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-6" class="callout-inner">
<h3 class="callout-title">Challenge:</h3>
<div class="callout-content">
<p>In R, we call converting objects from one class into another class
<em>coercion</em>. These conversions happen according to a hierarchy,
whereby some types get preferentially coerced into other types. Can you
draw a diagram that represents the hierarchy of how these data types are
coerced?</p>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" aria-labelledby="headingSolution5" data-bs-parent="#accordionSolution5">
<div class="accordion-body">
<p>logical → numeric → character ← logical</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="subsetting-vectors">Subsetting vectors<a class="anchor" aria-label="anchor" href="#subsetting-vectors"></a>
</h2>
<hr class="half-width">
<p>If we want to extract one or several values from a vector, we must
provide one or several indices in square brackets. For instance:</p>
<div class="codewrapper sourceCode" id="cb67">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">molecules</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">"dna"</span>, <span class="st">"rna"</span>, <span class="st">"peptide"</span>, <span class="st">"protein"</span><span class="op">)</span></span>
<span><span class="va">molecules</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "rna"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb69">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">molecules</span><span class="op">[</span><span class="fu">c</span><span class="op">(</span><span class="fl">3</span>, <span class="fl">2</span><span class="op">)</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "peptide" "rna"    </code></pre>
</div>
<p>We can also repeat the indices to create an object with more elements
than the original one:</p>
<div class="codewrapper sourceCode" id="cb71">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">more_molecules</span> <span class="op">&lt;-</span> <span class="va">molecules</span><span class="op">[</span><span class="fu">c</span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">2</span>, <span class="fl">1</span>, <span class="fl">4</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">more_molecules</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "dna"     "rna"     "peptide" "rna"     "dna"     "protein"</code></pre>
</div>
<p>R indices start at 1. Programming languages like Fortran, MATLAB,
Julia, and R start counting at 1, because that’s what human beings
typically do. Languages in the C family (including C++, Java, Perl, and
Python) count from 0 because that’s simpler for computers to do.</p>
<p>Finally, it is also possible to get all the elements of a vector
except some specified elements using negative indices:</p>
<div class="codewrapper sourceCode" id="cb73">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">molecules</span> <span class="co">## all molecules</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "dna"     "rna"     "peptide" "protein"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb75">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">molecules</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span> <span class="co">## all but the first one</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "rna"     "peptide" "protein"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb77">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">molecules</span><span class="op">[</span><span class="op">-</span><span class="fu">c</span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span><span class="op">]</span> <span class="co">## all but 1st/3rd ones</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "rna"     "protein"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb79">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">molecules</span><span class="op">[</span><span class="fu">c</span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="op">-</span><span class="fl">3</span><span class="op">)</span><span class="op">]</span> <span class="co">## all but 1st/3rd ones</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "rna"     "protein"</code></pre>
</div>
</section><section><h2 class="section-heading" id="conditional-subsetting">Conditional subsetting<a class="anchor" aria-label="anchor" href="#conditional-subsetting"></a>
</h2>
<hr class="half-width">
<p>Another common way of subsetting is by using a logical vector.
<code>TRUE</code> will select the element with the same index, while
<code>FALSE</code> will not:</p>
<div class="codewrapper sourceCode" id="cb81">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">weight_g</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fl">21</span>, <span class="fl">34</span>, <span class="fl">39</span>, <span class="fl">54</span>, <span class="fl">55</span><span class="op">)</span></span>
<span><span class="va">weight_g</span><span class="op">[</span><span class="fu">c</span><span class="op">(</span><span class="cn">TRUE</span>, <span class="cn">FALSE</span>, <span class="cn">TRUE</span>, <span class="cn">TRUE</span>, <span class="cn">FALSE</span><span class="op">)</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 21 39 54</code></pre>
</div>
<p>Typically, these logical vectors are not typed by hand, but are the
output of other functions or logical tests. For instance, if you wanted
to select only the values above 50:</p>
<div class="codewrapper sourceCode" id="cb83">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## will return logicals with TRUE for the indices that meet</span></span>
<span><span class="co">## the condition</span></span>
<span><span class="va">weight_g</span> <span class="op">&gt;</span> <span class="fl">50</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] FALSE FALSE FALSE  TRUE  TRUE</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb85">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## so we can use this to select only the values above 50</span></span>
<span><span class="va">weight_g</span><span class="op">[</span><span class="va">weight_g</span> <span class="op">&gt;</span> <span class="fl">50</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 54 55</code></pre>
</div>
<p>You can combine multiple tests using <code>&amp;</code> (both
conditions are true, AND) or <code>|</code> (at least one of the
conditions is true, OR):</p>
<div class="codewrapper sourceCode" id="cb87">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">weight_g</span><span class="op">[</span><span class="va">weight_g</span> <span class="op">&lt;</span> <span class="fl">30</span> <span class="op">|</span> <span class="va">weight_g</span> <span class="op">&gt;</span> <span class="fl">50</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 21 54 55</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb89">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">weight_g</span><span class="op">[</span><span class="va">weight_g</span> <span class="op">&gt;=</span> <span class="fl">30</span> <span class="op">&amp;</span> <span class="va">weight_g</span> <span class="op">==</span> <span class="fl">21</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="fu">numeric</span><span class="op">(</span><span class="fl">0</span><span class="op">)</span></span></code></pre>
</div>
<p>Here, <code>&lt;</code> stands for “less than”, <code>&gt;</code> for
“greater than”, <code>&gt;=</code> for “greater than or equal to”, and
<code>==</code> for “equal to”. The double equal sign <code>==</code> is
a test for numerical equality between the left and right hand sides, and
should not be confused with the single <code>=</code> sign, which
performs variable assignment (similar to <code>&lt;-</code>).</p>
<p>A common task is to search for certain strings in a vector. One could
use the “or” operator <code>|</code> to test for equality to multiple
values, but this can quickly become tedious. The function
<code>%in%</code> allows you to test if any of the elements of a search
vector are found:</p>
<div class="codewrapper sourceCode" id="cb91">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">molecules</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">"dna"</span>, <span class="st">"rna"</span>, <span class="st">"protein"</span>, <span class="st">"peptide"</span><span class="op">)</span></span>
<span><span class="va">molecules</span><span class="op">[</span><span class="va">molecules</span> <span class="op">==</span> <span class="st">"rna"</span> <span class="op">|</span> <span class="va">molecules</span> <span class="op">==</span> <span class="st">"dna"</span><span class="op">]</span> <span class="co"># returns both rna and dna</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "dna" "rna"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb93">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">molecules</span> <span class="op">%in%</span> <span class="fu">c</span><span class="op">(</span><span class="st">"rna"</span>, <span class="st">"dna"</span>, <span class="st">"metabolite"</span>, <span class="st">"peptide"</span>, <span class="st">"glycerol"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1]  TRUE  TRUE FALSE  TRUE</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb95">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">molecules</span><span class="op">[</span><span class="va">molecules</span> <span class="op">%in%</span> <span class="fu">c</span><span class="op">(</span><span class="st">"rna"</span>, <span class="st">"dna"</span>, <span class="st">"metabolite"</span>, <span class="st">"peptide"</span>, <span class="st">"glycerol"</span><span class="op">)</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "dna"     "rna"     "peptide"</code></pre>
</div>
<div id="challenge-7" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-7" class="callout-inner">
<h3 class="callout-title">Challenge:</h3>
<div class="callout-content">
<p>Can you figure out why <code>"four" &gt; "five"</code> returns
<code>TRUE</code>?</p>
</div>
</div>
</div>
<div id="accordionSolution6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution6" aria-expanded="false" aria-controls="collapseSolution6">
  <h4 class="accordion-header" id="headingSolution6"> Show me the solution </h4>
</button>
<div id="collapseSolution6" class="accordion-collapse collapse" aria-labelledby="headingSolution6" data-bs-parent="#accordionSolution6">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb97">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="st">"four"</span> <span class="op">&gt;</span> <span class="st">"five"</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] TRUE</code></pre>
</div>
<p>When using <code>&gt;</code> or <code>&lt;</code> on strings, R
compares their alphabetical order. Here <code>"four"</code> comes after
<code>"five"</code>, and therefore is <em>greater than</em> it.</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="names">Names<a class="anchor" aria-label="anchor" href="#names"></a>
</h2>
<hr class="half-width">
<p>It is possible to name each element of a vector. The code chunk below
shows an initial vector without any names, how names are set, and
retrieved.</p>
<div class="codewrapper sourceCode" id="cb99">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">3</span>, <span class="fl">5</span>, <span class="fl">10</span><span class="op">)</span></span>
<span><span class="fu">names</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="co">## no names</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="cn">NULL</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb101">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">names</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"C"</span>, <span class="st">"D"</span>, <span class="st">"E"</span><span class="op">)</span></span>
<span><span class="fu">names</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="co">## now we have names</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "A" "B" "C" "D" "E"</code></pre>
</div>
<p>When a vector has names, it is possible to access elements by their
name, in addition to their index.</p>
<div class="codewrapper sourceCode" id="cb103">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">[</span><span class="fu">c</span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>A C
1 3 </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb105">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">[</span><span class="fu">c</span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"C"</span><span class="op">)</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>A C
1 3 </code></pre>
</div>
</section><section><h2 class="section-heading" id="missing-data">Missing data<a class="anchor" aria-label="anchor" href="#missing-data"></a>
</h2>
<hr class="half-width">
<p>As R was designed to analyze datasets, it includes the concept of
missing data (which is uncommon in other programming languages). Missing
data are represented in vectors as <code>NA</code>.</p>
<p>When doing operations on numbers, most functions will return
<code>NA</code> if the data you are working with include missing values.
This feature makes it harder to overlook the cases where you are dealing
with missing data. You can add the argument <code>na.rm = TRUE</code> to
calculate the result while ignoring the missing values.</p>
<div class="codewrapper sourceCode" id="cb107">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">heights</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fl">2</span>, <span class="fl">4</span>, <span class="fl">4</span>, <span class="cn">NA</span>, <span class="fl">6</span><span class="op">)</span></span>
<span><span class="fu">mean</span><span class="op">(</span><span class="va">heights</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] NA</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb109">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">max</span><span class="op">(</span><span class="va">heights</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] NA</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb111">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mean</span><span class="op">(</span><span class="va">heights</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 4</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb113">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">max</span><span class="op">(</span><span class="va">heights</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 6</code></pre>
</div>
<p>If your data include missing values, you may want to become familiar
with the functions <code>is.na()</code>, <code>na.omit()</code>, and
<code>complete.cases()</code>. See below for examples.</p>
<div class="codewrapper sourceCode" id="cb115">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Extract those elements which are not missing values.</span></span>
<span><span class="va">heights</span><span class="op">[</span><span class="op">!</span><span class="fu">is.na</span><span class="op">(</span><span class="va">heights</span><span class="op">)</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 2 4 4 6</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb117">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Returns the object with incomplete cases removed.</span></span>
<span><span class="co">## The returned object is an atomic vector of type `"numeric"`</span></span>
<span><span class="co">## (or `"double"`).</span></span>
<span><span class="fu">na.omit</span><span class="op">(</span><span class="va">heights</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 2 4 4 6
attr(,"na.action")
[1] 4
attr(,"class")
[1] "omit"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb119">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Extract those elements which are complete cases.</span></span>
<span><span class="co">## The returned object is an atomic vector of type `"numeric"`</span></span>
<span><span class="co">## (or `"double"`).</span></span>
<span><span class="va">heights</span><span class="op">[</span><span class="fu">complete.cases</span><span class="op">(</span><span class="va">heights</span><span class="op">)</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 2 4 4 6</code></pre>
</div>
<div id="challenge-8" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-8" class="callout-inner">
<h3 class="callout-title">Challenge:</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li>Using this vector of heights in inches, create a new vector with the
NAs removed.</li>
</ol>
<div class="codewrapper sourceCode" id="cb121">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">heights</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fl">63</span>, <span class="fl">69</span>, <span class="fl">60</span>, <span class="fl">65</span>, <span class="cn">NA</span>, <span class="fl">68</span>, <span class="fl">61</span>, <span class="fl">70</span>, <span class="fl">61</span>, <span class="fl">59</span>, <span class="fl">64</span>, <span class="fl">69</span>, <span class="fl">63</span>, <span class="fl">63</span>, <span class="cn">NA</span>, <span class="fl">72</span>, <span class="fl">65</span>, <span class="fl">64</span>, <span class="fl">70</span>, <span class="fl">63</span>, <span class="fl">65</span><span class="op">)</span></span></code></pre>
</div>
<ol start="2" style="list-style-type: decimal">
<li>Use the function <code>median()</code> to calculate the median of
the <code>heights</code> vector.</li>
<li>Use R to figure out how many people in the set are taller than 67
inches.</li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution7" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution7" aria-expanded="false" aria-controls="collapseSolution7">
  <h4 class="accordion-header" id="headingSolution7"> Show me the solution </h4>
</button>
<div id="collapseSolution7" class="accordion-collapse collapse" aria-labelledby="headingSolution7" data-bs-parent="#accordionSolution7">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb122">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">heights_no_na</span> <span class="op">&lt;-</span> <span class="va">heights</span><span class="op">[</span><span class="op">!</span><span class="fu">is.na</span><span class="op">(</span><span class="va">heights</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">## or</span></span>
<span><span class="va">heights_no_na</span> <span class="op">&lt;-</span> <span class="fu">na.omit</span><span class="op">(</span><span class="va">heights</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb123">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">median</span><span class="op">(</span><span class="va">heights</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 64</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb125">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">heights_above_67</span> <span class="op">&lt;-</span> <span class="va">heights_no_na</span><span class="op">[</span><span class="va">heights_no_na</span> <span class="op">&gt;</span> <span class="fl">67</span><span class="op">]</span></span>
<span><span class="fu">length</span><span class="op">(</span><span class="va">heights_above_67</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 6</code></pre>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="sec:genvec">Generating vectors<a class="anchor" aria-label="anchor" href="#sec:genvec"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="constructors">Constructors<a class="anchor" aria-label="anchor" href="#constructors"></a>
</h3>
<p>There exists some functions to generate vectors of different type. To
generate a vector of numerics, one can use the <code>numeric()</code>
constructor, providing the length of the output vector as parameter. The
values will be initialised with 0.</p>
<div class="codewrapper sourceCode" id="cb127">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">numeric</span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0 0 0</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb129">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">numeric</span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] 0 0 0 0 0 0 0 0 0 0</code></pre>
</div>
<p>Note that if we ask for a vector of numerics of length 0, we obtain
exactly that:</p>
<div class="codewrapper sourceCode" id="cb131">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">numeric</span><span class="op">(</span><span class="fl">0</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="fu">numeric</span><span class="op">(</span><span class="fl">0</span><span class="op">)</span></span></code></pre>
</div>
<p>There are similar constructors for characters and logicals, named
<code>character()</code> and <code>logical()</code> respectively.</p>
<div id="challenge-9" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-9" class="callout-inner">
<h3 class="callout-title">Challenge:</h3>
<div class="callout-content">
<p>What are the defaults for character and logical vectors?</p>
</div>
</div>
</div>
<div id="accordionSolution8" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution8" aria-expanded="false" aria-controls="collapseSolution8">
  <h4 class="accordion-header" id="headingSolution8"> Show me the solution </h4>
</button>
<div id="collapseSolution8" class="accordion-collapse collapse" aria-labelledby="headingSolution8" data-bs-parent="#accordionSolution8">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb133">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">character</span><span class="op">(</span><span class="fl">2</span><span class="op">)</span> <span class="co">## the empty character</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "" ""</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb135">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">logical</span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>   <span class="co">## FALSE</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] FALSE FALSE</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="replicate-elements">Replicate elements<a class="anchor" aria-label="anchor" href="#replicate-elements"></a>
</h3>
<p>The <code>rep</code> function allow to repeat a value a certain
number of times. If we want to initiate a vector of numerics of length 5
with the value -1, for example, we could do the following:</p>
<div class="codewrapper sourceCode" id="cb137">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rep</span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">5</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] -1 -1 -1 -1 -1</code></pre>
</div>
<p>Similarly, to generate a vector populated with missing values, which
is often a good way to start, without setting assumptions on the data to
be collected:</p>
<div class="codewrapper sourceCode" id="cb139">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rep</span><span class="op">(</span><span class="cn">NA</span>, <span class="fl">5</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] NA NA NA NA NA</code></pre>
</div>
<p><code>rep</code> can take vectors of any length as input (above, we
used vectors of length 1) and any type. For example, if we want to
repeat the values 1, 2 and 3 five times, we would do the following:</p>
<div class="codewrapper sourceCode" id="cb141">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rep</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span>, <span class="fl">5</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] 1 2 3 1 2 3 1 2 3 1 2 3 1 2 3</code></pre>
</div>
<div id="challenge-10" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-10" class="callout-inner">
<h3 class="callout-title">Challenge:</h3>
<div class="callout-content">
<p>What if we wanted to repeat the values 1, 2 and 3 five times, but
obtain five 1s, five 2s and five 3s in that order? There are two
possibilities - see <code>?rep</code> or <code>?sort</code> for
help.</p>
</div>
</div>
</div>
<div id="accordionSolution9" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution9" aria-expanded="false" aria-controls="collapseSolution9">
  <h4 class="accordion-header" id="headingSolution9"> Show me the solution </h4>
</button>
<div id="collapseSolution9" class="accordion-collapse collapse" aria-labelledby="headingSolution9" data-bs-parent="#accordionSolution9">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb143">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rep</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] 1 1 1 1 1 2 2 2 2 2 3 3 3 3 3</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb145">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sort</span><span class="op">(</span><span class="fu">rep</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] 1 1 1 1 1 2 2 2 2 2 3 3 3 3 3</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="sequence-generation">Sequence generation<a class="anchor" aria-label="anchor" href="#sequence-generation"></a>
</h3>
<p>Another very useful function is <code>seq</code>, to generate a
sequence of numbers. For example, to generate a sequence of integers
from 1 to 20 by steps of 2, one would use:</p>
<div class="codewrapper sourceCode" id="cb147">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">seq</span><span class="op">(</span>from <span class="op">=</span> <span class="fl">1</span>, to <span class="op">=</span> <span class="fl">20</span>, by <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1]  1  3  5  7  9 11 13 15 17 19</code></pre>
</div>
<p>The default value of <code>by</code> is 1 and, given that the
generation of a sequence of one value to another with steps of 1 is
frequently used, there’s a shortcut:</p>
<div class="codewrapper sourceCode" id="cb149">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">seq</span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1 2 3 4 5</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb151">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">seq</span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span><span class="op">)</span> <span class="co">## default by</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1 2 3 4 5</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb153">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1 2 3 4 5</code></pre>
</div>
<p>To generate a sequence of numbers from 1 to 20 of final length of 3,
one would use:</p>
<div class="codewrapper sourceCode" id="cb155">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">seq</span><span class="op">(</span>from <span class="op">=</span> <span class="fl">1</span>, to <span class="op">=</span> <span class="fl">20</span>, length.out <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1]  1.0 10.5 20.0</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="random-samples-and-permutations">Random samples and permutations<a class="anchor" aria-label="anchor" href="#random-samples-and-permutations"></a>
</h3>
<p>A last group of useful functions are those that generate random data.
The first one, <code>sample</code>, generates a random permutation of
another vector. For example, to draw a random order to 10 students oral
exam, I first assign each student a number from 1 to ten (for instance
based on the alphabetic order of their name) and then:</p>
<div class="codewrapper sourceCode" id="cb157">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sample</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1]  9  4  7  1  2  5  3 10  6  8</code></pre>
</div>
<p>Without further arguments, <code>sample</code> will return a
permutation of all elements of the vector. If I want a random sample of
a certain size, I would set this value as the second argument. Below, I
sample 5 random letters from the alphabet contained in the pre-defined
<code>letters</code> vector:</p>
<div class="codewrapper sourceCode" id="cb159">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sample</span><span class="op">(</span><span class="va">letters</span>, <span class="fl">5</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "s" "a" "u" "x" "j"</code></pre>
</div>
<p>If I wanted an output larger than the input vector, or being able to
draw some elements multiple times, I would need to set the
<code>replace</code> argument to <code>TRUE</code>:</p>
<div class="codewrapper sourceCode" id="cb161">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sample</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">10</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] 2 1 5 5 1 1 5 5 2 2</code></pre>
</div>
<div id="challenge-11" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-11" class="callout-inner">
<h3 class="callout-title">Challenge:</h3>
<div class="callout-content">
<p>When trying the functions above out, you will have realised that the
samples are indeed random and that one doesn’t get the same permutation
twice. To be able to reproduce these random draws, one can set the
random number generation seed manually with <code>set.seed()</code>
before drawing the random sample.</p>
<p>Test this feature with your neighbour. First draw two random
permutations of <code>1:10</code> independently and observe that you get
different results.</p>
<p>Now set the seed with, for example, <code>set.seed(123)</code> and
repeat the random draw. Observe that you now get the same random
draws.</p>
<p>Repeat by setting a different seed.</p>
</div>
</div>
</div>
<div id="accordionSolution10" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution10" aria-expanded="false" aria-controls="collapseSolution10">
  <h4 class="accordion-header" id="headingSolution10"> Show me the solution </h4>
</button>
<div id="collapseSolution10" class="accordion-collapse collapse" aria-labelledby="headingSolution10" data-bs-parent="#accordionSolution10">
<div class="accordion-body">
<p>Different permutations</p>
<div class="codewrapper sourceCode" id="cb163">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sample</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1]  9  1  4  3  6  2  5  8 10  7</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb165">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sample</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1]  4  9  7  6  1 10  8  3  2  5</code></pre>
</div>
<p>Same permutations with seed 123</p>
<div class="codewrapper sourceCode" id="cb167">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="fu">sample</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1]  3 10  2  8  6  9  1  7  5  4</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb169">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="fu">sample</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1]  3 10  2  8  6  9  1  7  5  4</code></pre>
</div>
<p>A different seed</p>
<div class="codewrapper sourceCode" id="cb171">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu">sample</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1]  9  4  7  1  2  5  3 10  6  8</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb173">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu">sample</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1]  9  4  7  1  2  5  3 10  6  8</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="drawing-samples-from-a-normal-distribution">Drawing samples from a normal distribution<a class="anchor" aria-label="anchor" href="#drawing-samples-from-a-normal-distribution"></a>
</h3>
<p>The last function we are going to see is <code>rnorm</code>, that
draws a random sample from a normal distribution. Two normal
distributions of means 0 and 100 and standard deviations 1 and 5, noted
<em>N(0, 1)</em> and <em>N(100, 5)</em>, are shown below.</p>
<figure style="text-align: center"><img src="fig/23-starting-with-r-rendered-unnamed-chunk-61-1.png" alt="Two figures illustraing normal distributonis N(0,1) and N(100,5)." class="figure mx-auto d-block"><figcaption>
Two normal distributions: <em>N(0, 1)</em> on the left and <em>N(100,
5)</em> on the right.
</figcaption></figure><p>The three arguments, <code>n</code>, <code>mean</code> and
<code>sd</code>, define the size of the sample, and the parameters of
the normal distribution, i.e the mean and its standard deviation. The
defaults of the latter are 0 and 1.</p>
<div class="codewrapper sourceCode" id="cb175">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rnorm</span><span class="op">(</span><span class="fl">5</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1]  0.69641761  0.05351568 -1.31028350 -2.12306606 -0.20807859</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb177">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rnorm</span><span class="op">(</span><span class="fl">5</span>, <span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1]  1.3744268 -0.1164714  2.8344472  1.3690969  3.6510983</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb179">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rnorm</span><span class="op">(</span><span class="fl">5</span>, <span class="fl">100</span>, <span class="fl">5</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 106.45636  96.87448  95.62427 100.71678 107.12595</code></pre>
</div>
<p>Now that we have learned how to write scripts, and the basics of R’s
data structures, we are ready to start working with larger data, and
learn about data frames.</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>How to interact with R</li>
</ul>
</div>
</div>
</div>
</div>
</section></section><section id="aio-25-starting-with-data"><p>Content from <a href="25-starting-with-data.html">Starting with data</a></p>
<hr>
<p>Last updated on 2025-06-15 |

        <a href="https://github.com/carpentries-incubator/bioc-intro/edit/main/episodes/25-starting-with-data.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>First data analysis in R</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Describe what a <code>data.frame</code> is.</li>
<li>Load external data from a .csv file into a data frame.</li>
<li>Summarize the contents of a data frame.</li>
<li>Describe what a factor is.</li>
<li>Convert between strings and factors.</li>
<li>Reorder and rename factors.</li>
<li>Format dates.</li>
<li>Export and save data.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<blockquote>
<p>This episode is based on the Data Carpentries’s <em>Data Analysis and
Visualisation in R for Ecologists</em> lesson.</p>
</blockquote>
<section><h2 class="section-heading" id="presentation-of-the-gene-expression-data">Presentation of the gene expression data<a class="anchor" aria-label="anchor" href="#presentation-of-the-gene-expression-data"></a>
</h2>
<hr class="half-width">
<p>We are going to use part of the data published by <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5544260/" class="external-link">Blackmore
<em>et al.</em> (2017)</a>, <em>The effect of upper-respiratory
infection on transcriptomic changes in the CNS</em>. The goal of the
study was to determine the effect of an upper-respiratory infection on
changes in RNA transcription occurring in the cerebellum and spinal cord
post infection. Gender matched eight week old C57BL/6 mice were
inoculated with saline or with Influenza A by intranasal route and
transcriptomic changes in the cerebellum and spinal cord tissues were
evaluated by RNA-seq at days 0 (non-infected), 4 and 8.</p>
<p>The dataset is stored as a comma-separated values (CSV) file. Each
row holds information for a single RNA expression measurement, and the
first eleven columns represent:</p>
<table class="table">
<colgroup>
<col width="9%">
<col width="90%">
</colgroup>
<thead><tr class="header">
<th>Column</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>gene</td>
<td>The name of the gene that was measured</td>
</tr>
<tr class="even">
<td>sample</td>
<td>The name of the sample the gene expression was measured in</td>
</tr>
<tr class="odd">
<td>expression</td>
<td>The value of the gene expression</td>
</tr>
<tr class="even">
<td>organism</td>
<td>The organism/species - here all data stem from mice</td>
</tr>
<tr class="odd">
<td>age</td>
<td>The age of the mouse (all mice were 8 weeks here)</td>
</tr>
<tr class="even">
<td>sex</td>
<td>The sex of the mouse</td>
</tr>
<tr class="odd">
<td>infection</td>
<td>The infection state of the mouse, i.e. infected with Influenza A or
not infected.</td>
</tr>
<tr class="even">
<td>strain</td>
<td>The Influenza A strain.</td>
</tr>
<tr class="odd">
<td>time</td>
<td>The duration of the infection (in days).</td>
</tr>
<tr class="even">
<td>tissue</td>
<td>The tissue that was used for the gene expression experiment,
i.e. cerebellum or spinal cord.</td>
</tr>
<tr class="odd">
<td>mouse</td>
<td>The mouse unique identifier.</td>
</tr>
</tbody>
</table>
<p>We are going to use the R function <code>download.file()</code> to
download the CSV file that contains the gene expression data, and we
will use <code>read.csv()</code> to load into memory the content of the
CSV file as an object of class <code>data.frame</code>. Inside the
<code>download.file</code> command, the first entry is a character
string with the source URL. This source URL downloads a CSV file from a
GitHub repository. The text after the comma
(<code>"data/rnaseq.csv"</code>) is the destination of the file on your
local machine. You’ll need to have a folder on your machine called
<code>"data"</code> where you’ll download the file. So this command
downloads the remote file, names it <code>"rnaseq.csv"</code> and adds
it to a preexisting folder named <code>"data"</code>.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">download.file</span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://github.com/carpentries-incubator/bioc-intro/raw/main/episodes/data/rnaseq.csv"</span>,</span>
<span>              destfile <span class="op">=</span> <span class="st">"data/rnaseq.csv"</span><span class="op">)</span></span></code></pre>
</div>
<p>You are now ready to load the data:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span><span class="st">"data/rnaseq.csv"</span><span class="op">)</span></span></code></pre>
</div>
<p>This statement doesn’t produce any output because, as you might
recall, assignments don’t display anything. If we want to check that our
data has been loaded, we can see the contents of the data frame by
typing its name:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span></span></code></pre>
</div>
<p>Wow… that was a lot of output. At least it means the data loaded
properly. Let’s check the top (the first 6 lines) of this data frame
using the function <code>head()</code>:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">rna</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>     gene     sample expression     organism age    sex  infection  strain time
1     Asl GSM2545336       1170 Mus musculus   8 Female InfluenzaA C57BL/6    8
2    Apod GSM2545336      36194 Mus musculus   8 Female InfluenzaA C57BL/6    8
3 Cyp2d22 GSM2545336       4060 Mus musculus   8 Female InfluenzaA C57BL/6    8
4    Klk6 GSM2545336        287 Mus musculus   8 Female InfluenzaA C57BL/6    8
5   Fcrls GSM2545336         85 Mus musculus   8 Female InfluenzaA C57BL/6    8
6  Slc2a4 GSM2545336        782 Mus musculus   8 Female InfluenzaA C57BL/6    8
      tissue mouse ENTREZID
1 Cerebellum    14   109900
2 Cerebellum    14    11815
3 Cerebellum    14    56448
4 Cerebellum    14    19144
5 Cerebellum    14    80891
6 Cerebellum    14    20528
                                                                       product
1                               argininosuccinate lyase, transcript variant X1
2                                       apolipoprotein D, transcript variant 3
3 cytochrome P450, family 2, subfamily d, polypeptide 22, transcript variant 2
4                         kallikrein related-peptidase 6, transcript variant 2
5                Fc receptor-like S, scavenger receptor, transcript variant X1
6          solute carrier family 2 (facilitated glucose transporter), member 4
     ensembl_gene_id external_synonym chromosome_name   gene_biotype
1 ENSMUSG00000025533    2510006M18Rik               5 protein_coding
2 ENSMUSG00000022548             &lt;NA&gt;              16 protein_coding
3 ENSMUSG00000061740             2D22              15 protein_coding
4 ENSMUSG00000050063             Bssp               7 protein_coding
5 ENSMUSG00000015852    2810439C17Rik               3 protein_coding
6 ENSMUSG00000018566           Glut-4              11 protein_coding
                            phenotype_description
1           abnormal circulating amino acid level
2                      abnormal lipid homeostasis
3                        abnormal skin morphology
4                         abnormal cytokine level
5 decreased CD8-positive alpha-beta T cell number
6              abnormal circulating glucose level
  hsapiens_homolog_associated_gene_name
1                                   ASL
2                                  APOD
3                                CYP2D6
4                                  KLK6
5                                 FCRL2
6                                SLC2A4</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Try also</span></span>
<span><span class="co">## View(rna)</span></span></code></pre>
</div>
<p><strong>Note</strong></p>
<p><code>read.csv()</code> assumes that fields are delineated by commas,
however, in several countries, the comma is used as a decimal separator
and the semicolon (;) is used as a field delineator. If you want to read
in this type of files in R, you can use the <code>read.csv2()</code>
function. It behaves exactly like <code>read.csv()</code> but uses
different parameters for the decimal and the field separators. If you
are working with another format, they can be both specified by the user.
Check out the help for <code>read.csv()</code> by typing
<code>?read.csv</code> to learn more. There is also the
<code>read.delim()</code> function for reading tab separated data files.
It is important to note that all of these functions are actually wrapper
functions for the main <code>read.table()</code> function with different
arguments. As such, the data above could have also been loaded by using
<code>read.table()</code> with the separation argument as
<code>,</code>. The code is as follows:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">&lt;-</span> <span class="fu">read.table</span><span class="op">(</span>file <span class="op">=</span> <span class="st">"data/rnaseq.csv"</span>,</span>
<span>                  sep <span class="op">=</span> <span class="st">","</span>,</span>
<span>                  header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
</div>
<p>The header argument has to be set to TRUE to be able to read the
headers as by default <code>read.table()</code> has the header argument
set to FALSE.</p>
</section><section><h2 class="section-heading" id="what-are-data-frames">What are data frames?<a class="anchor" aria-label="anchor" href="#what-are-data-frames"></a>
</h2>
<hr class="half-width">
<p>Data frames are the <em>de facto</em> data structure for most tabular
data, and what we use for statistics and plotting.</p>
<p>A data frame can be created by hand, but most commonly they are
generated by the functions <code>read.csv()</code> or
<code>read.table()</code>; in other words, when importing spreadsheets
from your hard drive (or the web).</p>
<p>A data frame is the representation of data in the format of a table
where the columns are vectors that all have the same length. Because
columns are vectors, each column must contain a single type of data
(e.g., characters, integers, factors). For example, here is a figure
depicting a data frame comprising a numeric, a character, and a logical
vector.</p>
<figure style="text-align: center"><img src="fig/data-frame.svg" alt="Schematic representation of a tablular structure with different colours representing different columns and variables types." width="70%" class="figure mx-auto d-block"><figcaption>
Conceptual representation of a data.frame
</figcaption></figure><p>We can see this when inspecting the <b>str</b>ucture of a data frame
with the function <code>str()</code>:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">str</span><span class="op">(</span><span class="va">rna</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>'data.frame':	32428 obs. of  19 variables:
 $ gene                                 : chr  "Asl" "Apod" "Cyp2d22" "Klk6" ...
 $ sample                               : chr  "GSM2545336" "GSM2545336" "GSM2545336" "GSM2545336" ...
 $ expression                           : int  1170 36194 4060 287 85 782 1619 288 43217 1071 ...
 $ organism                             : chr  "Mus musculus" "Mus musculus" "Mus musculus" "Mus musculus" ...
 $ age                                  : int  8 8 8 8 8 8 8 8 8 8 ...
 $ sex                                  : chr  "Female" "Female" "Female" "Female" ...
 $ infection                            : chr  "InfluenzaA" "InfluenzaA" "InfluenzaA" "InfluenzaA" ...
 $ strain                               : chr  "C57BL/6" "C57BL/6" "C57BL/6" "C57BL/6" ...
 $ time                                 : int  8 8 8 8 8 8 8 8 8 8 ...
 $ tissue                               : chr  "Cerebellum" "Cerebellum" "Cerebellum" "Cerebellum" ...
 $ mouse                                : int  14 14 14 14 14 14 14 14 14 14 ...
 $ ENTREZID                             : int  109900 11815 56448 19144 80891 20528 97827 118454 18823 14696 ...
 $ product                              : chr  "argininosuccinate lyase, transcript variant X1" "apolipoprotein D, transcript variant 3" "cytochrome P450, family 2, subfamily d, polypeptide 22, transcript variant 2" "kallikrein related-peptidase 6, transcript variant 2" ...
 $ ensembl_gene_id                      : chr  "ENSMUSG00000025533" "ENSMUSG00000022548" "ENSMUSG00000061740" "ENSMUSG00000050063" ...
 $ external_synonym                     : chr  "2510006M18Rik" NA "2D22" "Bssp" ...
 $ chromosome_name                      : chr  "5" "16" "15" "7" ...
 $ gene_biotype                         : chr  "protein_coding" "protein_coding" "protein_coding" "protein_coding" ...
 $ phenotype_description                : chr  "abnormal circulating amino acid level" "abnormal lipid homeostasis" "abnormal skin morphology" "abnormal cytokine level" ...
 $ hsapiens_homolog_associated_gene_name: chr  "ASL" "APOD" "CYP2D6" "KLK6" ...</code></pre>
</div>
</section><section><h2 class="section-heading" id="inspecting-data-frame-objects">Inspecting <code>data.frame</code> Objects<a class="anchor" aria-label="anchor" href="#inspecting-data-frame-objects"></a>
</h2>
<hr class="half-width">
<p>We already saw how the functions <code>head()</code> and
<code>str()</code> can be useful to check the content and the structure
of a data frame. Here is a non-exhaustive list of functions to get a
sense of the content/structure of the data. Let’s try them out!</p>
<p><strong>Size</strong>:</p>
<ul>
<li>
<code>dim(rna)</code> - returns a vector with the number of rows as
the first element, and the number of columns as the second element (the
<strong>dim</strong>ensions of the object).</li>
<li>
<code>nrow(rna)</code> - returns the number of rows.</li>
<li>
<code>ncol(rna)</code> - returns the number of columns.</li>
</ul>
<p><strong>Content</strong>:</p>
<ul>
<li>
<code>head(rna)</code> - shows the first 6 rows.</li>
<li>
<code>tail(rna)</code> - shows the last 6 rows.</li>
</ul>
<p><strong>Names</strong>:</p>
<ul>
<li>
<code>names(rna)</code> - returns the column names (synonym of
<code>colnames()</code> for <code>data.frame</code> objects).</li>
<li>
<code>rownames(rna)</code> - returns the row names.</li>
</ul>
<p><strong>Summary</strong>:</p>
<ul>
<li>
<code>str(rna)</code> - structure of the object and information
about the class, length and content of each column.</li>
<li>
<code>summary(rna)</code> - summary statistics for each column.</li>
</ul>
<p>Note: most of these functions are “generic”, they can be used on
other types of objects besides <code>data.frame</code>.</p>
<div id="challenge" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge" class="callout-inner">
<h3 class="callout-title">Challenge:</h3>
<div class="callout-content">
<p>Based on the output of <code>str(rna)</code>, can you answer the
following questions?</p>
<ul>
<li>What is the class of the object <code>rna</code>?</li>
<li>How many rows and how many columns are in this object?</li>
</ul>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<ul>
<li>class: data frame</li>
<li>how many rows: 32428, how many columns: 19</li>
</ul>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="indexing-and-subsetting-data-frames">Indexing and subsetting data frames<a class="anchor" aria-label="anchor" href="#indexing-and-subsetting-data-frames"></a>
</h2>
<hr class="half-width">
<p>Our <code>rna</code> data frame has rows and columns (it has 2
dimensions); if we want to extract some specific data from it, we need
to specify the “coordinates” we want. Row numbers come first, followed
by column numbers. However, note that different ways of specifying these
coordinates lead to results with different classes.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># first element in the first column of the data frame (as a vector)</span></span>
<span><span class="va">rna</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">]</span></span>
<span><span class="co"># first element in the 6th column (as a vector)</span></span>
<span><span class="va">rna</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">6</span><span class="op">]</span></span>
<span><span class="co"># first column of the data frame (as a vector)</span></span>
<span><span class="va">rna</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span></span>
<span><span class="co"># first column of the data frame (as a data.frame)</span></span>
<span><span class="va">rna</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="co"># first three elements in the 7th column (as a vector)</span></span>
<span><span class="va">rna</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fl">7</span><span class="op">]</span></span>
<span><span class="co"># the 3rd row of the data frame (as a data.frame)</span></span>
<span><span class="va">rna</span><span class="op">[</span><span class="fl">3</span>, <span class="op">]</span></span>
<span><span class="co"># equivalent to head_rna &lt;- head(rna)</span></span>
<span><span class="va">head_rna</span> <span class="op">&lt;-</span> <span class="va">rna</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span>, <span class="op">]</span></span>
<span><span class="va">head_rna</span></span></code></pre>
</div>
<p><code>:</code> is a special function that creates numeric vectors of
integers in increasing or decreasing order, test <code>1:10</code> and
<code>10:1</code> for instance. See section <span class="citation">@ref</span>(sec:genvec) for details.</p>
<p>You can also exclude certain indices of a data frame using the
“<code>-</code>” sign:</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span>          <span class="co">## The whole data frame, except the first column</span></span>
<span><span class="va">rna</span><span class="op">[</span><span class="op">-</span><span class="fu">c</span><span class="op">(</span><span class="fl">7</span><span class="op">:</span><span class="fl">66465</span><span class="op">)</span>, <span class="op">]</span> <span class="co">## Equivalent to head(rna)</span></span></code></pre>
</div>
<p>Data frames can be subsetted by calling indices (as shown
previously), but also by calling their column names directly:</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span><span class="op">[</span><span class="st">"gene"</span><span class="op">]</span>       <span class="co"># Result is a data.frame</span></span>
<span><span class="va">rna</span><span class="op">[</span>, <span class="st">"gene"</span><span class="op">]</span>     <span class="co"># Result is a vector</span></span>
<span><span class="va">rna</span><span class="op">[[</span><span class="st">"gene"</span><span class="op">]</span><span class="op">]</span>     <span class="co"># Result is a vector</span></span>
<span><span class="va">rna</span><span class="op">$</span><span class="va">gene</span>          <span class="co"># Result is a vector</span></span></code></pre>
</div>
<p>In RStudio, you can use the autocompletion feature to get the full
and correct names of the columns.</p>
<div id="challenge-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li><p>Create a <code>data.frame</code> (<code>rna_200</code>)
containing only the data in row 200 of the <code>rna</code>
dataset.</p></li>
<li><p>Notice how <code>nrow()</code> gave you the number of rows in a
<code>data.frame</code>?</p></li>
</ol>
<ul>
<li><p>Use that number to pull out just that last row in the initial
<code>rna</code> data frame.</p></li>
<li><p>Compare that with what you see as the last row using
<code>tail()</code> to make sure it’s meeting expectations.</p></li>
<li><p>Pull out that last row using <code>nrow()</code> instead of the
row number.</p></li>
<li><p>Create a new data frame (<code>rna_last</code>) from that last
row.</p></li>
</ul>
<ol start="3" style="list-style-type: decimal">
<li><p>Use <code>nrow()</code> to extract the row that is in the middle
of the <code>rna</code> dataframe. Store the content of this row in an
object named <code>rna_middle</code>.</p></li>
<li><p>Combine <code>nrow()</code> with the <code>-</code> notation
above to reproduce the behavior of <code>head(rna)</code>, keeping just
the first through 6th rows of the rna dataset.</p></li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## 1.</span></span>
<span><span class="va">rna_200</span> <span class="op">&lt;-</span> <span class="va">rna</span><span class="op">[</span><span class="fl">200</span>, <span class="op">]</span></span>
<span><span class="co">## 2.</span></span>
<span><span class="co">## Saving `n_rows` to improve readability and reduce duplication</span></span>
<span><span class="va">n_rows</span> <span class="op">&lt;-</span> <span class="fu">nrow</span><span class="op">(</span><span class="va">rna</span><span class="op">)</span></span>
<span><span class="va">rna_last</span> <span class="op">&lt;-</span> <span class="va">rna</span><span class="op">[</span><span class="va">n_rows</span>, <span class="op">]</span></span>
<span><span class="co">## 3.</span></span>
<span><span class="va">rna_middle</span> <span class="op">&lt;-</span> <span class="va">rna</span><span class="op">[</span><span class="va">n_rows</span> <span class="op">/</span> <span class="fl">2</span>, <span class="op">]</span></span>
<span><span class="co">## 4.</span></span>
<span><span class="va">rna_head</span> <span class="op">&lt;-</span> <span class="va">rna</span><span class="op">[</span><span class="op">-</span><span class="op">(</span><span class="fl">7</span><span class="op">:</span><span class="va">n_rows</span><span class="op">)</span>, <span class="op">]</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="factors">Factors<a class="anchor" aria-label="anchor" href="#factors"></a>
</h2>
<hr class="half-width">
<p>Factors represent <strong>categorical data</strong>. They are stored
as integers associated with labels and they can be ordered or unordered.
While factors look (and often behave) like character vectors, they are
actually treated as integer vectors by R. So you need to be very careful
when treating them as strings.</p>
<p>Once created, factors can only contain a pre-defined set of values,
known as <em>levels</em>. By default, R always sorts levels in
alphabetical order. For instance, if you have a factor with 2
levels:</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sex</span> <span class="op">&lt;-</span> <span class="fu">factor</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="st">"male"</span>, <span class="st">"female"</span>, <span class="st">"female"</span>, <span class="st">"male"</span>, <span class="st">"female"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>R will assign <code>1</code> to the level <code>"female"</code> and
<code>2</code> to the level <code>"male"</code> (because <code>f</code>
comes before <code>m</code>, even though the first element in this
vector is <code>"male"</code>). You can see this by using the function
<code>levels()</code> and you can find the number of levels using
<code>nlevels()</code>:</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">levels</span><span class="op">(</span><span class="va">sex</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "female" "male"  </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">nlevels</span><span class="op">(</span><span class="va">sex</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 2</code></pre>
</div>
<p>Sometimes, the order of the factors does not matter, other times you
might want to specify the order because it is meaningful (e.g., “low”,
“medium”, “high”), it improves your visualization, or it is required by
a particular type of analysis. Here, one way to reorder our levels in
the <code>sex</code> vector would be:</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sex</span> <span class="co">## current order</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] male   female female male   female
Levels: female male</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sex</span> <span class="op">&lt;-</span> <span class="fu">factor</span><span class="op">(</span><span class="va">sex</span>, levels <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"male"</span>, <span class="st">"female"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sex</span> <span class="co">## after re-ordering</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] male   female female male   female
Levels: male female</code></pre>
</div>
<p>In R’s memory, these factors are represented by integers (1, 2, 3),
but are more informative than integers because factors are self
describing: <code>"female"</code>, <code>"male"</code> is more
descriptive than <code>1</code>, <code>2</code>. Which one is “male”?
You wouldn’t be able to tell just from the integer data. Factors, on the
other hand, have this information built-in. It is particularly helpful
when there are many levels (like the gene biotype in our example
dataset).</p>
<p>When your data is stored as a factor, you can use the
<code>plot()</code> function to get a quick glance at the number of
observations represented by each factor level. Let’s look at the number
of males and females in our data.</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">sex</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/25-starting-with-data-rendered-firstfactorplot-1.png" alt="Barplot showing the number of 'male' and 'female' occurrences." class="figure mx-auto d-block"><figcaption>
Bar plot of the number of females and males.
</figcaption></figure><div class="section level3">
<h3 id="converting-to-character">Converting to character<a class="anchor" aria-label="anchor" href="#converting-to-character"></a>
</h3>
<p>If you need to convert a factor to a character vector, you use
<code>as.character(x)</code>.</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">as.character</span><span class="op">(</span><span class="va">sex</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "male"   "female" "female" "male"   "female"</code></pre>
</div>
<!-- ### Numeric factors -->
<!-- Converting factors where the levels appear as numbers (such as -->
<!-- concentration levels, or years) to a numeric vector is a little -->
<!-- trickier. The `as.numeric()` function returns the index values of the -->
<!-- factor, not its levels, so it will result in an entirely new (and -->
<!-- unwanted in this case) set of numbers.  One method to avoid this is to -->
<!-- convert factors to characters, and then to numbers.  Another method is -->
<!-- to use the `levels()` function. Compare: -->
<!-- ```{r} -->
<!-- year_fct <- factor(c(1990, 1983, 1977, 1998, 1990)) -->
<!-- as.numeric(year_fct)  ## Wrong! And there is no warning... -->
<!-- as.numeric(as.character(year_fct)) ## Works... -->
<!-- as.numeric(levels(year_fct))[year_fct] ## The recommended way. -->
<!-- ```

<!-- Notice that in the `levels()` approach, three important steps occur: -->
<!-- * We obtain all the factor levels using `levels(year_fct)` -->
<!-- * We convert these levels to numeric values using `as.numeric(levels(year_fct))` -->
<!-- * We then access these numeric values using the underlying integers of the -->
<!--   vector `year_fct` inside the square brackets -->
</div>
<div class="section level3">
<h3 id="renaming-factors">Renaming factors<a class="anchor" aria-label="anchor" href="#renaming-factors"></a>
</h3>
<p>If we want to rename these factor, it is sufficient to change its
levels:</p>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">levels</span><span class="op">(</span><span class="va">sex</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "male"   "female"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">levels</span><span class="op">(</span><span class="va">sex</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">"M"</span>, <span class="st">"F"</span><span class="op">)</span></span>
<span><span class="va">sex</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] M F F M F
Levels: M F</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">sex</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/25-starting-with-data-rendered-unnamed-chunk-17-1.png" alt="Barplot showing the number of 'M' and 'F' occurrences." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div id="challenge-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2" class="callout-inner">
<h3 class="callout-title">Challenge:</h3>
<div class="callout-content">
<ul>
<li>Rename “F” and “M” to “Female” and “Male” respectively.</li>
</ul>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">levels</span><span class="op">(</span><span class="va">sex</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "M" "F"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">levels</span><span class="op">(</span><span class="va">sex</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">"Male"</span>, <span class="st">"Female"</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="challenge-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-3" class="callout-inner">
<h3 class="callout-title">Challenge:</h3>
<div class="callout-content">
<p>We have seen how data frames are created when using
<code>read.csv()</code>, but they can also be created by hand with the
<code>data.frame()</code> function. There are a few mistakes in this
hand-crafted <code>data.frame</code>. Can you spot and fix them? Don’t
hesitate to experiment!</p>
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" tabindex="-1"></a>animal_data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb34-2"><a href="#cb34-2" tabindex="-1"></a>       <span class="at">animal =</span> <span class="fu">c</span>(dog, cat, sea cucumber, sea urchin),</span>
<span id="cb34-3"><a href="#cb34-3" tabindex="-1"></a>       <span class="at">feel =</span> <span class="fu">c</span>(<span class="st">"furry"</span>, <span class="st">"squishy"</span>, <span class="st">"spiny"</span>),</span>
<span id="cb34-4"><a href="#cb34-4" tabindex="-1"></a>       <span class="at">weight =</span> <span class="fu">c</span>(<span class="dv">45</span>, <span class="dv">8</span> <span class="fl">1.1</span>, <span class="fl">0.8</span>))</span></code></pre>
</div>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<ul>
<li>missing quotations around the names of the animals</li>
<li>missing one entry in the “feel” column (probably for one of the
furry animals)</li>
<li>missing one comma in the weight column</li>
</ul>
</div>
</div>
</div>
</div>
<div id="challenge-4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-4" class="callout-inner">
<h3 class="callout-title">Challenge:</h3>
<div class="callout-content">
<p>Can you predict the class for each of the columns in the following
example?</p>
<p>Check your guesses using <code>str(country_climate)</code>:</p>
<ul>
<li><p>Are they what you expected? Why? Why not?</p></li>
<li><p>Try again by adding <code>stringsAsFactors = TRUE</code> after
the last variable when creating the data frame. What is happening now?
<code>stringsAsFactors</code> can also be set when reading text-based
spreadsheets into R using <code>read.csv()</code>.</p></li>
</ul>
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">country_climate</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span></span>
<span>       country <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"Canada"</span>, <span class="st">"Panama"</span>, <span class="st">"South Africa"</span>, <span class="st">"Australia"</span><span class="op">)</span>,</span>
<span>       climate <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"cold"</span>, <span class="st">"hot"</span>, <span class="st">"temperate"</span>, <span class="st">"hot/temperate"</span><span class="op">)</span>,</span>
<span>       temperature <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">10</span>, <span class="fl">30</span>, <span class="fl">18</span>, <span class="st">"15"</span><span class="op">)</span>,</span>
<span>       northern_hemisphere <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="cn">TRUE</span>, <span class="cn">TRUE</span>, <span class="cn">FALSE</span>, <span class="st">"FALSE"</span><span class="op">)</span>,</span>
<span>       has_kangaroo <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="cn">FALSE</span>, <span class="cn">FALSE</span>, <span class="cn">FALSE</span>, <span class="fl">1</span><span class="op">)</span></span>
<span>       <span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" data-bs-parent="#accordionSolution5" aria-labelledby="headingSolution5">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb36">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">country_climate</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span></span>
<span>       country <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"Canada"</span>, <span class="st">"Panama"</span>, <span class="st">"South Africa"</span>, <span class="st">"Australia"</span><span class="op">)</span>,</span>
<span>       climate <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"cold"</span>, <span class="st">"hot"</span>, <span class="st">"temperate"</span>, <span class="st">"hot/temperate"</span><span class="op">)</span>,</span>
<span>       temperature <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">10</span>, <span class="fl">30</span>, <span class="fl">18</span>, <span class="st">"15"</span><span class="op">)</span>,</span>
<span>       northern_hemisphere <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="cn">TRUE</span>, <span class="cn">TRUE</span>, <span class="cn">FALSE</span>, <span class="st">"FALSE"</span><span class="op">)</span>,</span>
<span>       has_kangaroo <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="cn">FALSE</span>, <span class="cn">FALSE</span>, <span class="cn">FALSE</span>, <span class="fl">1</span><span class="op">)</span></span>
<span>       <span class="op">)</span></span>
<span><span class="fu">str</span><span class="op">(</span><span class="va">country_climate</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>'data.frame':	4 obs. of  5 variables:
 $ country            : chr  "Canada" "Panama" "South Africa" "Australia"
 $ climate            : chr  "cold" "hot" "temperate" "hot/temperate"
 $ temperature        : chr  "10" "30" "18" "15"
 $ northern_hemisphere: chr  "TRUE" "TRUE" "FALSE" "FALSE"
 $ has_kangaroo       : num  0 0 0 1</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>The automatic conversion of data type is sometimes a blessing,
sometimes an annoyance. Be aware that it exists, learn the rules, and
double check that data you import in R are of the correct type within
your data frame. If not, use it to your advantage to detect mistakes
that might have been introduced during data entry (a letter in a column
that should only contain numbers for instance).</p>
<p>Learn more in this <a href="https://support.rstudio.com/hc/en-us/articles/218611977-Importing-Data-with-RStudio" class="external-link">RStudio
tutorial</a></p>
</div>
</section><section><h2 class="section-heading" id="matrices">Matrices<a class="anchor" aria-label="anchor" href="#matrices"></a>
</h2>
<hr class="half-width">
<p>Before proceeding, now that we have learnt about data frames, let’s
recap package installation and learn about a new data type, namely the
<code>matrix</code>. Like a <code>data.frame</code>, a matrix has two
dimensions, rows and columns. But the major difference is that all cells
in a <code>matrix</code> must be of the same type: <code>numeric</code>,
<code>character</code>, <code>logical</code>, … In that respect,
matrices are closer to a <code>vector</code> than a
<code>data.frame</code>.</p>
<p>The default constructor for a matrix is <code>matrix</code>. It takes
a vector of values to populate the matrix and the number of row and/or
columns<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>.
The values are sorted along the columns, as illustrated below.</p>
<div class="codewrapper sourceCode" id="cb38">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu">matrix</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">9</span>, ncol <span class="op">=</span> <span class="fl">3</span>, nrow <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">m</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>     [,1] [,2] [,3]
[1,]    1    4    7
[2,]    2    5    8
[3,]    3    6    9</code></pre>
</div>
<div id="challenge-5" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-5" class="callout-inner">
<h3 class="callout-title">Challenge:</h3>
<div class="callout-content">
<p>Using the function <code>installed.packages()</code>, create a
<code>character</code> matrix containing the information about all
packages currently installed on your computer. Explore it.</p>
</div>
</div>
</div>
<div id="accordionSolution6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution6" aria-expanded="false" aria-controls="collapseSolution6">
  <h4 class="accordion-header" id="headingSolution6"> Solution: </h4>
</button>
<div id="collapseSolution6" class="accordion-collapse collapse" data-bs-parent="#accordionSolution6" aria-labelledby="headingSolution6">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb40">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## create the matrix</span></span>
<span><span class="va">ip</span> <span class="op">&lt;-</span> <span class="fu">installed.packages</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">ip</span><span class="op">)</span></span>
<span><span class="co">## try also View(ip)</span></span>
<span><span class="co">## number of package</span></span>
<span><span class="fu">nrow</span><span class="op">(</span><span class="va">ip</span><span class="op">)</span></span>
<span><span class="co">## names of all installed packages</span></span>
<span><span class="fu">rownames</span><span class="op">(</span><span class="va">ip</span><span class="op">)</span></span>
<span><span class="co">## type of information we have about each package</span></span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">ip</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p>It is often useful to create large random data matrices as test data.
The exercise below asks you to create such a matrix with random data
drawn from a normal distribution of mean 0 and standard deviation 1,
which can be done with the <code>rnorm()</code> function.</p>
<div id="challenge-6" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-6" class="callout-inner">
<h3 class="callout-title">Challenge:</h3>
<div class="callout-content">
<p>Construct a matrix of dimension 1000 by 3 of normally distributed
data (mean 0, standard deviation 1)</p>
</div>
</div>
</div>
<div id="accordionSolution7" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution7" aria-expanded="false" aria-controls="collapseSolution7">
  <h4 class="accordion-header" id="headingSolution7"> Show me the solution </h4>
</button>
<div id="collapseSolution7" class="accordion-collapse collapse" data-bs-parent="#accordionSolution7" aria-labelledby="headingSolution7">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu">matrix</span><span class="op">(</span><span class="fu">rnorm</span><span class="op">(</span><span class="fl">3000</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="fu">dim</span><span class="op">(</span><span class="va">m</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1000    3</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb43">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">m</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>            [,1]        [,2]       [,3]
[1,] -0.56047565 -0.99579872 -0.5116037
[2,] -0.23017749 -1.03995504  0.2369379
[3,]  1.55870831 -0.01798024 -0.5415892
[4,]  0.07050839 -0.13217513  1.2192276
[5,]  0.12928774 -2.54934277  0.1741359
[6,]  1.71506499  1.04057346 -0.6152683</code></pre>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="formatting-dates">Formatting Dates<a class="anchor" aria-label="anchor" href="#formatting-dates"></a>
</h2>
<hr class="half-width">
<p>One of the most common issues that new (and experienced!) R users
have is converting date and time information into a variable that is
appropriate and usable during analyses.</p>
<div class="section level3">
<h3 id="note-on-dates-in-spreadsheet-programs">Note on dates in spreadsheet programs<a class="anchor" aria-label="anchor" href="#note-on-dates-in-spreadsheet-programs"></a>
</h3>
<p>Dates in spreadsheets are generally stored in a single column. While
this seems the most natural way to record dates, it actually is not best
practice. A spreadsheet application will display the dates in a
seemingly correct way (to a human observer) but how it actually handles
and stores the dates may be problematic. It is often much safer to store
dates with YEAR, MONTH and DAY in separate columns or as YEAR and
DAY-OF-YEAR in separate columns.</p>
<p>Spreadsheet programs such as LibreOffice, Microsoft Excel,
OpenOffice, Gnumeric, … have different (and often incompatible) ways of
encoding dates (even for the same program between versions and operating
systems). Additionally, Excel can <a href="https://nsaunders.wordpress.com/2012/10/22/gene-name-errors-and-excel-lessons-not-learned/" class="external-link">turn
things that aren’t dates into dates</a> (<span class="citation">@Zeeberg:2004</span>), for example names or identifiers
like MAR1, DEC1, OCT4. So if you’re avoiding the date format overall,
it’s easier to identify these issues.</p>
<p>The <a href="https://datacarpentry.org/spreadsheet-ecology-lesson/03-dates-as-data/index.html" class="external-link">Dates
as data</a> section of the Data Carpentry lesson provides additional
insights about pitfalls of dates with spreadsheets.</p>
<p>We are going to use the <code>ymd()</code> function from the package
<strong><code>lubridate</code></strong> (which belongs to the
<strong><code>tidyverse</code></strong>; learn more on the <a href="https://www.tidyverse.org/" class="external-link">tidyverse web site</a>).
<strong><code>lubridate</code></strong> gets installed as part of the
<strong><code>tidyverse</code></strong> installation. When you load the
<strong><code>tidyverse</code></strong>
(<code><a href="https://tidyverse.tidyverse.org" class="external-link">library(tidyverse)</a></code>), the core packages (the packages used
in most data analyses) get loaded.
<strong><code>lubridate</code></strong> however does not belong to the
core tidyverse, so you have to load it explicitly with
<code><a href="https://lubridate.tidyverse.org" class="external-link">library(lubridate)</a></code>.</p>
<p>Start by loading the required package:</p>
<div class="codewrapper sourceCode" id="cb45">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="st">"lubridate"</span><span class="op">)</span></span></code></pre>
</div>
<p><code>ymd()</code> takes a vector representing year, month, and day,
and converts it to a <code>Date</code> vector. <code>Date</code> is a
class of data recognized by R as being a date and can be manipulated as
such. The argument that the function requires is flexible, but, as a
best practice, is a character vector formatted as “YYYY-MM-DD”.</p>
<p>Let’s create a date object and inspect the structure:</p>
<div class="codewrapper sourceCode" id="cb46">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_date</span> <span class="op">&lt;-</span> <span class="fu">ymd</span><span class="op">(</span><span class="st">"2015-01-01"</span><span class="op">)</span></span>
<span><span class="fu">str</span><span class="op">(</span><span class="va">my_date</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> Date[1:1], format: "2015-01-01"</code></pre>
</div>
<p>Now let’s paste the year, month, and day separately - we get the same
result:</p>
<div class="codewrapper sourceCode" id="cb48">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># sep indicates the character to use to separate each component</span></span>
<span><span class="va">my_date</span> <span class="op">&lt;-</span> <span class="fu">ymd</span><span class="op">(</span><span class="fu">paste</span><span class="op">(</span><span class="st">"2015"</span>, <span class="st">"1"</span>, <span class="st">"1"</span>, sep <span class="op">=</span> <span class="st">"-"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">str</span><span class="op">(</span><span class="va">my_date</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> Date[1:1], format: "2015-01-01"</code></pre>
</div>
<p>Let’s now familiarise ourselves with a typical date manipulation
pipeline. The small data below has stored dates in different
<code>year</code>, <code>month</code> and <code>day</code> columns.</p>
<div class="codewrapper sourceCode" id="cb50">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span>year <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1996</span>, <span class="fl">1992</span>, <span class="fl">1987</span>, <span class="fl">1986</span>, <span class="fl">2000</span>, <span class="fl">1990</span>, <span class="fl">2002</span>, <span class="fl">1994</span>, <span class="fl">1997</span>, <span class="fl">1985</span><span class="op">)</span>,</span>
<span>                month <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">2</span>,  <span class="fl">3</span>,  <span class="fl">3</span>, <span class="fl">10</span>,  <span class="fl">1</span>,  <span class="fl">8</span>,  <span class="fl">3</span>,  <span class="fl">4</span>,  <span class="fl">5</span>,  <span class="fl">5</span><span class="op">)</span>,</span>
<span>                day <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">24</span>,  <span class="fl">8</span>,  <span class="fl">1</span>,  <span class="fl">5</span>,  <span class="fl">8</span>, <span class="fl">17</span>, <span class="fl">13</span>, <span class="fl">10</span>, <span class="fl">11</span>, <span class="fl">24</span><span class="op">)</span>,</span>
<span>                value <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">4</span>,  <span class="fl">5</span>,  <span class="fl">1</span>,  <span class="fl">9</span>,  <span class="fl">3</span>,  <span class="fl">8</span>, <span class="fl">10</span>,  <span class="fl">2</span>,  <span class="fl">6</span>,  <span class="fl">7</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">x</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>   year month day value
1  1996     2  24     4
2  1992     3   8     5
3  1987     3   1     1
4  1986    10   5     9
5  2000     1   8     3
6  1990     8  17     8
7  2002     3  13    10
8  1994     4  10     2
9  1997     5  11     6
10 1985     5  24     7</code></pre>
</div>
<p>Now we apply this function to the <code>x</code> dataset. We first
create a character vector from the <code>year</code>,
<code>month</code>, and <code>day</code> columns of <code>x</code> using
<code>paste()</code>:</p>
<div class="codewrapper sourceCode" id="cb52">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">paste</span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">year</span>, <span class="va">x</span><span class="op">$</span><span class="va">month</span>, <span class="va">x</span><span class="op">$</span><span class="va">day</span>, sep <span class="op">=</span> <span class="st">"-"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "1996-2-24" "1992-3-8"  "1987-3-1"  "1986-10-5" "2000-1-8"  "1990-8-17"
 [7] "2002-3-13" "1994-4-10" "1997-5-11" "1985-5-24"</code></pre>
</div>
<p>This character vector can be used as the argument for
<code>ymd()</code>:</p>
<div class="codewrapper sourceCode" id="cb54">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ymd</span><span class="op">(</span><span class="fu">paste</span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">year</span>, <span class="va">x</span><span class="op">$</span><span class="va">month</span>, <span class="va">x</span><span class="op">$</span><span class="va">day</span>, sep <span class="op">=</span> <span class="st">"-"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "1996-02-24" "1992-03-08" "1987-03-01" "1986-10-05" "2000-01-08"
 [6] "1990-08-17" "2002-03-13" "1994-04-10" "1997-05-11" "1985-05-24"</code></pre>
</div>
<p>The resulting <code>Date</code> vector can be added to <code>x</code>
as a new column called <code>date</code>:</p>
<div class="codewrapper sourceCode" id="cb56">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="va">date</span> <span class="op">&lt;-</span> <span class="fu">ymd</span><span class="op">(</span><span class="fu">paste</span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">year</span>, <span class="va">x</span><span class="op">$</span><span class="va">month</span>, <span class="va">x</span><span class="op">$</span><span class="va">day</span>, sep <span class="op">=</span> <span class="st">"-"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">str</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="co"># notice the new column, with 'date' as the class</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>'data.frame':	10 obs. of  5 variables:
 $ year : num  1996 1992 1987 1986 2000 ...
 $ month: num  2 3 3 10 1 8 3 4 5 5
 $ day  : num  24 8 1 5 8 17 13 10 11 24
 $ value: num  4 5 1 9 3 8 10 2 6 7
 $ date : Date, format: "1996-02-24" "1992-03-08" ...</code></pre>
</div>
<p>Let’s make sure everything worked correctly. One way to inspect the
new column is to use <code>summary()</code>:</p>
<div class="codewrapper sourceCode" id="cb58">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">summary</span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">date</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>        Min.      1st Qu.       Median         Mean      3rd Qu.         Max.
"1985-05-24" "1988-01-11" "1993-03-24" "1993-03-18" "1997-01-20" "2002-03-13" </code></pre>
</div>
<p>Note that <code>ymd()</code> expects to have the year, month and day,
in that order. If you have for instance day, month and year, you would
need <code>dmy()</code>.</p>
<div class="codewrapper sourceCode" id="cb60">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dmy</span><span class="op">(</span><span class="fu">paste</span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">day</span>, <span class="va">x</span><span class="op">$</span><span class="va">month</span>, <span class="va">x</span><span class="op">$</span><span class="va">year</span>, sep <span class="op">=</span> <span class="st">"-"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "1996-02-24" "1992-03-08" "1987-03-01" "1986-10-05" "2000-01-08"
 [6] "1990-08-17" "2002-03-13" "1994-04-10" "1997-05-11" "1985-05-24"</code></pre>
</div>
<p><code>lubdridate</code> has many functions to address all date
variations.</p>
</div>
</section><section><h2 class="section-heading" id="summary-of-r-objects">Summary of R objects<a class="anchor" aria-label="anchor" href="#summary-of-r-objects"></a>
</h2>
<hr class="half-width">
<p>So far, we have seen several types of R object varying in the number
of dimensions and whether they could store a single or multiple data
types:</p>
<ul>
<li>
<strong><code>vector</code></strong>: one dimension (they have a
length), single type of data.</li>
<li>
<strong><code>matrix</code></strong>: two dimensions, single type of
data.</li>
<li>
<strong><code>data.frame</code></strong>: two dimensions, one type
per column.</li>
</ul></section><section><h2 class="section-heading" id="lists">Lists<a class="anchor" aria-label="anchor" href="#lists"></a>
</h2>
<hr class="half-width">
<p>A data type that we haven’t seen yet, but that is useful to know, and
follows from the summary that we have just seen are lists:</p>
<ul>
<li>
<strong><code>list</code></strong>: one dimension, every item can be
of a different data type.</li>
</ul>
<p>Below, let’s create a list containing a vector of numbers,
characters, a matrix, a dataframe and another list:</p>
<div class="codewrapper sourceCode" id="cb62">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">l</span> <span class="op">&lt;-</span> <span class="fu">list</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="co">## numeric</span></span>
<span>          <span class="va">letters</span>, <span class="co">## character</span></span>
<span>          <span class="fu">installed.packages</span><span class="op">(</span><span class="op">)</span>, <span class="co">## a matrix</span></span>
<span>          <span class="va">cars</span>, <span class="co">## a data.frame</span></span>
<span>          <span class="fu">list</span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span> <span class="co">## a list</span></span>
<span><span class="fu">length</span><span class="op">(</span><span class="va">l</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 5</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb64">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">str</span><span class="op">(</span><span class="va">l</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>List of 5
 $ : int [1:10] 1 2 3 4 5 6 7 8 9 10
 $ : chr [1:26] "a" "b" "c" "d" ...
 $ : chr [1:165, 1:16] "abind" "askpass" "backports" "base64enc" ...
  ..- attr(*, "dimnames")=List of 2
  .. ..$ : chr [1:165] "abind" "askpass" "backports" "base64enc" ...
  .. ..$ : chr [1:16] "Package" "LibPath" "Version" "Priority" ...
 $ :'data.frame':	50 obs. of  2 variables:
  ..$ speed: num [1:50] 4 4 7 7 8 9 10 10 10 11 ...
  ..$ dist : num [1:50] 2 10 4 22 16 10 18 26 34 17 ...
 $ :List of 3
  ..$ : num 1
  ..$ : num 2
  ..$ : num 3</code></pre>
</div>
<p>List subsetting is done using <code>[]</code> to subset a new
sub-list or <code>[[]]</code> to extract a single element of that list
(using indices or names, if the list is named).</p>
<div class="codewrapper sourceCode" id="cb66">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">l</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="co">## first element</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1]  1  2  3  4  5  6  7  8  9 10</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb68">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">l</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span> <span class="co">## a list of length 2</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[[1]]
 [1]  1  2  3  4  5  6  7  8  9 10

[[2]]
 [1] "a" "b" "c" "d" "e" "f" "g" "h" "i" "j" "k" "l" "m" "n" "o" "p" "q" "r" "s"
[20] "t" "u" "v" "w" "x" "y" "z"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb70">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">l</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>   <span class="co">## a list of length 1</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[[1]]
 [1]  1  2  3  4  5  6  7  8  9 10</code></pre>
</div>
</section><section><h2 class="section-heading" id="sec:exportandsave">Exporting and saving tabular data<a class="anchor" aria-label="anchor" href="#sec:exportandsave"></a>
</h2>
<hr class="half-width">
<p>We have seen how to read a text-based spreadsheet into R using the
<code>read.table</code> family of functions. To export a
<code>data.frame</code> to a text-based spreadsheet, we can use the
<code>write.table</code> set of functions (<code>write.csv</code>,
<code>write.delim</code>, …). They all take the variable to be exported
and the file to be exported to. For example, to export the
<code>rna</code> data to the <code>my_rna.csv</code> file in the
<code>data_output</code> directory, we would execute:</p>
<div class="codewrapper sourceCode" id="cb72">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">write.csv</span><span class="op">(</span><span class="va">rna</span>, file <span class="op">=</span> <span class="st">"data_output/my_rna.csv"</span><span class="op">)</span></span></code></pre>
</div>
<p>This new csv file can now be shared with other collaborators who
aren’t familiar with R. Note that even though there are commas in some
of the fields in the <code>data.frame</code> (see for example the
“product” column), R will by default surround each field with quotes,
and thus we will be able to read it back into R correctly, despite also
using commas as column separators.</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>Tabular data in R</li>
</ul>
</div>
</div>
</div>
</section><div class="footnotes footnotes-end-of-document">
<hr>
<ol>
<li id="fn1"><p>Either the number of rows or columns are enough, as the
other one can be deduced from the length of the values. Try out what
happens if the values and number of rows/columns don’t add up.<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
</ol>
</div></section><section id="aio-30-dplyr"><p>Content from <a href="30-dplyr.html">Manipulating and analysing data with dplyr</a></p>
<hr>
<p>Last updated on 2025-06-15 |

        <a href="https://github.com/carpentries-incubator/bioc-intro/edit/main/episodes/30-dplyr.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>Data analysis in R using the tidyverse meta-package</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Describe the purpose of the <strong><code>dplyr</code></strong> and
<strong><code>tidyr</code></strong> packages.</li>
<li>Describe several of their functions that are extremely useful to
manipulate data.</li>
<li>Describe the concept of a wide and a long table format, and see how
to reshape a data frame from one format to the other one.</li>
<li>Demonstrate how to join tables.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<blockquote>
<p>This episode is based on the Data Carpentries’s <em>Data Analysis and
Visualisation in R for Ecologists</em> lesson.</p>
</blockquote>
<section><h2 class="section-heading" id="data-manipulation-using-dplyr-and-tidyr">Data manipulation using <strong><code>dplyr</code></strong> and
<strong><code>tidyr</code></strong>
<a class="anchor" aria-label="anchor" href="#data-manipulation-using-dplyr-and-tidyr"></a>
</h2>
<hr class="half-width">
<p>Bracket subsetting is handy, but it can be cumbersome and difficult
to read, especially for complicated operations.</p>
<p>Some packages can greatly facilitate our task when we manipulate
data. Packages in R are basically sets of additional functions that let
you do more stuff. The functions we’ve been using so far, like
<code>str()</code> or <code>data.frame()</code>, come built into R;
Loading packages can give you access to other specific functions. Before
you use a package for the first time you need to install it on your
machine, and then you should import it in every subsequent R session
when you need it.</p>
<ul>
<li><p>The package <strong><code>dplyr</code></strong> provides powerful
tools for data manipulation tasks. It is built to work directly with
data frames, with many manipulation tasks optimised.</p></li>
<li><p>As we will see latter on, sometimes we want a data frame to be
reshaped to be able to do some specific analyses or for visualisation.
The package <strong><code>tidyr</code></strong> addresses this common
problem of reshaping data and provides tools for manipulating data in a
tidy way.</p></li>
</ul>
<p>To learn more about <strong><code>dplyr</code></strong> and
<strong><code>tidyr</code></strong> after the workshop, you may want to
check out this <a href="https://raw.githubusercontent.com/rstudio/cheatsheets/main/data-transformation.pdf" class="external-link">handy
data transformation with <strong><code>dplyr</code></strong>
cheatsheet</a> and this <a href="https://raw.githubusercontent.com/rstudio/cheatsheets/main/tidyr.pdf" class="external-link">one
about <strong><code>tidyr</code></strong></a>.</p>
<ul>
<li>The <strong><code>tidyverse</code></strong> package is an
“umbrella-package” that installs several useful packages for data
analysis which work well together, such as
<strong><code>tidyr</code></strong>,
<strong><code>dplyr</code></strong>,
<strong><code>ggplot2</code></strong>,
<strong><code>tibble</code></strong>, etc. These packages help us to
work and interact with the data. They allow us to do many things with
your data, such as subsetting, transforming, visualising, etc.</li>
</ul>
<p>If you did the set up, you should have already installed the
tidyverse package. Check to see if you have it by trying to load in from
the library:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## load the tidyverse packages, incl. dplyr</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="st">"tidyverse"</span><span class="op">)</span></span></code></pre>
</div>
<p>If you got an error message
<code>there is no package called ‘tidyverse’</code> then you have not
installed the package yet for this version of R. To install the
<strong><code>tidyverse</code></strong> package type:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"tidyverse"</span><span class="op">)</span></span></code></pre>
</div>
<p>If you had to install the <strong><code>tidyverse</code></strong>
package, do not forget to load it in this R session by using the
<code>library()</code> command above!</p>
</section><section><h2 class="section-heading" id="loading-data-with-tidyverse">Loading data with tidyverse<a class="anchor" aria-label="anchor" href="#loading-data-with-tidyverse"></a>
</h2>
<hr class="half-width">
<p>Instead of <code>read.csv()</code>, we will read in our data using
the <code>read_csv()</code> function (notice the <code>_</code> instead
of the <code>.</code>), from the tidyverse package
<strong><code>readr</code></strong>.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">&lt;-</span> <span class="fu">read_csv</span><span class="op">(</span><span class="st">"data/rnaseq.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## view the data</span></span>
<span><span class="va">rna</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 32,428 × 19
   gene    sample  expression organism   age sex   infection strain  time tissue
   &lt;chr&gt;   &lt;chr&gt;        &lt;dbl&gt; &lt;chr&gt;    &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;     &lt;chr&gt;  &lt;dbl&gt; &lt;chr&gt;
 1 Asl     GSM254…       1170 Mus mus…     8 Fema… Influenz… C57BL…     8 Cereb…
 2 Apod    GSM254…      36194 Mus mus…     8 Fema… Influenz… C57BL…     8 Cereb…
 3 Cyp2d22 GSM254…       4060 Mus mus…     8 Fema… Influenz… C57BL…     8 Cereb…
 4 Klk6    GSM254…        287 Mus mus…     8 Fema… Influenz… C57BL…     8 Cereb…
 5 Fcrls   GSM254…         85 Mus mus…     8 Fema… Influenz… C57BL…     8 Cereb…
 6 Slc2a4  GSM254…        782 Mus mus…     8 Fema… Influenz… C57BL…     8 Cereb…
 7 Exd2    GSM254…       1619 Mus mus…     8 Fema… Influenz… C57BL…     8 Cereb…
 8 Gjc2    GSM254…        288 Mus mus…     8 Fema… Influenz… C57BL…     8 Cereb…
 9 Plp1    GSM254…      43217 Mus mus…     8 Fema… Influenz… C57BL…     8 Cereb…
10 Gnb4    GSM254…       1071 Mus mus…     8 Fema… Influenz… C57BL…     8 Cereb…
# ℹ 32,418 more rows
# ℹ 9 more variables: mouse &lt;dbl&gt;, ENTREZID &lt;dbl&gt;, product &lt;chr&gt;,
#   ensembl_gene_id &lt;chr&gt;, external_synonym &lt;chr&gt;, chromosome_name &lt;chr&gt;,
#   gene_biotype &lt;chr&gt;, phenotype_description &lt;chr&gt;,
#   hsapiens_homolog_associated_gene_name &lt;chr&gt;</code></pre>
</div>
<p>Notice that the class of the data is now referred to as a
“tibble”.</p>
<p>Tibbles tweak some of the behaviors of the data frame objects we
introduced in the previously. The data structure is very similar to a
data frame. For our purposes the only differences are that:</p>
<ol style="list-style-type: decimal">
<li><p>It displays the data type of each column under its name. Note
that &lt;<code>dbl</code>&gt; is a data type defined to hold numeric
values with decimal points.</p></li>
<li><p>It only prints the first few rows of data and only as many
columns as fit on one screen.</p></li>
</ol>
<p>We are now going to learn some of the most common
<strong><code>dplyr</code></strong> functions:</p>
<ul>
<li>
<code>select()</code>: subset columns</li>
<li>
<code>filter()</code>: subset rows on conditions</li>
<li>
<code>mutate()</code>: create new columns by using information from
other columns</li>
<li>
<code>group_by()</code> and <code>summarise()</code>: create summary
statistics on grouped data</li>
<li>
<code>arrange()</code>: sort results</li>
<li>
<code>count()</code>: count discrete values</li>
</ul></section><section><h2 class="section-heading" id="selecting-columns-and-filtering-rows">Selecting columns and filtering rows<a class="anchor" aria-label="anchor" href="#selecting-columns-and-filtering-rows"></a>
</h2>
<hr class="half-width">
<p>To select columns of a data frame, use <code>select()</code>. The
first argument to this function is the data frame (<code>rna</code>),
and the subsequent arguments are the columns to keep.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">select</span><span class="op">(</span><span class="va">rna</span>, <span class="va">gene</span>, <span class="va">sample</span>, <span class="va">tissue</span>, <span class="va">expression</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 32,428 × 4
   gene    sample     tissue     expression
   &lt;chr&gt;   &lt;chr&gt;      &lt;chr&gt;           &lt;dbl&gt;
 1 Asl     GSM2545336 Cerebellum       1170
 2 Apod    GSM2545336 Cerebellum      36194
 3 Cyp2d22 GSM2545336 Cerebellum       4060
 4 Klk6    GSM2545336 Cerebellum        287
 5 Fcrls   GSM2545336 Cerebellum         85
 6 Slc2a4  GSM2545336 Cerebellum        782
 7 Exd2    GSM2545336 Cerebellum       1619
 8 Gjc2    GSM2545336 Cerebellum        288
 9 Plp1    GSM2545336 Cerebellum      43217
10 Gnb4    GSM2545336 Cerebellum       1071
# ℹ 32,418 more rows</code></pre>
</div>
<p>To select all columns <em>except</em> certain ones, put a “-” in
front of the variable to exclude it.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">select</span><span class="op">(</span><span class="va">rna</span>, <span class="op">-</span><span class="va">tissue</span>, <span class="op">-</span><span class="va">organism</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 32,428 × 17
   gene    sample   expression   age sex   infection strain  time mouse ENTREZID
   &lt;chr&gt;   &lt;chr&gt;         &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;     &lt;chr&gt;  &lt;dbl&gt; &lt;dbl&gt;    &lt;dbl&gt;
 1 Asl     GSM2545…       1170     8 Fema… Influenz… C57BL…     8    14   109900
 2 Apod    GSM2545…      36194     8 Fema… Influenz… C57BL…     8    14    11815
 3 Cyp2d22 GSM2545…       4060     8 Fema… Influenz… C57BL…     8    14    56448
 4 Klk6    GSM2545…        287     8 Fema… Influenz… C57BL…     8    14    19144
 5 Fcrls   GSM2545…         85     8 Fema… Influenz… C57BL…     8    14    80891
 6 Slc2a4  GSM2545…        782     8 Fema… Influenz… C57BL…     8    14    20528
 7 Exd2    GSM2545…       1619     8 Fema… Influenz… C57BL…     8    14    97827
 8 Gjc2    GSM2545…        288     8 Fema… Influenz… C57BL…     8    14   118454
 9 Plp1    GSM2545…      43217     8 Fema… Influenz… C57BL…     8    14    18823
10 Gnb4    GSM2545…       1071     8 Fema… Influenz… C57BL…     8    14    14696
# ℹ 32,418 more rows
# ℹ 7 more variables: product &lt;chr&gt;, ensembl_gene_id &lt;chr&gt;,
#   external_synonym &lt;chr&gt;, chromosome_name &lt;chr&gt;, gene_biotype &lt;chr&gt;,
#   phenotype_description &lt;chr&gt;, hsapiens_homolog_associated_gene_name &lt;chr&gt;</code></pre>
</div>
<p>This will select all the variables in <code>rna</code> except
<code>tissue</code> and <code>organism</code>.</p>
<p>To choose rows based on a specific criteria, use
<code>filter()</code>:</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">filter</span><span class="op">(</span><span class="va">rna</span>, <span class="va">sex</span> <span class="op">==</span> <span class="st">"Male"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 14,740 × 19
   gene    sample  expression organism   age sex   infection strain  time tissue
   &lt;chr&gt;   &lt;chr&gt;        &lt;dbl&gt; &lt;chr&gt;    &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;     &lt;chr&gt;  &lt;dbl&gt; &lt;chr&gt;
 1 Asl     GSM254…        626 Mus mus…     8 Male  Influenz… C57BL…     4 Cereb…
 2 Apod    GSM254…      13021 Mus mus…     8 Male  Influenz… C57BL…     4 Cereb…
 3 Cyp2d22 GSM254…       2171 Mus mus…     8 Male  Influenz… C57BL…     4 Cereb…
 4 Klk6    GSM254…        448 Mus mus…     8 Male  Influenz… C57BL…     4 Cereb…
 5 Fcrls   GSM254…        180 Mus mus…     8 Male  Influenz… C57BL…     4 Cereb…
 6 Slc2a4  GSM254…        313 Mus mus…     8 Male  Influenz… C57BL…     4 Cereb…
 7 Exd2    GSM254…       2366 Mus mus…     8 Male  Influenz… C57BL…     4 Cereb…
 8 Gjc2    GSM254…        310 Mus mus…     8 Male  Influenz… C57BL…     4 Cereb…
 9 Plp1    GSM254…      53126 Mus mus…     8 Male  Influenz… C57BL…     4 Cereb…
10 Gnb4    GSM254…       1355 Mus mus…     8 Male  Influenz… C57BL…     4 Cereb…
# ℹ 14,730 more rows
# ℹ 9 more variables: mouse &lt;dbl&gt;, ENTREZID &lt;dbl&gt;, product &lt;chr&gt;,
#   ensembl_gene_id &lt;chr&gt;, external_synonym &lt;chr&gt;, chromosome_name &lt;chr&gt;,
#   gene_biotype &lt;chr&gt;, phenotype_description &lt;chr&gt;,
#   hsapiens_homolog_associated_gene_name &lt;chr&gt;</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">filter</span><span class="op">(</span><span class="va">rna</span>, <span class="va">sex</span> <span class="op">==</span> <span class="st">"Male"</span> <span class="op">&amp;</span> <span class="va">infection</span> <span class="op">==</span> <span class="st">"NonInfected"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 4,422 × 19
   gene    sample  expression organism   age sex   infection strain  time tissue
   &lt;chr&gt;   &lt;chr&gt;        &lt;dbl&gt; &lt;chr&gt;    &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;     &lt;chr&gt;  &lt;dbl&gt; &lt;chr&gt;
 1 Asl     GSM254…        535 Mus mus…     8 Male  NonInfec… C57BL…     0 Cereb…
 2 Apod    GSM254…      13668 Mus mus…     8 Male  NonInfec… C57BL…     0 Cereb…
 3 Cyp2d22 GSM254…       2008 Mus mus…     8 Male  NonInfec… C57BL…     0 Cereb…
 4 Klk6    GSM254…       1101 Mus mus…     8 Male  NonInfec… C57BL…     0 Cereb…
 5 Fcrls   GSM254…        375 Mus mus…     8 Male  NonInfec… C57BL…     0 Cereb…
 6 Slc2a4  GSM254…        249 Mus mus…     8 Male  NonInfec… C57BL…     0 Cereb…
 7 Exd2    GSM254…       3126 Mus mus…     8 Male  NonInfec… C57BL…     0 Cereb…
 8 Gjc2    GSM254…        791 Mus mus…     8 Male  NonInfec… C57BL…     0 Cereb…
 9 Plp1    GSM254…      98658 Mus mus…     8 Male  NonInfec… C57BL…     0 Cereb…
10 Gnb4    GSM254…       2437 Mus mus…     8 Male  NonInfec… C57BL…     0 Cereb…
# ℹ 4,412 more rows
# ℹ 9 more variables: mouse &lt;dbl&gt;, ENTREZID &lt;dbl&gt;, product &lt;chr&gt;,
#   ensembl_gene_id &lt;chr&gt;, external_synonym &lt;chr&gt;, chromosome_name &lt;chr&gt;,
#   gene_biotype &lt;chr&gt;, phenotype_description &lt;chr&gt;,
#   hsapiens_homolog_associated_gene_name &lt;chr&gt;</code></pre>
</div>
<p>Now let’s imagine we are interested in the human homologs of the
mouse genes analysed in this dataset. This information can be found in
the last column of the <code>rna</code> tibble, named
<code>hsapiens_homolog_associated_gene_name</code>. To visualise it
easily, we will create a new table containing just the 2 columns
<code>gene</code> and
<code>hsapiens_homolog_associated_gene_name</code>.</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu">select</span><span class="op">(</span><span class="va">rna</span>, <span class="va">gene</span>, <span class="va">hsapiens_homolog_associated_gene_name</span><span class="op">)</span></span>
<span><span class="va">genes</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 32,428 × 2
   gene    hsapiens_homolog_associated_gene_name
   &lt;chr&gt;   &lt;chr&gt;
 1 Asl     ASL
 2 Apod    APOD
 3 Cyp2d22 CYP2D6
 4 Klk6    KLK6
 5 Fcrls   FCRL2
 6 Slc2a4  SLC2A4
 7 Exd2    EXD2
 8 Gjc2    GJC2
 9 Plp1    PLP1
10 Gnb4    GNB4
# ℹ 32,418 more rows</code></pre>
</div>
<p>Some mouse genes have no human homologs. These can be retrieved using
<code>filter()</code> and the <code>is.na()</code> function, that
determines whether something is an <code>NA</code>.</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">filter</span><span class="op">(</span><span class="va">genes</span>, <span class="fu">is.na</span><span class="op">(</span><span class="va">hsapiens_homolog_associated_gene_name</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 4,290 × 2
   gene     hsapiens_homolog_associated_gene_name
   &lt;chr&gt;    &lt;chr&gt;
 1 Prodh    &lt;NA&gt;
 2 Tssk5    &lt;NA&gt;
 3 Vmn2r1   &lt;NA&gt;
 4 Gm10654  &lt;NA&gt;
 5 Hexa     &lt;NA&gt;
 6 Sult1a1  &lt;NA&gt;
 7 Gm6277   &lt;NA&gt;
 8 Tmem198b &lt;NA&gt;
 9 Adam1a   &lt;NA&gt;
10 Ebp      &lt;NA&gt;
# ℹ 4,280 more rows</code></pre>
</div>
<p>If we want to keep only mouse genes that have a human homolog, we can
insert a “!” symbol that negates the result, so we’re asking for every
row where hsapiens_homolog_associated_gene_name <em>is not</em> an
<code>NA</code>.</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">filter</span><span class="op">(</span><span class="va">genes</span>, <span class="op">!</span><span class="fu">is.na</span><span class="op">(</span><span class="va">hsapiens_homolog_associated_gene_name</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 28,138 × 2
   gene    hsapiens_homolog_associated_gene_name
   &lt;chr&gt;   &lt;chr&gt;
 1 Asl     ASL
 2 Apod    APOD
 3 Cyp2d22 CYP2D6
 4 Klk6    KLK6
 5 Fcrls   FCRL2
 6 Slc2a4  SLC2A4
 7 Exd2    EXD2
 8 Gjc2    GJC2
 9 Plp1    PLP1
10 Gnb4    GNB4
# ℹ 28,128 more rows</code></pre>
</div>
</section><section><h2 class="section-heading" id="pipes">Pipes<a class="anchor" aria-label="anchor" href="#pipes"></a>
</h2>
<hr class="half-width">
<p>What if you want to select and filter at the same time? There are
three ways to do this: use intermediate steps, nested functions, or
pipes.</p>
<p>With intermediate steps, you create a temporary data frame and use
that as input to the next function, like this:</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna2</span> <span class="op">&lt;-</span> <span class="fu">filter</span><span class="op">(</span><span class="va">rna</span>, <span class="va">sex</span> <span class="op">==</span> <span class="st">"Male"</span><span class="op">)</span></span>
<span><span class="va">rna3</span> <span class="op">&lt;-</span> <span class="fu">select</span><span class="op">(</span><span class="va">rna2</span>, <span class="va">gene</span>, <span class="va">sample</span>, <span class="va">tissue</span>, <span class="va">expression</span><span class="op">)</span></span>
<span><span class="va">rna3</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 14,740 × 4
   gene    sample     tissue     expression
   &lt;chr&gt;   &lt;chr&gt;      &lt;chr&gt;           &lt;dbl&gt;
 1 Asl     GSM2545340 Cerebellum        626
 2 Apod    GSM2545340 Cerebellum      13021
 3 Cyp2d22 GSM2545340 Cerebellum       2171
 4 Klk6    GSM2545340 Cerebellum        448
 5 Fcrls   GSM2545340 Cerebellum        180
 6 Slc2a4  GSM2545340 Cerebellum        313
 7 Exd2    GSM2545340 Cerebellum       2366
 8 Gjc2    GSM2545340 Cerebellum        310
 9 Plp1    GSM2545340 Cerebellum      53126
10 Gnb4    GSM2545340 Cerebellum       1355
# ℹ 14,730 more rows</code></pre>
</div>
<p>This is readable, but can clutter up your workspace with lots of
intermediate objects that you have to name individually. With multiple
steps, that can be hard to keep track of.</p>
<p>You can also nest functions (i.e. one function inside of another),
like this:</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna3</span> <span class="op">&lt;-</span> <span class="fu">select</span><span class="op">(</span><span class="fu">filter</span><span class="op">(</span><span class="va">rna</span>, <span class="va">sex</span> <span class="op">==</span> <span class="st">"Male"</span><span class="op">)</span>, <span class="va">gene</span>, <span class="va">sample</span>, <span class="va">tissue</span>, <span class="va">expression</span><span class="op">)</span></span>
<span><span class="va">rna3</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 14,740 × 4
   gene    sample     tissue     expression
   &lt;chr&gt;   &lt;chr&gt;      &lt;chr&gt;           &lt;dbl&gt;
 1 Asl     GSM2545340 Cerebellum        626
 2 Apod    GSM2545340 Cerebellum      13021
 3 Cyp2d22 GSM2545340 Cerebellum       2171
 4 Klk6    GSM2545340 Cerebellum        448
 5 Fcrls   GSM2545340 Cerebellum        180
 6 Slc2a4  GSM2545340 Cerebellum        313
 7 Exd2    GSM2545340 Cerebellum       2366
 8 Gjc2    GSM2545340 Cerebellum        310
 9 Plp1    GSM2545340 Cerebellum      53126
10 Gnb4    GSM2545340 Cerebellum       1355
# ℹ 14,730 more rows</code></pre>
</div>
<p>This is handy, but can be difficult to read if too many functions are
nested, as R evaluates the expression from the inside out (in this case,
filtering, then selecting).</p>
<p>The last option, <em>pipes</em>, are a recent addition to R. Pipes
let you take the output of one function and send it directly to the
next, which is useful when you need to do many things to the same
dataset.</p>
<p>Pipes in R look like <code>%&gt;%</code> (made available via the
<strong><code>magrittr</code></strong> package) or <code>|&gt;</code>
(through base R). If you use RStudio, you can type the pipe with
<kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>M</kbd> if you have a PC or
<kbd>Cmd</kbd> + <kbd>Shift</kbd> + <kbd>M</kbd> if you have a Mac.</p>
<p>In the above code, we use the pipe to send the <code>rna</code>
dataset first through <code>filter()</code> to keep rows where
<code>sex</code> is Male, then through <code>select()</code> to keep
only the <code>gene</code>, <code>sample</code>, <code>tissue</code>,
and <code>expression</code>columns.</p>
<p>The pipe <code>|&gt;</code> takes the object on its left and passes
it directly as the first argument to the function on its right, we don’t
need to explicitly include the data frame as an argument to the
<code>filter()</code> and <code>select()</code> functions any more.</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">sex</span> <span class="op">==</span> <span class="st">"Male"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">gene</span>, <span class="va">sample</span>, <span class="va">tissue</span>, <span class="va">expression</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 14,740 × 4
   gene    sample     tissue     expression
   &lt;chr&gt;   &lt;chr&gt;      &lt;chr&gt;           &lt;dbl&gt;
 1 Asl     GSM2545340 Cerebellum        626
 2 Apod    GSM2545340 Cerebellum      13021
 3 Cyp2d22 GSM2545340 Cerebellum       2171
 4 Klk6    GSM2545340 Cerebellum        448
 5 Fcrls   GSM2545340 Cerebellum        180
 6 Slc2a4  GSM2545340 Cerebellum        313
 7 Exd2    GSM2545340 Cerebellum       2366
 8 Gjc2    GSM2545340 Cerebellum        310
 9 Plp1    GSM2545340 Cerebellum      53126
10 Gnb4    GSM2545340 Cerebellum       1355
# ℹ 14,730 more rows</code></pre>
</div>
<p>Some may find it helpful to read the pipe like the word “then”. For
instance, in the above example, we took the data frame <code>rna</code>,
<em>then</em> we <code>filter</code>ed for rows with
<code>sex == "Male"</code>, <em>then</em> we <code>select</code>ed
columns <code>gene</code>, <code>sample</code>, <code>tissue</code>, and
<code>expression</code>.</p>
<p>The <strong><code>dplyr</code></strong> functions by themselves are
somewhat simple, but by combining them into linear workflows with the
pipe, we can accomplish more complex manipulations of data frames.</p>
<p>If we want to create a new object with this smaller version of the
data, we can assign it a new name:</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna3</span> <span class="op">&lt;-</span> <span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">sex</span> <span class="op">==</span> <span class="st">"Male"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">gene</span>, <span class="va">sample</span>, <span class="va">tissue</span>, <span class="va">expression</span><span class="op">)</span></span>
<span></span>
<span><span class="va">rna3</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 14,740 × 4
   gene    sample     tissue     expression
   &lt;chr&gt;   &lt;chr&gt;      &lt;chr&gt;           &lt;dbl&gt;
 1 Asl     GSM2545340 Cerebellum        626
 2 Apod    GSM2545340 Cerebellum      13021
 3 Cyp2d22 GSM2545340 Cerebellum       2171
 4 Klk6    GSM2545340 Cerebellum        448
 5 Fcrls   GSM2545340 Cerebellum        180
 6 Slc2a4  GSM2545340 Cerebellum        313
 7 Exd2    GSM2545340 Cerebellum       2366
 8 Gjc2    GSM2545340 Cerebellum        310
 9 Plp1    GSM2545340 Cerebellum      53126
10 Gnb4    GSM2545340 Cerebellum       1355
# ℹ 14,730 more rows</code></pre>
</div>
<div id="challenge" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge" class="callout-inner">
<h3 class="callout-title">Challenge:</h3>
<div class="callout-content">
<p>Using pipes, subset the <code>rna</code> data to keep observations in
female mice at time 0, where the gene has an expression higher than
50000, and retain only the columns <code>gene</code>,
<code>sample</code>, <code>time</code>, <code>expression</code> and
<code>age</code>.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">expression</span> <span class="op">&gt;</span> <span class="fl">50000</span>,</span>
<span>         <span class="va">sex</span> <span class="op">==</span> <span class="st">"Female"</span>,</span>
<span>         <span class="va">time</span> <span class="op">==</span> <span class="fl">0</span> <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">gene</span>, <span class="va">sample</span>, <span class="va">time</span>, <span class="va">expression</span>, <span class="va">age</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 9 × 5
  gene   sample      time expression   age
  &lt;chr&gt;  &lt;chr&gt;      &lt;dbl&gt;      &lt;dbl&gt; &lt;dbl&gt;
1 Plp1   GSM2545337     0     101241     8
2 Atp1b1 GSM2545337     0      53260     8
3 Plp1   GSM2545338     0      96534     8
4 Atp1b1 GSM2545338     0      50614     8
5 Plp1   GSM2545348     0     102790     8
6 Atp1b1 GSM2545348     0      59544     8
7 Plp1   GSM2545353     0      71237     8
8 Glul   GSM2545353     0      52451     8
9 Atp1b1 GSM2545353     0      61451     8</code></pre>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="mutate">Mutate<a class="anchor" aria-label="anchor" href="#mutate"></a>
</h2>
<hr class="half-width">
<p>Frequently you’ll want to create new columns based on the values of
existing columns, for example to do unit conversions, or to find the
ratio of values in two columns. For this we’ll use
<code>mutate()</code>.</p>
<p>To create a new column of time in hours:</p>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span>time_hours <span class="op">=</span> <span class="va">time</span> <span class="op">*</span> <span class="fl">24</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">time</span>, <span class="va">time_hours</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 32,428 × 2
    time time_hours
   &lt;dbl&gt;      &lt;dbl&gt;
 1     8        192
 2     8        192
 3     8        192
 4     8        192
 5     8        192
 6     8        192
 7     8        192
 8     8        192
 9     8        192
10     8        192
# ℹ 32,418 more rows</code></pre>
</div>
<p>You can also create a second new column based on the first new column
within the same call of <code>mutate()</code>:</p>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span>time_hours <span class="op">=</span> <span class="va">time</span> <span class="op">*</span> <span class="fl">24</span>,</span>
<span>         time_mn <span class="op">=</span> <span class="va">time_hours</span> <span class="op">*</span> <span class="fl">60</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">time</span>, <span class="va">time_hours</span>, <span class="va">time_mn</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 32,428 × 3
    time time_hours time_mn
   &lt;dbl&gt;      &lt;dbl&gt;   &lt;dbl&gt;
 1     8        192   11520
 2     8        192   11520
 3     8        192   11520
 4     8        192   11520
 5     8        192   11520
 6     8        192   11520
 7     8        192   11520
 8     8        192   11520
 9     8        192   11520
10     8        192   11520
# ℹ 32,418 more rows</code></pre>
</div>
<div id="challenge-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Create a new data frame from the <code>rna</code> data that meets the
following criteria: contains only the <code>gene</code>,
<code>chromosome_name</code>, <code>phenotype_description</code>,
<code>sample</code>, and <code>expression</code> columns. The expression
values should be log-transformed. This data frame must only contain
genes located on sex chromosomes, associated with a
phenotype_description, and with a log expression higher than 5.</p>
<p><strong>Hint</strong>: think about how the commands should be ordered
to produce this data frame!</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span>expression <span class="op">=</span> <span class="fu">log</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">gene</span>, <span class="va">chromosome_name</span>, <span class="va">phenotype_description</span>, <span class="va">sample</span>, <span class="va">expression</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">chromosome_name</span> <span class="op">==</span> <span class="st">"X"</span> <span class="op">|</span> <span class="va">chromosome_name</span> <span class="op">==</span> <span class="st">"Y"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="op">!</span><span class="fu">is.na</span><span class="op">(</span><span class="va">phenotype_description</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">expression</span> <span class="op">&gt;</span> <span class="fl">5</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 649 × 5
   gene   chromosome_name phenotype_description                sample expression
   &lt;chr&gt;  &lt;chr&gt;           &lt;chr&gt;                                &lt;chr&gt;       &lt;dbl&gt;
 1 Plp1   X               abnormal CNS glial cell morphology   GSM25…      10.7
 2 Slc7a3 X               decreased body length                GSM25…       5.46
 3 Plxnb3 X               abnormal coat appearance             GSM25…       6.58
 4 Rbm3   X               abnormal liver morphology            GSM25…       9.32
 5 Cfp    X               abnormal cardiovascular system phys… GSM25…       6.18
 6 Ebp    X               abnormal embryonic erythrocyte morp… GSM25…       6.68
 7 Cd99l2 X               abnormal cellular extravasation      GSM25…       8.04
 8 Piga   X               abnormal brain development           GSM25…       6.06
 9 Pim2   X               decreased T cell proliferation       GSM25…       7.11
10 Itm2a  X               no abnormal phenotype detected       GSM25…       7.48
# ℹ 639 more rows</code></pre>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="split-apply-combine-data-analysis">Split-apply-combine data analysis<a class="anchor" aria-label="anchor" href="#split-apply-combine-data-analysis"></a>
</h2>
<hr class="half-width">
<p>Many data analysis tasks can be approached using the
<em>split-apply-combine</em> paradigm: split the data into groups, apply
some analysis to each group, and then combine the results.
<strong><code>dplyr</code></strong> makes this very easy through the use
of the <code>group_by()</code> function.</p>
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 32,428 × 19
# Groups:   gene [1,474]
   gene    sample  expression organism   age sex   infection strain  time tissue
   &lt;chr&gt;   &lt;chr&gt;        &lt;dbl&gt; &lt;chr&gt;    &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;     &lt;chr&gt;  &lt;dbl&gt; &lt;chr&gt;
 1 Asl     GSM254…       1170 Mus mus…     8 Fema… Influenz… C57BL…     8 Cereb…
 2 Apod    GSM254…      36194 Mus mus…     8 Fema… Influenz… C57BL…     8 Cereb…
 3 Cyp2d22 GSM254…       4060 Mus mus…     8 Fema… Influenz… C57BL…     8 Cereb…
 4 Klk6    GSM254…        287 Mus mus…     8 Fema… Influenz… C57BL…     8 Cereb…
 5 Fcrls   GSM254…         85 Mus mus…     8 Fema… Influenz… C57BL…     8 Cereb…
 6 Slc2a4  GSM254…        782 Mus mus…     8 Fema… Influenz… C57BL…     8 Cereb…
 7 Exd2    GSM254…       1619 Mus mus…     8 Fema… Influenz… C57BL…     8 Cereb…
 8 Gjc2    GSM254…        288 Mus mus…     8 Fema… Influenz… C57BL…     8 Cereb…
 9 Plp1    GSM254…      43217 Mus mus…     8 Fema… Influenz… C57BL…     8 Cereb…
10 Gnb4    GSM254…       1071 Mus mus…     8 Fema… Influenz… C57BL…     8 Cereb…
# ℹ 32,418 more rows
# ℹ 9 more variables: mouse &lt;dbl&gt;, ENTREZID &lt;dbl&gt;, product &lt;chr&gt;,
#   ensembl_gene_id &lt;chr&gt;, external_synonym &lt;chr&gt;, chromosome_name &lt;chr&gt;,
#   gene_biotype &lt;chr&gt;, phenotype_description &lt;chr&gt;,
#   hsapiens_homolog_associated_gene_name &lt;chr&gt;</code></pre>
</div>
<p>The <code>group_by()</code> function doesn’t perform any data
processing, it groups the data into subsets: in the example above, our
initial <code>tibble</code> of 32428 observations is split into 1474
groups based on the <code>gene</code> variable.</p>
<p>We could similarly decide to group the tibble by the samples:</p>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">sample</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 32,428 × 19
# Groups:   sample [22]
   gene    sample  expression organism   age sex   infection strain  time tissue
   &lt;chr&gt;   &lt;chr&gt;        &lt;dbl&gt; &lt;chr&gt;    &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;     &lt;chr&gt;  &lt;dbl&gt; &lt;chr&gt;
 1 Asl     GSM254…       1170 Mus mus…     8 Fema… Influenz… C57BL…     8 Cereb…
 2 Apod    GSM254…      36194 Mus mus…     8 Fema… Influenz… C57BL…     8 Cereb…
 3 Cyp2d22 GSM254…       4060 Mus mus…     8 Fema… Influenz… C57BL…     8 Cereb…
 4 Klk6    GSM254…        287 Mus mus…     8 Fema… Influenz… C57BL…     8 Cereb…
 5 Fcrls   GSM254…         85 Mus mus…     8 Fema… Influenz… C57BL…     8 Cereb…
 6 Slc2a4  GSM254…        782 Mus mus…     8 Fema… Influenz… C57BL…     8 Cereb…
 7 Exd2    GSM254…       1619 Mus mus…     8 Fema… Influenz… C57BL…     8 Cereb…
 8 Gjc2    GSM254…        288 Mus mus…     8 Fema… Influenz… C57BL…     8 Cereb…
 9 Plp1    GSM254…      43217 Mus mus…     8 Fema… Influenz… C57BL…     8 Cereb…
10 Gnb4    GSM254…       1071 Mus mus…     8 Fema… Influenz… C57BL…     8 Cereb…
# ℹ 32,418 more rows
# ℹ 9 more variables: mouse &lt;dbl&gt;, ENTREZID &lt;dbl&gt;, product &lt;chr&gt;,
#   ensembl_gene_id &lt;chr&gt;, external_synonym &lt;chr&gt;, chromosome_name &lt;chr&gt;,
#   gene_biotype &lt;chr&gt;, phenotype_description &lt;chr&gt;,
#   hsapiens_homolog_associated_gene_name &lt;chr&gt;</code></pre>
</div>
<p>Here our initial <code>tibble</code> of 32428 observations is split
into 22 groups based on the <code>sample</code> variable.</p>
<p>Once the data has been grouped, subsequent operations will be applied
on each group independently.</p>
<div class="section level3">
<h3 id="the-summarise-function">The <code>summarise()</code> function<a class="anchor" aria-label="anchor" href="#the-summarise-function"></a>
</h3>
<p><code>group_by()</code> is often used together with
<code>summarise()</code>, which collapses each group into a single-row
summary of that group.</p>
<p><code>group_by()</code> takes as arguments the column names that
contain the <strong>categorical</strong> variables for which you want to
calculate the summary statistics. So to compute the mean
<code>expression</code> by gene:</p>
<div class="codewrapper sourceCode" id="cb39">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean_expression <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 1,474 × 2
   gene     mean_expression
   &lt;chr&gt;              &lt;dbl&gt;
 1 AI504432         1053.
 2 AW046200          131.
 3 AW551984          295.
 4 Aamp             4751.
 5 Abca12              4.55
 6 Abcc8            2498.
 7 Abhd14a           525.
 8 Abi2             4909.
 9 Abi3bp           1002.
10 Abl2             2124.
# ℹ 1,464 more rows</code></pre>
</div>
<p>We could also want to calculate the mean expression levels of all
genes in each sample:</p>
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">sample</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean_expression <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 22 × 2
   sample     mean_expression
   &lt;chr&gt;                &lt;dbl&gt;
 1 GSM2545336           2062.
 2 GSM2545337           1766.
 3 GSM2545338           1668.
 4 GSM2545339           1696.
 5 GSM2545340           1682.
 6 GSM2545341           1638.
 7 GSM2545342           1594.
 8 GSM2545343           2107.
 9 GSM2545344           1712.
10 GSM2545345           1700.
# ℹ 12 more rows</code></pre>
</div>
<p>But we can can also group by multiple columns:</p>
<div class="codewrapper sourceCode" id="cb43">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene</span>, <span class="va">infection</span>, <span class="va">time</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean_expression <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>`summarise()` has grouped output by 'gene', 'infection'. You can override using
the `.groups` argument.</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 4,422 × 4
# Groups:   gene, infection [2,948]
   gene     infection    time mean_expression
   &lt;chr&gt;    &lt;chr&gt;       &lt;dbl&gt;           &lt;dbl&gt;
 1 AI504432 InfluenzaA      4           1104.
 2 AI504432 InfluenzaA      8           1014
 3 AI504432 NonInfected     0           1034.
 4 AW046200 InfluenzaA      4            152.
 5 AW046200 InfluenzaA      8             81
 6 AW046200 NonInfected     0            155.
 7 AW551984 InfluenzaA      4            302.
 8 AW551984 InfluenzaA      8            342.
 9 AW551984 NonInfected     0            238
10 Aamp     InfluenzaA      4           4870
# ℹ 4,412 more rows</code></pre>
</div>
<p>Once the data is grouped, you can also summarise multiple variables
at the same time (and not necessarily on the same variable). For
instance, we could add a column indicating the median
<code>expression</code> by gene and by condition:</p>
<div class="codewrapper sourceCode" id="cb46">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene</span>, <span class="va">infection</span>, <span class="va">time</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean_expression <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span>,</span>
<span>            median_expression <span class="op">=</span> <span class="fu">median</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>`summarise()` has grouped output by 'gene', 'infection'. You can override using
the `.groups` argument.</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 4,422 × 5
# Groups:   gene, infection [2,948]
   gene     infection    time mean_expression median_expression
   &lt;chr&gt;    &lt;chr&gt;       &lt;dbl&gt;           &lt;dbl&gt;             &lt;dbl&gt;
 1 AI504432 InfluenzaA      4           1104.             1094.
 2 AI504432 InfluenzaA      8           1014               985
 3 AI504432 NonInfected     0           1034.             1016
 4 AW046200 InfluenzaA      4            152.              144.
 5 AW046200 InfluenzaA      8             81                82
 6 AW046200 NonInfected     0            155.              163
 7 AW551984 InfluenzaA      4            302.              245
 8 AW551984 InfluenzaA      8            342.              287
 9 AW551984 NonInfected     0            238               265
10 Aamp     InfluenzaA      4           4870              4708
# ℹ 4,412 more rows</code></pre>
</div>
<div id="challenge-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Calculate the mean expression level of gene “Dok3” by timepoints.</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb49">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">gene</span> <span class="op">==</span> <span class="st">"Dok3"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">time</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 3 × 2
   time  mean
  &lt;dbl&gt; &lt;dbl&gt;
1     0  169
2     4  156.
3     8   61 </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="counting">Counting<a class="anchor" aria-label="anchor" href="#counting"></a>
</h3>
<p>When working with data, we often want to know the number of
observations found for each factor or combination of factors. For this
task, <strong><code>dplyr</code></strong> provides <code>count()</code>.
For example, if we wanted to count the number of rows of data for each
infected and non-infected samples, we would do:</p>
<div class="codewrapper sourceCode" id="cb51">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">count</span><span class="op">(</span><span class="va">infection</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 2 × 2
  infection       n
  &lt;chr&gt;       &lt;int&gt;
1 InfluenzaA  22110
2 NonInfected 10318</code></pre>
</div>
<p>The <code>count()</code> function is shorthand for something we’ve
already seen: grouping by a variable, and summarising it by counting the
number of observations in that group. In other words,
<code>rna |&gt; count(infection)</code> is equivalent to:</p>
<div class="codewrapper sourceCode" id="cb53">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">group_by</span><span class="op">(</span><span class="va">infection</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">summarise</span><span class="op">(</span>n <span class="op">=</span> <span class="fu">n</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 2 × 2
  infection       n
  &lt;chr&gt;       &lt;int&gt;
1 InfluenzaA  22110
2 NonInfected 10318</code></pre>
</div>
<p>The previous example shows the use of <code>count()</code> to count
the number of rows/observations for <em>one</em> factor (i.e.,
<code>infection</code>). If we wanted to count a <em>combination of
factors</em>, such as <code>infection</code> and <code>time</code>, we
would specify the first and the second factor as the arguments of
<code>count()</code>:</p>
<div class="codewrapper sourceCode" id="cb55">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">count</span><span class="op">(</span><span class="va">infection</span>, <span class="va">time</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 3 × 3
  infection    time     n
  &lt;chr&gt;       &lt;dbl&gt; &lt;int&gt;
1 InfluenzaA      4 11792
2 InfluenzaA      8 10318
3 NonInfected     0 10318</code></pre>
</div>
<p>which is equivalent to this:</p>
<div class="codewrapper sourceCode" id="cb57">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">infection</span>, <span class="va">time</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>n <span class="op">=</span> <span class="fu">n</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>`summarise()` has grouped output by 'infection'. You can override using the
`.groups` argument.</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 3 × 3
# Groups:   infection [2]
  infection    time     n
  &lt;chr&gt;       &lt;dbl&gt; &lt;int&gt;
1 InfluenzaA      4 11792
2 InfluenzaA      8 10318
3 NonInfected     0 10318</code></pre>
</div>
<p>It is sometimes useful to sort the result to facilitate the
comparisons. We can use <code>arrange()</code> to sort the table. For
instance, we might want to arrange the table above by time:</p>
<div class="codewrapper sourceCode" id="cb60">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">count</span><span class="op">(</span><span class="va">infection</span>, <span class="va">time</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="va">time</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 3 × 3
  infection    time     n
  &lt;chr&gt;       &lt;dbl&gt; &lt;int&gt;
1 NonInfected     0 10318
2 InfluenzaA      4 11792
3 InfluenzaA      8 10318</code></pre>
</div>
<p>or by counts:</p>
<div class="codewrapper sourceCode" id="cb62">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">count</span><span class="op">(</span><span class="va">infection</span>, <span class="va">time</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="va">n</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 3 × 3
  infection    time     n
  &lt;chr&gt;       &lt;dbl&gt; &lt;int&gt;
1 InfluenzaA      8 10318
2 NonInfected     0 10318
3 InfluenzaA      4 11792</code></pre>
</div>
<p>To sort in descending order, we need to add the <code>desc()</code>
function:</p>
<div class="codewrapper sourceCode" id="cb64">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">count</span><span class="op">(</span><span class="va">infection</span>, <span class="va">time</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="fu">desc</span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 3 × 3
  infection    time     n
  &lt;chr&gt;       &lt;dbl&gt; &lt;int&gt;
1 InfluenzaA      4 11792
2 InfluenzaA      8 10318
3 NonInfected     0 10318</code></pre>
</div>
<div id="challenge-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-3" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li>How many genes were analysed in each sample?</li>
<li>Use <code>group_by()</code> and <code>summarise()</code> to evaluate
the sequencing depth (the sum of all counts) in each sample. Which
sample has the highest sequencing depth?</li>
<li>Pick one sample and evaluate the number of genes by biotype.</li>
<li>Identify genes associated with the “abnormal DNA methylation”
phenotype description, and calculate their mean expression (in log) at
time 0, time 4 and time 8.</li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb66">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## 1.</span></span>
<span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">count</span><span class="op">(</span><span class="va">sample</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 22 × 2
   sample         n
   &lt;chr&gt;      &lt;int&gt;
 1 GSM2545336  1474
 2 GSM2545337  1474
 3 GSM2545338  1474
 4 GSM2545339  1474
 5 GSM2545340  1474
 6 GSM2545341  1474
 7 GSM2545342  1474
 8 GSM2545343  1474
 9 GSM2545344  1474
10 GSM2545345  1474
# ℹ 12 more rows</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb68">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## 2.</span></span>
<span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">sample</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>seq_depth <span class="op">=</span> <span class="fu">sum</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="fu">desc</span><span class="op">(</span><span class="va">seq_depth</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 22 × 2
   sample     seq_depth
   &lt;chr&gt;          &lt;dbl&gt;
 1 GSM2545350   3255566
 2 GSM2545352   3216163
 3 GSM2545343   3105652
 4 GSM2545336   3039671
 5 GSM2545380   3036098
 6 GSM2545353   2953249
 7 GSM2545348   2913678
 8 GSM2545362   2913517
 9 GSM2545351   2782464
10 GSM2545349   2758006
# ℹ 12 more rows</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb70">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## 3.</span></span>
<span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">sample</span> <span class="op">==</span> <span class="st">"GSM2545336"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">count</span><span class="op">(</span><span class="va">gene_biotype</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="fu">desc</span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 13 × 2
   gene_biotype                           n
   &lt;chr&gt;                              &lt;int&gt;
 1 protein_coding                      1321
 2 lncRNA                                69
 3 processed_pseudogene                  59
 4 miRNA                                  7
 5 snoRNA                                 5
 6 TEC                                    4
 7 polymorphic_pseudogene                 2
 8 unprocessed_pseudogene                 2
 9 IG_C_gene                              1
10 scaRNA                                 1
11 transcribed_processed_pseudogene       1
12 transcribed_unitary_pseudogene         1
13 transcribed_unprocessed_pseudogene     1</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb72">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## 4.</span></span>
<span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">phenotype_description</span> <span class="op">==</span> <span class="st">"abnormal DNA methylation"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene</span>, <span class="va">time</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean_expression <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="fu">log</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>`summarise()` has grouped output by 'gene'. You can override using the
`.groups` argument.</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 6 × 3
# Groups:   gene [2]
  gene   time mean_expression
  &lt;chr&gt; &lt;dbl&gt;           &lt;dbl&gt;
1 Xist      0            6.95
2 Xist      4            6.34
3 Xist      8            7.13
4 Zdbf2     0            6.27
5 Zdbf2     4            6.27
6 Zdbf2     8            6.19</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="reshaping-data">Reshaping data<a class="anchor" aria-label="anchor" href="#reshaping-data"></a>
</h2>
<hr class="half-width">
<p>In the <code>rna</code> tibble, the rows contain expression values
(the unit) that are associated with a combination of 2 other variables:
<code>gene</code> and <code>sample</code>.</p>
<p>All the other columns correspond to variables describing either the
sample (organism, age, sex, …) or the gene (gene_biotype, ENTREZ_ID,
product, …). The variables that don’t change with genes or with samples
will have the same value in all the rows.</p>
<div class="codewrapper sourceCode" id="cb75">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="va">gene</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 32,428 × 19
   gene     sample expression organism   age sex   infection strain  time tissue
   &lt;chr&gt;    &lt;chr&gt;       &lt;dbl&gt; &lt;chr&gt;    &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;     &lt;chr&gt;  &lt;dbl&gt; &lt;chr&gt;
 1 AI504432 GSM25…       1230 Mus mus…     8 Fema… Influenz… C57BL…     8 Cereb…
 2 AI504432 GSM25…       1085 Mus mus…     8 Fema… NonInfec… C57BL…     0 Cereb…
 3 AI504432 GSM25…        969 Mus mus…     8 Fema… NonInfec… C57BL…     0 Cereb…
 4 AI504432 GSM25…       1284 Mus mus…     8 Fema… Influenz… C57BL…     4 Cereb…
 5 AI504432 GSM25…        966 Mus mus…     8 Male  Influenz… C57BL…     4 Cereb…
 6 AI504432 GSM25…        918 Mus mus…     8 Male  Influenz… C57BL…     8 Cereb…
 7 AI504432 GSM25…        985 Mus mus…     8 Fema… Influenz… C57BL…     8 Cereb…
 8 AI504432 GSM25…        972 Mus mus…     8 Male  NonInfec… C57BL…     0 Cereb…
 9 AI504432 GSM25…       1000 Mus mus…     8 Fema… Influenz… C57BL…     4 Cereb…
10 AI504432 GSM25…        816 Mus mus…     8 Male  Influenz… C57BL…     4 Cereb…
# ℹ 32,418 more rows
# ℹ 9 more variables: mouse &lt;dbl&gt;, ENTREZID &lt;dbl&gt;, product &lt;chr&gt;,
#   ensembl_gene_id &lt;chr&gt;, external_synonym &lt;chr&gt;, chromosome_name &lt;chr&gt;,
#   gene_biotype &lt;chr&gt;, phenotype_description &lt;chr&gt;,
#   hsapiens_homolog_associated_gene_name &lt;chr&gt;</code></pre>
</div>
<p>This structure is called a <code>long-format</code>, as one column
contains all the values, and other column(s) list(s) the context of the
value.</p>
<p>In certain cases, the <code>long-format</code> is not really
“human-readable”, and another format, a <code>wide-format</code> is
preferred, as a more compact way of representing the data. This is
typically the case with gene expression values that scientists are used
to look as matrices, were rows represent genes and columns represent
samples.</p>
<p>In this format, it would therefore become straightforward to explore
the relationship between the gene expression levels within, and between,
the samples.</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 1,474 × 23
   gene    GSM2545336 GSM2545337 GSM2545338 GSM2545339 GSM2545340 GSM2545341
   &lt;chr&gt;        &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;
 1 Asl           1170        361        400        586        626        988
 2 Apod         36194      10347       9173      10620      13021      29594
 3 Cyp2d22       4060       1616       1603       1901       2171       3349
 4 Klk6           287        629        641        578        448        195
 5 Fcrls           85        233        244        237        180         38
 6 Slc2a4         782        231        248        265        313        786
 7 Exd2          1619       2288       2235       2513       2366       1359
 8 Gjc2           288        595        568        551        310        146
 9 Plp1         43217     101241      96534      58354      53126      27173
10 Gnb4          1071       1791       1867       1430       1355        798
# ℹ 1,464 more rows
# ℹ 16 more variables: GSM2545342 &lt;dbl&gt;, GSM2545343 &lt;dbl&gt;, GSM2545344 &lt;dbl&gt;,
#   GSM2545345 &lt;dbl&gt;, GSM2545346 &lt;dbl&gt;, GSM2545347 &lt;dbl&gt;, GSM2545348 &lt;dbl&gt;,
#   GSM2545349 &lt;dbl&gt;, GSM2545350 &lt;dbl&gt;, GSM2545351 &lt;dbl&gt;, GSM2545352 &lt;dbl&gt;,
#   GSM2545353 &lt;dbl&gt;, GSM2545354 &lt;dbl&gt;, GSM2545362 &lt;dbl&gt;, GSM2545363 &lt;dbl&gt;,
#   GSM2545380 &lt;dbl&gt;</code></pre>
</div>
<p>To convert the gene expression values from <code>rna</code> into a
wide-format, we need to create a new table where the values of the
<code>sample</code> column would become the names of column
variables.</p>
<p>The key point here is that we are still following a tidy data
structure, but we have <strong>reshaped</strong> the data according to
the observations of interest: expression levels per gene instead of
recording them per gene and per sample.</p>
<p>The opposite transformation would be to transform column names into
values of a new variable.</p>
<p>We can do both these of transformations with two <code>tidyr</code>
functions, <code>pivot_longer()</code> and <code>pivot_wider()</code>
(see the article about <a href="https://tidyr.tidyverse.org/dev/articles/pivot.html" class="external-link">pivots on the
tidyr website</a> for details).</p>
<div class="section level3">
<h3 id="pivoting-the-data-into-a-wider-format">Pivoting the data into a wider format<a class="anchor" aria-label="anchor" href="#pivoting-the-data-into-a-wider-format"></a>
</h3>
<p>Let’s select the first 3 columns of <code>rna</code> and use
<code>pivot_wider()</code> to transform the data into a wide-format.</p>
<div class="codewrapper sourceCode" id="cb78">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_exp</span> <span class="op">&lt;-</span> <span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">gene</span>, <span class="va">sample</span>, <span class="va">expression</span><span class="op">)</span></span>
<span><span class="va">rna_exp</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 32,428 × 3
   gene    sample     expression
   &lt;chr&gt;   &lt;chr&gt;           &lt;dbl&gt;
 1 Asl     GSM2545336       1170
 2 Apod    GSM2545336      36194
 3 Cyp2d22 GSM2545336       4060
 4 Klk6    GSM2545336        287
 5 Fcrls   GSM2545336         85
 6 Slc2a4  GSM2545336        782
 7 Exd2    GSM2545336       1619
 8 Gjc2    GSM2545336        288
 9 Plp1    GSM2545336      43217
10 Gnb4    GSM2545336       1071
# ℹ 32,418 more rows</code></pre>
</div>
<p><code>pivot_wider</code> takes three main arguments:</p>
<ol style="list-style-type: decimal">
<li>the data to be transformed;</li>
<li>the <code>names_from</code> : the column whose values will become
new column names;</li>
<li>the <code>values_from</code>: the column whose values will fill the
new columns.</li>
</ol>
<figure style="text-align: center"><img src="fig/pivot_wider.png" alt="Figure showing the long table format on the left and wide table format on the right and arrows illustrating how the values in the 'sample' column on the left have become column names on the right and how the values in the 'expression' column on the left have become values on the the right. Below is the call to 'pivot_wider()' with annotations pointing to the 'sample' and 'expression' function arguments." class="figure mx-auto d-block"><figcaption>
Wide pivot of the <code>rna</code> data.
</figcaption></figure><div class="codewrapper sourceCode" id="cb80">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_wide</span> <span class="op">&lt;-</span> <span class="va">rna_exp</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">sample</span>,</span>
<span>              values_from <span class="op">=</span> <span class="va">expression</span><span class="op">)</span></span>
<span><span class="va">rna_wide</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 1,474 × 23
   gene    GSM2545336 GSM2545337 GSM2545338 GSM2545339 GSM2545340 GSM2545341
   &lt;chr&gt;        &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;
 1 Asl           1170        361        400        586        626        988
 2 Apod         36194      10347       9173      10620      13021      29594
 3 Cyp2d22       4060       1616       1603       1901       2171       3349
 4 Klk6           287        629        641        578        448        195
 5 Fcrls           85        233        244        237        180         38
 6 Slc2a4         782        231        248        265        313        786
 7 Exd2          1619       2288       2235       2513       2366       1359
 8 Gjc2           288        595        568        551        310        146
 9 Plp1         43217     101241      96534      58354      53126      27173
10 Gnb4          1071       1791       1867       1430       1355        798
# ℹ 1,464 more rows
# ℹ 16 more variables: GSM2545342 &lt;dbl&gt;, GSM2545343 &lt;dbl&gt;, GSM2545344 &lt;dbl&gt;,
#   GSM2545345 &lt;dbl&gt;, GSM2545346 &lt;dbl&gt;, GSM2545347 &lt;dbl&gt;, GSM2545348 &lt;dbl&gt;,
#   GSM2545349 &lt;dbl&gt;, GSM2545350 &lt;dbl&gt;, GSM2545351 &lt;dbl&gt;, GSM2545352 &lt;dbl&gt;,
#   GSM2545353 &lt;dbl&gt;, GSM2545354 &lt;dbl&gt;, GSM2545362 &lt;dbl&gt;, GSM2545363 &lt;dbl&gt;,
#   GSM2545380 &lt;dbl&gt;</code></pre>
</div>
<p>Note that by default, the <code>pivot_wider()</code> function will
add <code>NA</code> for missing values.</p>
<p>Let’s imagine that for some reason, we had some missing expression
values for some genes in certain samples. In the following fictive
example, the gene Cyp2d22 has only one expression value, in GSM2545338
sample.</p>
<div class="codewrapper sourceCode" id="cb82">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_with_missing_values</span> <span class="op">&lt;-</span> <span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">gene</span>, <span class="va">sample</span>, <span class="va">expression</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">gene</span> <span class="op">%in%</span> <span class="fu">c</span><span class="op">(</span><span class="st">"Asl"</span>, <span class="st">"Apod"</span>, <span class="st">"Cyp2d22"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">sample</span> <span class="op">%in%</span> <span class="fu">c</span><span class="op">(</span><span class="st">"GSM2545336"</span>, <span class="st">"GSM2545337"</span>, <span class="st">"GSM2545338"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="va">sample</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="op">!</span><span class="op">(</span><span class="va">gene</span> <span class="op">==</span> <span class="st">"Cyp2d22"</span> <span class="op">&amp;</span> <span class="va">sample</span> <span class="op">!=</span> <span class="st">"GSM2545338"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">rna_with_missing_values</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 7 × 3
  gene    sample     expression
  &lt;chr&gt;   &lt;chr&gt;           &lt;dbl&gt;
1 Asl     GSM2545336       1170
2 Apod    GSM2545336      36194
3 Asl     GSM2545337        361
4 Apod    GSM2545337      10347
5 Asl     GSM2545338        400
6 Apod    GSM2545338       9173
7 Cyp2d22 GSM2545338       1603</code></pre>
</div>
<p>By default, the <code>pivot_wider()</code> function will add
<code>NA</code> for missing values. This can be parameterised with the
<code>values_fill</code> argument of the <code>pivot_wider()</code>
function.</p>
<div class="codewrapper sourceCode" id="cb84">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_with_missing_values</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">sample</span>,</span>
<span>              values_from <span class="op">=</span> <span class="va">expression</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 3 × 4
  gene    GSM2545336 GSM2545337 GSM2545338
  &lt;chr&gt;        &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;
1 Asl           1170        361        400
2 Apod         36194      10347       9173
3 Cyp2d22         NA         NA       1603</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb86">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_with_missing_values</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">sample</span>,</span>
<span>              values_from <span class="op">=</span> <span class="va">expression</span>,</span>
<span>              values_fill <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 3 × 4
  gene    GSM2545336 GSM2545337 GSM2545338
  &lt;chr&gt;        &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;
1 Asl           1170        361        400
2 Apod         36194      10347       9173
3 Cyp2d22          0          0       1603</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="pivoting-data-into-a-longer-format">Pivoting data into a longer format<a class="anchor" aria-label="anchor" href="#pivoting-data-into-a-longer-format"></a>
</h3>
<p>In the opposite situation we are using the column names and turning
them into a pair of new variables. One variable represents the column
names as values, and the other variable contains the values previously
associated with the column names.</p>
<p><code>pivot_longer()</code> takes four main arguments:</p>
<ol style="list-style-type: decimal">
<li>the data to be transformed;</li>
<li>the <code>names_to</code>: the new column name we wish to create and
populate with the current column names;</li>
<li>the <code>values_to</code>: the new column name we wish to create
and populate with current values;</li>
<li>the names of the columns to be used to populate the
<code>names_to</code> and <code>values_to</code> variables (or to
drop).</li>
</ol>
<!-- on the right and arrows illustrating how the values in the 'sample' --><!-- column on the left have become column names on the right and how the --><!-- values in the 'expression' column on the left have become values on --><!-- the the right. --><figure style="text-align: center"><img src="fig/pivot_longer.png" alt="Figure showing the long table format on the left and wide table format on the right and arrows illustrating how the column names on the left have become a new column 'sample' on the left and the values in the wide table on the right have become a new column 'expression' on the left.  Below is the call to 'pivot_wider()' with annotations pointing to the 'sample', 'expression' and the '-gene' arguments." class="figure mx-auto d-block"><figcaption>
Long pivot of the <code>rna</code> data.
</figcaption></figure><p>To recreate <code>rna_long</code> from <code>rna_wide</code> we would
create a key called <code>sample</code> and value called
<code>expression</code> and use all columns except <code>gene</code> for
the key variable. Here we drop <code>gene</code> column with a minus
sign.</p>
<p>Notice how the new variable names are to be quoted here.</p>
<div class="codewrapper sourceCode" id="cb88">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_long</span> <span class="op">&lt;-</span> <span class="va">rna_wide</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">pivot_longer</span><span class="op">(</span>names_to <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>                 values_to <span class="op">=</span> <span class="st">"expression"</span>,</span>
<span>                 <span class="op">-</span><span class="va">gene</span><span class="op">)</span></span>
<span><span class="va">rna_long</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 32,428 × 3
   gene  sample     expression
   &lt;chr&gt; &lt;chr&gt;           &lt;dbl&gt;
 1 Asl   GSM2545336       1170
 2 Asl   GSM2545337        361
 3 Asl   GSM2545338        400
 4 Asl   GSM2545339        586
 5 Asl   GSM2545340        626
 6 Asl   GSM2545341        988
 7 Asl   GSM2545342        836
 8 Asl   GSM2545343        535
 9 Asl   GSM2545344        586
10 Asl   GSM2545345        597
# ℹ 32,418 more rows</code></pre>
</div>
<p>We could also have used a specification for what columns to include.
This can be useful if you have a large number of identifying columns,
and it’s easier to specify what to gather than what to leave alone. Here
the <code>starts_with()</code> function can help to retrieve sample
names without having to list them all! Another possibility would be to
use the <code>:</code> operator!</p>
<div class="codewrapper sourceCode" id="cb90">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_wide</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">pivot_longer</span><span class="op">(</span>names_to <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>                 values_to <span class="op">=</span> <span class="st">"expression"</span>,</span>
<span>                 cols <span class="op">=</span> <span class="fu">starts_with</span><span class="op">(</span><span class="st">"GSM"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 32,428 × 3
   gene  sample     expression
   &lt;chr&gt; &lt;chr&gt;           &lt;dbl&gt;
 1 Asl   GSM2545336       1170
 2 Asl   GSM2545337        361
 3 Asl   GSM2545338        400
 4 Asl   GSM2545339        586
 5 Asl   GSM2545340        626
 6 Asl   GSM2545341        988
 7 Asl   GSM2545342        836
 8 Asl   GSM2545343        535
 9 Asl   GSM2545344        586
10 Asl   GSM2545345        597
# ℹ 32,418 more rows</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb92">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_wide</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">pivot_longer</span><span class="op">(</span>names_to <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>                 values_to <span class="op">=</span> <span class="st">"expression"</span>,</span>
<span>                 <span class="va">GSM2545336</span><span class="op">:</span><span class="va">GSM2545380</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 32,428 × 3
   gene  sample     expression
   &lt;chr&gt; &lt;chr&gt;           &lt;dbl&gt;
 1 Asl   GSM2545336       1170
 2 Asl   GSM2545337        361
 3 Asl   GSM2545338        400
 4 Asl   GSM2545339        586
 5 Asl   GSM2545340        626
 6 Asl   GSM2545341        988
 7 Asl   GSM2545342        836
 8 Asl   GSM2545343        535
 9 Asl   GSM2545344        586
10 Asl   GSM2545345        597
# ℹ 32,418 more rows</code></pre>
</div>
<p>Note that if we had missing values in the wide-format, the
<code>NA</code> would be included in the new long format.</p>
<p>Remember our previous fictive tibble containing missing values:</p>
<div class="codewrapper sourceCode" id="cb94">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_with_missing_values</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 7 × 3
  gene    sample     expression
  &lt;chr&gt;   &lt;chr&gt;           &lt;dbl&gt;
1 Asl     GSM2545336       1170
2 Apod    GSM2545336      36194
3 Asl     GSM2545337        361
4 Apod    GSM2545337      10347
5 Asl     GSM2545338        400
6 Apod    GSM2545338       9173
7 Cyp2d22 GSM2545338       1603</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb96">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">wide_with_NA</span> <span class="op">&lt;-</span> <span class="va">rna_with_missing_values</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">sample</span>,</span>
<span>              values_from <span class="op">=</span> <span class="va">expression</span><span class="op">)</span></span>
<span><span class="va">wide_with_NA</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 3 × 4
  gene    GSM2545336 GSM2545337 GSM2545338
  &lt;chr&gt;        &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;
1 Asl           1170        361        400
2 Apod         36194      10347       9173
3 Cyp2d22         NA         NA       1603</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb98">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">wide_with_NA</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">pivot_longer</span><span class="op">(</span>names_to <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>                 values_to <span class="op">=</span> <span class="st">"expression"</span>,</span>
<span>                 <span class="op">-</span><span class="va">gene</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 9 × 3
  gene    sample     expression
  &lt;chr&gt;   &lt;chr&gt;           &lt;dbl&gt;
1 Asl     GSM2545336       1170
2 Asl     GSM2545337        361
3 Asl     GSM2545338        400
4 Apod    GSM2545336      36194
5 Apod    GSM2545337      10347
6 Apod    GSM2545338       9173
7 Cyp2d22 GSM2545336         NA
8 Cyp2d22 GSM2545337         NA
9 Cyp2d22 GSM2545338       1603</code></pre>
</div>
<p>Pivoting to wider and longer formats can be a useful way to balance
out a dataset so every replicate has the same composition.</p>
<div id="question" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="question" class="callout-inner">
<h3 class="callout-title">Question</h3>
<div class="callout-content">
<p>Starting from the rna table, use the <code>pivot_wider()</code>
function to create a wide-format table giving the gene expression levels
in each mouse. Then use the <code>pivot_longer()</code> function to
restore a long-format table.</p>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" aria-labelledby="headingSolution5" data-bs-parent="#accordionSolution5">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb100">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna1</span> <span class="op">&lt;-</span> <span class="va">rna</span> <span class="op">|&gt;</span></span>
<span><span class="fu">select</span><span class="op">(</span><span class="va">gene</span>, <span class="va">mouse</span>, <span class="va">expression</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">mouse</span>, values_from <span class="op">=</span> <span class="va">expression</span><span class="op">)</span></span>
<span><span class="va">rna1</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 1,474 × 23
   gene     `14`    `9`  `10`  `15`  `18`   `6`   `5`  `11`  `22`  `13`  `23`
   &lt;chr&gt;   &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
 1 Asl      1170    361   400   586   626   988   836   535   586   597   938
 2 Apod    36194  10347  9173 10620 13021 29594 24959 13668 13230 15868 27769
 3 Cyp2d22  4060   1616  1603  1901  2171  3349  3122  2008  2254  2277  2985
 4 Klk6      287    629   641   578   448   195   186  1101   537   567   327
 5 Fcrls      85    233   244   237   180    38    68   375   199   177    89
 6 Slc2a4    782    231   248   265   313   786   528   249   266   357   654
 7 Exd2     1619   2288  2235  2513  2366  1359  1474  3126  2379  2173  1531
 8 Gjc2      288    595   568   551   310   146   186   791   454   370   240
 9 Plp1    43217 101241 96534 58354 53126 27173 28728 98658 61356 61647 38019
10 Gnb4     1071   1791  1867  1430  1355   798   806  2437  1394  1554   960
# ℹ 1,464 more rows
# ℹ 11 more variables: `24` &lt;dbl&gt;, `8` &lt;dbl&gt;, `7` &lt;dbl&gt;, `1` &lt;dbl&gt;, `16` &lt;dbl&gt;,
#   `21` &lt;dbl&gt;, `4` &lt;dbl&gt;, `2` &lt;dbl&gt;, `20` &lt;dbl&gt;, `12` &lt;dbl&gt;, `19` &lt;dbl&gt;</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb102">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna1</span> <span class="op">|&gt;</span></span>
<span><span class="fu">pivot_longer</span><span class="op">(</span>names_to <span class="op">=</span> <span class="st">"mouse_id"</span>, values_to <span class="op">=</span> <span class="st">"counts"</span>, <span class="op">-</span><span class="va">gene</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 32,428 × 3
   gene  mouse_id counts
   &lt;chr&gt; &lt;chr&gt;     &lt;dbl&gt;
 1 Asl   14         1170
 2 Asl   9           361
 3 Asl   10          400
 4 Asl   15          586
 5 Asl   18          626
 6 Asl   6           988
 7 Asl   5           836
 8 Asl   11          535
 9 Asl   22          586
10 Asl   13          597
# ℹ 32,418 more rows</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="question-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="question-1" class="callout-inner">
<h3 class="callout-title">Question</h3>
<div class="callout-content">
<p>Subset genes located on X and Y chromosomes from the <code>rna</code>
data frame and spread the data frame with <code>sex</code> as columns,
<code>chromosome_name</code> as rows, and the mean expression of genes
located in each chromosome as the values, as in the following
tibble:</p>
<figure><img src="fig/Exercise_pivot_W.png" alt="A two by two table with  X and Y row names and Female and Male column names showing the total counts for each X/Y and Female/Male combination. The Y/Female combinations shows 3 counts, while all other counts are above 2000 counts." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>You will need to summarise before reshaping!</p>
</div>
</div>
</div>
<div id="accordionSolution6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution6" aria-expanded="false" aria-controls="collapseSolution6">
  <h4 class="accordion-header" id="headingSolution6"> Show me the solution </h4>
</button>
<div id="collapseSolution6" class="accordion-collapse collapse" aria-labelledby="headingSolution6" data-bs-parent="#accordionSolution6">
<div class="accordion-body">
<p>Let’s first calculate the mean expression level of X and Y linked
genes from male and female samples…</p>
<div class="codewrapper sourceCode" id="cb104">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span> <span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">chromosome_name</span> <span class="op">==</span> <span class="st">"Y"</span> <span class="op">|</span> <span class="va">chromosome_name</span> <span class="op">==</span> <span class="st">"X"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">sex</span>, <span class="va">chromosome_name</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>`summarise()` has grouped output by 'sex'. You can override using the `.groups`
argument.</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 4 × 3
# Groups:   sex [2]
  sex    chromosome_name  mean
  &lt;chr&gt;  &lt;chr&gt;           &lt;dbl&gt;
1 Female X               3504.
2 Female Y                  3
3 Male   X               2497.
4 Male   Y               2117.</code></pre>
</div>
<p>And pivot the table to wide format</p>
<div class="codewrapper sourceCode" id="cb107">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_1</span> <span class="op">&lt;-</span> <span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">chromosome_name</span> <span class="op">==</span> <span class="st">"Y"</span> <span class="op">|</span> <span class="va">chromosome_name</span> <span class="op">==</span> <span class="st">"X"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">sex</span>, <span class="va">chromosome_name</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">sex</span>,</span>
<span>              values_from <span class="op">=</span> <span class="va">mean</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>`summarise()` has grouped output by 'sex'. You can override using the `.groups`
argument.</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb109">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_1</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 2 × 3
  chromosome_name Female  Male
  &lt;chr&gt;            &lt;dbl&gt; &lt;dbl&gt;
1 X                3504. 2497.
2 Y                   3  2117.</code></pre>
</div>
<p>Now take that data frame and transform it with
<code>pivot_longer()</code> so each row is a unique
<code>chromosome_name</code> by <code>gender</code> combination.</p>
<div class="codewrapper sourceCode" id="cb111">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_1</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span>names_to <span class="op">=</span> <span class="st">"gender"</span>,</span>
<span>               values_to <span class="op">=</span> <span class="st">"mean"</span>,</span>
<span>               <span class="op">-</span><span class="va">chromosome_name</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 4 × 3
  chromosome_name gender  mean
  &lt;chr&gt;           &lt;chr&gt;  &lt;dbl&gt;
1 X               Female 3504.
2 X               Male   2497.
3 Y               Female    3
4 Y               Male   2117.</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="question-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="question-2" class="callout-inner">
<h3 class="callout-title">Question</h3>
<div class="callout-content">
<p>Use the <code>rna</code> dataset to create an expression matrix where
each row represents the mean expression levels of genes and columns
represent the different timepoints.</p>
</div>
</div>
</div>
<div id="accordionSolution7" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution7" aria-expanded="false" aria-controls="collapseSolution7">
  <h4 class="accordion-header" id="headingSolution7"> Show me the solution </h4>
</button>
<div id="collapseSolution7" class="accordion-collapse collapse" aria-labelledby="headingSolution7" data-bs-parent="#accordionSolution7">
<div class="accordion-body">
<p>Let’s first calculate the mean expression by gene and by time</p>
<div class="codewrapper sourceCode" id="cb113">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene</span>, <span class="va">time</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean_exp <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>`summarise()` has grouped output by 'gene'. You can override using the
`.groups` argument.</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 4,422 × 3
# Groups:   gene [1,474]
   gene      time mean_exp
   &lt;chr&gt;    &lt;dbl&gt;    &lt;dbl&gt;
 1 AI504432     0    1034.
 2 AI504432     4    1104.
 3 AI504432     8    1014
 4 AW046200     0     155.
 5 AW046200     4     152.
 6 AW046200     8      81
 7 AW551984     0     238
 8 AW551984     4     302.
 9 AW551984     8     342.
10 Aamp         0    4603.
# ℹ 4,412 more rows</code></pre>
</div>
<p>before using the pivot_wider() function</p>
<div class="codewrapper sourceCode" id="cb116">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_time</span> <span class="op">&lt;-</span> <span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene</span>, <span class="va">time</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean_exp <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">time</span>,</span>
<span>              values_from <span class="op">=</span> <span class="va">mean_exp</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>`summarise()` has grouped output by 'gene'. You can override using the
`.groups` argument.</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb118">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_time</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 1,474 × 4
# Groups:   gene [1,474]
   gene         `0`     `4`     `8`
   &lt;chr&gt;      &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;
 1 AI504432 1034.   1104.   1014
 2 AW046200  155.    152.     81
 3 AW551984  238     302.    342.
 4 Aamp     4603.   4870    4763.
 5 Abca12      5.29    4.25    4.14
 6 Abcc8    2576.   2609.   2292.
 7 Abhd14a   591.    547.    432.
 8 Abi2     4881.   4903.   4945.
 9 Abi3bp   1175.   1061.    762.
10 Abl2     2170.   2078.   2131.
# ℹ 1,464 more rows</code></pre>
</div>
<p>Notice that this generates a tibble with some column names starting
by a number. If we wanted to select the column corresponding to the
timepoints, we could not use the column names directly… What happens
when we select the column 4?</p>
<div class="codewrapper sourceCode" id="cb120">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene</span>, <span class="va">time</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean_exp <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">time</span>,</span>
<span>              values_from <span class="op">=</span> <span class="va">mean_exp</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">gene</span>, <span class="fl">4</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>`summarise()` has grouped output by 'gene'. You can override using the
`.groups` argument.</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 1,474 × 2
# Groups:   gene [1,474]
   gene         `8`
   &lt;chr&gt;      &lt;dbl&gt;
 1 AI504432 1014
 2 AW046200   81
 3 AW551984  342.
 4 Aamp     4763.
 5 Abca12      4.14
 6 Abcc8    2292.
 7 Abhd14a   432.
 8 Abi2     4945.
 9 Abi3bp    762.
10 Abl2     2131.
# ℹ 1,464 more rows</code></pre>
</div>
<p>To select the timepoint 4, we would have to quote the column name,
with backticks “`”</p>
<div class="codewrapper sourceCode" id="cb123">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene</span>, <span class="va">time</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean_exp <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">time</span>,</span>
<span>              values_from <span class="op">=</span> <span class="va">mean_exp</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">gene</span>, <span class="va">`4`</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>`summarise()` has grouped output by 'gene'. You can override using the
`.groups` argument.</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 1,474 × 2
# Groups:   gene [1,474]
   gene         `4`
   &lt;chr&gt;      &lt;dbl&gt;
 1 AI504432 1104.
 2 AW046200  152.
 3 AW551984  302.
 4 Aamp     4870
 5 Abca12      4.25
 6 Abcc8    2609.
 7 Abhd14a   547.
 8 Abi2     4903.
 9 Abi3bp   1061.
10 Abl2     2078.
# ℹ 1,464 more rows</code></pre>
</div>
<p>Another possibility would be to rename the column, choosing a name
that doesn’t start by a number :</p>
<div class="codewrapper sourceCode" id="cb126">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene</span>, <span class="va">time</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean_exp <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">time</span>,</span>
<span>              values_from <span class="op">=</span> <span class="va">mean_exp</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">rename</span><span class="op">(</span><span class="st">"time0"</span> <span class="op">=</span> <span class="va">`0`</span>, <span class="st">"time4"</span> <span class="op">=</span> <span class="va">`4`</span>, <span class="st">"time8"</span> <span class="op">=</span> <span class="va">`8`</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">gene</span>, <span class="va">time4</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>`summarise()` has grouped output by 'gene'. You can override using the
`.groups` argument.</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 1,474 × 2
# Groups:   gene [1,474]
   gene       time4
   &lt;chr&gt;      &lt;dbl&gt;
 1 AI504432 1104.
 2 AW046200  152.
 3 AW551984  302.
 4 Aamp     4870
 5 Abca12      4.25
 6 Abcc8    2609.
 7 Abhd14a   547.
 8 Abi2     4903.
 9 Abi3bp   1061.
10 Abl2     2078.
# ℹ 1,464 more rows</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="question-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="question-3" class="callout-inner">
<h3 class="callout-title">Question</h3>
<div class="callout-content">
<p>Use the previous data frame containing mean expression levels per
timepoint and create a new column containing fold-changes between
timepoint 8 and timepoint 0, and fold-changes between timepoint 8 and
timepoint 4. Convert this table into a long-format table gathering the
fold-changes calculated.</p>
</div>
</div>
</div>
<div id="accordionSolution8" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution8" aria-expanded="false" aria-controls="collapseSolution8">
  <h4 class="accordion-header" id="headingSolution8"> Show me the solution </h4>
</button>
<div id="collapseSolution8" class="accordion-collapse collapse" aria-labelledby="headingSolution8" data-bs-parent="#accordionSolution8">
<div class="accordion-body">
<p>Starting from the rna_time tibble:</p>
<div class="codewrapper sourceCode" id="cb129">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_time</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 1,474 × 4
# Groups:   gene [1,474]
   gene         `0`     `4`     `8`
   &lt;chr&gt;      &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;
 1 AI504432 1034.   1104.   1014
 2 AW046200  155.    152.     81
 3 AW551984  238     302.    342.
 4 Aamp     4603.   4870    4763.
 5 Abca12      5.29    4.25    4.14
 6 Abcc8    2576.   2609.   2292.
 7 Abhd14a   591.    547.    432.
 8 Abi2     4881.   4903.   4945.
 9 Abi3bp   1175.   1061.    762.
10 Abl2     2170.   2078.   2131.
# ℹ 1,464 more rows</code></pre>
</div>
<p>Calculate fold-changes:</p>
<div class="codewrapper sourceCode" id="cb131">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_time</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span>time_8_vs_0 <span class="op">=</span> <span class="va">`8`</span> <span class="op">/</span> <span class="va">`0`</span>, time_8_vs_4 <span class="op">=</span> <span class="va">`8`</span> <span class="op">/</span> <span class="va">`4`</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 1,474 × 6
# Groups:   gene [1,474]
   gene         `0`     `4`     `8` time_8_vs_0 time_8_vs_4
   &lt;chr&gt;      &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;       &lt;dbl&gt;       &lt;dbl&gt;
 1 AI504432 1034.   1104.   1014          0.981       0.918
 2 AW046200  155.    152.     81          0.522       0.532
 3 AW551984  238     302.    342.         1.44        1.13
 4 Aamp     4603.   4870    4763.         1.03        0.978
 5 Abca12      5.29    4.25    4.14       0.784       0.975
 6 Abcc8    2576.   2609.   2292.         0.889       0.878
 7 Abhd14a   591.    547.    432.         0.731       0.791
 8 Abi2     4881.   4903.   4945.         1.01        1.01
 9 Abi3bp   1175.   1061.    762.         0.649       0.719
10 Abl2     2170.   2078.   2131.         0.982       1.03
# ℹ 1,464 more rows</code></pre>
</div>
<p>And use the pivot_longer() function:</p>
<div class="codewrapper sourceCode" id="cb133">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_time</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span>time_8_vs_0 <span class="op">=</span> <span class="va">`8`</span> <span class="op">/</span> <span class="va">`0`</span>, time_8_vs_4 <span class="op">=</span> <span class="va">`8`</span> <span class="op">/</span> <span class="va">`4`</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span>names_to <span class="op">=</span> <span class="st">"comparisons"</span>,</span>
<span>               values_to <span class="op">=</span> <span class="st">"Fold_changes"</span>,</span>
<span>               <span class="va">time_8_vs_0</span><span class="op">:</span><span class="va">time_8_vs_4</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 2,948 × 6
# Groups:   gene [1,474]
   gene         `0`     `4`     `8` comparisons Fold_changes
   &lt;chr&gt;      &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt; &lt;chr&gt;              &lt;dbl&gt;
 1 AI504432 1034.   1104.   1014    time_8_vs_0        0.981
 2 AI504432 1034.   1104.   1014    time_8_vs_4        0.918
 3 AW046200  155.    152.     81    time_8_vs_0        0.522
 4 AW046200  155.    152.     81    time_8_vs_4        0.532
 5 AW551984  238     302.    342.   time_8_vs_0        1.44
 6 AW551984  238     302.    342.   time_8_vs_4        1.13
 7 Aamp     4603.   4870    4763.   time_8_vs_0        1.03
 8 Aamp     4603.   4870    4763.   time_8_vs_4        0.978
 9 Abca12      5.29    4.25    4.14 time_8_vs_0        0.784
10 Abca12      5.29    4.25    4.14 time_8_vs_4        0.975
# ℹ 2,938 more rows</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="joining-tables">Joining tables<a class="anchor" aria-label="anchor" href="#joining-tables"></a>
</h2>
<hr class="half-width">
<p>In many real life situations, data are spread across multiple tables.
Usually this occurs because different types of information are collected
from different sources.</p>
<p>It may be desirable for some analyses to combine data from two or
more tables into a single data frame based on a column that would be
common to all the tables.</p>
<p>The <code>dplyr</code> package provides a set of join functions for
combining two data frames based on matches within specified columns.
Here, we provide a short introduction to joins. For further reading,
please refer to the chapter about <a href="https://uclouvain-cbio.github.io/WSBIM1207/sec-join.html" class="external-link">table
joins</a>. The <a href="https://github.com/rstudio/cheatsheets/raw/master/data-transformation.pdf" class="external-link">Data
Transformation Cheat Sheet</a> also provides a short overview on table
joins.</p>
<p>We are going to illustrate join using a small table,
<code>rna_mini</code> that we will create by subsetting the original
<code>rna</code> table, keeping only 3 columns and 10 lines.</p>
<div class="codewrapper sourceCode" id="cb135">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_mini</span> <span class="op">&lt;-</span> <span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>   <span class="fu">select</span><span class="op">(</span><span class="va">gene</span>, <span class="va">sample</span>, <span class="va">expression</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>   <span class="fu">head</span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">rna_mini</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 10 × 3
   gene    sample     expression
   &lt;chr&gt;   &lt;chr&gt;           &lt;dbl&gt;
 1 Asl     GSM2545336       1170
 2 Apod    GSM2545336      36194
 3 Cyp2d22 GSM2545336       4060
 4 Klk6    GSM2545336        287
 5 Fcrls   GSM2545336         85
 6 Slc2a4  GSM2545336        782
 7 Exd2    GSM2545336       1619
 8 Gjc2    GSM2545336        288
 9 Plp1    GSM2545336      43217
10 Gnb4    GSM2545336       1071</code></pre>
</div>
<p>The second table, <code>annot1</code>, contains 2 columns, gene and
gene_description. You can either <a href="https://carpentries-incubator.github.io/bioc-intro/data/annot1.csv" class="external-link">download
annot1.csv</a> by clicking on the link and then moving it to the
<code>data/</code> folder, or you can use the R code below to download
it directly to the folder.</p>
<div class="codewrapper sourceCode" id="cb137">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">download.file</span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://carpentries-incubator.github.io/bioc-intro/data/annot1.csv"</span>,</span>
<span>              destfile <span class="op">=</span> <span class="st">"data/annot1.csv"</span><span class="op">)</span></span>
<span><span class="va">annot1</span> <span class="op">&lt;-</span> <span class="fu">read_csv</span><span class="op">(</span>file <span class="op">=</span> <span class="st">"data/annot1.csv"</span><span class="op">)</span></span>
<span><span class="va">annot1</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 10 × 2
   gene    gene_description
   &lt;chr&gt;   &lt;chr&gt;
 1 Cyp2d22 cytochrome P450, family 2, subfamily d, polypeptide 22 [Source:MGI S…
 2 Klk6    kallikrein related-peptidase 6 [Source:MGI Symbol;Acc:MGI:1343166]
 3 Fcrls   Fc receptor-like S, scavenger receptor [Source:MGI Symbol;Acc:MGI:19…
 4 Plp1    proteolipid protein (myelin) 1 [Source:MGI Symbol;Acc:MGI:97623]
 5 Exd2    exonuclease 3'-5' domain containing 2 [Source:MGI Symbol;Acc:MGI:192…
 6 Apod    apolipoprotein D [Source:MGI Symbol;Acc:MGI:88056]
 7 Gnb4    guanine nucleotide binding protein (G protein), beta 4 [Source:MGI S…
 8 Slc2a4  solute carrier family 2 (facilitated glucose transporter), member 4 …
 9 Asl     argininosuccinate lyase [Source:MGI Symbol;Acc:MGI:88084]
10 Gjc2    gap junction protein, gamma 2 [Source:MGI Symbol;Acc:MGI:2153060]    </code></pre>
</div>
<p>We now want to join these two tables into a single one containing all
variables using the <code>full_join()</code> function from the
<code>dplyr</code> package. The function will automatically find the
common variable to match columns from the first and second table. In
this case, <code>gene</code> is the common variable. Such variables are
called keys. Keys are used to match observations across different
tables.</p>
<div class="codewrapper sourceCode" id="cb139">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">full_join</span><span class="op">(</span><span class="va">rna_mini</span>, <span class="va">annot1</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Joining with `by = join_by(gene)`</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 10 × 4
   gene    sample     expression gene_description
   &lt;chr&gt;   &lt;chr&gt;           &lt;dbl&gt; &lt;chr&gt;
 1 Asl     GSM2545336       1170 argininosuccinate lyase [Source:MGI Symbol;Acc…
 2 Apod    GSM2545336      36194 apolipoprotein D [Source:MGI Symbol;Acc:MGI:88…
 3 Cyp2d22 GSM2545336       4060 cytochrome P450, family 2, subfamily d, polype…
 4 Klk6    GSM2545336        287 kallikrein related-peptidase 6 [Source:MGI Sym…
 5 Fcrls   GSM2545336         85 Fc receptor-like S, scavenger receptor [Source…
 6 Slc2a4  GSM2545336        782 solute carrier family 2 (facilitated glucose t…
 7 Exd2    GSM2545336       1619 exonuclease 3'-5' domain containing 2 [Source:…
 8 Gjc2    GSM2545336        288 gap junction protein, gamma 2 [Source:MGI Symb…
 9 Plp1    GSM2545336      43217 proteolipid protein (myelin) 1 [Source:MGI Sym…
10 Gnb4    GSM2545336       1071 guanine nucleotide binding protein (G protein)…</code></pre>
</div>
<p>In real life, gene annotations are sometimes labelled
differently.</p>
<p>The <code>annot2</code> table is exactly the same than
<code>annot1</code> except that the variable containing gene names is
labelled differently. Again, either <a href="https://carpentries-incubator.github.io/bioc-intro/data/annot2.csv" class="external-link">download
annot2.csv</a> yourself and move it to <code>data/</code> or use the R
code below.</p>
<div class="codewrapper sourceCode" id="cb142">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">download.file</span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://carpentries-incubator.github.io/bioc-intro/data/annot2.csv"</span>,</span>
<span>              destfile <span class="op">=</span> <span class="st">"data/annot2.csv"</span><span class="op">)</span></span>
<span><span class="va">annot2</span> <span class="op">&lt;-</span> <span class="fu">read_csv</span><span class="op">(</span>file <span class="op">=</span> <span class="st">"data/annot2.csv"</span><span class="op">)</span></span>
<span><span class="va">annot2</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 10 × 2
   external_gene_name description
   &lt;chr&gt;              &lt;chr&gt;
 1 Cyp2d22            cytochrome P450, family 2, subfamily d, polypeptide 22 [S…
 2 Klk6               kallikrein related-peptidase 6 [Source:MGI Symbol;Acc:MGI…
 3 Fcrls              Fc receptor-like S, scavenger receptor [Source:MGI Symbol…
 4 Plp1               proteolipid protein (myelin) 1 [Source:MGI Symbol;Acc:MGI…
 5 Exd2               exonuclease 3'-5' domain containing 2 [Source:MGI Symbol;…
 6 Apod               apolipoprotein D [Source:MGI Symbol;Acc:MGI:88056]
 7 Gnb4               guanine nucleotide binding protein (G protein), beta 4 [S…
 8 Slc2a4             solute carrier family 2 (facilitated glucose transporter)…
 9 Asl                argininosuccinate lyase [Source:MGI Symbol;Acc:MGI:88084]
10 Gjc2               gap junction protein, gamma 2 [Source:MGI Symbol;Acc:MGI:…</code></pre>
</div>
<p>In case none of the variable names match, we can set manually the
variables to use for the matching. These variables can be set using the
<code>by</code> argument, as shown below with <code>rna_mini</code> and
<code>annot2</code> tables.</p>
<div class="codewrapper sourceCode" id="cb144">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">full_join</span><span class="op">(</span><span class="va">rna_mini</span>, <span class="va">annot2</span>, by <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"gene"</span> <span class="op">=</span> <span class="st">"external_gene_name"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 10 × 4
   gene    sample     expression description
   &lt;chr&gt;   &lt;chr&gt;           &lt;dbl&gt; &lt;chr&gt;
 1 Asl     GSM2545336       1170 argininosuccinate lyase [Source:MGI Symbol;Acc…
 2 Apod    GSM2545336      36194 apolipoprotein D [Source:MGI Symbol;Acc:MGI:88…
 3 Cyp2d22 GSM2545336       4060 cytochrome P450, family 2, subfamily d, polype…
 4 Klk6    GSM2545336        287 kallikrein related-peptidase 6 [Source:MGI Sym…
 5 Fcrls   GSM2545336         85 Fc receptor-like S, scavenger receptor [Source…
 6 Slc2a4  GSM2545336        782 solute carrier family 2 (facilitated glucose t…
 7 Exd2    GSM2545336       1619 exonuclease 3'-5' domain containing 2 [Source:…
 8 Gjc2    GSM2545336        288 gap junction protein, gamma 2 [Source:MGI Symb…
 9 Plp1    GSM2545336      43217 proteolipid protein (myelin) 1 [Source:MGI Sym…
10 Gnb4    GSM2545336       1071 guanine nucleotide binding protein (G protein)…</code></pre>
</div>
<p>As can be seen above, the variable name of the first table is
retained in the joined one.</p>
<div id="challenge-4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-4" class="callout-inner">
<h3 class="callout-title">Challenge:</h3>
<div class="callout-content">
<p>Download the <code>annot3</code> table by <a href="https://carpentries-incubator.github.io/bioc-intro/data/annot3.csv" class="external-link">clicking
this link</a> and put the table in your data/ repository. Using the
<code>full_join()</code> function, join tables <code>rna_mini</code> and
<code>annot3</code>. What has happened for genes <em>Klk6</em>,
<em>mt-Tf</em>, <em>mt-Rnr1</em>, <em>mt-Tv</em>, <em>mt-Rnr2</em>, and
<em>mt-Tl1</em> ?</p>
</div>
</div>
</div>
<div id="accordionSolution9" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution9" aria-expanded="false" aria-controls="collapseSolution9">
  <h4 class="accordion-header" id="headingSolution9"> Show me the solution </h4>
</button>
<div id="collapseSolution9" class="accordion-collapse collapse" aria-labelledby="headingSolution9" data-bs-parent="#accordionSolution9">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb146">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">annot3</span> <span class="op">&lt;-</span> <span class="fu">read_csv</span><span class="op">(</span><span class="st">"data/annot3.csv"</span><span class="op">)</span></span>
<span><span class="fu">full_join</span><span class="op">(</span><span class="va">rna_mini</span>, <span class="va">annot3</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 15 × 4
   gene    sample     expression gene_description
   &lt;chr&gt;   &lt;chr&gt;           &lt;dbl&gt; &lt;chr&gt;
 1 Asl     GSM2545336       1170 argininosuccinate lyase [Source:MGI Symbol;Acc…
 2 Apod    GSM2545336      36194 apolipoprotein D [Source:MGI Symbol;Acc:MGI:88…
 3 Cyp2d22 GSM2545336       4060 cytochrome P450, family 2, subfamily d, polype…
 4 Klk6    GSM2545336        287 &lt;NA&gt;
 5 Fcrls   GSM2545336         85 Fc receptor-like S, scavenger receptor [Source…
 6 Slc2a4  GSM2545336        782 solute carrier family 2 (facilitated glucose t…
 7 Exd2    GSM2545336       1619 exonuclease 3'-5' domain containing 2 [Source:…
 8 Gjc2    GSM2545336        288 gap junction protein, gamma 2 [Source:MGI Symb…
 9 Plp1    GSM2545336      43217 proteolipid protein (myelin) 1 [Source:MGI Sym…
10 Gnb4    GSM2545336       1071 guanine nucleotide binding protein (G protein)…
11 mt-Tf   &lt;NA&gt;               NA mitochondrially encoded tRNA phenylalanine [So…
12 mt-Rnr1 &lt;NA&gt;               NA mitochondrially encoded 12S rRNA [Source:MGI S…
13 mt-Tv   &lt;NA&gt;               NA mitochondrially encoded tRNA valine [Source:MG…
14 mt-Rnr2 &lt;NA&gt;               NA mitochondrially encoded 16S rRNA [Source:MGI S…
15 mt-Tl1  &lt;NA&gt;               NA mitochondrially encoded tRNA leucine 1 [Source…</code></pre>
</div>
<p>Genes <em>Klk6</em> is only present in <code>rna_mini</code>, while
genes <em>mt-Tf</em>, <em>mt-Rnr1</em>, <em>mt-Tv</em>,
<em>mt-Rnr2</em>, and <em>mt-Tl1</em> are only present in
<code>annot3</code> table. Their respective values for the variables of
the table have been encoded as missing.</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="exporting-data">Exporting data<a class="anchor" aria-label="anchor" href="#exporting-data"></a>
</h2>
<hr class="half-width">
<p>Now that you have learned how to use <code>dplyr</code> to extract
information from or summarise your raw data, you may want to export
these new data sets to share them with your collaborators or for
archival.</p>
<p>Similar to the <code>read_csv()</code> function used for reading CSV
files into R, there is a <code>write_csv()</code> function that
generates CSV files from data frames.</p>
<p>Before using <code>write_csv()</code>, we are going to create a new
folder, <code>data_output</code>, in our working directory that will
store this generated dataset. We don’t want to write generated datasets
in the same directory as our raw data. It’s good practice to keep them
separate. The <code>data</code> folder should only contain the raw,
unaltered data, and should be left alone to make sure we don’t delete or
modify it. In contrast, our script will generate the contents of the
<code>data_output</code> directory, so even if the files it contains are
deleted, we can always re-generate them.</p>
<p>Let’s use <code>write_csv()</code> to save the rna_wide table that we
have created previously.</p>
<div class="codewrapper sourceCode" id="cb148">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">write_csv</span><span class="op">(</span><span class="va">rna_wide</span>, file <span class="op">=</span> <span class="st">"data_output/rna_wide.csv"</span><span class="op">)</span></span></code></pre>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>Tabular data in R using the tidyverse meta-package</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-40-visualization"><p>Content from <a href="40-visualization.html">Data visualization</a></p>
<hr>
<p>Last updated on 2025-06-15 |

        <a href="https://github.com/carpentries-incubator/bioc-intro/edit/main/episodes/40-visualization.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>Visualization in R</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Produce scatter plots, boxplots, line plots, etc. using ggplot.</li>
<li>Set universal plot settings.</li>
<li>Describe what faceting is and apply faceting in ggplot.</li>
<li>Modify the aesthetics of an existing ggplot plot (including axis
labels and color).</li>
<li>Build complex and customized plots from data in a data frame.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<blockquote>
<p>This episode is based on the Data Carpentries’s <em>Data Analysis and
Visualisation in R for Ecologists</em> lesson.</p>
</blockquote>
<section><h2 class="section-heading" id="data-visualization">Data Visualization<a class="anchor" aria-label="anchor" href="#data-visualization"></a>
</h2>
<hr class="half-width">
<p>We start by loading the required packages.
<strong><code>ggplot2</code></strong> is included in the
<strong><code>tidyverse</code></strong> package.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="st">"tidyverse"</span><span class="op">)</span></span></code></pre>
</div>
<p>If not still in the workspace, load the data we saved in the previous
lesson.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span><span class="st">"data/rnaseq.csv"</span><span class="op">)</span></span></code></pre>
</div>
<p>The <a href="https://raw.githubusercontent.com/rstudio/cheatsheets/main/data-visualization.pdf" class="external-link">Data
Visualization Cheat Sheet</a> will cover the basics and more advanced
features of <code>ggplot2</code> and will help, in addition to serve as
a reminder, getting an overview of the many data representations
available in the package. The following video tutorials (<a href="https://www.youtube.com/watch?v=h29g21z0a68" class="external-link">part 1</a> and <a href="https://www.youtube.com/watch?v=0m4yywqNPVY" class="external-link">part 2</a>) by Thomas
Lin Pedersen are also very instructive.</p>
</section><section><h2 class="section-heading" id="plotting-with-ggplot2">Plotting with <code>ggplot2</code>
<a class="anchor" aria-label="anchor" href="#plotting-with-ggplot2"></a>
</h2>
<hr class="half-width">
<p><code>ggplot2</code> is a plotting package that makes it simple to
create complex plots from data in a data frame. It provides a more
programmatic interface for specifying what variables to plot, how they
are displayed, and general visual properties. The theoretical foundation
that supports the <code>ggplot2</code> is the <em>Grammar of
Graphics</em> (<span class="citation">@Wilkinson:2005</span>). Using
this approach, we only need minimal changes if the underlying data
change or if we decide to change from a bar plot to a scatterplot. This
helps in creating publication quality plots with minimal amounts of
adjustments and tweaking.</p>
<p>There is a book about <code>ggplot2</code> (<span class="citation">@ggplot2book</span>) that provides a good overview, but
it is outdated. The 3rd edition is in preparation and will be <a href="https://ggplot2-book.org/" class="external-link">freely available online</a>. The
<code>ggplot2</code> webpage (<a href="https://ggplot2.tidyverse.org" class="external-link">https://ggplot2.tidyverse.org</a>)
provides ample documentation.</p>
<p><code>ggplot2</code> functions like data in the ‘long’ format, i.e.,
a column for every dimension, and a row for every observation.
Well-structured data will save you lots of time when making figures with
<code>ggplot2</code>.</p>
<p>ggplot graphics are built step by step by adding new elements. Adding
layers in this fashion allows for extensive flexibility and
customization of plots.</p>
<blockquote>
<p>The idea behind the Grammar of Graphics it is that you can build
every graph from the same 3 components: (1) a data set, (2) a coordinate
system, and (3) geoms — i.e. visual marks that represent data points <a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a></p>
</blockquote>
<p>To build a ggplot, we will use the following basic template that can
be used for different types of plots:</p>
<pre><code>ggplot(data = &lt;DATA&gt;, mapping = aes(&lt;MAPPINGS&gt;)) +  &lt;GEOM_FUNCTION&gt;()</code></pre>
<ul>
<li>use the <code>ggplot()</code> function and bind the plot to a
specific <strong>data frame</strong> using the <code>data</code>
argument</li>
</ul>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">rna</span><span class="op">)</span></span></code></pre>
</div>
<ul>
<li>define a <strong>mapping</strong> (using the aesthetic
(<code>aes</code>) function), by selecting the variables to be plotted
and specifying how to present them in the graph, e.g. as x/y positions
or characteristics such as size, shape, color, etc.</li>
</ul>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">rna</span>, mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">expression</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<ul>
<li>
<p>add ‘<strong>geoms</strong>’ - geometries, or graphical
representations of the data in the plot (points, lines, bars).
<code>ggplot2</code> offers many different geoms; we will use some
common ones today, including:</p>
<pre><code>* `geom_point()` for scatter plots, dot plots, etc.
* `geom_histogram()` for histograms
* `geom_boxplot()` for, well, boxplots!
* `geom_line()` for trend lines, time series, etc.</code></pre>
</li>
</ul>
<p>To add a geom(etry) to the plot use the <code>+</code> operator.
Let’s use <code>geom_histogram()</code> first:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">rna</span>, mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_histogram</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-first-ggplot-1.png" alt="Default histogram produced by ggplot() and geom_histogram() for the expression data." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>The <code>+</code> in the <code>ggplot2</code> package is
particularly useful because it allows you to modify existing
<code>ggplot</code> objects. This means you can easily set up plot
templates and conveniently explore different types of plots, so the
above plot can also be generated with code like this:</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Assign plot to a variable</span></span>
<span><span class="va">rna_plot</span> <span class="op">&lt;-</span> <span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">rna</span>,</span>
<span>                   mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">expression</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Draw the plot</span></span>
<span><span class="va">rna_plot</span> <span class="op">+</span> <span class="fu">geom_histogram</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div id="challenge" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>You have probably noticed an automatic message that appears when
drawing the histogram:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
</div>
<p>Change the arguments <code>bins</code> or <code>binwidth</code> of
<code>geom_histogram()</code> to change the number or width of the
bins.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># change bins</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">rna</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_histogram</span><span class="op">(</span>bins <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-unnamed-chunk-5-1.png" alt="Histograms produced by ggplot() and geom_histogram() for the expression data with bin set of 15 (top) and binwith set to 2000 (bottom)." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># change binwidth</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">rna</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_histogram</span><span class="op">(</span>binwidth <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-unnamed-chunk-5-2.png" alt="Histograms produced by ggplot() and geom_histogram() for the expression data with bin set of 15 (top) and binwith set to 2000 (bottom)." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
</div>
<p>We can observe here that the data are skewed to the right. We can
apply log2 transformation to have a more symmetric distribution. Note
that we add here a small constant value (<code>+1</code>) to avoid
having <code>-Inf</code> values returned for expression values equal to
0.</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">&lt;-</span> <span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span>expression_log <span class="op">=</span> <span class="fu">log2</span><span class="op">(</span><span class="va">expression</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>If we now draw the histogram of the log2-transformed expressions, the
distribution is indeed closer to a normal distribution.</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span><span class="va">rna</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">expression_log</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_histogram</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-second-ggplot-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>From now on we will work on the log-transformed expression
values.</p>
<div id="challenge-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Another way to visualize this transformation is to consider the scale
of the observations. For example, it may be worth changing the scale of
the axis to better distribute the observations in the space of the plot.
Changing the scale of the axes is done similarly to adding/modifying
other components (i.e., by incrementally adding commands). Try making
this modification:</p>
<ul>
<li>Represent the un-transformed expression on the log10 scale; see
<code>scale_x_log10()</code>. Compare it with the previous graph. Why do
you now have warning messages appearing?</li>
</ul>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">rna</span>,mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">expression</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu">geom_histogram</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_x_log10</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning in scale_x_log10(): log-10 transformation introduced infinite values.</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning: Removed 507 rows containing non-finite outside the scale range
(`stat_bin()`).</code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-unnamed-chunk-6-1.png" alt="Histograms produced by ggplot(), geom_histogram() and scale_x_log10() for the log of expression." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
</div>
<p><strong>Notes</strong></p>
<ul>
<li>Anything you put in the <code>ggplot()</code> function can be seen
by any geom layers that you add (i.e., these are global plot settings).
This includes the x- and y-axis mapping you set up in
<code>aes()</code>.</li>
<li>You can also specify mappings for a given geom independently of the
mappings defined globally in the <code>ggplot()</code> function.</li>
<li>The <code>+</code> sign used to add new layers must be placed at the
end of the line containing the <em>previous</em> layer. If, instead, the
<code>+</code> sign is added at the beginning of the line containing the
new layer, <code>ggplot2</code> will not add the new layer and will
return an error message.</li>
</ul>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># This is the correct syntax for adding layers</span></span>
<span><span class="va">rna_plot</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_histogram</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># This will not add the new layer and will return an error message</span></span>
<span><span class="va">rna_plot</span></span>
<span>  <span class="op">+</span> <span class="fu">geom_histogram</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="building-your-plots-iteratively">Building your plots iteratively<a class="anchor" aria-label="anchor" href="#building-your-plots-iteratively"></a>
</h2>
<hr class="half-width">
<p>We will now draw a scatter plot with two continuous variables and the
<code>geom_point()</code> function. This graph will represent the log2
fold changes of expression comparing time 8 versus time 0, and time 4
versus time 0. To this end, we first need to compute the means of the
log-transformed expression values by gene and time, then the log fold
changes by subtracting the mean log expressions between time 8 and time
0 and between time 4 and time 0. Note that we also include here the gene
biotype that we will use later on to represent the genes. We will save
the fold changes in a new data frame called <code>rna_fc.</code></p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_fc</span> <span class="op">&lt;-</span> <span class="va">rna</span> <span class="op">|&gt;</span> <span class="fu">select</span><span class="op">(</span><span class="va">gene</span>, <span class="va">time</span>,</span>
<span>                         <span class="va">gene_biotype</span>, <span class="va">expression_log</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene</span>, <span class="va">time</span>, <span class="va">gene_biotype</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">summarize</span><span class="op">(</span>mean_exp <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">expression_log</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">time</span>,</span>
<span>              values_from <span class="op">=</span> <span class="va">mean_exp</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span>time_8_vs_0 <span class="op">=</span> <span class="va">`8`</span> <span class="op">-</span> <span class="va">`0`</span>, time_4_vs_0 <span class="op">=</span> <span class="va">`4`</span> <span class="op">-</span> <span class="va">`0`</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>`summarise()` has grouped output by 'gene', 'time'. You can override using the
`.groups` argument.</code></pre>
</div>
<p>We can then build a ggplot with the newly created dataset
<code>rna_fc</code>. Building plots with <code>ggplot2</code> is
typically an iterative process. We start by defining the dataset we’ll
use, lay out the axes, and choose a geom:</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">rna_fc</span>, mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">time_4_vs_0</span>, y <span class="op">=</span> <span class="va">time_8_vs_0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-create-ggplot-object-1.png" alt="Scatter plot produced by ggplot() and geom_point() comparing the log-foldchanges computed above. All dots are black." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Then, we start modifying this plot to extract more information from
it. For instance, we can add transparency (<code>alpha</code>) to avoid
overplotting:</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">rna_fc</span>, mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">time_4_vs_0</span>, y <span class="op">=</span> <span class="va">time_8_vs_0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-adding-transparency-1.png" alt="Scatter plot produced by ggplot() and geom_point() comparing the log-foldchanges computed above. All dots are semi-transparent black." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We can also add colors for all the points:</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">rna_fc</span>, mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">time_4_vs_0</span>, y <span class="op">=</span> <span class="va">time_8_vs_0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.3</span>, color <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-adding-colors-1.png" alt="Scatter plot produced by ggplot() and geom_point() comparing the log-foldchanges computed above. All dots are semi-transparent blue." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Or to color each gene in the plot differently, you could use a vector
as an input to the argument <strong>color</strong>. <code>ggplot2</code>
will provide a different color corresponding to different values in the
vector. Here is an example where we color with
<code>gene_biotype</code>:</p>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">rna_fc</span>, mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">time_4_vs_0</span>, y <span class="op">=</span> <span class="va">time_8_vs_0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.3</span>, <span class="fu">aes</span><span class="op">(</span>color <span class="op">=</span> <span class="va">gene_biotype</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-color-by-gene_biotype1-1.png" alt="Scatter plot produced by ggplot() and geom_point() comparing the log-foldchanges computed above. Dots are colour-coded based on the gene's biotype." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We can also specify the colors directly inside the mapping provided
in the <code>ggplot()</code> function. This will be seen by any geom
layers and the mapping will be determined by the x- and y-axis set up in
<code>aes()</code>.</p>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">rna_fc</span>, mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">time_4_vs_0</span>, y <span class="op">=</span> <span class="va">time_8_vs_0</span>,</span>
<span>                                color <span class="op">=</span> <span class="va">gene_biotype</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-color-by-gene_biotype2-1.png" alt="Scatter plot produced by ggplot() and geom_point() comparing the log-foldchanges computed above. Dots are colour-coded based on the gene's biotype." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Finally, we could also add a diagonal line with the
<code>geom_abline()</code> function:</p>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">rna_fc</span>, mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">time_4_vs_0</span>, y <span class="op">=</span> <span class="va">time_8_vs_0</span>,</span>
<span>                                color <span class="op">=</span> <span class="va">gene_biotype</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_abline</span><span class="op">(</span>intercept <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-adding-diag-1.png" alt="Scatter plot produced by ggplot() and geom_point() comparing the log-foldchanges computed above. Dots are colour-coded based on the gene's biotype. A black line of slope 1 crossing the origin was added by geom_abline()." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Notice that we can change the geom layer from <code>geom_point</code>
to <code>geom_jitter</code> and colors will still be determined by
<code>gene_biotype</code>.</p>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">rna_fc</span>, mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">time_4_vs_0</span>, y <span class="op">=</span> <span class="va">time_8_vs_0</span>,</span>
<span>                                color <span class="op">=</span> <span class="va">gene_biotype</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_jitter</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_abline</span><span class="op">(</span>intercept <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-color-by-gene_biotype3-1.png" alt="Scatter plot produced by ggplot() and geom_point() comparing the log-foldchanges computed above. Dots are colour-coded based on the gene's biotype. A black line of slope 1 crossing the origin was added by geom_abline()." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div id="challenge-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Scatter plots can be useful exploratory tools for small datasets. For
data sets with large numbers of observations, such as the
<code>rna_fc</code> data set, overplotting of points can be a limitation
of scatter plots. One strategy for handling such settings is to use
hexagonal binning of observations. The plot space is tessellated into
hexagons. Each hexagon is assigned a color based on the number of
observations that fall within its boundaries.</p>
<ul>
<li><p>To use hexagonal binning in <code>ggplot2</code>, first install
the R package <code>hexbin</code> from CRAN and load it.</p></li>
<li><p>Then use the <code>geom_hex()</code> function to produce the
hexbin figure.</p></li>
<li><p>What are the relative strengths and weaknesses of a hexagonal bin
plot compared to a scatter plot? Examine the above scatter plot and
compare it with the hexagonal bin plot that you created.</p></li>
</ul>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"hexbin"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="st">"hexbin"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">rna_fc</span>, mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">time_4_vs_0</span>, y <span class="op">=</span> <span class="va">time_8_vs_0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_hex</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_abline</span><span class="op">(</span>intercept <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-unnamed-chunk-10-1.png" alt="Scatter plot produced by ggplot() and geom_hexbin() comparing the log-foldchanges computed above shows hexagons coloured based on the underlying dot density. A black line of slope 1 crossing the origin was added by geom_abline()." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
</div>
<div id="challenge-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-3" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Use what you just learned to create a scatter plot of
<code>expression_log</code> over <code>sample</code> from the
<code>rna</code> dataset with the time showing in different colors. Is
this a good way to show this type of data?</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">rna</span>, mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>y <span class="op">=</span> <span class="va">expression_log</span>, x <span class="op">=</span> <span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_point</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>color <span class="op">=</span> <span class="va">time</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-unnamed-chunk-11-1.png" alt="Figures showing a stretch of overlapping points indicating the log of expression + 1 for each sample. The points are coloured with different shades of blue for samples collected at different time points." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="boxplot">Boxplot<a class="anchor" aria-label="anchor" href="#boxplot"></a>
</h2>
<hr class="half-width">
<p>We can use boxplots to visualize the distribution of gene expressions
within each sample:</p>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">rna</span>,</span>
<span>         mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>y <span class="op">=</span> <span class="va">expression_log</span>, x <span class="op">=</span> <span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_boxplot</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-boxplot-1.png" alt="Boxplot showing the distribution of log expression + 1 values for each sample, as produced by geom_boxlpot(). Each boxplot is filled with white colour." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>By adding points to boxplot, we can have a better idea of the number
of measurements and of their distribution:</p>
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">rna</span>,</span>
<span>         mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>y <span class="op">=</span> <span class="va">expression_log</span>, x <span class="op">=</span> <span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_jitter</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.2</span>, color <span class="op">=</span> <span class="st">"tomato"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_boxplot</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-boxplot-with-points-1.png" alt="Boxplot and dots showing the distribution of log expression + 1 values for each sample, as produced by geom_boxlpot(). Each boxplot is transparent and the jittered dots are semi-transparent tomato-coloured and behind the boxpots." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div id="challenge-4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-4" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Note how the boxplot layer is in front of the jitter layer? What do
you need to change in the code to put the boxplot below the points?</p>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" aria-labelledby="headingSolution5" data-bs-parent="#accordionSolution5">
<div class="accordion-body">
<p>We should switch the order of these two geoms:</p>
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">rna</span>,</span>
<span>         mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>y <span class="op">=</span> <span class="va">expression_log</span>, x <span class="op">=</span> <span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_boxplot</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_jitter</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.2</span>, color <span class="op">=</span> <span class="st">"tomato"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-boxplot-with-points2-1.png" alt="Boxplot and dots showing the distribution of log expression + 1 values for each sample, as produced by geom_boxlpot(). Each boxplot is transparent and the jittered dots are semi-transparent tomato-coloured. This time, the boxplots are behind the dots." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
</div>
<p>You may notice that the values on the x-axis are still not properly
readable. Let’s change the orientation of the labels and adjust them
vertically and horizontally so they don’t overlap. You can use a
90-degree angle, or experiment to find the appropriate angle for
diagonally oriented labels:</p>
<div class="codewrapper sourceCode" id="cb36">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">rna</span>,</span>
<span>         mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>y <span class="op">=</span> <span class="va">expression_log</span>, x <span class="op">=</span> <span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_jitter</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.2</span>, color <span class="op">=</span> <span class="st">"tomato"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_boxplot</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>,  hjust <span class="op">=</span> <span class="fl">0.5</span>, vjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-boxplot-xaxis-rotated-1.png" alt="Boxplot and dots showing the distribution of log expression + 1 values for each sample, as produced by geom_boxlpot(). Each boxplot is transparent and the jittered dots are semi-transparent tomato-coloured. The sample labels are displayed vertically and readable." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div id="challenge-5" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-5" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Add color to the data points on your boxplot according to the
duration of the infection (<code>time</code>).</p>
<p><em>Hint:</em> Check the class for <code>time</code>. Consider
changing the class of <code>time</code> from integer to factor directly
in the ggplot mapping. Why does this change how R makes the graph?</p>
</div>
</div>
</div>
<div id="accordionSolution6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution6" aria-expanded="false" aria-controls="collapseSolution6">
  <h4 class="accordion-header" id="headingSolution6"> Show me the solution </h4>
</button>
<div id="collapseSolution6" class="accordion-collapse collapse" aria-labelledby="headingSolution6" data-bs-parent="#accordionSolution6">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># time as integer</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">rna</span>,</span>
<span>         mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>y <span class="op">=</span> <span class="va">expression_log</span>,</span>
<span>                       x <span class="op">=</span> <span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_jitter</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.2</span>, <span class="fu">aes</span><span class="op">(</span>color <span class="op">=</span> <span class="va">time</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_boxplot</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>,  hjust <span class="op">=</span> <span class="fl">0.5</span>, vjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-boxplot-color-time-1.png" alt="Boxplot and dots showing the distribution of log expression + 1 values for each sample, as produced by geom_boxlpot(). On the first figure, each boxplot is transparent and the jittered dots are semi-transparent and coloured in different shares of blue. On the second figures, each boxplot is transparent and the jittered dots are semi-transparent and coloured red, green and blue." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb38">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># time as factor</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">rna</span>,</span>
<span>         mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>y <span class="op">=</span> <span class="va">expression_log</span>,</span>
<span>                       x <span class="op">=</span> <span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_jitter</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.2</span>, <span class="fu">aes</span><span class="op">(</span>color <span class="op">=</span> <span class="fu">as.factor</span><span class="op">(</span><span class="va">time</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_boxplot</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>,  hjust <span class="op">=</span> <span class="fl">0.5</span>, vjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-boxplot-color-time-2.png" alt="Boxplot and dots showing the distribution of log expression + 1 values for each sample, as produced by geom_boxlpot(). On the first figure, each boxplot is transparent and the jittered dots are semi-transparent and coloured in different shares of blue. On the second figures, each boxplot is transparent and the jittered dots are semi-transparent and coloured red, green and blue." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
</div>
<div id="challenge-6" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-6" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Boxplots are useful summaries, but hide the <em>shape</em> of the
distribution. For example, if the distribution is bimodal, we would not
see it in a boxplot. An alternative to the boxplot is the violin plot,
where the shape (of the density of points) is drawn.</p>
<ul>
<li>Replace the box plot with a violin plot; see
<code>geom_violin()</code>. Fill in the violins according to the time
with the argument <code>fill</code>.</li>
</ul>
</div>
</div>
</div>
<div id="accordionSolution7" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution7" aria-expanded="false" aria-controls="collapseSolution7">
  <h4 class="accordion-header" id="headingSolution7"> Show me the solution </h4>
</button>
<div id="collapseSolution7" class="accordion-collapse collapse" aria-labelledby="headingSolution7" data-bs-parent="#accordionSolution7">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb39">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">rna</span>,</span>
<span>         mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>y <span class="op">=</span> <span class="va">expression_log</span>, x <span class="op">=</span> <span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_violin</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>fill <span class="op">=</span> <span class="fu">as.factor</span><span class="op">(</span><span class="va">time</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>,  hjust <span class="op">=</span> <span class="fl">0.5</span>, vjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-unnamed-chunk-12-1.png" alt="Violin plot showing the distribution of log expression + 1 values for each sample, as produced by geom_violin(). Each boxplot is transparent and the jittered dots are semi-transparent and coloured red, green and blue." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
</div>
<div id="challenge-7" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-7" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<ul>
<li>Modify the violin plot to fill in the violins by
<code>sex</code>.</li>
</ul>
</div>
</div>
</div>
<div id="accordionSolution8" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution8" aria-expanded="false" aria-controls="collapseSolution8">
  <h4 class="accordion-header" id="headingSolution8"> Show me the solution </h4>
</button>
<div id="collapseSolution8" class="accordion-collapse collapse" aria-labelledby="headingSolution8" data-bs-parent="#accordionSolution8">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb40">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">rna</span>,</span>
<span>         mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>y <span class="op">=</span> <span class="va">expression_log</span>, x <span class="op">=</span> <span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_violin</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>fill <span class="op">=</span> <span class="va">sex</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>,  hjust <span class="op">=</span> <span class="fl">0.5</span>, vjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-unnamed-chunk-13-1.png" alt="Violin plot showing the distribution of log expression + 1 values for each sample, as produced by geom_violon(). Each violin plot is coloured in red or blue depending on the sex variable." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="line-plots">Line plots<a class="anchor" aria-label="anchor" href="#line-plots"></a>
</h2>
<hr class="half-width">
<p>Let’s calculate the mean expression per duration of the infection for
the 10 genes having the highest log fold changes comparing time 8 versus
time 0. First, we need to select the genes and create a subset of
<code>rna</code> called <code>sub_rna</code> containing the 10 selected
genes, then we need to group the data and calculate the mean gene
expression within each group:</p>
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_fc</span> <span class="op">&lt;-</span> <span class="va">rna_fc</span> <span class="op">|&gt;</span> <span class="fu">arrange</span><span class="op">(</span><span class="fu">desc</span><span class="op">(</span><span class="va">time_8_vs_0</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">genes_selected</span> <span class="op">&lt;-</span> <span class="va">rna_fc</span><span class="op">$</span><span class="va">gene</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span></span>
<span><span class="va">sub_rna</span> <span class="op">&lt;-</span> <span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">filter</span><span class="op">(</span><span class="va">gene</span> <span class="op">%in%</span> <span class="va">genes_selected</span><span class="op">)</span></span>
<span></span>
<span><span class="va">mean_exp_by_time</span> <span class="op">&lt;-</span> <span class="va">sub_rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene</span>,<span class="va">time</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">summarize</span><span class="op">(</span>mean_exp <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">expression_log</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>`summarise()` has grouped output by 'gene'. You can override using the
`.groups` argument.</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb43">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mean_exp_by_time</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 30 × 3
# Groups:   gene [10]
   gene   time mean_exp
   &lt;chr&gt; &lt;int&gt;    &lt;dbl&gt;
 1 Acr       0     5.07
 2 Acr       4     5.54
 3 Acr       8     7.31
 4 Aipl1     0     3.70
 5 Aipl1     4     3.89
 6 Aipl1     8     6.56
 7 Bst1      0     3.20
 8 Bst1      4     3.77
 9 Bst1      8     5.22
10 Chil3     0     4.00
# ℹ 20 more rows</code></pre>
</div>
<p>We can build the line plot with duration of the infection on the
x-axis and the mean expression on the y-axis:</p>
<div class="codewrapper sourceCode" id="cb45">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mean_exp_by_time</span>, mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">time</span>, y <span class="op">=</span> <span class="va">mean_exp</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_line</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-first-time-series-1.png" alt="Line plot, as produced by geom_line(), but the lines are clearly not show what we expect." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Unfortunately, this does not work because we plotted data for all the
genes together. We need to tell ggplot to draw a line for each gene by
modifying the aesthetic function to include
<code>group = gene</code>:</p>
<div class="codewrapper sourceCode" id="cb46">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mean_exp_by_time</span>,</span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">time</span>, y <span class="op">=</span> <span class="va">mean_exp</span>, group <span class="op">=</span> <span class="va">gene</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_line</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-time-series-by-gene-1.png" alt="Line plot, as produced by geom_line(), with 10 lines showing increasing expression values over time." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We will be able to distinguish genes in the plot if we add colors
(using <code>color</code> also automatically groups the data):</p>
<div class="codewrapper sourceCode" id="cb47">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mean_exp_by_time</span>,</span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">time</span>, y <span class="op">=</span> <span class="va">mean_exp</span>, color <span class="op">=</span> <span class="va">gene</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_line</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-time-series-with-colors-1.png" alt="Line plot, as produced by geom_line(), with 10 colour-coded lines showing increasing expression values over time." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></section><section><h2 class="section-heading" id="faceting">Faceting<a class="anchor" aria-label="anchor" href="#faceting"></a>
</h2>
<hr class="half-width">
<p><code>ggplot2</code> has a special technique called <em>faceting</em>
that allows the user to split one plot into multiple (sub) plots based
on a factor included in the dataset. These different subplots inherit
the same properties (axes limits, ticks, …) to facilitate their direct
comparison. We will use it to make a line plot across time for each
gene:</p>
<div class="codewrapper sourceCode" id="cb48">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mean_exp_by_time</span>,</span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">time</span>, y <span class="op">=</span> <span class="va">mean_exp</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_line</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-first-facet-1.png" alt="Line plot, as produced by geom_line(), with 10 sub-plots/facets, each showing one line with increasing expression values over time. All y-scales are identical." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Here both x- and y-axis have the same scale for all the subplots. You
can change this default behavior by modifying <code>scales</code> in
order to allow a free scale for the y-axis:</p>
<div class="codewrapper sourceCode" id="cb49">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mean_exp_by_time</span>,</span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">time</span>, y <span class="op">=</span> <span class="va">mean_exp</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_line</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene</span>, scales <span class="op">=</span> <span class="st">"free_y"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-first-facet-scales-1.png" alt="Line plot, as produced by geom_line(), with 10 sub-plots/facets, each showing one line with increasing expression values over time. y-scales are now adapted to the expression range for each facet/gene." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Now we would like to split the line in each plot by the sex of the
mice. To do that we need to calculate the mean expression in the data
frame grouped by <code>gene</code>, <code>time</code>, and
<code>sex</code>:</p>
<div class="codewrapper sourceCode" id="cb50">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mean_exp_by_time_sex</span> <span class="op">&lt;-</span> <span class="va">sub_rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene</span>, <span class="va">time</span>, <span class="va">sex</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">summarize</span><span class="op">(</span>mean_exp <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">expression_log</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>`summarise()` has grouped output by 'gene', 'time'. You can override using the
`.groups` argument.</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb52">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mean_exp_by_time_sex</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 60 × 4
# Groups:   gene, time [30]
   gene   time sex    mean_exp
   &lt;chr&gt; &lt;int&gt; &lt;chr&gt;     &lt;dbl&gt;
 1 Acr       0 Female     5.13
 2 Acr       0 Male       5.00
 3 Acr       4 Female     5.93
 4 Acr       4 Male       5.15
 5 Acr       8 Female     7.27
 6 Acr       8 Male       7.36
 7 Aipl1     0 Female     3.67
 8 Aipl1     0 Male       3.73
 9 Aipl1     4 Female     4.07
10 Aipl1     4 Male       3.72
# ℹ 50 more rows</code></pre>
</div>
<p>We can now make the faceted plot by splitting further by sex using
<code>color</code> (within a single plot):</p>
<div class="codewrapper sourceCode" id="cb54">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mean_exp_by_time_sex</span>,</span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">time</span>, y <span class="op">=</span> <span class="va">mean_exp</span>, color <span class="op">=</span> <span class="va">sex</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_line</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene</span>, scales <span class="op">=</span> <span class="st">"free_y"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-facet-by-gene-and-sex-1.png" alt="Line plot, as produced by geom_line(), with 10 sub-plots/facets, each showing two coloured lines (red for Female and blue for Make) with increasing expression values over time." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Usually plots with white background look more readable when printed.
We can set the background to white using the function
<code>theme_bw()</code>. Additionally, we can remove the grid:</p>
<div class="codewrapper sourceCode" id="cb55">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mean_exp_by_time_sex</span>,</span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">time</span>, y <span class="op">=</span> <span class="va">mean_exp</span>, color <span class="op">=</span> <span class="va">sex</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_line</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene</span>, scales <span class="op">=</span> <span class="st">"free_y"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>panel.grid <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-facet-by-gene-and-sex-white-bg-1.png" alt="Line plot, as produced by geom_line(), with 10 sub-plots/facets, each showing two coloured lines (red for Female and blue for Make) with increasing expression values over time. The figure background is now white." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div id="challenge-8" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-8" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Use what you just learned to create a plot that depicts how the
average expression of each chromosome changes through the duration of
infection.</p>
</div>
</div>
</div>
<div id="accordionSolution9" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution9" aria-expanded="false" aria-controls="collapseSolution9">
  <h4 class="accordion-header" id="headingSolution9"> Show me the solution </h4>
</button>
<div id="collapseSolution9" class="accordion-collapse collapse" aria-labelledby="headingSolution9" data-bs-parent="#accordionSolution9">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb56">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mean_exp_by_chromosome</span> <span class="op">&lt;-</span> <span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">chromosome_name</span>, <span class="va">time</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">summarize</span><span class="op">(</span>mean_exp <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">expression_log</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>`summarise()` has grouped output by 'chromosome_name'. You can override using
the `.groups` argument.</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb58">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mean_exp_by_chromosome</span>, mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">time</span>,</span>
<span>                                y <span class="op">=</span> <span class="va">mean_exp</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_line</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">chromosome_name</span>, scales <span class="op">=</span> <span class="st">"free_y"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-mean-exp-chromosome-time-series-1.png" alt="Line plot, as produced by geom_line(), with 21 sub-plots/facets, each showing one line with expression values over time for each chromosome." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
</div>
<p>The <code>facet_wrap</code> geometry extracts plots into an arbitrary
number of dimensions to allow them to cleanly fit on one page. On the
other hand, the <code>facet_grid</code> geometry allows you to
explicitly specify how you want your plots to be arranged via formula
notation (<code>rows ~ columns</code>; a <code>.</code> can be used as a
placeholder that indicates only one row or column).</p>
<p>Let’s modify the previous plot to compare how the mean gene
expression of males and females has changed through time:</p>
<div class="codewrapper sourceCode" id="cb59">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># One column, facet by rows</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mean_exp_by_time_sex</span>,</span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">time</span>, y <span class="op">=</span> <span class="va">mean_exp</span>, color <span class="op">=</span> <span class="va">gene</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_line</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_grid</span><span class="op">(</span><span class="va">sex</span> <span class="op">~</span> <span class="va">.</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-mean-exp-time-facet-sex-rows-1.png" alt="Two line plots above each other, each showing 10 lines coloured by gene. The top facet shows the expression values for Female sample, the bottom one for Male samples." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb60">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># One row, facet by column</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mean_exp_by_time_sex</span>,</span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">time</span>, y <span class="op">=</span> <span class="va">mean_exp</span>, color <span class="op">=</span> <span class="va">gene</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_line</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_grid</span><span class="op">(</span><span class="va">.</span> <span class="op">~</span> <span class="va">sex</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-mean-exp-time-facet-sex-columns-1.png" alt="Two line plots next to each other, each showing 10 lines coloured by gene. The left facet shows the expression values for Female sample, the right one for Male samples." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></section><section><h2 class="section-heading" id="ggplot2-themes">
<code>ggplot2</code> themes<a class="anchor" aria-label="anchor" href="#ggplot2-themes"></a>
</h2>
<hr class="half-width">
<p>In addition to <code>theme_bw()</code>, which changes the plot
background to white, <code>ggplot2</code> comes with several other
themes which can be useful to quickly change the look of your
visualization. The complete list of themes is available at <a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">https://ggplot2.tidyverse.org/reference/ggtheme.html</a>.
<code>theme_minimal()</code> and <code>theme_light()</code> are popular,
and <code>theme_void()</code> can be useful as a starting point to
create a new hand-crafted theme.</p>
<p>The <a href="https://jrnold.github.io/ggthemes/reference/index.html" class="external-link">ggthemes</a>
package provides a wide variety of options (including an Excel 2003
theme). The <a href="https://exts.ggplot2.tidyverse.org/" class="external-link"><code>ggplot2</code>
extensions website</a> provides a list of packages that extend the
capabilities of <code>ggplot2</code>, including additional themes.</p>
</section><section><h2 class="section-heading" id="customisation">Customisation<a class="anchor" aria-label="anchor" href="#customisation"></a>
</h2>
<hr class="half-width">
<p>Let’s come back to the faceted plot of mean expression by time and
gene, colored by sex.</p>
<p>Take a look at the <a href="https://raw.githubusercontent.com/rstudio/cheatsheets/main/data-visualization.pdf" class="external-link"><code>ggplot2</code>
cheat sheet</a>, and think of ways you could improve the plot.</p>
<p>Now, we can change names of axes to something more informative than
‘time’ and ‘mean_exp’, and add a title to the figure:</p>
<div class="codewrapper sourceCode" id="cb61">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mean_exp_by_time_sex</span>,</span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">time</span>, y <span class="op">=</span> <span class="va">mean_exp</span>, color <span class="op">=</span> <span class="va">sex</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_line</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene</span>, scales <span class="op">=</span> <span class="st">"free_y"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>panel.grid <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Mean gene expression by duration of the infection"</span>,</span>
<span>       x <span class="op">=</span> <span class="st">"Duration of the infection (in days)"</span>,</span>
<span>       y <span class="op">=</span> <span class="st">"Mean gene expression"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-mean_exp-time-with-right-labels-1.png" alt="Line plot with white background and custom title and axis legends." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>The axes have more informative names, but their readability can be
improved by increasing the font size:</p>
<div class="codewrapper sourceCode" id="cb62">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mean_exp_by_time_sex</span>,</span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">time</span>, y <span class="op">=</span> <span class="va">mean_exp</span>, color <span class="op">=</span> <span class="va">sex</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_line</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene</span>, scales <span class="op">=</span> <span class="st">"free_y"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>panel.grid <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Mean gene expression by duration of the infection"</span>,</span>
<span>       x <span class="op">=</span> <span class="st">"Duration of the infection (in days)"</span>,</span>
<span>       y <span class="op">=</span> <span class="st">"Mean gene expression"</span><span class="op">)</span>  <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>text <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">16</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-mean_exp-time-with-right-labels-xfont-size-1.png" alt="Line plot with white background and larger custom title and axis legends." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Note that it is also possible to change the fonts of your plots. If
you are on Windows, you may have to install the <a href="https://cran.r-project.org/web/packages/extrafont/index.html" class="external-link"><strong><code>extrafont</code></strong>
package</a>.</p>
<p>We can further customize the color of x- and y-axis text, the color
of the grid, etc. We can also for example move the legend to the top by
setting <code>legend.position</code> to <code>"top"</code>.</p>
<div class="codewrapper sourceCode" id="cb63">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mean_exp_by_time_sex</span>,</span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">time</span>, y <span class="op">=</span> <span class="va">mean_exp</span>, color <span class="op">=</span> <span class="va">sex</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_line</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene</span>, scales <span class="op">=</span> <span class="st">"free_y"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>panel.grid <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Mean gene expression by duration of the infection"</span>,</span>
<span>       x <span class="op">=</span> <span class="st">"Duration of the infection (in days)"</span>,</span>
<span>       y <span class="op">=</span> <span class="st">"Mean gene expression"</span><span class="op">)</span>  <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>text <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">16</span><span class="op">)</span>,</span>
<span>        axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>colour <span class="op">=</span> <span class="st">"royalblue4"</span>, size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>,</span>
<span>        axis.text.y <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>colour <span class="op">=</span> <span class="st">"royalblue4"</span>, size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>,</span>
<span>        panel.grid <span class="op">=</span> <span class="fu">element_line</span><span class="op">(</span>colour<span class="op">=</span><span class="st">"lightsteelblue1"</span><span class="op">)</span>,</span>
<span>        legend.position <span class="op">=</span> <span class="st">"top"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-mean_exp-time-with-theme-1.png" alt="Line plot with white background and larger custom title and axis legends and blue grid." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>If you like the changes you created better than the default theme,
you can save them as an object to be able to easily apply them to other
plots you may create. Here is an example with the histogram we have
previously created.</p>
<div class="codewrapper sourceCode" id="cb64">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">blue_theme</span> <span class="op">&lt;-</span> <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>colour <span class="op">=</span> <span class="st">"royalblue4"</span>,</span>
<span>                                               size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>,</span>
<span>                    axis.text.y <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>colour <span class="op">=</span> <span class="st">"royalblue4"</span>,</span>
<span>                                               size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>,</span>
<span>                    text <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">16</span><span class="op">)</span>,</span>
<span>                    panel.grid <span class="op">=</span> <span class="fu">element_line</span><span class="op">(</span>colour<span class="op">=</span><span class="st">"lightsteelblue1"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">rna</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">expression_log</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_histogram</span><span class="op">(</span>bins <span class="op">=</span> <span class="fl">20</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">blue_theme</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in blue_theme(): could not find function "blue_theme"</code></pre>
</div>
<div id="challenge-9" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-9" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>With all of this information in hand, please take another five
minutes to either improve one of the plots generated in this exercise or
create a beautiful graph of your own. Use the RStudio <a href="https://github.com/rstudio/cheatsheets/raw/master/data-visualization-2.1.pdf" class="external-link"><code>ggplot2</code>
cheat sheet</a> for inspiration. Here are some ideas:</p>
<ul>
<li>See if you can change the thickness of the lines.</li>
<li>Can you find a way to change the name of the legend? What about its
labels? (hint: look for a ggplot function starting with
<code>scale_</code>)</li>
<li>Try using a different color palette or manually specifying the
colors for the lines (see <a href="https://www.cookbook-r.com/Graphs/Colors_(ggplot2)/" class="external-link">https://www.cookbook-r.com/Graphs/Colors_(ggplot2)/</a>).</li>
</ul>
</div>
</div>
</div>
<div id="accordionSolution10" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution10" aria-expanded="false" aria-controls="collapseSolution10">
  <h4 class="accordion-header" id="headingSolution10"> Show me the solution </h4>
</button>
<div id="collapseSolution10" class="accordion-collapse collapse" aria-labelledby="headingSolution10" data-bs-parent="#accordionSolution10">
<div class="accordion-body">
<p>For example, based on this plot:</p>
<div class="codewrapper sourceCode" id="cb66">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mean_exp_by_time_sex</span>,</span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">time</span>, y <span class="op">=</span> <span class="va">mean_exp</span>, color <span class="op">=</span> <span class="va">sex</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_line</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene</span>, scales <span class="op">=</span> <span class="st">"free_y"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>panel.grid <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-unnamed-chunk-15-1.png" alt="Plain facetted line plot, as produced by theme_bw() and a blank grid above." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We can customize it the following ways:</p>
<div class="codewrapper sourceCode" id="cb67">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># change the thickness of the lines</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mean_exp_by_time_sex</span>,</span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">time</span>, y <span class="op">=</span> <span class="va">mean_exp</span>, color <span class="op">=</span> <span class="va">sex</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_line</span><span class="op">(</span>size<span class="op">=</span><span class="fl">1.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene</span>, scales <span class="op">=</span> <span class="st">"free_y"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>panel.grid <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning: Using `size` aesthetic for lines was deprecated in ggplot2 3.4.0.
ℹ Please use `linewidth` instead.
This warning is displayed once every 8 hours.
Call `lifecycle::last_lifecycle_warnings()` to see where this warning was
generated.</code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-unnamed-chunk-16-1.png" alt="Plain facetted line plot, as produced by theme_bw() and a blank grid with wide lines." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb69">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># change the name of the legend and the labels</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mean_exp_by_time_sex</span>,</span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">time</span>, y <span class="op">=</span> <span class="va">mean_exp</span>, color <span class="op">=</span> <span class="va">sex</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_line</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene</span>, scales <span class="op">=</span> <span class="st">"free_y"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>panel.grid <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_color_discrete</span><span class="op">(</span>name <span class="op">=</span> <span class="st">"Gender"</span>, labels <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"F"</span>, <span class="st">"M"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-unnamed-chunk-17-1.png" alt="Plain facetted line plot, as produced by theme_bw() and a blank grid with renamed colour labels." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb70">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># using a different color palette</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mean_exp_by_time_sex</span>,</span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">time</span>, y <span class="op">=</span> <span class="va">mean_exp</span>, color <span class="op">=</span> <span class="va">sex</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_line</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene</span>, scales <span class="op">=</span> <span class="st">"free_y"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>panel.grid <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_color_brewer</span><span class="op">(</span>name <span class="op">=</span> <span class="st">"Gender"</span>, labels <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"F"</span>, <span class="st">"M"</span><span class="op">)</span>, palette <span class="op">=</span> <span class="st">"Dark2"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-unnamed-chunk-18-1.png" alt="Plain facetted line plot, as produced by theme_bw() and a blank grid with a different color palette and renamed colour labels." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb71">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># manually specifying the colors</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mean_exp_by_time_sex</span>,</span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">time</span>, y <span class="op">=</span> <span class="va">mean_exp</span>, color <span class="op">=</span> <span class="va">sex</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_line</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene</span>, scales <span class="op">=</span> <span class="st">"free_y"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>panel.grid <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_color_manual</span><span class="op">(</span>name <span class="op">=</span> <span class="st">"Gender"</span>,  labels <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"F"</span>, <span class="st">"M"</span><span class="op">)</span>,</span>
<span>                     values <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"royalblue"</span>, <span class="st">"deeppink"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-unnamed-chunk-19-1.png" alt="Plain facetted line plot, as produced by theme_bw() and a blank grid with a manuelly-set colors and renamed colour labels.." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="composing-plots">Composing plots<a class="anchor" aria-label="anchor" href="#composing-plots"></a>
</h2>
<hr class="half-width">
<p>Faceting is a great tool for splitting one plot into multiple
subplots, but sometimes you may want to produce a single figure that
contains multiple independent plots, i.e. plots that are based on
different variables or even different data frames.</p>
<p>Let’s start by creating the two plots that we want to arrange next to
each other:</p>
<p>The first graph counts the number of unique genes per chromosome. We
first need to reorder the levels of <code>chromosome_name</code> and
filter the unique genes per chromosome. We also change the scale of the
y-axis to a log10 scale for better readability.</p>
<div class="codewrapper sourceCode" id="cb72">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span><span class="op">$</span><span class="va">chromosome_name</span> <span class="op">&lt;-</span> <span class="fu">factor</span><span class="op">(</span><span class="va">rna</span><span class="op">$</span><span class="va">chromosome_name</span>,</span>
<span>                               levels <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">19</span>,<span class="st">"X"</span>,<span class="st">"Y"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">count_gene_chromosome</span> <span class="op">&lt;-</span> <span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">chromosome_name</span>, <span class="va">gene</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">distinct</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">ggplot</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_bar</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">chromosome_name</span><span class="op">)</span>, fill <span class="op">=</span> <span class="st">"seagreen"</span>,</span>
<span>           position <span class="op">=</span> <span class="st">"dodge"</span>, stat <span class="op">=</span> <span class="st">"count"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>y <span class="op">=</span> <span class="st">"log10(n genes)"</span>, x <span class="op">=</span> <span class="st">"chromosome"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_y_log10</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">count_gene_chromosome</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-sub1-1.png" alt="A simple histogram with greed bars showing the log10 number of genes per chromosome." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Below, we also remove the legend altogether by setting the
<code>legend.position</code> to <code>"none"</code>.</p>
<div class="codewrapper sourceCode" id="cb73">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exp_boxplot_sex</span> <span class="op">&lt;-</span> <span class="fu">ggplot</span><span class="op">(</span><span class="va">rna</span>, <span class="fu">aes</span><span class="op">(</span>y<span class="op">=</span><span class="va">expression_log</span>, x <span class="op">=</span> <span class="fu">as.factor</span><span class="op">(</span><span class="va">time</span><span class="op">)</span>,</span>
<span>                 color<span class="op">=</span><span class="va">sex</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>   <span class="fu">geom_boxplot</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Mean gene exp"</span>,</span>
<span>       x <span class="op">=</span> <span class="st">"time"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">exp_boxplot_sex</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-sub2-1.png" alt="Transparently-filled red and blue boxplots for Female and Male expression values for different time points." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>The <a href="https://github.com/thomasp85/patchwork" class="external-link"><strong>patchwork</strong></a>
package provides an elegant approach to combining figures using the
<code>+</code> to arrange figures (typically side by side). More
specifically the <code>|</code> explicitly arranges them side by side
and <code>/</code> stacks them on top of each other.</p>
<div class="codewrapper sourceCode" id="cb74">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"patchwork"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb75">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="st">"patchwork"</span><span class="op">)</span></span>
<span><span class="va">count_gene_chromosome</span> <span class="op">+</span> <span class="va">exp_boxplot_sex</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-patchworkplot1-1.png" alt="A composed figure show the histogram (left) and boxplots (right) next to each other." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb76">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## or count_gene_chromosome | exp_boxplot_sex</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb77">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">count_gene_chromosome</span> <span class="op">/</span> <span class="va">exp_boxplot_sex</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-patchwork2-1.png" alt="A composed figure show the histogram (top) and boxplots (bottom) on top or each other." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We can combine further control the layout of the final composition
with <code>plot_layout</code> to create more complex layouts:</p>
<div class="codewrapper sourceCode" id="cb78">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">count_gene_chromosome</span> <span class="op">+</span> <span class="va">exp_boxplot_sex</span> <span class="op">+</span> <span class="fu">plot_layout</span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-patchwork3-1.png" alt="A composed figure show the histogram (top) and boxplots (bottom) on top or each other." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb79">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">count_gene_chromosome</span> <span class="op">+</span></span>
<span> <span class="op">(</span><span class="va">count_gene_chromosome</span> <span class="op">+</span> <span class="va">exp_boxplot_sex</span><span class="op">)</span> <span class="op">+</span></span>
<span> <span class="va">exp_boxplot_sex</span> <span class="op">+</span></span>
<span> <span class="fu">plot_layout</span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-patchwork4-1.png" alt="A figure composed of 4 plots with on top the histogram, in the middle the histogram and the boxpot, side-by-side, and at the bottom, the boxplot." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>The last plot can also be created using the <code>|</code> and
<code>/</code> composers:</p>
<div class="codewrapper sourceCode" id="cb80">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">count_gene_chromosome</span> <span class="op">/</span></span>
<span> <span class="op">(</span><span class="va">count_gene_chromosome</span> <span class="op">|</span> <span class="va">exp_boxplot_sex</span><span class="op">)</span> <span class="op">/</span></span>
<span> <span class="va">exp_boxplot_sex</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-patchwork5-1.png" alt="As above, a figure composed of 4 plots with on top the histogram, in the middle the histogram and the boxpot, side-by-side, and at the bottom, the boxplot." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Learn more about <code>patchwork</code> on its <a href="https://patchwork.data-imaginist.com/" class="external-link">webpage</a> or in this <a href="https://www.youtube.com/watch?v=0m4yywqNPVY" class="external-link">video</a>.</p>
<p>Another option is the <strong><code>gridExtra</code></strong> package
that allows to combine separate ggplots into a single figure using
<code>grid.arrange()</code>:</p>
<div class="codewrapper sourceCode" id="cb81">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"gridExtra"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb82">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="st">"gridExtra"</span><span class="op">)</span></span>
<span><span class="fu">grid.arrange</span><span class="op">(</span><span class="va">count_gene_chromosome</span>, <span class="va">exp_boxplot_sex</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/40-visualization-rendered-gridarrange-example-1.png" alt="A composed figure with the histogram (left) and boxplot (right) side-by-side." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>In addition to the <code>ncol</code> and <code>nrow</code> arguments,
used to make simple arrangements, there are tools for <a href="https://cran.r-project.org/web/packages/gridExtra/vignettes/arrangeGrob.html" class="external-link">constructing
more complex layouts</a>.</p>
</section><section><h2 class="section-heading" id="exporting-plots">Exporting plots<a class="anchor" aria-label="anchor" href="#exporting-plots"></a>
</h2>
<hr class="half-width">
<p>After creating your plot, you can save it to a file in your favorite
format. The Export tab in the <strong>Plot</strong> pane in RStudio will
save your plots at low resolution, which will not be accepted by many
journals and will not scale well for posters.</p>
<p>Instead, use the <code>ggsave()</code> function, which allows you
easily change the dimension and resolution of your plot by adjusting the
appropriate arguments (<code>width</code>, <code>height</code> and
<code>dpi</code>).</p>
<p>Make sure you have the <code>fig_output/</code> folder in your
working directory.</p>
<div class="codewrapper sourceCode" id="cb83">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_plot</span> <span class="op">&lt;-</span> <span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mean_exp_by_time_sex</span>,</span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">time</span>, y <span class="op">=</span> <span class="va">mean_exp</span>, color <span class="op">=</span> <span class="va">sex</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_line</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene</span>, scales <span class="op">=</span> <span class="st">"free_y"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Mean gene expression by duration of the infection"</span>,</span>
<span>         x <span class="op">=</span> <span class="st">"Duration of the infection (in days)"</span>,</span>
<span>         y <span class="op">=</span> <span class="st">"Mean gene expression"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">guides</span><span class="op">(</span>color<span class="op">=</span><span class="fu">guide_legend</span><span class="op">(</span>title<span class="op">=</span><span class="st">"Gender"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>colour <span class="op">=</span> <span class="st">"royalblue4"</span>, size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>,</span>
<span>        axis.text.y <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>colour <span class="op">=</span> <span class="st">"royalblue4"</span>, size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>,</span>
<span>        text <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">16</span><span class="op">)</span>,</span>
<span>        panel.grid <span class="op">=</span> <span class="fu">element_line</span><span class="op">(</span>colour<span class="op">=</span><span class="st">"lightsteelblue1"</span><span class="op">)</span>,</span>
<span>        legend.position <span class="op">=</span> <span class="st">"top"</span><span class="op">)</span></span>
<span><span class="fu">ggsave</span><span class="op">(</span><span class="st">"fig_output/mean_exp_by_time_sex.png"</span>, <span class="va">my_plot</span>, width <span class="op">=</span> <span class="fl">15</span>,</span>
<span>       height <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># This also works for grid.arrange() plots</span></span>
<span><span class="va">combo_plot</span> <span class="op">&lt;-</span> <span class="fu">grid.arrange</span><span class="op">(</span><span class="va">count_gene_chromosome</span>, <span class="va">exp_boxplot_sex</span>,</span>
<span>                           ncol <span class="op">=</span> <span class="fl">2</span>, widths <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">4</span>, <span class="fl">6</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">ggsave</span><span class="op">(</span><span class="st">"fig_output/combo_plot_chromosome_sex.png"</span>, <span class="va">combo_plot</span>,</span>
<span>       width <span class="op">=</span> <span class="fl">10</span>, dpi <span class="op">=</span> <span class="fl">300</span><span class="op">)</span></span></code></pre>
</div>
<p>Note: The parameters <code>width</code> and <code>height</code> also
determine the font size in the saved plot.</p>
</section><section><h2 class="section-heading" id="other-packages-for-visualisation">Other packages for visualisation<a class="anchor" aria-label="anchor" href="#other-packages-for-visualisation"></a>
</h2>
<hr class="half-width">
<p><code>ggplot2</code> is a very powerful package that fits very nicely
in our <em>tidy data</em> and <em>tidy tools</em> pipeline. There are
other visualization packages in R that shouldn’t be ignored.</p>
<div class="section level3">
<h3 id="base-graphics">Base graphics<a class="anchor" aria-label="anchor" href="#base-graphics"></a>
</h3>
<p>The default graphics system that comes with R, often called <em>base
R graphics</em> is simple and fast. It is based on the <em>painter’s or
canvas model</em>, where different output are directly overlaid on top
of each other (see figure <span class="citation">@ref</span>(fig:paintermodel)). This is a fundamental
difference with <code>ggplot2</code> (and with <code>lattice</code>,
described below), that returns dedicated objects, that are rendered on
screen or in a file, and that can even be updated.</p>
<div class="codewrapper sourceCode" id="cb84">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">par</span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, main <span class="op">=</span> <span class="st">"First layer, produced with plot(1:20)"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, main <span class="op">=</span> <span class="st">"A horizontal red line, added with abline(h = 10)"</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>h <span class="op">=</span> <span class="fl">10</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, main <span class="op">=</span> <span class="st">"A rectangle, added with rect(5, 5, 15, 15)"</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>h <span class="op">=</span> <span class="fl">10</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span>
<span><span class="fu">rect</span><span class="op">(</span><span class="fl">5</span>, <span class="fl">5</span>, <span class="fl">15</span>, <span class="fl">15</span>, lwd <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/40-visualization-rendered-paintermodel-1.png" alt="The 'base' figures, side-by-side, showing (from left to right) 20 empty dots along the diagonal, same with a vertical red line, and same as the second one with a rectangle overlaid in the middle of the plot." class="figure mx-auto d-block"><figcaption>
Successive layers added on top of each other.
</figcaption></figure><p>Another main difference is that base graphics’ plotting function try
to do <em>the right</em> thing based on their input type, i.e. they will
adapt their behaviour based on the class of their input. This is again
very different from what we have in <code>ggplot2</code>, that only
accepts dataframes as input, and that requires plots to be constructed
bit by bit.</p>
<div class="codewrapper sourceCode" id="cb85">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">par</span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">boxplot</span><span class="op">(</span><span class="fu">rnorm</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>,</span>
<span>        main <span class="op">=</span> <span class="st">"Boxplot of rnorm(100)"</span><span class="op">)</span></span>
<span><span class="fu">boxplot</span><span class="op">(</span><span class="fu">matrix</span><span class="op">(</span><span class="fu">rnorm</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>,</span>
<span>        main <span class="op">=</span> <span class="st">"Boxplot of matrix(rnorm(100), ncol = 10)"</span><span class="op">)</span></span>
<span><span class="fu">hist</span><span class="op">(</span><span class="fu">rnorm</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">hist</span><span class="op">(</span><span class="fu">matrix</span><span class="op">(</span><span class="fu">rnorm</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/40-visualization-rendered-plotmethod-1.png" alt="A two-by-two composition of 2 boxplots at the top, and 2 histograms at the bottom." class="figure mx-auto d-block"><figcaption>
Plotting boxplots (top) and histograms (bottom) vectors (left) or a
matrices (right).
</figcaption></figure><p>The out-of-the-box approach in base graphics can be very efficient
for simple, standard figures, that can be produced very quickly with a
single line of code and a single function such as <code>plot</code>, or
<code>hist</code>, or <code>boxplot</code>, … The defaults are however
not always the most appealing and tuning of figures, especially when
they become more complex (for example to produce facets), can become
lengthy and cumbersome.</p>
</div>
<div class="section level3">
<h3 id="the-lattice-package">The lattice package<a class="anchor" aria-label="anchor" href="#the-lattice-package"></a>
</h3>
<p>The <strong><code>lattice</code></strong> package is similar to
<code>ggplot2</code> in that is uses dataframes as input, returns
graphical objects and supports faceting. <code>lattice</code> however
isn’t based on the grammar of graphics and has a more convoluted
interface.</p>
<p>A good reference for the <code>lattice</code> package is <span class="citation">@latticebook</span>.</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>Visualization in R</li>
</ul>
</div>
</div>
</div>
</div>
</section><div class="footnotes footnotes-end-of-document">
<hr>
<ol>
<li id="fn1"><p>Source: <a href="https://raw.githubusercontent.com/rstudio/cheatsheets/main/data-visualization.pdf" class="external-link">Data
Visualization Cheat Sheet</a>.<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
</ol>
</div></section><section id="aio-60-next-steps"><p>Content from <a href="60-next-steps.html">Next steps</a></p>
<hr>
<p>Last updated on 2025-06-15 |

        <a href="https://github.com/carpentries-incubator/bioc-intro/edit/main/episodes/60-next-steps.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What is a <code>SummarizedExperiment</code>?</li>
<li>What is Bioconductor?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Introduce the Bioconductor project.</li>
<li>Introduce the notion of data containers.</li>
<li>Give an overview of the <code>SummarizedExperiment</code>,
extensively used in omics analyses.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="next-steps">Next steps<a class="anchor" aria-label="anchor" href="#next-steps"></a>
</h2>
<hr class="half-width">
<p>Data in bioinformatics is often complex. To deal with this,
developers define specialised data containers (termed classes) that
match the properties of the data they need to handle.</p>
<p>This aspect is central to the <strong>Bioconductor</strong><a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a> project
which uses the same <strong>core data infrastructure</strong> across
packages. This certainly contributed to Bioconductor’s success.
Bioconductor package developers are advised to make use of existing
infrastructure to provide coherence, interoperability, and stability to
the project as a whole.</p>
<p>To illustrate such an omics data container, we’ll present the
<code>SummarizedExperiment</code> class.</p>
</section><section><h2 class="section-heading" id="summarizedexperiment">SummarizedExperiment<a class="anchor" aria-label="anchor" href="#summarizedexperiment"></a>
</h2>
<hr class="half-width">
<p>The figure below represents the anatomy of the SummarizedExperiment
class.</p>
<figure><img src="https://uclouvain-cbio.github.io/WSBIM1322/figs/SE.svg" alt="Schematic representation of the SummarizedExperiment class illustrating the following slots: fowData and rowRanges, one assay or multiple assays, colData and metadata." width="80%" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Objects of the class SummarizedExperiment contain :</p>
<ul>
<li><p><strong>One (or more) assay(s)</strong> containing the
quantitative omics data (expression data), stored as a matrix-like
object. Features (genes, transcripts, proteins, …) are defined along the
rows, and samples along the columns.</p></li>
<li><p>A <strong>sample metadata</strong> slot containing sample
co-variates, stored as a data frame. Rows from this table represent
samples (rows match exactly the columns of the expression
data).</p></li>
<li><p>A <strong>feature metadata</strong> slot containing feature
co-variates, stored as a data frame. The rows of this data frame match
exactly the rows of the expression data.</p></li>
</ul>
<p>The coordinated nature of the <code>SummarizedExperiment</code>
guarantees that during data manipulation, the dimensions of the
different slots will always match (i.e the columns in the expression
data and then rows in the sample metadata, as well as the rows in the
expression data and feature metadata) during data manipulation. For
example, if we had to exclude one sample from the assay, it would be
automatically removed from the sample metadata in the same
operation.</p>
<p>The metadata slots can grow additional co-variates (columns) without
affecting the other structures.</p>
<div class="section level3">
<h3 id="creating-a-summarizedexperiment">Creating a SummarizedExperiment<a class="anchor" aria-label="anchor" href="#creating-a-summarizedexperiment"></a>
</h3>
<p>In order to create a <code>SummarizedExperiment</code>, we will
create the individual components, i.e the count matrix, the sample and
gene metadata from csv files. These are typically how RNA-Seq data are
provided (after raw data have been processed).</p>
<ul>
<li>
<strong>An expression matrix</strong>: we load the count matrix,
specifying that the first columns contains row/gene names, and convert
the <code>data.frame</code> to a <code>matrix</code>. You can download
it <a href="https://carpentries-incubator.github.io/bioc-intro/data/count_matrix.csv" class="external-link">by
clicking this link</a>.</li>
</ul>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">count_matrix</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span><span class="st">"data/count_matrix.csv"</span>,</span>
<span>                         row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">as.matrix</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">count_matrix</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>        GSM2545336 GSM2545337 GSM2545338 GSM2545339 GSM2545340 GSM2545341
Asl           1170        361        400        586        626        988
Apod         36194      10347       9173      10620      13021      29594
Cyp2d22       4060       1616       1603       1901       2171       3349
Klk6           287        629        641        578        448        195
Fcrls           85        233        244        237        180         38
        GSM2545342 GSM2545343 GSM2545344 GSM2545345 GSM2545346 GSM2545347
Asl            836        535        586        597        938       1035
Apod         24959      13668      13230      15868      27769      34301
Cyp2d22       3122       2008       2254       2277       2985       3452
Klk6           186       1101        537        567        327        233
Fcrls           68        375        199        177         89         67
        GSM2545348 GSM2545349 GSM2545350 GSM2545351 GSM2545352 GSM2545353
Asl            494        481        666        937        803        541
Apod         11258      11812      15816      29242      20415      13682
Cyp2d22       1883       2014       2417       3678       2920       2216
Klk6           742        881        828        250        798        710
Fcrls          300        233        231         81        303        285
        GSM2545354 GSM2545362 GSM2545363 GSM2545380
Asl            473        748        576       1192
Apod         11088      15916      11166      38148
Cyp2d22       1821       2842       2011       4019
Klk6           894        501        598        259
Fcrls          248        179        184         68</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dim</span><span class="op">(</span><span class="va">count_matrix</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1474   22</code></pre>
</div>
<ul>
<li>
<strong>A table describing the samples</strong>, available <a href="https://carpentries-incubator.github.io/bioc-intro/data/sample_metadata.csv" class="external-link">at
this link</a>.</li>
</ul>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sample_metadata</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span><span class="st">"data/sample_metadata.csv"</span><span class="op">)</span></span>
<span><span class="va">sample_metadata</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>       sample     organism age    sex   infection  strain time     tissue mouse
1  GSM2545336 Mus musculus   8 Female  InfluenzaA C57BL/6    8 Cerebellum    14
2  GSM2545337 Mus musculus   8 Female NonInfected C57BL/6    0 Cerebellum     9
3  GSM2545338 Mus musculus   8 Female NonInfected C57BL/6    0 Cerebellum    10
4  GSM2545339 Mus musculus   8 Female  InfluenzaA C57BL/6    4 Cerebellum    15
5  GSM2545340 Mus musculus   8   Male  InfluenzaA C57BL/6    4 Cerebellum    18
6  GSM2545341 Mus musculus   8   Male  InfluenzaA C57BL/6    8 Cerebellum     6
7  GSM2545342 Mus musculus   8 Female  InfluenzaA C57BL/6    8 Cerebellum     5
8  GSM2545343 Mus musculus   8   Male NonInfected C57BL/6    0 Cerebellum    11
9  GSM2545344 Mus musculus   8 Female  InfluenzaA C57BL/6    4 Cerebellum    22
10 GSM2545345 Mus musculus   8   Male  InfluenzaA C57BL/6    4 Cerebellum    13
11 GSM2545346 Mus musculus   8   Male  InfluenzaA C57BL/6    8 Cerebellum    23
12 GSM2545347 Mus musculus   8   Male  InfluenzaA C57BL/6    8 Cerebellum    24
13 GSM2545348 Mus musculus   8 Female NonInfected C57BL/6    0 Cerebellum     8
14 GSM2545349 Mus musculus   8   Male NonInfected C57BL/6    0 Cerebellum     7
15 GSM2545350 Mus musculus   8   Male  InfluenzaA C57BL/6    4 Cerebellum     1
16 GSM2545351 Mus musculus   8 Female  InfluenzaA C57BL/6    8 Cerebellum    16
17 GSM2545352 Mus musculus   8 Female  InfluenzaA C57BL/6    4 Cerebellum    21
18 GSM2545353 Mus musculus   8 Female NonInfected C57BL/6    0 Cerebellum     4
19 GSM2545354 Mus musculus   8   Male NonInfected C57BL/6    0 Cerebellum     2
20 GSM2545362 Mus musculus   8 Female  InfluenzaA C57BL/6    4 Cerebellum    20
21 GSM2545363 Mus musculus   8   Male  InfluenzaA C57BL/6    4 Cerebellum    12
22 GSM2545380 Mus musculus   8 Female  InfluenzaA C57BL/6    8 Cerebellum    19</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dim</span><span class="op">(</span><span class="va">sample_metadata</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 22  9</code></pre>
</div>
<ul>
<li>
<strong>A table describing the genes</strong>, available <a href="https://carpentries-incubator.github.io/bioc-intro/data/gene_metadata.csv" class="external-link">at
this link</a>.</li>
</ul>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_metadata</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span><span class="st">"data/gene_metadata.csv"</span><span class="op">)</span></span>
<span><span class="va">gene_metadata</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>      gene ENTREZID
1      Asl   109900
2     Apod    11815
3  Cyp2d22    56448
4     Klk6    19144
5    Fcrls    80891
6   Slc2a4    20528
7     Exd2    97827
8     Gjc2   118454
9     Plp1    18823
10    Gnb4    14696
                                                                         product
1                                 argininosuccinate lyase, transcript variant X1
2                                         apolipoprotein D, transcript variant 3
3   cytochrome P450, family 2, subfamily d, polypeptide 22, transcript variant 2
4                           kallikrein related-peptidase 6, transcript variant 2
5                  Fc receptor-like S, scavenger receptor, transcript variant X1
6            solute carrier family 2 (facilitated glucose transporter), member 4
7                                          exonuclease 3'-5' domain containing 2
8                            gap junction protein, gamma 2, transcript variant 1
9                           proteolipid protein (myelin) 1, transcript variant 1
10 guanine nucleotide binding protein (G protein), beta 4, transcript variant X2
      ensembl_gene_id
1  ENSMUSG00000025533
2  ENSMUSG00000022548
3  ENSMUSG00000061740
4  ENSMUSG00000050063
5  ENSMUSG00000015852
6  ENSMUSG00000018566
7  ENSMUSG00000032705
8  ENSMUSG00000043448
9  ENSMUSG00000031425
10 ENSMUSG00000027669</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dim</span><span class="op">(</span><span class="va">gene_metadata</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1474    9</code></pre>
</div>
<p>We will create a <code>SummarizedExperiment</code> from these
tables:</p>
<ul>
<li><p>The count matrix that will be used as the
<strong><code>assay</code></strong></p></li>
<li><p>The table describing the samples will be used as the
<strong>sample metadata</strong> slot</p></li>
<li><p>The table describing the genes will be used as the
<strong>features metadata</strong> slot</p></li>
</ul>
<p>To do this we can put the different parts together using the
<code>SummarizedExperiment</code> constructor:</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## BiocManager::install("SummarizedExperiment")</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="st">"SummarizedExperiment"</span><span class="op">)</span></span></code></pre>
</div>
<p>First, we make sure that the samples are in the same order in the
count matrix and the sample annotation, and the same for the genes in
the count matrix and the gene annotation.</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">stopifnot</span><span class="op">(</span><span class="fu">rownames</span><span class="op">(</span><span class="va">count_matrix</span><span class="op">)</span> <span class="op">==</span> <span class="va">gene_metadata</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span></span>
<span><span class="fu">stopifnot</span><span class="op">(</span><span class="fu">colnames</span><span class="op">(</span><span class="va">count_matrix</span><span class="op">)</span> <span class="op">==</span> <span class="va">sample_metadata</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="op">(</span>assays <span class="op">=</span> <span class="fu">list</span><span class="op">(</span>counts <span class="op">=</span> <span class="va">count_matrix</span><span class="op">)</span>,</span>
<span>                           colData <span class="op">=</span> <span class="va">sample_metadata</span>,</span>
<span>                           rowData <span class="op">=</span> <span class="va">gene_metadata</span><span class="op">)</span></span>
<span><span class="va">se</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>class: SummarizedExperiment
dim: 1474 22
metadata(0):
assays(1): counts
rownames(1474): Asl Apod ... Lmx1a Pbx1
rowData names(9): gene ENTREZID ... phenotype_description
  hsapiens_homolog_associated_gene_name
colnames(22): GSM2545336 GSM2545337 ... GSM2545363 GSM2545380
colData names(9): sample organism ... tissue mouse</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="saving-data">Saving data<a class="anchor" aria-label="anchor" href="#saving-data"></a>
</h3>
<p>Exporting data to a spreadsheet, as we did in a previous episode, has
several limitations, such as those described in the first chapter
(possible inconsistencies with <code>,</code> and <code>.</code> for
decimal separators and lack of variable type definitions). Furthermore,
exporting data to a spreadsheet is only relevant for rectangular data
such as dataframes and matrices.</p>
<p>A more general way to save data, that is specific to R and is
guaranteed to work on any operating system, is to use the
<code>saveRDS</code> function. Saving objects like this will generate a
binary representation on disk (using the <code>rds</code> file extension
here), which can be loaded back into R using the <code>readRDS</code>
function.</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">saveRDS</span><span class="op">(</span><span class="va">se</span>, file <span class="op">=</span> <span class="st">"data_output/se.rds"</span><span class="op">)</span></span>
<span><span class="fu">rm</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">readRDS</span><span class="op">(</span><span class="st">"data_output/se.rds"</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></code></pre>
</div>
<p>To conclude, when it comes to saving data from R that will be loaded
again in R, saving and loading with <code>saveRDS</code> and
<code>readRDS</code> is the preferred approach. If tabular data need to
be shared with somebody that is not using R, then exporting to a
text-based spreadsheet is a good alternative.</p>
<p>Using this data structure, we can access the expression matrix with
the <code>assay</code> function:</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>        GSM2545336 GSM2545337 GSM2545338 GSM2545339 GSM2545340 GSM2545341
Asl           1170        361        400        586        626        988
Apod         36194      10347       9173      10620      13021      29594
Cyp2d22       4060       1616       1603       1901       2171       3349
Klk6           287        629        641        578        448        195
Fcrls           85        233        244        237        180         38
Slc2a4         782        231        248        265        313        786
        GSM2545342 GSM2545343 GSM2545344 GSM2545345 GSM2545346 GSM2545347
Asl            836        535        586        597        938       1035
Apod         24959      13668      13230      15868      27769      34301
Cyp2d22       3122       2008       2254       2277       2985       3452
Klk6           186       1101        537        567        327        233
Fcrls           68        375        199        177         89         67
Slc2a4         528        249        266        357        654        693
        GSM2545348 GSM2545349 GSM2545350 GSM2545351 GSM2545352 GSM2545353
Asl            494        481        666        937        803        541
Apod         11258      11812      15816      29242      20415      13682
Cyp2d22       1883       2014       2417       3678       2920       2216
Klk6           742        881        828        250        798        710
Fcrls          300        233        231         81        303        285
Slc2a4         271        304        349        715        513        320
        GSM2545354 GSM2545362 GSM2545363 GSM2545380
Asl            473        748        576       1192
Apod         11088      15916      11166      38148
Cyp2d22       1821       2842       2011       4019
Klk6           894        501        598        259
Fcrls          248        179        184         68
Slc2a4         248        350        317        796</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dim</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1474   22</code></pre>
</div>
<p>We can access the sample metadata using the <code>colData</code>
function:</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 22 rows and 9 columns
                sample     organism       age         sex   infection
           &lt;character&gt;  &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt;
GSM2545336  GSM2545336 Mus musculus         8      Female  InfluenzaA
GSM2545337  GSM2545337 Mus musculus         8      Female NonInfected
GSM2545338  GSM2545338 Mus musculus         8      Female NonInfected
GSM2545339  GSM2545339 Mus musculus         8      Female  InfluenzaA
GSM2545340  GSM2545340 Mus musculus         8        Male  InfluenzaA
...                ...          ...       ...         ...         ...
GSM2545353  GSM2545353 Mus musculus         8      Female NonInfected
GSM2545354  GSM2545354 Mus musculus         8        Male NonInfected
GSM2545362  GSM2545362 Mus musculus         8      Female  InfluenzaA
GSM2545363  GSM2545363 Mus musculus         8        Male  InfluenzaA
GSM2545380  GSM2545380 Mus musculus         8      Female  InfluenzaA
                strain      time      tissue     mouse
           &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;integer&gt;
GSM2545336     C57BL/6         8  Cerebellum        14
GSM2545337     C57BL/6         0  Cerebellum         9
GSM2545338     C57BL/6         0  Cerebellum        10
GSM2545339     C57BL/6         4  Cerebellum        15
GSM2545340     C57BL/6         4  Cerebellum        18
...                ...       ...         ...       ...
GSM2545353     C57BL/6         0  Cerebellum         4
GSM2545354     C57BL/6         0  Cerebellum         2
GSM2545362     C57BL/6         4  Cerebellum        20
GSM2545363     C57BL/6         4  Cerebellum        12
GSM2545380     C57BL/6         8  Cerebellum        19</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dim</span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 22  9</code></pre>
</div>
<p>We can also access the feature metadata using the
<code>rowData</code> function:</p>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 6 rows and 9 columns
               gene  ENTREZID                product    ensembl_gene_id
        &lt;character&gt; &lt;integer&gt;            &lt;character&gt;        &lt;character&gt;
Asl             Asl    109900 argininosuccinate ly.. ENSMUSG00000025533
Apod           Apod     11815 apolipoprotein D, tr.. ENSMUSG00000022548
Cyp2d22     Cyp2d22     56448 cytochrome P450, fam.. ENSMUSG00000061740
Klk6           Klk6     19144 kallikrein related-p.. ENSMUSG00000050063
Fcrls         Fcrls     80891 Fc receptor-like S, .. ENSMUSG00000015852
Slc2a4       Slc2a4     20528 solute carrier famil.. ENSMUSG00000018566
        external_synonym chromosome_name   gene_biotype  phenotype_description
             &lt;character&gt;     &lt;character&gt;    &lt;character&gt;            &lt;character&gt;
Asl        2510006M18Rik               5 protein_coding abnormal circulating..
Apod                  NA              16 protein_coding abnormal lipid homeo..
Cyp2d22             2D22              15 protein_coding abnormal skin morpho..
Klk6                Bssp               7 protein_coding abnormal cytokine le..
Fcrls      2810439C17Rik               3 protein_coding decreased CD8-positi..
Slc2a4            Glut-4              11 protein_coding abnormal circulating..
        hsapiens_homolog_associated_gene_name
                                  &lt;character&gt;
Asl                                       ASL
Apod                                     APOD
Cyp2d22                                CYP2D6
Klk6                                     KLK6
Fcrls                                   FCRL2
Slc2a4                                 SLC2A4</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dim</span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1474    9</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="subsetting-a-summarizedexperiment">Subsetting a SummarizedExperiment<a class="anchor" aria-label="anchor" href="#subsetting-a-summarizedexperiment"></a>
</h3>
<p>SummarizedExperiment can be subset just like with data frames, with
numerics or with characters of logicals.</p>
<p>Below, we create a new instance of class SummarizedExperiment that
contains only the 5 first features for the 3 first samples.</p>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se1</span> <span class="op">&lt;-</span> <span class="va">se</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span>
<span><span class="va">se1</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>class: SummarizedExperiment
dim: 5 3
metadata(0):
assays(1): counts
rownames(5): Asl Apod Cyp2d22 Klk6 Fcrls
rowData names(9): gene ENTREZID ... phenotype_description
  hsapiens_homolog_associated_gene_name
colnames(3): GSM2545336 GSM2545337 GSM2545338
colData names(9): sample organism ... tissue mouse</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colData</span><span class="op">(</span><span class="va">se1</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 3 rows and 9 columns
                sample     organism       age         sex   infection
           &lt;character&gt;  &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt;
GSM2545336  GSM2545336 Mus musculus         8      Female  InfluenzaA
GSM2545337  GSM2545337 Mus musculus         8      Female NonInfected
GSM2545338  GSM2545338 Mus musculus         8      Female NonInfected
                strain      time      tissue     mouse
           &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;integer&gt;
GSM2545336     C57BL/6         8  Cerebellum        14
GSM2545337     C57BL/6         0  Cerebellum         9
GSM2545338     C57BL/6         0  Cerebellum        10</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rowData</span><span class="op">(</span><span class="va">se1</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 5 rows and 9 columns
               gene  ENTREZID                product    ensembl_gene_id
        &lt;character&gt; &lt;integer&gt;            &lt;character&gt;        &lt;character&gt;
Asl             Asl    109900 argininosuccinate ly.. ENSMUSG00000025533
Apod           Apod     11815 apolipoprotein D, tr.. ENSMUSG00000022548
Cyp2d22     Cyp2d22     56448 cytochrome P450, fam.. ENSMUSG00000061740
Klk6           Klk6     19144 kallikrein related-p.. ENSMUSG00000050063
Fcrls         Fcrls     80891 Fc receptor-like S, .. ENSMUSG00000015852
        external_synonym chromosome_name   gene_biotype  phenotype_description
             &lt;character&gt;     &lt;character&gt;    &lt;character&gt;            &lt;character&gt;
Asl        2510006M18Rik               5 protein_coding abnormal circulating..
Apod                  NA              16 protein_coding abnormal lipid homeo..
Cyp2d22             2D22              15 protein_coding abnormal skin morpho..
Klk6                Bssp               7 protein_coding abnormal cytokine le..
Fcrls      2810439C17Rik               3 protein_coding decreased CD8-positi..
        hsapiens_homolog_associated_gene_name
                                  &lt;character&gt;
Asl                                       ASL
Apod                                     APOD
Cyp2d22                                CYP2D6
Klk6                                     KLK6
Fcrls                                   FCRL2</code></pre>
</div>
<p>We can also use the <code>colData()</code> function to subset on
something from the sample metadata or the <code>rowData()</code> to
subset on something from the feature metadata. For example, here we keep
only miRNAs and the non infected samples:</p>
<div class="codewrapper sourceCode" id="cb36">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se1</span> <span class="op">&lt;-</span> <span class="va">se</span><span class="op">[</span><span class="fu">rowData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">$</span><span class="va">gene_biotype</span> <span class="op">==</span> <span class="st">"miRNA"</span>,</span>
<span>          <span class="fu">colData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">$</span><span class="va">infection</span> <span class="op">==</span> <span class="st">"NonInfected"</span><span class="op">]</span></span>
<span><span class="va">se1</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>class: SummarizedExperiment
dim: 7 7
metadata(0):
assays(1): counts
rownames(7): Mir1901 Mir378a ... Mir128-1 Mir7682
rowData names(9): gene ENTREZID ... phenotype_description
  hsapiens_homolog_associated_gene_name
colnames(7): GSM2545337 GSM2545338 ... GSM2545353 GSM2545354
colData names(9): sample organism ... tissue mouse</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb38">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">assay</span><span class="op">(</span><span class="va">se1</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>         GSM2545337 GSM2545338 GSM2545343 GSM2545348 GSM2545349 GSM2545353
Mir1901          45         44         74         55         68         33
Mir378a          11          7          9          4         12          4
Mir133b           4          6          5          4          6          7
Mir30c-2         10          6         16         12          8         17
Mir149            1          2          0          0          0          0
Mir128-1          4          1          2          2          1          2
Mir7682           2          0          4          1          3          5
         GSM2545354
Mir1901          60
Mir378a           8
Mir133b           3
Mir30c-2         15
Mir149            2
Mir128-1          1
Mir7682           5</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb40">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colData</span><span class="op">(</span><span class="va">se1</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 7 rows and 9 columns
                sample     organism       age         sex   infection
           &lt;character&gt;  &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt;
GSM2545337  GSM2545337 Mus musculus         8      Female NonInfected
GSM2545338  GSM2545338 Mus musculus         8      Female NonInfected
GSM2545343  GSM2545343 Mus musculus         8        Male NonInfected
GSM2545348  GSM2545348 Mus musculus         8      Female NonInfected
GSM2545349  GSM2545349 Mus musculus         8        Male NonInfected
GSM2545353  GSM2545353 Mus musculus         8      Female NonInfected
GSM2545354  GSM2545354 Mus musculus         8        Male NonInfected
                strain      time      tissue     mouse
           &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;integer&gt;
GSM2545337     C57BL/6         0  Cerebellum         9
GSM2545338     C57BL/6         0  Cerebellum        10
GSM2545343     C57BL/6         0  Cerebellum        11
GSM2545348     C57BL/6         0  Cerebellum         8
GSM2545349     C57BL/6         0  Cerebellum         7
GSM2545353     C57BL/6         0  Cerebellum         4
GSM2545354     C57BL/6         0  Cerebellum         2</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb42">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rowData</span><span class="op">(</span><span class="va">se1</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 7 rows and 9 columns
                gene  ENTREZID        product    ensembl_gene_id
         &lt;character&gt; &lt;integer&gt;    &lt;character&gt;        &lt;character&gt;
Mir1901      Mir1901 100316686  microRNA 1901 ENSMUSG00000084565
Mir378a      Mir378a    723889  microRNA 378a ENSMUSG00000105200
Mir133b      Mir133b    723817  microRNA 133b ENSMUSG00000065480
Mir30c-2    Mir30c-2    723964 microRNA 30c-2 ENSMUSG00000065567
Mir149        Mir149    387167   microRNA 149 ENSMUSG00000065470
Mir128-1    Mir128-1    387147 microRNA 128-1 ENSMUSG00000065520
Mir7682      Mir7682 102466847  microRNA 7682 ENSMUSG00000106406
         external_synonym chromosome_name gene_biotype  phenotype_description
              &lt;character&gt;     &lt;character&gt;  &lt;character&gt;            &lt;character&gt;
Mir1901          Mirn1901              18        miRNA                     NA
Mir378a           Mirn378              18        miRNA abnormal mitochondri..
Mir133b          mir 133b               1        miRNA no abnormal phenotyp..
Mir30c-2        mir 30c-2               1        miRNA                     NA
Mir149            Mirn149               1        miRNA increased circulatin..
Mir128-1          Mirn128               1        miRNA no abnormal phenotyp..
Mir7682      mmu-mir-7682               1        miRNA                     NA
         hsapiens_homolog_associated_gene_name
                                   &lt;character&gt;
Mir1901                                     NA
Mir378a                                MIR378A
Mir133b                                MIR133B
Mir30c-2                               MIR30C2
Mir149                                      NA
Mir128-1                              MIR128-1
Mir7682                                     NA</code></pre>
</div>
<!--For the following exercise, you should download the SE.rda object
(that contains the `se` object), and open the file using the 'load()'
function.-->
<!-- ```{r, eval = FALSE, echo = FALSE} -->
<!-- download.file(url = "https://raw.githubusercontent.com/UCLouvain-CBIO/bioinfo-training-01-intro-r/master/data/SE.rda", -->
<!--               destfile = "data/SE.rda") -->
<!-- load("data/SE.rda") -->
<!-- ``` -->
<div id="challenge" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Extract the gene expression levels of the 3 first genes in samples at
time 0 and at time 8.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb44">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">assay</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fu">colData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">$</span><span class="va">time</span> <span class="op">!=</span> <span class="fl">4</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>        GSM2545336 GSM2545337 GSM2545338 GSM2545341 GSM2545342 GSM2545343
Asl           1170        361        400        988        836        535
Apod         36194      10347       9173      29594      24959      13668
Cyp2d22       4060       1616       1603       3349       3122       2008
        GSM2545346 GSM2545347 GSM2545348 GSM2545349 GSM2545351 GSM2545353
Asl            938       1035        494        481        937        541
Apod         27769      34301      11258      11812      29242      13682
Cyp2d22       2985       3452       1883       2014       3678       2216
        GSM2545354 GSM2545380
Asl            473       1192
Apod         11088      38148
Cyp2d22       1821       4019</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb46">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Equivalent to</span></span>
<span><span class="fu">assay</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fu">colData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">$</span><span class="va">time</span> <span class="op">==</span> <span class="fl">0</span> <span class="op">|</span> <span class="fu">colData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">$</span><span class="va">time</span> <span class="op">==</span> <span class="fl">8</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>        GSM2545336 GSM2545337 GSM2545338 GSM2545341 GSM2545342 GSM2545343
Asl           1170        361        400        988        836        535
Apod         36194      10347       9173      29594      24959      13668
Cyp2d22       4060       1616       1603       3349       3122       2008
        GSM2545346 GSM2545347 GSM2545348 GSM2545349 GSM2545351 GSM2545353
Asl            938       1035        494        481        937        541
Apod         27769      34301      11258      11812      29242      13682
Cyp2d22       2985       3452       1883       2014       3678       2216
        GSM2545354 GSM2545380
Asl            473       1192
Apod         11088      38148
Cyp2d22       1821       4019</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="challenge-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Verify that you get the same values using the long <code>rna</code>
table.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb48">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">filter</span><span class="op">(</span><span class="va">gene</span> <span class="op">%in%</span> <span class="fu">c</span><span class="op">(</span><span class="st">"Asl"</span>, <span class="st">"Apod"</span>, <span class="st">"Cyd2d22"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">filter</span><span class="op">(</span><span class="va">time</span> <span class="op">!=</span> <span class="fl">4</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">select</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 28 × 1
   expression
        &lt;dbl&gt;
 1       1170
 2      36194
 3        361
 4      10347
 5        400
 6       9173
 7        988
 8      29594
 9        836
10      24959
# ℹ 18 more rows</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>The long table and the <code>SummarizedExperiment</code> contain the
same information, but are simply structured differently. Each approach
has its own advantages: the former is a good fit for the
<code>tidyverse</code> packages, while the latter is the preferred
structure for many bioinformatics and statistical processing steps. For
example, a typical RNA-Seq analyses using the <code>DESeq2</code>
package.</p>
<div class="section level4">
<h4 id="adding-variables-to-metadata">Adding variables to metadata<a class="anchor" aria-label="anchor" href="#adding-variables-to-metadata"></a>
</h4>
<p>We can also add information to the metadata. Suppose that you want to
add the center where the samples were collected…</p>
<div class="codewrapper sourceCode" id="cb50">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">$</span><span class="va">center</span> <span class="op">&lt;-</span> <span class="fu">rep</span><span class="op">(</span><span class="st">"University of Illinois"</span>, <span class="fu">nrow</span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 22 rows and 10 columns
                sample     organism       age         sex   infection
           &lt;character&gt;  &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt;
GSM2545336  GSM2545336 Mus musculus         8      Female  InfluenzaA
GSM2545337  GSM2545337 Mus musculus         8      Female NonInfected
GSM2545338  GSM2545338 Mus musculus         8      Female NonInfected
GSM2545339  GSM2545339 Mus musculus         8      Female  InfluenzaA
GSM2545340  GSM2545340 Mus musculus         8        Male  InfluenzaA
...                ...          ...       ...         ...         ...
GSM2545353  GSM2545353 Mus musculus         8      Female NonInfected
GSM2545354  GSM2545354 Mus musculus         8        Male NonInfected
GSM2545362  GSM2545362 Mus musculus         8      Female  InfluenzaA
GSM2545363  GSM2545363 Mus musculus         8        Male  InfluenzaA
GSM2545380  GSM2545380 Mus musculus         8      Female  InfluenzaA
                strain      time      tissue     mouse                 center
           &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;integer&gt;            &lt;character&gt;
GSM2545336     C57BL/6         8  Cerebellum        14 University of Illinois
GSM2545337     C57BL/6         0  Cerebellum         9 University of Illinois
GSM2545338     C57BL/6         0  Cerebellum        10 University of Illinois
GSM2545339     C57BL/6         4  Cerebellum        15 University of Illinois
GSM2545340     C57BL/6         4  Cerebellum        18 University of Illinois
...                ...       ...         ...       ...                    ...
GSM2545353     C57BL/6         0  Cerebellum         4 University of Illinois
GSM2545354     C57BL/6         0  Cerebellum         2 University of Illinois
GSM2545362     C57BL/6         4  Cerebellum        20 University of Illinois
GSM2545363     C57BL/6         4  Cerebellum        12 University of Illinois
GSM2545380     C57BL/6         8  Cerebellum        19 University of Illinois</code></pre>
</div>
<p>This illustrates that the metadata slots can grow indefinitely
without affecting the other structures!</p>
</div>
</div>
<div class="section level3">
<h3 id="tidysummarizedexperiment">tidySummarizedExperiment<a class="anchor" aria-label="anchor" href="#tidysummarizedexperiment"></a>
</h3>
<p>You may be wondering, can we use tidyverse commands to interact with
<code>SummarizedExperiment</code> objects? The answer is yes, we can
with the <code>tidySummarizedExperiment</code> package.</p>
<p>Remember what our SummarizedExperiment object looks like:</p>
<div class="codewrapper sourceCode" id="cb52">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>class: SummarizedExperiment
dim: 1474 22
metadata(0):
assays(1): counts
rownames(1474): Asl Apod ... Lmx1a Pbx1
rowData names(9): gene ENTREZID ... phenotype_description
  hsapiens_homolog_associated_gene_name
colnames(22): GSM2545336 GSM2545337 ... GSM2545363 GSM2545380
colData names(10): sample organism ... mouse center</code></pre>
</div>
<p>Load <code>tidySummarizedExperiment</code> and then take a look at
the se object again.</p>
<div class="codewrapper sourceCode" id="cb54">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#BiocManager::install("tidySummarizedExperiment")</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="st">"tidySummarizedExperiment"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">se</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A SummarizedExperiment-tibble abstraction: 32,428 × 22
# Features=1474 | Samples=22 | Assays=counts
   .feature .sample    counts sample organism   age sex   infection strain  time
   &lt;chr&gt;    &lt;chr&gt;       &lt;int&gt; &lt;chr&gt;  &lt;chr&gt;    &lt;int&gt; &lt;chr&gt; &lt;chr&gt;     &lt;chr&gt;  &lt;int&gt;
 1 Asl      GSM2545336   1170 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
 2 Apod     GSM2545336  36194 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
 3 Cyp2d22  GSM2545336   4060 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
 4 Klk6     GSM2545336    287 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
 5 Fcrls    GSM2545336     85 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
 6 Slc2a4   GSM2545336    782 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
 7 Exd2     GSM2545336   1619 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
 8 Gjc2     GSM2545336    288 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
 9 Plp1     GSM2545336  43217 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
10 Gnb4     GSM2545336   1071 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
# ℹ 40 more rows
# ℹ 12 more variables: tissue &lt;chr&gt;, mouse &lt;int&gt;, center &lt;chr&gt;, gene &lt;chr&gt;,
#   ENTREZID &lt;int&gt;, product &lt;chr&gt;, ensembl_gene_id &lt;chr&gt;,
#   external_synonym &lt;chr&gt;, chromosome_name &lt;chr&gt;, gene_biotype &lt;chr&gt;,
#   phenotype_description &lt;chr&gt;, hsapiens_homolog_associated_gene_name &lt;chr&gt;</code></pre>
</div>
<p>It’s still a <code>SummarizedExperiment</code> object, so maintains
the efficient structure, but now we can view it as a tibble. Note the
first line of the output says this, it’s a
<code>SummarizedExperiment</code>-<code>tibble</code> abstraction. We
can also see in the second line of the output the number of transcripts
and samples.</p>
<p>If we want to revert to the standard
<code>SummarizedExperiment</code> view, we can do that.</p>
<div class="codewrapper sourceCode" id="cb56">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">options</span><span class="op">(</span><span class="st">"restore_SummarizedExperiment_show"</span> <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">se</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>class: SummarizedExperiment
dim: 1474 22
metadata(0):
assays(1): counts
rownames(1474): Asl Apod ... Lmx1a Pbx1
rowData names(9): gene ENTREZID ... phenotype_description
  hsapiens_homolog_associated_gene_name
colnames(22): GSM2545336 GSM2545337 ... GSM2545363 GSM2545380
colData names(10): sample organism ... mouse center</code></pre>
</div>
<p>But here we use the tibble view.</p>
<div class="codewrapper sourceCode" id="cb58">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">options</span><span class="op">(</span><span class="st">"restore_SummarizedExperiment_show"</span> <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">se</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A SummarizedExperiment-tibble abstraction: 32,428 × 22
# Features=1474 | Samples=22 | Assays=counts
   .feature .sample    counts sample organism   age sex   infection strain  time
   &lt;chr&gt;    &lt;chr&gt;       &lt;int&gt; &lt;chr&gt;  &lt;chr&gt;    &lt;int&gt; &lt;chr&gt; &lt;chr&gt;     &lt;chr&gt;  &lt;int&gt;
 1 Asl      GSM2545336   1170 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
 2 Apod     GSM2545336  36194 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
 3 Cyp2d22  GSM2545336   4060 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
 4 Klk6     GSM2545336    287 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
 5 Fcrls    GSM2545336     85 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
 6 Slc2a4   GSM2545336    782 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
 7 Exd2     GSM2545336   1619 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
 8 Gjc2     GSM2545336    288 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
 9 Plp1     GSM2545336  43217 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
10 Gnb4     GSM2545336   1071 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
# ℹ 40 more rows
# ℹ 12 more variables: tissue &lt;chr&gt;, mouse &lt;int&gt;, center &lt;chr&gt;, gene &lt;chr&gt;,
#   ENTREZID &lt;int&gt;, product &lt;chr&gt;, ensembl_gene_id &lt;chr&gt;,
#   external_synonym &lt;chr&gt;, chromosome_name &lt;chr&gt;, gene_biotype &lt;chr&gt;,
#   phenotype_description &lt;chr&gt;, hsapiens_homolog_associated_gene_name &lt;chr&gt;</code></pre>
</div>
<p>We can now use tidyverse commands to interact with the
<code>SummarizedExperiment</code> object.</p>
<p>We can use <code>filter</code> to filter for rows using a condition
e.g. to view all rows for one sample.</p>
<div class="codewrapper sourceCode" id="cb60">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">|&gt;</span> <span class="fu">filter</span><span class="op">(</span><span class="va">.sample</span> <span class="op">==</span> <span class="st">"GSM2545336"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A SummarizedExperiment-tibble abstraction: 1,474 × 22
# Features=1474 | Samples=1 | Assays=counts
   .feature .sample    counts sample organism   age sex   infection strain  time
   &lt;chr&gt;    &lt;chr&gt;       &lt;int&gt; &lt;chr&gt;  &lt;chr&gt;    &lt;int&gt; &lt;chr&gt; &lt;chr&gt;     &lt;chr&gt;  &lt;int&gt;
 1 Asl      GSM2545336   1170 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
 2 Apod     GSM2545336  36194 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
 3 Cyp2d22  GSM2545336   4060 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
 4 Klk6     GSM2545336    287 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
 5 Fcrls    GSM2545336     85 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
 6 Slc2a4   GSM2545336    782 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
 7 Exd2     GSM2545336   1619 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
 8 Gjc2     GSM2545336    288 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
 9 Plp1     GSM2545336  43217 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
10 Gnb4     GSM2545336   1071 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
# ℹ 40 more rows
# ℹ 12 more variables: tissue &lt;chr&gt;, mouse &lt;int&gt;, center &lt;chr&gt;, gene &lt;chr&gt;,
#   ENTREZID &lt;int&gt;, product &lt;chr&gt;, ensembl_gene_id &lt;chr&gt;,
#   external_synonym &lt;chr&gt;, chromosome_name &lt;chr&gt;, gene_biotype &lt;chr&gt;,
#   phenotype_description &lt;chr&gt;, hsapiens_homolog_associated_gene_name &lt;chr&gt;</code></pre>
</div>
<p>We can use <code>select</code> to specify columns we want to
view.</p>
<div class="codewrapper sourceCode" id="cb62">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">|&gt;</span> <span class="fu">select</span><span class="op">(</span><span class="va">.sample</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>tidySummarizedExperiment says: Key columns are missing. A data frame is returned for independent data analysis.</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 32,428 × 1
   .sample
   &lt;chr&gt;
 1 GSM2545336
 2 GSM2545336
 3 GSM2545336
 4 GSM2545336
 5 GSM2545336
 6 GSM2545336
 7 GSM2545336
 8 GSM2545336
 9 GSM2545336
10 GSM2545336
# ℹ 32,418 more rows</code></pre>
</div>
<p>We can use <code>mutate</code> to add metadata info.</p>
<div class="codewrapper sourceCode" id="cb65">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">|&gt;</span> <span class="fu">mutate</span><span class="op">(</span>center <span class="op">=</span> <span class="st">"Heidelberg University"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A SummarizedExperiment-tibble abstraction: 32,428 × 22
# Features=1474 | Samples=22 | Assays=counts
   .feature .sample    counts sample organism   age sex   infection strain  time
   &lt;chr&gt;    &lt;chr&gt;       &lt;int&gt; &lt;chr&gt;  &lt;chr&gt;    &lt;int&gt; &lt;chr&gt; &lt;chr&gt;     &lt;chr&gt;  &lt;int&gt;
 1 Asl      GSM2545336   1170 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
 2 Apod     GSM2545336  36194 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
 3 Cyp2d22  GSM2545336   4060 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
 4 Klk6     GSM2545336    287 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
 5 Fcrls    GSM2545336     85 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
 6 Slc2a4   GSM2545336    782 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
 7 Exd2     GSM2545336   1619 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
 8 Gjc2     GSM2545336    288 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
 9 Plp1     GSM2545336  43217 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
10 Gnb4     GSM2545336   1071 GSM25… Mus mus…     8 Fema… Influenz… C57BL…     8
# ℹ 40 more rows
# ℹ 12 more variables: tissue &lt;chr&gt;, mouse &lt;int&gt;, center &lt;chr&gt;, gene &lt;chr&gt;,
#   ENTREZID &lt;int&gt;, product &lt;chr&gt;, ensembl_gene_id &lt;chr&gt;,
#   external_synonym &lt;chr&gt;, chromosome_name &lt;chr&gt;, gene_biotype &lt;chr&gt;,
#   phenotype_description &lt;chr&gt;, hsapiens_homolog_associated_gene_name &lt;chr&gt;</code></pre>
</div>
<p>We can also combine commands with the tidyverse pipe
<code>|&gt;</code>. For example, we could combine <code>group_by</code>
and <code>summarise</code> to get the total counts for each sample.</p>
<div class="codewrapper sourceCode" id="cb67">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">group_by</span><span class="op">(</span><span class="va">.sample</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">summarise</span><span class="op">(</span>total_counts<span class="op">=</span><span class="fu">sum</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>tidySummarizedExperiment says: A data frame is returned for independent data analysis.</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 22 × 2
   .sample    total_counts
   &lt;chr&gt;             &lt;int&gt;
 1 GSM2545336      3039671
 2 GSM2545337      2602360
 3 GSM2545338      2458618
 4 GSM2545339      2500082
 5 GSM2545340      2479024
 6 GSM2545341      2413723
 7 GSM2545342      2349728
 8 GSM2545343      3105652
 9 GSM2545344      2524137
10 GSM2545345      2506038
# ℹ 12 more rows</code></pre>
</div>
<p>We can treat the tidy SummarizedExperiment object as a normal tibble
for plotting.</p>
<p>Here we plot the distribution of counts per sample.</p>
<div class="codewrapper sourceCode" id="cb70">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span><span class="va">counts</span> <span class="op">+</span> <span class="fl">1</span>, group<span class="op">=</span><span class="va">.sample</span>, color<span class="op">=</span><span class="va">infection</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_density</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">scale_x_log10</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/60-next-steps-rendered-tidySE-plot-1.png" alt="Density plot showing log of expression counts + 1 density lines, one per samples, coloured based on the infection status." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>For more information on <code>tidySummarizedExperiment</code>, see
the <a href="https://stemangiola.github.io/tidySummarizedExperiment/" class="external-link">package
website</a>.</p>
<p><strong>Take-home message</strong></p>
<ul>
<li><p><code>SummarizedExperiment</code> represents an efficient way to
store and handle omics data.</p></li>
<li><p>They are used in many Bioconductor packages.</p></li>
</ul>
<p>If you follow the next training focused on RNA sequencing analysis,
you will learn to use the Bioconductor <code>DESeq2</code> package to do
some differential expression analyses. The whole analysis of the
<code>DESeq2</code> package is handled in a
<code>SummarizedExperiment</code>.</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>Bioconductor is a project provide support and packages for the
comprehension of high high-throughput biology data.</li>
<li>A <code>SummarizedExperiment</code> is a type of object useful to
store and manage high-throughput omics data.</li>
</ul>
</div>
</div>
</div>
</div>
</section><div class="footnotes footnotes-end-of-document">
<hr>
<ol>
<li id="fn1"><p>The <a href="https://www.bioconductor.org" class="external-link">Bioconductor</a> was initiated by
Robert Gentleman, one of the two creators of the R language.
Bioconductor provides tools dedicated to omics data analysis.
Bioconductor uses the R statistical programming language and is open
source and open development.<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
</ol>
</div></section>
</div>
    </main>
</div>
<!-- END  : inst/pkgdown/templates/content-extra.html -->

      </div>
<!--/div.row-->
      		<footer class="row footer mx-md-3"><hr>
<div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/carpentries-incubator/bioc-intro/edit/main/README.md" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/carpentries-incubator/bioc-intro/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/carpentries-incubator/bioc-intro/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries-incubator/bioc-intro/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:laurent.gatto@uclouvain.be">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.12" class="external-link">sandpaper (0.16.12)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.6" class="external-link">varnish (1.0.6)</a></p>
			</div>
		</footer>
</div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://carpentries-incubator.github.io/bioc-intro/aio.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "All in One View",
  "creativeWorkStatus": "active",
  "url": "https://carpentries-incubator.github.io/bioc-intro/aio.html",
  "identifier": "https://carpentries-incubator.github.io/bioc-intro/aio.html",
  "dateCreated": "2020-09-14",
  "dateModified": "2025-06-17",
  "datePublished": "2025-06-17"
}

  </script><script>
		feather.replace();
	</script>
</body>
</html><!-- END:   inst/pkgdown/templates/layout.html-->

