<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<title>Introduction to data analysis with R and Bioconductor: All Images</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<script src="assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="assets/styles.css">
<script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="favicons/incubator/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicons/incubator/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="favicons/incubator/favicon-16x16.png">
<link rel="manifest" href="favicons/incubator/site.webmanifest">
<link rel="mask-icon" href="favicons/incubator/safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" media="(prefers-color-scheme: light)" content="white">
<meta name="theme-color" media="(prefers-color-scheme: dark)" content="black">
</head>
<body>
    <header id="top" class="navbar navbar-expand-md top-nav incubator"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="incubator-logo" alt="Carpentries Incubator" src="assets/images/incubator-logo.svg">
</div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text">
<li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul>
</li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1">
<li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/images.html';">Instructor View</button></li>
        </ul>
</div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="assets/images/incubator-logo-sm.svg">
</div>
    <div class="lesson-title-md">
      Introduction to data analysis with R and Bioconductor
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0">
<li class="nav-item">
          <span class="lesson-title">
            Introduction to data analysis with R and Bioconductor
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown">
<li><a class="dropdown-item" href="discuss.html">Discussion</a></li>
<li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul>
</li>
      </ul>
</div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div>
<!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Introduction to data analysis with R and Bioconductor
</div>

<aside class="col-md-12 lesson-progress"><div style="width: %" class="percentage">
    %
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: %" aria-valuenow="" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text">
<li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul>
</li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/images.html">Instructor View</a>
                      </div>
                    </div>
                  </div>
<!--/div.accordion-item-->
                </div>
<!--/div.accordion-flush-->
              </div>
<!--div.sidenav-view-selector -->
            </div>
<!--/div.col -->

            <hr>
</div>
<!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="10-data-organisation.html">1. Data organisation with spreadsheets</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="20-r-rstudio.html">2. R and RStudio</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="23-starting-with-r.html">3. Introduction to R</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="25-starting-with-data.html">4. Starting with data</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="30-dplyr.html">5. Manipulating and analysing data with dplyr</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="40-visualization.html">6. Data visualization</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="60-next-steps.html">7. Next steps</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width">
<div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul>
<li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="discuss.html">Discussion</a></li>
<li><a href="reference.html">Reference</a></li>
                    </ul>
</div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources">
<a href="aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none">
<div class="d-grid gap-1">

            </div>
          </div>
<!-- /div.accordion -->
        </div>
<!-- /div.sidebar-inner -->
      </nav>
</div>
<!-- /div.sidebar -->
  </div>
<!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-extra.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <main id="main-content" class="main-content"><div class="container lesson-content">
        
        
<section id="10-data-organisation"><h2 class="section-heading"><a href="10-data-organisation.html">Data organisation with spreadsheets</a></h2>
<hr class="half-width">
<h3 id="10-data-organisation-figure-1">Figure 1</h3>
<p aria-hidden="true">Image 1 of 1: ‘SCreenshot of a spreadsheet and a README.txt file.’</p>
<figure style="text-align: center"><img src="fig/spreadsheet-setup-updated.png" alt="SCreenshot of a spreadsheet and a README.txt file." width="70%" class="figure mx-auto d-block"><figcaption>
Spreadsheet example setip
</figcaption></figure><hr>
<h3 id="10-data-organisation-figure-2">Figure 2</h3>
<p aria-hidden="true">Image 1 of 1: ‘A table showing the blood and rhesus types encoded together in one column.’</p>
<figure style="text-align: center"><img src="fig/multiple-info.png" alt="A table showing the blood and rhesus types encoded together in one column." width="70%" class="figure mx-auto d-block"><figcaption>
Multiple variable in a single column.
</figcaption></figure><hr>
<h3 id="10-data-organisation-figure-3">Figure 3</h3>
<p aria-hidden="true">Image 1 of 1: ‘A table showing the blood and rhesus types encoded together in different columns.’</p>
<figure style="text-align: center"><img src="fig/single-info.png" alt="A table showing the blood and rhesus types encoded together in different columns." width="70%" class="figure mx-auto d-block"><figcaption>
Variables encoded in a single column each.
</figcaption></figure><hr>
<h3 id="10-data-organisation-figure-4">Figure 4</h3>
<p aria-hidden="true">Image 1 of 1: ‘A sheet with many subtable.’</p>
<figure style="text-align: center"><img src="fig/2_datasheet_example.jpg" alt="A sheet with many subtable." width="70%" class="figure mx-auto d-block"><figcaption>
A sheet with many subtable.
</figcaption></figure><hr>
<h3 id="10-data-organisation-figure-5">Figure 5</h3>
<p aria-hidden="true">Image 1 of 1: ‘A table extracted from a paper showing commonly used null values.’</p>
<figure style="text-align: center"><img src="fig/3_white_table_1.jpg" alt="A table extracted from a paper showing commonly used null values." width="70%" class="figure mx-auto d-block"><figcaption>
Commonly used null values.
</figcaption></figure><hr>
<h3 id="10-data-organisation-figure-6">Figure 6</h3>
<p aria-hidden="true">Image 1 of 1: ‘A table showing several cells highlighted in yellow used to encode sample contamination.’</p>
<figure style="text-align: center"><img src="fig/formatting.png" alt="A table showing several cells highlighted in yellow used to encode sample contamination." width="70%" class="figure mx-auto d-block"><figcaption>
Colours used to ecode information.
</figcaption></figure><hr>
<h3 id="10-data-organisation-figure-7">Figure 7</h3>
<p aria-hidden="true">Image 1 of 1: ‘Same table as above with a new variable encoding contambination.’</p>
<figure style="text-align: center"><img src="fig/good_formatting.png" alt="Same table as above with a new variable encoding contambination." width="70%" class="figure mx-auto d-block"><figcaption>
A new variable encoding sample contamination
</figcaption></figure><hr>
<h3 id="10-data-organisation-figure-8">Figure 8</h3>
<p aria-hidden="true">Image 1 of 1: ‘Excel menu to export data in a csv format on MacOS ’</p>
<figure style="text-align: center"><img src="fig/excel-to-csv.png" alt="Excel menu to export data in a csv format on MacOS " width="70%" class="figure mx-auto d-block"><figcaption>
Saving an Excel file to CSV.
</figcaption></figure><hr>
<h3 id="10-data-organisation-figure-9">Figure 9</h3>
<p aria-hidden="true">Image 1 of 1: ‘A table illustrating ill-imported data that contained a comma in a cell. That cell has been split in two cells.’</p>
<figure style="text-align: center"><img src="fig/csv-mistake.png" alt="A table illustrating ill-imported data that contained a comma in a cell. That cell has been split in two cells." width="80%" class="figure mx-auto d-block"><figcaption>
The risks of having commas inside comma-separated data.
</figcaption></figure><hr>
<h3 id="10-data-organisation-figure-10">Figure 10</h3>
<p aria-hidden="true">Image 1 of 1: ‘The tidy data analysis circle, iterating over data transformation, visualisation and modelling.’</p>
<figure style="text-align: center"><img src="fig/analysis.png" alt="The tidy data analysis circle, iterating over data transformation, visualisation and modelling." width="70%" class="figure mx-auto d-block"><figcaption>
A typical data analysis workflow.
</figcaption></figure><hr></section><section id="20-r-rstudio"><h2 class="section-heading"><a href="20-r-rstudio.html">R and RStudio</a></h2>
<hr class="half-width">
<h3 id="20-r-rstudio-figure-1">Figure 1</h3>
<p aria-hidden="true">Image 1 of 1: ‘Two graphs showing the exponential increase of the number of CRAN packages from 2000 to 2018. The figure on the right shows linear trend for the log of number of packages.’</p>
<figure style="text-align: center"><img src="fig/cran.png" alt="Two graphs showing the exponential increase of the number of CRAN packages from 2000 to 2018. The figure on the right shows linear trend for the log of number of packages." class="figure mx-auto d-block"><figcaption>
Exponential increase of the number of packages available on <a href="https://cran.r-project.org/" class="external-link">CRAN</a>, the Comprehensive R Archive
Network. From the R Journal, Volume 10/2, December 2018.
</figcaption></figure><hr>
<h3 id="20-r-rstudio-figure-2">Figure 2</h3>
<p aria-hidden="true">Image 1 of 1: ‘Screenshot of the RStudio interface showing the typical four panel setup.’</p>
<figure style="text-align: center"><img src="fig/rstudio-screenshot.png" alt="Screenshot of the RStudio interface showing the typical four panel setup." width="100%" class="figure mx-auto d-block"><figcaption>
RStudio interface screenshot. Clockwise from top left: Source,
Environment/History, Files/Plots/Packages/Help/Viewer, Console.
</figcaption></figure><hr>
<h3 id="20-r-rstudio-figure-3">Figure 3</h3>
<p aria-hidden="true">Image 1 of 1: ‘Screenshot of the RStudio General Options panel.’</p>
<figure style="text-align: center"><img src="fig/rstudio-preferences.png" alt="Screenshot of the RStudio General Options panel." width="70%" class="figure mx-auto d-block"><figcaption>
Set ‘Save workspace to .RData on exit’ to ‘Never’
</figcaption></figure><hr>
<h3 id="20-r-rstudio-figure-4">Figure 4</h3>
<p aria-hidden="true">Image 1 of 1: ‘Screenshot of the RStudio Code Options panel.’</p>
<figure style="text-align: center"><img src="fig/utf8.png" alt="Screenshot of the RStudio Code Options panel." width="70%" class="figure mx-auto d-block"><figcaption>
Set the default text encoding to UTF-8 to save us headache in the coming
future. (Figure from the link above).
</figcaption></figure><hr>
<h3 id="20-r-rstudio-figure-5">Figure 5</h3>
<p aria-hidden="true">Image 1 of 1: ‘Flow diagram showing the structure of the working directory and various directory of a typical RStuio project.’</p>
<figure style="text-align: center"><img src="fig/working-directory-structure.png" alt="Flow diagram showing the structure of the working directory and various directory of a typical RStuio project." width="100%" class="figure mx-auto d-block"><figcaption>
Example of a working directory structure.
</figcaption></figure><hr>
<h3 id="20-r-rstudio-figure-6">Figure 6</h3>
<p aria-hidden="true">Image 1 of 1: ‘Screenshot of RStudio's Files panel.’</p>
<figure style="text-align: center"><img src="fig/r-starting-how-it-should-look-like.png" alt="Screenshot of RStudio's Files panel." width="100%" class="figure mx-auto d-block"><figcaption>
How it should look like at the beginning of this lesson
</figcaption></figure><hr>
<h3 id="20-r-rstudio-figure-7">Figure 7</h3>
<p aria-hidden="true">Image 1 of 1: ‘Figure form Nobel et al. (2009) showing a complex bioinformatics project structure.’</p>
<figure style="text-align: center"><img src="fig/noble-bioinfo-project.png" alt="Figure form Nobel et al. (2009) showing a complex bioinformatics project structure." width="100%" class="figure mx-auto d-block"><figcaption>
Directory structure for a sample bioinformatics project.
</figcaption></figure><hr>
<h3 id="20-r-rstudio-figure-8">Figure 8</h3>
<p aria-hidden="true">Image 1 of 1: ‘Fake book cover showing a kitten with title 'Changing Stuff and Seeing What Happens.'’</p>
<figure><img src="fig/kitten-try-things.jpg" alt="Fake book cover showing a kitten with title 'Changing Stuff and Seeing What Happens.'" width="400px" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="20-r-rstudio-figure-9">Figure 9</h3>
<p aria-hidden="true">Image 1 of 1: ‘Screenshot of RStudio's Help panel.’</p>
<figure style="text-align: center"><img src="fig/rstudiohelp.png" alt="Screenshot of RStudio's Help panel." width="70%" class="figure mx-auto d-block"><figcaption>
RStudio help interface.
</figcaption></figure><hr></section><section id="23-starting-with-r"><h2 class="section-heading"><a href="23-starting-with-r.html">Introduction to R</a></h2>
<hr class="half-width">
<h3 id="23-starting-with-r-figure-1">Figure 1</h3>
<p aria-hidden="true">Image 1 of 1: ‘Two figures illustraing normal distributonis N(0,1) and N(100,5).’</p>
<figure style="text-align: center"><img src="fig/23-starting-with-r-rendered-unnamed-chunk-61-1.png" alt="Two figures illustraing normal distributonis N(0,1) and N(100,5)." class="figure mx-auto d-block"><figcaption>
Two normal distributions: <em>N(0, 1)</em> on the left and <em>N(100,
5)</em> on the right.
</figcaption></figure><hr></section><section id="25-starting-with-data"><h2 class="section-heading"><a href="25-starting-with-data.html">Starting with data</a></h2>
<hr class="half-width">
<h3 id="25-starting-with-data-figure-1">Figure 1</h3>
<p aria-hidden="true">Image 1 of 1: ‘Schematic representation of a tablular structure with different colours representing different columns and variables types.’</p>
<figure style="text-align: center"><img src="fig/data-frame.svg" alt="Schematic representation of a tablular structure with different colours representing different columns and variables types." width="70%" class="figure mx-auto d-block"><figcaption>
Conceptual representation of a data.frame
</figcaption></figure><hr>
<h3 id="25-starting-with-data-figure-2">Figure 2</h3>
<p aria-hidden="true">Image 1 of 1: ‘Barplot showing the number of 'male' and 'female' occurrences.’</p>
<figure style="text-align: center"><img src="fig/25-starting-with-data-rendered-firstfactorplot-1.png" alt="Barplot showing the number of 'male' and 'female' occurrences." class="figure mx-auto d-block"><figcaption>
Bar plot of the number of females and males.
</figcaption></figure><hr>
<h3 id="25-starting-with-data-figure-3">Figure 3</h3>
<p aria-hidden="true">Image 1 of 1: ‘Barplot showing the number of 'M' and 'F' occurrences.’</p>
<figure><img src="fig/25-starting-with-data-rendered-unnamed-chunk-17-1.png" alt="Barplot showing the number of 'M' and 'F' occurrences." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr></section><section id="30-dplyr"><h2 class="section-heading"><a href="30-dplyr.html">Manipulating and analysing data with dplyr</a></h2>
<hr class="half-width">
<h3 id="30-dplyr-figure-1">Figure 1</h3>
<p aria-hidden="true">Image 1 of 1: ‘Figure showing the long table format on the left and wide table format on the right and arrows illustrating how the values in the 'sample' column on the left have become column names on the right and how the values in the 'expression' column on the left have become values on the the right. Below is the call to 'pivot_wider()' with annotations pointing to the 'sample' and 'expression' function arguments.’</p>
<figure style="text-align: center"><img src="fig/pivot_wider.png" alt="Figure showing the long table format on the left and wide table format on the right and arrows illustrating how the values in the 'sample' column on the left have become column names on the right and how the values in the 'expression' column on the left have become values on the the right. Below is the call to 'pivot_wider()' with annotations pointing to the 'sample' and 'expression' function arguments." class="figure mx-auto d-block"><figcaption>
Wide pivot of the <code>rna</code> data.
</figcaption></figure><hr>
<h3 id="30-dplyr-figure-2">Figure 2</h3>
<p aria-hidden="true">Image 1 of 1: ‘Figure showing the long table format on the left and wide table format on the right and arrows illustrating how the column names on the left have become a new column 'sample' on the left and the values in the wide table on the right have become a new column 'expression' on the left.  Below is the call to 'pivot_wider()' with annotations pointing to the 'sample', 'expression' and the '-gene' arguments.’</p>
<figure style="text-align: center"><img src="fig/pivot_longer.png" alt="Figure showing the long table format on the left and wide table format on the right and arrows illustrating how the column names on the left have become a new column 'sample' on the left and the values in the wide table on the right have become a new column 'expression' on the left.  Below is the call to 'pivot_wider()' with annotations pointing to the 'sample', 'expression' and the '-gene' arguments." class="figure mx-auto d-block"><figcaption>
Long pivot of the <code>rna</code> data.
</figcaption></figure><hr>
<h3 id="30-dplyr-figure-3">Figure 3</h3>
<p aria-hidden="true">Image 1 of 1: ‘A two by two table with  X and Y row names and Female and Male column names showing the total counts for each X/Y and Female/Male combination. The Y/Female combinations shows 3 counts, while all other counts are above 2000 counts.’</p>
<figure><img src="fig/Exercise_pivot_W.png" alt="A two by two table with  X and Y row names and Female and Male column names showing the total counts for each X/Y and Female/Male combination. The Y/Female combinations shows 3 counts, while all other counts are above 2000 counts." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr></section><section id="40-visualization"><h2 class="section-heading"><a href="40-visualization.html">Data visualization</a></h2>
<hr class="half-width">
<h3 id="40-visualization-figure-1">Figure 1</h3>
<p aria-hidden="true">Image 1 of 1: ‘Default histogram produced by ggplot() and geom_histogram() for the expression data.’</p>
<figure><img src="fig/40-visualization-rendered-first-ggplot-1.png" alt="Default histogram produced by ggplot() and geom_histogram() for the expression data." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-2">Figure 2</h3>
<p aria-hidden="true">Image 1 of 1: ‘Histograms produced by ggplot() and geom_histogram() for the expression data with bin set of 15 (top) and binwith set to 2000 (bottom).’</p>
<figure><img src="fig/40-visualization-rendered-unnamed-chunk-5-1.png" alt="Histograms produced by ggplot() and geom_histogram() for the expression data with bin set of 15 (top) and binwith set to 2000 (bottom)." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-3">Figure 3</h3>
<p aria-hidden="true">Image 1 of 1: ‘Histograms produced by ggplot() and geom_histogram() for the expression data with bin set of 15 (top) and binwith set to 2000 (bottom).’</p>
<figure><img src="fig/40-visualization-rendered-unnamed-chunk-5-2.png" alt="Histograms produced by ggplot() and geom_histogram() for the expression data with bin set of 15 (top) and binwith set to 2000 (bottom)." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-4">Figure 4</h3>
<figure><img src="fig/40-visualization-rendered-second-ggplot-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-5">Figure 5</h3>
<p aria-hidden="true">Image 1 of 1: ‘Histograms produced by ggplot(), geom_histogram() and scale_x_log10() for the log of expression.’</p>
<figure><img src="fig/40-visualization-rendered-unnamed-chunk-6-1.png" alt="Histograms produced by ggplot(), geom_histogram() and scale_x_log10() for the log of expression." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-6">Figure 6</h3>
<p aria-hidden="true">Image 1 of 1: ‘Scatter plot produced by ggplot() and geom_point() comparing the log-foldchanges computed above. All dots are black.’</p>
<figure><img src="fig/40-visualization-rendered-create-ggplot-object-1.png" alt="Scatter plot produced by ggplot() and geom_point() comparing the log-foldchanges computed above. All dots are black." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-7">Figure 7</h3>
<p aria-hidden="true">Image 1 of 1: ‘Scatter plot produced by ggplot() and geom_point() comparing the log-foldchanges computed above. All dots are semi-transparent black.’</p>
<figure><img src="fig/40-visualization-rendered-adding-transparency-1.png" alt="Scatter plot produced by ggplot() and geom_point() comparing the log-foldchanges computed above. All dots are semi-transparent black." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-8">Figure 8</h3>
<p aria-hidden="true">Image 1 of 1: ‘Scatter plot produced by ggplot() and geom_point() comparing the log-foldchanges computed above. All dots are semi-transparent blue.’</p>
<figure><img src="fig/40-visualization-rendered-adding-colors-1.png" alt="Scatter plot produced by ggplot() and geom_point() comparing the log-foldchanges computed above. All dots are semi-transparent blue." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-9">Figure 9</h3>
<p aria-hidden="true">Image 1 of 1: ‘Scatter plot produced by ggplot() and geom_point() comparing the log-foldchanges computed above. Dots are colour-coded based on the gene's biotype.’</p>
<figure><img src="fig/40-visualization-rendered-color-by-gene_biotype1-1.png" alt="Scatter plot produced by ggplot() and geom_point() comparing the log-foldchanges computed above. Dots are colour-coded based on the gene's biotype." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-10">Figure 10</h3>
<p aria-hidden="true">Image 1 of 1: ‘Scatter plot produced by ggplot() and geom_point() comparing the log-foldchanges computed above. Dots are colour-coded based on the gene's biotype.’</p>
<figure><img src="fig/40-visualization-rendered-color-by-gene_biotype2-1.png" alt="Scatter plot produced by ggplot() and geom_point() comparing the log-foldchanges computed above. Dots are colour-coded based on the gene's biotype." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-11">Figure 11</h3>
<p aria-hidden="true">Image 1 of 1: ‘Scatter plot produced by ggplot() and geom_point() comparing the log-foldchanges computed above. Dots are colour-coded based on the gene's biotype. A black line of slope 1 crossing the origin was added by geom_abline().’</p>
<figure><img src="fig/40-visualization-rendered-adding-diag-1.png" alt="Scatter plot produced by ggplot() and geom_point() comparing the log-foldchanges computed above. Dots are colour-coded based on the gene's biotype. A black line of slope 1 crossing the origin was added by geom_abline()." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-12">Figure 12</h3>
<p aria-hidden="true">Image 1 of 1: ‘Scatter plot produced by ggplot() and geom_point() comparing the log-foldchanges computed above. Dots are colour-coded based on the gene's biotype. A black line of slope 1 crossing the origin was added by geom_abline().’</p>
<figure><img src="fig/40-visualization-rendered-color-by-gene_biotype3-1.png" alt="Scatter plot produced by ggplot() and geom_point() comparing the log-foldchanges computed above. Dots are colour-coded based on the gene's biotype. A black line of slope 1 crossing the origin was added by geom_abline()." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-13">Figure 13</h3>
<p aria-hidden="true">Image 1 of 1: ‘Scatter plot produced by ggplot() and geom_hexbin() comparing the log-foldchanges computed above shows hexagons coloured based on the underlying dot density. A black line of slope 1 crossing the origin was added by geom_abline().’</p>
<figure><img src="fig/40-visualization-rendered-unnamed-chunk-10-1.png" alt="Scatter plot produced by ggplot() and geom_hexbin() comparing the log-foldchanges computed above shows hexagons coloured based on the underlying dot density. A black line of slope 1 crossing the origin was added by geom_abline()." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-14">Figure 14</h3>
<p aria-hidden="true">Image 1 of 1: ‘Figures showing a stretch of overlapping points indicating the log of expression + 1 for each sample. The points are coloured with different shades of blue for samples collected at different time points.’</p>
<figure><img src="fig/40-visualization-rendered-unnamed-chunk-11-1.png" alt="Figures showing a stretch of overlapping points indicating the log of expression + 1 for each sample. The points are coloured with different shades of blue for samples collected at different time points." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-15">Figure 15</h3>
<p aria-hidden="true">Image 1 of 1: ‘Boxplot showing the distribution of log expression + 1 values for each sample, as produced by geom_boxlpot(). Each boxplot is filled with white colour.’</p>
<figure><img src="fig/40-visualization-rendered-boxplot-1.png" alt="Boxplot showing the distribution of log expression + 1 values for each sample, as produced by geom_boxlpot(). Each boxplot is filled with white colour." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-16">Figure 16</h3>
<p aria-hidden="true">Image 1 of 1: ‘Boxplot and dots showing the distribution of log expression + 1 values for each sample, as produced by geom_boxlpot(). Each boxplot is transparent and the jittered dots are semi-transparent tomato-coloured and behind the boxpots.’</p>
<figure><img src="fig/40-visualization-rendered-boxplot-with-points-1.png" alt="Boxplot and dots showing the distribution of log expression + 1 values for each sample, as produced by geom_boxlpot(). Each boxplot is transparent and the jittered dots are semi-transparent tomato-coloured and behind the boxpots." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-17">Figure 17</h3>
<p aria-hidden="true">Image 1 of 1: ‘Boxplot and dots showing the distribution of log expression + 1 values for each sample, as produced by geom_boxlpot(). Each boxplot is transparent and the jittered dots are semi-transparent tomato-coloured. This time, the boxplots are behind the dots.’</p>
<figure><img src="fig/40-visualization-rendered-boxplot-with-points2-1.png" alt="Boxplot and dots showing the distribution of log expression + 1 values for each sample, as produced by geom_boxlpot(). Each boxplot is transparent and the jittered dots are semi-transparent tomato-coloured. This time, the boxplots are behind the dots." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-18">Figure 18</h3>
<p aria-hidden="true">Image 1 of 1: ‘Boxplot and dots showing the distribution of log expression + 1 values for each sample, as produced by geom_boxlpot(). Each boxplot is transparent and the jittered dots are semi-transparent tomato-coloured. The sample labels are displayed vertically and readable.’</p>
<figure><img src="fig/40-visualization-rendered-boxplot-xaxis-rotated-1.png" alt="Boxplot and dots showing the distribution of log expression + 1 values for each sample, as produced by geom_boxlpot(). Each boxplot is transparent and the jittered dots are semi-transparent tomato-coloured. The sample labels are displayed vertically and readable." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-19">Figure 19</h3>
<p aria-hidden="true">Image 1 of 1: ‘Boxplot and dots showing the distribution of log expression + 1 values for each sample, as produced by geom_boxlpot(). On the first figure, each boxplot is transparent and the jittered dots are semi-transparent and coloured in different shares of blue. On the second figures, each boxplot is transparent and the jittered dots are semi-transparent and coloured red, green and blue.’</p>
<figure><img src="fig/40-visualization-rendered-boxplot-color-time-1.png" alt="Boxplot and dots showing the distribution of log expression + 1 values for each sample, as produced by geom_boxlpot(). On the first figure, each boxplot is transparent and the jittered dots are semi-transparent and coloured in different shares of blue. On the second figures, each boxplot is transparent and the jittered dots are semi-transparent and coloured red, green and blue." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-20">Figure 20</h3>
<p aria-hidden="true">Image 1 of 1: ‘Boxplot and dots showing the distribution of log expression + 1 values for each sample, as produced by geom_boxlpot(). On the first figure, each boxplot is transparent and the jittered dots are semi-transparent and coloured in different shares of blue. On the second figures, each boxplot is transparent and the jittered dots are semi-transparent and coloured red, green and blue.’</p>
<figure><img src="fig/40-visualization-rendered-boxplot-color-time-2.png" alt="Boxplot and dots showing the distribution of log expression + 1 values for each sample, as produced by geom_boxlpot(). On the first figure, each boxplot is transparent and the jittered dots are semi-transparent and coloured in different shares of blue. On the second figures, each boxplot is transparent and the jittered dots are semi-transparent and coloured red, green and blue." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-21">Figure 21</h3>
<p aria-hidden="true">Image 1 of 1: ‘Violin plot showing the distribution of log expression + 1 values for each sample, as produced by geom_violin(). Each boxplot is transparent and the jittered dots are semi-transparent and coloured red, green and blue.’</p>
<figure><img src="fig/40-visualization-rendered-unnamed-chunk-12-1.png" alt="Violin plot showing the distribution of log expression + 1 values for each sample, as produced by geom_violin(). Each boxplot is transparent and the jittered dots are semi-transparent and coloured red, green and blue." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-22">Figure 22</h3>
<p aria-hidden="true">Image 1 of 1: ‘Violin plot showing the distribution of log expression + 1 values for each sample, as produced by geom_violon(). Each violin plot is coloured in red or blue depending on the sex variable.’</p>
<figure><img src="fig/40-visualization-rendered-unnamed-chunk-13-1.png" alt="Violin plot showing the distribution of log expression + 1 values for each sample, as produced by geom_violon(). Each violin plot is coloured in red or blue depending on the sex variable." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-23">Figure 23</h3>
<p aria-hidden="true">Image 1 of 1: ‘Line plot, as produced by geom_line(), but the lines are clearly not show what we expect.’</p>
<figure><img src="fig/40-visualization-rendered-first-time-series-1.png" alt="Line plot, as produced by geom_line(), but the lines are clearly not show what we expect." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-24">Figure 24</h3>
<p aria-hidden="true">Image 1 of 1: ‘Line plot, as produced by geom_line(), with 10 lines showing increasing expression values over time.’</p>
<figure><img src="fig/40-visualization-rendered-time-series-by-gene-1.png" alt="Line plot, as produced by geom_line(), with 10 lines showing increasing expression values over time." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-25">Figure 25</h3>
<p aria-hidden="true">Image 1 of 1: ‘Line plot, as produced by geom_line(), with 10 colour-coded lines showing increasing expression values over time.’</p>
<figure><img src="fig/40-visualization-rendered-time-series-with-colors-1.png" alt="Line plot, as produced by geom_line(), with 10 colour-coded lines showing increasing expression values over time." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-26">Figure 26</h3>
<p aria-hidden="true">Image 1 of 1: ‘Line plot, as produced by geom_line(), with 10 sub-plots/facets, each showing one line with increasing expression values over time. All y-scales are identical.’</p>
<figure><img src="fig/40-visualization-rendered-first-facet-1.png" alt="Line plot, as produced by geom_line(), with 10 sub-plots/facets, each showing one line with increasing expression values over time. All y-scales are identical." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-27">Figure 27</h3>
<p aria-hidden="true">Image 1 of 1: ‘Line plot, as produced by geom_line(), with 10 sub-plots/facets, each showing one line with increasing expression values over time. y-scales are now adapted to the expression range for each facet/gene.’</p>
<figure><img src="fig/40-visualization-rendered-first-facet-scales-1.png" alt="Line plot, as produced by geom_line(), with 10 sub-plots/facets, each showing one line with increasing expression values over time. y-scales are now adapted to the expression range for each facet/gene." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-28">Figure 28</h3>
<p aria-hidden="true">Image 1 of 1: ‘Line plot, as produced by geom_line(), with 10 sub-plots/facets, each showing two coloured lines (red for Female and blue for Make) with increasing expression values over time.’</p>
<figure><img src="fig/40-visualization-rendered-facet-by-gene-and-sex-1.png" alt="Line plot, as produced by geom_line(), with 10 sub-plots/facets, each showing two coloured lines (red for Female and blue for Make) with increasing expression values over time." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-29">Figure 29</h3>
<p aria-hidden="true">Image 1 of 1: ‘Line plot, as produced by geom_line(), with 10 sub-plots/facets, each showing two coloured lines (red for Female and blue for Make) with increasing expression values over time. The figure background is now white.’</p>
<figure><img src="fig/40-visualization-rendered-facet-by-gene-and-sex-white-bg-1.png" alt="Line plot, as produced by geom_line(), with 10 sub-plots/facets, each showing two coloured lines (red for Female and blue for Make) with increasing expression values over time. The figure background is now white." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-30">Figure 30</h3>
<p aria-hidden="true">Image 1 of 1: ‘Line plot, as produced by geom_line(), with 21 sub-plots/facets, each showing one line with expression values over time for each chromosome.’</p>
<figure><img src="fig/40-visualization-rendered-mean-exp-chromosome-time-series-1.png" alt="Line plot, as produced by geom_line(), with 21 sub-plots/facets, each showing one line with expression values over time for each chromosome." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-31">Figure 31</h3>
<p aria-hidden="true">Image 1 of 1: ‘Two line plots above each other, each showing 10 lines coloured by gene. The top facet shows the expression values for Female sample, the bottom one for Male samples.’</p>
<figure><img src="fig/40-visualization-rendered-mean-exp-time-facet-sex-rows-1.png" alt="Two line plots above each other, each showing 10 lines coloured by gene. The top facet shows the expression values for Female sample, the bottom one for Male samples." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-32">Figure 32</h3>
<p aria-hidden="true">Image 1 of 1: ‘Two line plots next to each other, each showing 10 lines coloured by gene. The left facet shows the expression values for Female sample, the right one for Male samples.’</p>
<figure><img src="fig/40-visualization-rendered-mean-exp-time-facet-sex-columns-1.png" alt="Two line plots next to each other, each showing 10 lines coloured by gene. The left facet shows the expression values for Female sample, the right one for Male samples." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-33">Figure 33</h3>
<figure><img src="fig/40-visualization-rendered-mean_exp-time-with-right-labels-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-34">Figure 34</h3>
<figure><img src="fig/40-visualization-rendered-mean_exp-time-with-right-labels-xfont-size-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-35">Figure 35</h3>
<figure><img src="fig/40-visualization-rendered-mean_exp-time-with-theme-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-36">Figure 36</h3>
<figure><img src="fig/40-visualization-rendered-mean_exp-time-with-right-labels-xfont-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-37">Figure 37</h3>
<figure><img src="fig/40-visualization-rendered-unnamed-chunk-15-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-38">Figure 38</h3>
<figure><img src="fig/40-visualization-rendered-unnamed-chunk-16-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-39">Figure 39</h3>
<figure><img src="fig/40-visualization-rendered-unnamed-chunk-16-2.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-40">Figure 40</h3>
<figure><img src="fig/40-visualization-rendered-unnamed-chunk-16-3.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-41">Figure 41</h3>
<figure><img src="fig/40-visualization-rendered-unnamed-chunk-16-4.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-42">Figure 42</h3>
<figure><img src="fig/40-visualization-rendered-sub1-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-43">Figure 43</h3>
<figure><img src="fig/40-visualization-rendered-sub2-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-44">Figure 44</h3>
<figure><img src="fig/40-visualization-rendered-patchworkplot1-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-45">Figure 45</h3>
<figure><img src="fig/40-visualization-rendered-patchwork2-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-46">Figure 46</h3>
<figure><img src="fig/40-visualization-rendered-patchwork3-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-47">Figure 47</h3>
<figure><img src="fig/40-visualization-rendered-patchwork4-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-48">Figure 48</h3>
<figure><img src="fig/40-visualization-rendered-patchwork5-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-49">Figure 49</h3>
<figure><img src="fig/40-visualization-rendered-gridarrange-example-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="40-visualization-figure-50">Figure 50</h3>
<p aria-hidden="true">Image 1 of 1: ‘Successive layers added on top of each other.’</p>
<figure style="text-align: center"><img src="fig/40-visualization-rendered-paintermodel-1.png" alt="Successive layers added on top of each other." class="figure mx-auto d-block"><figcaption>
Successive layers added on top of each other.
</figcaption></figure><hr>
<h3 id="40-visualization-figure-51">Figure 51</h3>
<p aria-hidden="true">Image 1 of 1: ‘Plotting boxplots (top) and histograms (bottom) vectors (left) or a matrices (right).’</p>
<figure style="text-align: center"><img src="fig/40-visualization-rendered-plotmethod-1.png" alt="Plotting boxplots (top) and histograms (bottom) vectors (left) or a matrices (right)." class="figure mx-auto d-block"><figcaption>
Plotting boxplots (top) and histograms (bottom) vectors (left) or a
matrices (right).
</figcaption></figure><hr></section><section id="60-next-steps"><h2 class="section-heading"><a href="60-next-steps.html">Next steps</a></h2>
<hr class="half-width">
<h3 id="60-next-steps-figure-1">Figure 1</h3>
<figure><img src="https://uclouvain-cbio.github.io/WSBIM1322/figs/SE.svg" width="80%" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr>
<h3 id="60-next-steps-figure-2">Figure 2</h3>
<figure><img src="fig/60-next-steps-rendered-tidySE-plot-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr></section>
</div>
    </main>
</div>
<!-- END  : inst/pkgdown/templates/content-extra.html -->

      </div>
<!--/div.row-->
      		<footer class="row footer mx-md-3"><hr>
<div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/carpentries-incubator/bioc-intro/edit/main/README.md" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/carpentries-incubator/bioc-intro/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/carpentries-incubator/bioc-intro/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries-incubator/bioc-intro/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:laurent.gatto@uclouvain.be">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.12" class="external-link">sandpaper (0.16.12)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.6" class="external-link">varnish (1.0.6)</a></p>
			</div>
		</footer>
</div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://carpentries-incubator.github.io/bioc-intro/images.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "All Images",
  "creativeWorkStatus": "active",
  "url": "https://carpentries-incubator.github.io/bioc-intro/images.html",
  "identifier": "https://carpentries-incubator.github.io/bioc-intro/images.html",
  "dateCreated": "2020-09-14",
  "dateModified": "2025-06-15",
  "datePublished": "2025-06-15"
}

  </script><script>
		feather.replace();
	</script>
</body>
</html><!-- END:   inst/pkgdown/templates/layout.html-->

